<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta name="Generator" content="Kate, the KDE Advanced Text Editor" />
<title>course_commands.r</title>
</head>
<!-- Highlighting: "R Script" -->
<body>
<pre style='font-weight:bold;color:#b2b2b2;background-color:#000000;'>
<b><span style='color:#b00000;'>## Getting help</span></b>
<b><span style='color:#dede49;'>help</span></b><b>(</b>rep<b>)</b> <b><span style='color:#54ffff;'># Help for particular function (package must be loaded)</span></b>
?rep <b><span style='color:#54ffff;'># Help for particular function (package must be loaded)</span></b>
??ANOVA <b><span style='color:#54ffff;'># Search for the term within all installed packages</span></b>
<b><span style='color:#dede49;'>help.search</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;analysis of variance&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Search for the phrase within all installed packages - return list of hits sorted according to type and package (i.e. package::function)</span></b>
<b><span style='color:#dede49;'>install.packages</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;sos&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># More comprehensive search from packages</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>sos<b>)</b>
<b><span style='color:#dede49;'>findFn</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;DAPC&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Search for function name</span></b>

<b><span style='color:#b00000;'>## Set working directory</span></b>
<b><span style='color:#dede49;'>setwd</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;/home/vojta/dokumenty/fakulta/vyuka/r_mol_data/examples/&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Create YOUR OWN empty directory and modify the path accordingly!</span></b>
<b><span style='color:#dede49;'>getwd</span></b><b>()</b> <b><span style='color:#54ffff;'># Verifies where we are</span></b>
<b><span style='color:#dede49;'>dir</span></b><b>()</b> <b><span style='color:#54ffff;'># Lists files and folders on the disk</span></b>
<b><span style='color:#dede49;'>ls</span></b><b>()</b> <b><span style='color:#54ffff;'># Lists currently available R objects</span></b>

<b><span style='color:#b00000;'>## Get information about object classes</span></b>
<b><span style='color:#dede49;'>getClassDef</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;genind&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Or select any other class name</span></b>

<b><span style='color:#b00000;'>## Basic operations</span></b>
x <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>5</span></b>, <b><span style='color:#ff54ff;'>6</span></b>, <b><span style='color:#ff54ff;'>7</span></b>, <b><span style='color:#ff54ff;'>8</span></b>, <b><span style='color:#ff54ff;'>9</span></b><b>)</b> <b><span style='color:#54ffff;'># Creates vector (see also ?rep)</span></b>
x <b><span style='color:#54ffff;'># Print &quot;x&quot; content</span></b>
<b><span style='color:#dede49;'>c</span></b><b>()</b> <b><span style='color:#54ffff;'># Is generic function to concatenate objects into new one</span></b>
<b><span style='color:#dede49;'>length</span></b><b>(</b>x<b>)</b> <b><span style='color:#54ffff;'># Length of the object - for matrices and DF use dim()</span></b>
<b><span style='color:#dede49;'>str</span></b><b>(</b>x<b>)</b> <b><span style='color:#54ffff;'># Information about structure of the object</span></b>
<b><span style='color:#dede49;'>mode</span></b><b>(</b>x<b>)</b> <b><span style='color:#54ffff;'># Gets type of storage mode of the object</span></b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>x<b>)</b> <b><span style='color:#54ffff;'># Shows class of the object</span></b>
x<b>[</b><b><span style='color:#ff54ff;'>2</span></b><b>]</b> <b><span style='color:#54ffff;'># Shows second element of the object</span></b>
x <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>x<b>[</b><b>-</b><b><span style='color:#ff54ff;'>5</span></b><b>]</b> <b><span style='color:#54ffff;'># Removes fifth element</span></b>
x <b><span style='color:#54ffff;'># See modified object</span></b>
y <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>matrix</span></b><b>(</b><b><span style='color:#54ff54;'>data=</span></b><b><span style='color:#ff54ff;'>5</span></b><b>:</b><b><span style='color:#ff54ff;'>20</span></b>, <b><span style='color:#54ff54;'>nrow=</span></b><b><span style='color:#ff54ff;'>4</span></b>, <b><span style='color:#54ff54;'>ncol=</span></b><b><span style='color:#ff54ff;'>4</span></b><b>)</b> <b><span style='color:#54ffff;'># Creates a matrix</span></b>
<b><span style='color:#dede49;'>is.matrix</span></b><b>(</b>y<b>)</b> <b><span style='color:#54ffff;'># Is it matrix? Try is.&lt;TAB&gt;&lt;TAB&gt; TAB key shows available functions and objects starting by typed text</span></b>
y <b><span style='color:#54ffff;'># Prints the matrix</span></b>
y<b>[</b>,<b><span style='color:#ff54ff;'>2</span></b><b>]</b> <b><span style='color:#54ffff;'># Prints second column</span></b>
y<b>[</b><b><span style='color:#ff54ff;'>3</span></b>,<b>]</b> <b><span style='color:#54ffff;'># Prints third row</span></b>
y<b>[</b><b><span style='color:#ff54ff;'>4</span></b>,<b><span style='color:#ff54ff;'>3</span></b><b>]</b> <b><span style='color:#54ffff;'># Prints element from fourth row and third column</span></b>
x <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>y<b>[</b><b><span style='color:#ff54ff;'>2</span></b>,<b>]</b> <b><span style='color:#54ffff;'># Replaces &quot;x&quot; by second row of &quot;y&quot; (no warning) - R doesn't ask neither notifies when overwriting objects! Be careful!</span></b>
x <b><span style='color:#54ffff;'># See modified object</span></b>
<b><span style='color:#dede49;'>rm</span></b><b>(</b>x<b>)</b> <b><span style='color:#54ffff;'># Deletes x</span></b>
y<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>3</span></b><b>]</b> <b><span style='color:#54ffff;'># Prints first through third column of the matrix</span></b>
y<b>[</b><b><span style='color:#ff54ff;'>3</span></b>,<b>]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>rep</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>20</span></b>, <b><span style='color:#54ff54;'>each=</span></b><b><span style='color:#ff54ff;'>4</span></b><b>)</b> <b><span style='color:#54ffff;'># Replaces third line by value of 20</span></b>
y <b><span style='color:#54ffff;'># See modified object</span></b>
y<b>[</b>y<b>==</b><b><span style='color:#ff54ff;'>20</span></b><b>]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>10</span></b> <b><span style='color:#54ffff;'># If value of y's element is 20, replace it by 10</span></b>
y <b><span style='color:#54ffff;'># See modified object</span></b>
<b><span style='color:#dede49;'>summary</span></b><b>(</b>y<b>)</b> <b><span style='color:#54ffff;'># Basic statistics - according to columns</span></b>
<b><span style='color:#dede49;'>colnames</span></b><b>(</b>y<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;A&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;B&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;C&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;D&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Set column names - Objects and functions are without quotation marks; files, text with</span></b>
<b><span style='color:#dede49;'>colnames</span></b><b>(</b>y<b>)</b> <b><span style='color:#54ffff;'># Prints column names, use rownames() in same way</span></b>
y<b>[</b>,<b><span style='color:#ff54ff;'>&quot;C&quot;</span></b><b>]</b> <b><span style='color:#54ffff;'># Prints column C (R is case sensitive!)</span></b>
<b><span style='color:#dede49;'>t</span></b><b>(</b>y<b>)</b> <b><span style='color:#54ffff;'># Transposes the matrix</span></b>
y <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>as.data.frame</span></b><b>(</b>y<b>)</b> <b><span style='color:#54ffff;'># Turns into DF (see other functions as.*)</span></b>
y<b>[</b>y<b>==</b><b><span style='color:#ff54ff;'>17</span></b><b>]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> &quot;NA&quot;</span></b> <b><span style='color:#54ffff;'># Removes values of 17</span></b>
y <b><span style='color:#54ffff;'># See modified object</span></b>
y<b>$</b>B <b><span style='color:#54ffff;'># Gets variable B of data frame y ($ works similarly in S3 objects)</span></b>
<b><span style='color:#dede49;'>save</span></b><b>(</b><b><span style='color:#54ff54;'>list=</span></b><b><span style='color:#dede49;'>ls</span></b><b>()</b>, <b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;test.RData&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Saves all objects during the work</span></b>
<b><span style='color:#dede49;'>load</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;test.RData&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Loads saved R environment with all objects</span></b>
<b><span style='color:#54ffff;'># When loading saved project, you have to load again libraries and scripts (see further), data objects are restored</span></b>
<b><span style='color:#dede49;'>rm</span></b><b>(</b>y<b>)</b>

<b><span style='color:#b00000;'>## Packages and repositories</span></b>

<b><span style='color:#54ffff;'># Set repositories</span></b>
<b><span style='color:#54ffff;'>#  We will need extra repositories. For Bioconductor keep same version as is version of your R installation (Bioconductor 3.4 for R 3.3, see https://bioconductor.org/).</span></b>
<b><span style='color:#dede49;'>options</span></b><b>(</b><b><span style='color:#54ff54;'>repos=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;https://mirrors.nic.cz/R/&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;https://bioconductor.statistik.tu-dortmund.de/packages/3.6/bioc&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;https://bioconductor.statistik.tu-dortmund.de/packages/3.6/data/annotation&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;https://bioconductor.statistik.tu-dortmund.de/packages/3.6/data/experiment&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;https://r-forge.r-project.org/&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;https://rforge.net/&quot;</span></b><b>))</b>
<b><span style='color:#dede49;'>getOption</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;repos&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Shows actual repositories</span></b>
<b><span style='color:#dede49;'>options</span></b><b>()</b> <b><span style='color:#54ffff;'># Generic function to modify various settings</span></b>
?options <b><span style='color:#54ffff;'># Gives details</span></b>
<b><span style='color:#54ffff;'># Install packages</span></b>
<b><span style='color:#54ffff;'># Installation of multiple packages may sometimes fail - install then packages in smaller groups or one by one</span></b>
<b><span style='color:#dede49;'>install.packages</span></b><b>(</b><b><span style='color:#54ff54;'>pkgs=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;ade4&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;adegenet&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;adephylo&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;akima&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;animation&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;ape&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;BiocGenerics&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Biostrings&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;boot&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;brew&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;caTools&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;caper&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;cluster&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;clusterGeneration&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;coda&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;colorspace&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;combinat&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;corrplot&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;deSolve&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;deldir&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;devtools&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;digest&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;doParallel&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;dplyr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;expm&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;fastmatch&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;fields&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;foreach&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;foreign&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;gdata&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;geiger&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;ggplot2&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;git2r&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;gmodels&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;gplots&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;gtable&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;gtools&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;hierfstat&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;httr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;igraph&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;ips&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;IRanges&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;jsonlite&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KernSmooth&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;lattice&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;lazyeval&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;LearnBayes&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;magrittr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mapdata&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mapplots&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mapproj&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;maps&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;maptools&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;MASS&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Matrix&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;memoise&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;memuse&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mgcv&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mnormt&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;msm&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;muscle&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mvtnorm&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;ncbit&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;nlme&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;numDeriv&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;ParallelStructure&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;PBSmapping&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;pegas&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;permute&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;phangorn&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;philentropy&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;phylobase&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;phytools&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;picante&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;plotrix&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;plyr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;png&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;poppr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;RandomFields&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;RandomFieldsUtils&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Rcpp&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;reshape2&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;rgdal&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;rgeos&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;RgoogleMaps&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;rlang&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Rmpi&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;rncl&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;RNeXML&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;rworldmap&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;S4Vectors&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;scales&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;scatterplot3d&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;segmented&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;seqinr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;shapefiles&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;shiny&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;snow&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;sos&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;sp&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;spData&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;spam&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;spdep&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;StAMPP&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;stringr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;subplex&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;TeachingDemos&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;tibble&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;tidyr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;tripack&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;vcfR&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;vegan&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;viridisLite&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;whisker&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;withr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;XML&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;XVector&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>repos=</span></b><b><span style='color:#dede49;'>getOption</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;repos&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>dependencies=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
?install.packages <b><span style='color:#54ffff;'># See for more options</span></b>
<b><span style='color:#dede49;'>update.packages</span></b><b>(</b><b><span style='color:#54ff54;'>repos=</span></b><b><span style='color:#dede49;'>getOption</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;repos&quot;</span></b><b>))</b> <b><span style='color:#54ffff;'># Updates installed packages</span></b>

<b><span style='color:#54ffff;'># Install packages without setting the repositories</span></b>
<b><span style='color:#54ffff;'># If repositories are not set (for any reason), it is possible to install in several steps packages from main repository and from another sources</span></b>
<b><span style='color:#dede49;'>install.packages</span></b><b>(</b><b><span style='color:#54ff54;'>pkgs=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;ade4&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;adegenet&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;adephylo&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;akima&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;animation&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;ape&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;boot&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;brew&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;caTools&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;caper&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;cluster&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;clusterGeneration&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;coda&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;colorspace&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;combinat&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;corrplot&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;deSolve&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;deldir&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;devtools&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;digest&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;doParallel&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;dplyr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;expm&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;fastmatch&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;fields&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;foreach&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;foreign&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;gdata&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;geiger&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;ggplot2&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;git2r&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;gmodels&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;gplots&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;gtable&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;gtools&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;hierfstat&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;httr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;igraph&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;ips&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;jsonlite&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KernSmooth&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;lattice&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;lazyeval&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;LearnBayes&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;magrittr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mapdata&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mapplots&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mapproj&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;maps&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;maptools&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;MASS&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Matrix&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;memoise&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;memuse&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mgcv&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mnormt&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;msm&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mvtnorm&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;ncbit&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;nlme&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;numDeriv&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;PBSmapping&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;pegas&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;permute&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;phangorn&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;philentropy&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;phylobase&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;phytools&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;picante&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;plotrix&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;plyr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;png&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;poppr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;RandomFields&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;RandomFieldsUtils&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Rcpp&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;reshape2&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;rgdal&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;rgeos&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;RgoogleMaps&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;rlang&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Rmpi&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;rncl&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;RNeXML&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;rworldmap&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;scales&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;scatterplot3d&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;segmented&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;seqinr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;shapefiles&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;shiny&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;snow&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;sos&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;sp&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;spData&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;spam&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;spdep&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;StAMPP&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;stringr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;subplex&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;TeachingDemos&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;tibble&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;tidyr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;tripack&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;vcfR&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;vegan&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;viridisLite&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;whisker&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;withr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;XML&quot;</span></b><b>))</b>
<b><span style='color:#dede49;'>update.packages</span></b><b>(</b><b><span style='color:#54ff54;'>ask=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b> <b><span style='color:#54ffff;'># Update installed (CRAN by default) packages</span></b>

<b><span style='color:#54ffff;'># Install package phyloch not available in any repository</span></b>
<b><span style='color:#54ffff;'># If not done already, install required packages first</span></b>
<b><span style='color:#dede49;'>install.packages</span></b><b>(</b><b><span style='color:#54ff54;'>pkgs=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;ape&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;colorspace&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;XML&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>dependencies=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># It is possible to specify direct path (local or web URL) to package source</span></b>
<b><span style='color:#dede49;'>install.packages</span></b><b>(</b><b><span style='color:#54ff54;'>pkgs=</span></b><b><span style='color:#ff54ff;'>&quot;http://www.christophheibl.de/phyloch_1.5-3.tar.gz&quot;</span></b>, <b><span style='color:#54ff54;'>repos=</span></b><b><span style='color:#00ff00;'>NULL</span></b>, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;source&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># If above command fails on Windows, try</span></b>
<b><span style='color:#dede49;'>install.packages</span></b><b>(</b><b><span style='color:#54ff54;'>pkgs=</span></b><b><span style='color:#ff54ff;'>&quot;http://www.christophheibl.de/phyloch_1.5-3.zip&quot;</span></b>, <b><span style='color:#54ff54;'>repos=</span></b><b><span style='color:#00ff00;'>NULL</span></b><b>)</b>

<b><span style='color:#54ffff;'># Install package Geneland (since version 4 not available in CRAN anymore)</span></b>
<b><span style='color:#54ffff;'># It is possible to specify direct path (local or web URL) to package source</span></b>
<b><span style='color:#dede49;'>install.packages</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;https://www2.imm.dtu.dk/~gigu/Geneland/distrib/Geneland_4.0.8.tar.gz&quot;</span></b>, <b><span style='color:#54ff54;'>repos=</span></b><b><span style='color:#00ff00;'>NULL</span></b>, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;source&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># If above command fails on Windows, try</span></b>
<b><span style='color:#dede49;'>install.packages</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;https://www2.imm.dtu.dk/~gigu/Geneland/distrib/Geneland_4.0.8.zip&quot;</span></b>, <b><span style='color:#54ff54;'>repos=</span></b><b><span style='color:#00ff00;'>NULL</span></b><b>)</b>
<b><span style='color:#54ffff;'># Other packages used when using Geneland</span></b>
<b><span style='color:#54ffff;'># Needed is PBSmapping or mapproj for conversion of coordinates</span></b>
<b><span style='color:#54ffff;'># GUI uses for parallelisation snow and Rmpi</span></b>
<b><span style='color:#54ffff;'># RgoogleMaps (requires rgdal) can be used to plot Geneland output on top of Google map, maptools (requires rgeos and sp), shapefiles (requires foreign) and tripack on GIS layer</span></b>
<b><span style='color:#dede49;'>install.packages</span></b><b>(</b><b><span style='color:#54ff54;'>pkgs=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;PBSmapping&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;RgoogleMaps&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Rmpi&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;foreign&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;mapproj&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;maptools&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;rgdal&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;rgeos&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;shapefiles&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;snow&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;sp&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;tripack&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>dependencies=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>

<b><span style='color:#54ffff;'># We will load packages by library(package) one by one when needed and plugins...</span></b>

<b><span style='color:#54ffff;'># Standard installation</span></b>
<b><span style='color:#dede49;'>install.pacakges</span></b><b>(</b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;adegenet&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;poppr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;phytools&quot;</span></b><b>))</b>
<b><span style='color:#dede49;'>update.packages</span></b><b>()</b> <b><span style='color:#54ffff;'># Update packages</span></b>

<b><span style='color:#54ffff;'># Installation from custom repository</span></b>
<b><span style='color:#dede49;'>install.packages</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;ParallelStructure&quot;</span></b>, <b><span style='color:#54ff54;'>repos=</span></b><b><span style='color:#ff54ff;'>&quot;https://r-forge.r-project.org/&quot;</span></b><b>)</b>
?install.packages <b><span style='color:#54ffff;'># See help for details</span></b>

<b><span style='color:#b00000;'>## Bioconductor</span></b>
<b><span style='color:#54ffff;'># Bioconductor - if https fails, use http</span></b>
<b><span style='color:#dede49;'>source</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;https://bioconductor.org/biocLite.R&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Get help how to use it</span></b>
?biocLite
<b><span style='color:#54ffff;'># Install package(s)</span></b>
<b><span style='color:#dede49;'>biocLite</span></b><b>(</b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Biostrings&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;muscle&quot;</span></b><b>))</b>
<b><span style='color:#54ffff;'># Install Bioconductor packages used during the course</span></b>
<b><span style='color:#dede49;'>biocLite</span></b><b>(</b><b><span style='color:#54ff54;'>pkgs=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;BiocGenerics&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Biostrings&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;IRanges&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;muscle&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;S4Vectors&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;XVector&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>ask=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#dede49;'>biocLite</span></b><b>()</b> <b><span style='color:#54ffff;'># Update Bioconductor packages</span></b>
<b><span style='color:#54ffff;'># Upgrades installed Bioconductor packages to new R release (e.g. from 3.3 to 3.4)</span></b>
<b><span style='color:#dede49;'>biocLite</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;BiocUpgrade&quot;</span></b><b>)</b>

<b><span style='color:#b00000;'>## Libraries for population genetic analysis</span></b>

<b><span style='color:#54ffff;'># Load needed libraries</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>ape<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>ade4<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>adegenet<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>pegas<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>poppr<b>)</b>

<b><span style='color:#b00000;'>## Data</span></b>

<b><span style='color:#54ffff;'># R training data</span></b>
<b><span style='color:#54ffff;'># List data available in currently loaded packages</span></b>
<b><span style='color:#dede49;'>data</span></b><b>()</b>
<b><span style='color:#54ffff;'># List data available in all installed packages (can be very long)</span></b>
<b><span style='color:#dede49;'>data</span></b><b>(</b><b><span style='color:#54ff54;'>package=</span></b><b><span style='color:#dede49;'>rownames</span></b><b>(</b><b><span style='color:#dede49;'>installed.packages</span></b><b>()))</b>
<b><span style='color:#54ffff;'># Load selected data set</span></b>
<b><span style='color:#54ffff;'># For example phylogeny and species traits of shorebirds from package caper (we will use it much later)</span></b>
<b><span style='color:#dede49;'>data</span></b><b>(</b>shorebird, <b><span style='color:#54ff54;'>package=</span></b><b><span style='color:#ff54ff;'>&quot;caper&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Optional method (load respective library and then data)</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>caper<b>)</b> <b><span style='color:#54ffff;'># Library containing (among others) desired data</span></b>
<b><span style='color:#dede49;'>data</span></b><b>(</b>shorebird<b>)</b> <b><span style='color:#54ffff;'># Loading the data</span></b>
<b><span style='color:#54ffff;'># See content of the data set</span></b>
?shorebird <b><span style='color:#54ffff;'># Or</span></b>
?caper<b>::</b>shorebird
<b><span style='color:#54ffff;'># R training data we will use during the course</span></b>
?adegenet<b>::</b>microbov
?adegenet<b>::</b>rupica
?ape<b>::</b>carnivora
?caper<b>::</b>shorebird

<b><span style='color:#54ffff;'># Load training data</span></b>
hauss.loci <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.loci</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;https://soubory.trapa.cz/rcourse/haussknechtii_ssrs.txt&quot;</span></b>, <b><span style='color:#54ff54;'>header=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>loci.sep=</span></b><b><span style='color:#ff54ff;'>&quot;</span></b><b><span style='color:#ff54ff;'>\t</span></b><b><span style='color:#ff54ff;'>&quot;</span></b>, <b><span style='color:#54ff54;'>allele.sep=</span></b><b><span style='color:#ff54ff;'>&quot;/&quot;</span></b>, <b><span style='color:#54ff54;'>col.pop=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>col.loci=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>:</b><b><span style='color:#ff54ff;'>14</span></b>, <b><span style='color:#54ff54;'>row.names=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>
<b><span style='color:#54ffff;'># Data control</span></b>
hauss.loci
<b><span style='color:#dede49;'>print</span></b><b>(</b>hauss.loci, <b><span style='color:#54ff54;'>details=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>

<b><span style='color:#54ffff;'># Conversion of loci to genind - used for many analysis</span></b>
hauss.genind <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>loci2genind</span></b><b>(</b>hauss.loci<b>)</b>
<b><span style='color:#54ffff;'># See population names</span></b>
<b><span style='color:#dede49;'>pop</span></b><b>(</b>hauss.genind<b>)</b>
hauss.genind<b>$</b>pop

<b><span style='color:#54ffff;'># Read coordinates</span></b>
hauss.coord <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.csv</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;https://soubory.trapa.cz/rcourse/haussknechtii_coordinates.csv&quot;</span></b>, <b><span style='color:#54ff54;'>header=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;</span></b><b><span style='color:#ff54ff;'>\t</span></b><b><span style='color:#ff54ff;'>&quot;</span></b>, <b><span style='color:#54ff54;'>quote=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b>, <b><span style='color:#54ff54;'>dec=</span></b><b><span style='color:#ff54ff;'>&quot;.&quot;</span></b>, <b><span style='color:#54ff54;'>row.names=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>
hauss.coord

<b><span style='color:#54ffff;'># Add coordinates - note identification of slots within object</span></b>
hauss.genind<b>$</b>other<b>$</b>xy <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>hauss.coord
<b><span style='color:#54ffff;'># See result</span></b>
hauss.genind<b>$</b>other<b>$</b>xy
hauss.genind

<b><span style='color:#54ffff;'># Conversion to genpop - for population-level analysis</span></b>
hauss.genpop <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>genind2genpop</span></b><b>(</b>hauss.genind, <b><span style='color:#54ff54;'>process.other=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># See result</span></b>
hauss.genpop

<b><span style='color:#54ffff;'># Removes missing data - see ?missingno for types of dealing with them</span></b>
hauss.genind.cor <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>missingno</span></b><b>(</b><b><span style='color:#54ff54;'>pop=</span></b>hauss.genind, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;mean&quot;</span></b>, <b><span style='color:#54ff54;'>cutoff=</span></b><b><span style='color:#ff54ff;'>0.1</span></b>, <b><span style='color:#54ff54;'>quiet=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b> <b><span style='color:#54ffff;'># Use with caution! It modifies original data!</span></b>
?missingno <b><span style='color:#54ffff;'># See other options of handling missing data</span></b>
<b><span style='color:#54ffff;'># Convert corrected genind to loci</span></b>
hauss.loci.cor <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>genind2loci</span></b><b>(</b>hauss.genind.cor<b>)</b>

<b><span style='color:#54ffff;'># Writes loci file to the disk</span></b>
<b><span style='color:#dede49;'>write.loci</span></b><b>(</b>hauss.loci.cor, <b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;hauss.loci.cor.txt&quot;</span></b>, <b><span style='color:#54ff54;'>loci.sep=</span></b><b><span style='color:#ff54ff;'>&quot;</span></b><b><span style='color:#ff54ff;'>\t</span></b><b><span style='color:#ff54ff;'>&quot;</span></b>, <b><span style='color:#54ff54;'>allele.sep=</span></b><b><span style='color:#ff54ff;'>&quot;/&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># Read data sets from various software</span></b>
<b><span style='color:#dede49;'>read.genalex</span></b><b>()</b> <b><span style='color:#54ffff;'># poppr - reads *.csv file</span></b>
<b><span style='color:#dede49;'>read.fstat</span></b><b>()</b> <b><span style='color:#54ffff;'># adegenet - reads *.dat files, only haploid/diploid data</span></b>
<b><span style='color:#dede49;'>read.genetix</span></b><b>()</b> <b><span style='color:#54ffff;'># adegenet - reads *.gtx files, only haploid/diploid data</span></b>
<b><span style='color:#dede49;'>read.genepop</span></b><b>()</b> <b><span style='color:#54ffff;'># adegenet - reads *.gen files, only haploid/diploid data</span></b>
<b><span style='color:#dede49;'>read.structure</span></b><b>()</b> <b><span style='color:#54ffff;'># adegenet - reads *.str files, only haploid/diploid data</span></b>
<b><span style='color:#dede49;'>import2genind</span></b><b>()</b> <b><span style='color:#54ffff;'># adegenet - more automated version of above functions</span></b>

<b><span style='color:#54ffff;'># Import of triploid (polyploid) microsattelites</span></b>
tarax3n.table <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.table</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;https://soubory.trapa.cz/rcourse/tarax3n.txt&quot;</span></b>, <b><span style='color:#54ff54;'>header=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;</span></b><b><span style='color:#ff54ff;'>\t</span></b><b><span style='color:#ff54ff;'>&quot;</span></b>, <b><span style='color:#54ff54;'>quote=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b>, <b><span style='color:#54ff54;'>row.names=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>
<b><span style='color:#54ffff;'># Check the data</span></b>
tarax3n.table
<b><span style='color:#dede49;'>class</span></b><b>(</b>tarax3n.table<b>)</b>
<b><span style='color:#dede49;'>dim</span></b><b>(</b>tarax3n.table<b>)</b>

<b><span style='color:#54ffff;'># Convert to genind</span></b>
tarax3n.genind <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>df2genind</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>tarax3n.table<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>6</span></b><b>]</b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;/&quot;</span></b>, <b><span style='color:#54ff54;'>ncode=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>pop=</span></b>tarax3n.table<b>[[</b><b><span style='color:#ff54ff;'>&quot;pop&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>ploidy=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;codom&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># See resulting genind object</span></b>
tarax3n.genind
<b><span style='color:#dede49;'>summary</span></b><b>(</b>tarax3n.genind<b>)</b>

<b><span style='color:#54ffff;'># Import of AFLP data</span></b>
amara.aflp <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.table</span></b><b>(</b><b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;https://soubory.trapa.cz/rcourse/amara_aflp.txt&quot;</span></b>, <b><span style='color:#54ff54;'>header=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;</span></b><b><span style='color:#ff54ff;'>\t</span></b><b><span style='color:#ff54ff;'>&quot;</span></b>, <b><span style='color:#54ff54;'>quote=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b><b>)</b>
amara.aflp
<b><span style='color:#dede49;'>dim</span></b><b>(</b>amara.aflp<b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>amara.aflp<b>)</b> <b><span style='color:#54ffff;'># Must be matrix or data frame</span></b>
<b><span style='color:#54ffff;'># Populations - just one column with population names for all inds</span></b>
amara.pop <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.table</span></b><b>(</b><b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;https://soubory.trapa.cz/rcourse/amara_pop.txt&quot;</span></b>, <b><span style='color:#54ff54;'>header=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;</span></b><b><span style='color:#ff54ff;'>\t</span></b><b><span style='color:#ff54ff;'>&quot;</span></b>, <b><span style='color:#54ff54;'>quote=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b><b>)</b>
amara.pop
<b><span style='color:#dede49;'>dim</span></b><b>(</b>amara.pop<b>)</b>
<b><span style='color:#54ffff;'># Create genind object - ind.names and loc.names are taken from X</span></b>
aflp.genind <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>df2genind</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>amara.aflp, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b>, <b><span style='color:#54ff54;'>ind.names=</span></b><b><span style='color:#00ff00;'>NULL</span></b>, <b><span style='color:#54ff54;'>loc.names=</span></b><b><span style='color:#00ff00;'>NULL</span></b>, <b><span style='color:#54ff54;'>pop=</span></b>amara.pop<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>]</b>, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;PA&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>indNames</span></b><b>(</b>aflp.genind<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>amara.aflp<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>]</b> <b><span style='color:#54ffff;'># Add individual names</span></b>
aflp.genind
<b><span style='color:#dede49;'>summary</span></b><b>(</b>amara.aflp<b>)</b>

<b><span style='color:#54ffff;'># Some useful functions for data manipulations</span></b>
<b><span style='color:#dede49;'>genind2df</span></b><b>()</b> <b><span style='color:#54ffff;'># adegenet - export into data frame</span></b>
<b><span style='color:#dede49;'>genind2genalex</span></b><b>()</b> <b><span style='color:#54ffff;'># poppr - export for genalex</span></b>
<b><span style='color:#dede49;'>splitcombine</span></b><b>()</b> <b><span style='color:#54ffff;'># poppr - edits population hierarchy</span></b>
<b><span style='color:#dede49;'>popsub</span></b><b>()</b> <b><span style='color:#54ffff;'># poppr - extracts only selected population(s)</span></b>
<b><span style='color:#dede49;'>clonecorrect</span></b><b>()</b> <b><span style='color:#54ffff;'># poppr - corrects for clones</span></b>
<b><span style='color:#dede49;'>informloci</span></b><b>()</b> <b><span style='color:#54ffff;'># poppr - removes uninformative loci</span></b>
<b><span style='color:#dede49;'>seppop</span></b><b>()</b> <b><span style='color:#54ffff;'># adegenet - separates populations from genind or genlight object</span></b>
<b><span style='color:#dede49;'>seploc</span></b><b>()</b> <b><span style='color:#54ffff;'># adegenet - splits genind, genpop or genlight by markers</span></b>
<b><span style='color:#dede49;'>alleles2loci</span></b><b>()</b> <b><span style='color:#54ffff;'># pegas - transforms a matrix of alleles into &quot;loci&quot;</span></b>
<b><span style='color:#54ffff;'># seppop and seploc return lists of objects - for further analysis</span></b>
<b><span style='color:#54ffff;'># read manual pages (?...) of the functions before usage</span></b>

<b><span style='color:#54ffff;'># Reading FASTA (reads also another formats, see ?read.dna), sequences of flu viruses from various years from USA (Adegenet toy data)</span></b>
usflu.dna <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.dna</span></b><b>(</b><b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;http://adegenet.r-forge.r-project.org/files/usflu.fasta&quot;</span></b>, <b><span style='color:#54ff54;'>format=</span></b><b><span style='color:#ff54ff;'>&quot;fasta&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Check the object</span></b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>usflu.dna<b>)</b>
usflu.dna
<b><span style='color:#54ffff;'># Another possibility (only for FASTA alignments, same result):</span></b>
usflu.dna2 <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>fasta2DNAbin</span></b><b>(</b><b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;http://adegenet.r-forge.r-project.org/files/usflu.fasta&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Normally keeps only SNP - see ?fasta2DNAbin</span></b>
<b><span style='color:#54ffff;'># Check the object</span></b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>usflu.dna2<b>)</b>
usflu.dna2
<b><span style='color:#dede49;'>as.character</span></b><b>(</b>usflu.dna2<b>)[</b><b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>5</span></b>,<b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>10</span></b><b>]</b>
<b><span style='color:#dede49;'>dim</span></b><b>(</b>usflu.dna2<b>)</b> <b><span style='color:#54ffff;'># Does it have correct size?</span></b>
<b><span style='color:#54ffff;'># Read anotations</span></b>
usflu.annot <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.csv</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;http://adegenet.r-forge.r-project.org/files/usflu.annot.csv&quot;</span></b>, <b><span style='color:#54ff54;'>header=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>row.names=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>
<b><span style='color:#dede49;'>head</span></b><b>(</b>usflu.annot<b>)</b>
<b><span style='color:#54ffff;'># Convert DNAbin to genind - only polymorphic loci are retained</span></b>
usflu.genind <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>DNAbin2genind</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.dna, <b><span style='color:#54ff54;'>pop=</span></b>usflu.annot<b>[[</b><b><span style='color:#ff54ff;'>&quot;year&quot;</span></b><b>]])</b>
usflu.genind <b><span style='color:#54ffff;'># Check it</span></b>
<b><span style='color:#54ffff;'># read.fasta() from seqinr package reads DNA or AA in FASTA format - returns a list (DNAbin is for us now better choice)</span></b>
usflu.dna3 <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>seqinr<b>::</b><b><span style='color:#dede49;'>read.fasta</span></b><b>(</b><b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;http://adegenet.r-forge.r-project.org/files/usflu.fasta&quot;</span></b>, <b><span style='color:#54ff54;'>seqtype=</span></b><b><span style='color:#ff54ff;'>&quot;DNA&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>usflu.dna3<b>)</b>
<b><span style='color:#dede49;'>length</span></b><b>(</b>usflu.dna3<b>)</b> <b><span style='color:#54ffff;'># How many sequences we have in the list</span></b>
usflu.dna3 <b><span style='color:#54ffff;'># Check it</span></b>
<b><span style='color:#54ffff;'># Convert into DNAbin class (technically, DNAbin is a list)</span></b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>usflu.dna3<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> &quot;DNAbin&quot;</span></b>
<b><span style='color:#54ffff;'># Read sequence data in NEXUS</span></b>
<b><span style='color:#dede49;'>read.nexus.data</span></b><b>(</b><b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;sequences.nex&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># Importing DNA sequences from GeneBank - according to sequence ID, data from http://www.ncbi.nlm.nih.gov/popset/608602125</span></b>
meles.dna <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.GenBank</span></b><b>(</b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;KJ161355.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161354.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161353.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161352.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161351.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161350.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161349.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161348.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161347.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161346.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161345.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161344.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161343.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161342.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161341.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161340.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161339.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161338.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161337.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161336.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161335.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161334.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161333.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161332.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161331.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161330.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161329.1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;KJ161328.1&quot;</span></b><b>))</b>
meles.dna
<b><span style='color:#dede49;'>class</span></b><b>(</b>meles.dna<b>)</b>
<b><span style='color:#54ffff;'># Converts DNAbin to genind - extracts SNP - for large data sets can be computationally very intensive</span></b>
meles.genind <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>DNAbin2genind</span></b><b>(</b>meles.dna<b>)</b>
meles.genind <b><span style='color:#54ffff;'># Check it</span></b>

<b><span style='color:#54ffff;'># Query on-line sequence databases</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>seqinr<b>)</b>
<b><span style='color:#dede49;'>choosebank</span></b><b>()</b> <b><span style='color:#54ffff;'># Genetic banks available for seqinr</span></b>
<b><span style='color:#dede49;'>choosebank</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;embl&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Choose some bank</span></b>
<b><span style='color:#54ffff;'># Query selected database - there are much possibilities</span></b>
?query <b><span style='color:#54ffff;'># See how to construct the query</span></b>
nothofagus <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>query</span></b><b>(</b><b><span style='color:#54ff54;'>listname=</span></b><b><span style='color:#ff54ff;'>&quot;nothofagus&quot;</span></b>, <b><span style='color:#54ff54;'>query=</span></b><b><span style='color:#ff54ff;'>&quot;SP=Nothofagus AND K=rbcl&quot;</span></b>, <b><span style='color:#54ff54;'>verbose=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># See the sequences information</span></b>
nothofagus<b>$</b>req
<b><span style='color:#54ffff;'># Get the sequences as a list</span></b>
nothofagus.sequences <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>getSequence</span></b><b>(</b>nothofagus<b>$</b>req<b>)</b>
<b><span style='color:#54ffff;'># See sequences</span></b>
nothofagus.sequences
<b><span style='color:#54ffff;'># Get annotations</span></b>
nothofagus.annot <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>getAnnot</span></b><b>(</b>nothofagus<b>$</b>req<b>)</b>
nothofagus.annot
<b><span style='color:#54ffff;'># Close the bank when work is over</span></b>
<b><span style='color:#dede49;'>closebank</span></b><b>()</b>
<b><span style='color:#54ffff;'># Convert sequences from a list to DNAbin (functions as.DNAbin*)</span></b>
nothofagus.dna <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>as.DNAbin.list</span></b><b>(</b>nothofagus.sequences<b>)</b>
nothofagus.dna <b><span style='color:#54ffff;'># See it</span></b>

<b><span style='color:#54ffff;'># Importing SNP</span></b>
<b><span style='color:#dede49;'>read.PLINK</span></b><b>(</b><b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;PLINKfile&quot;</span></b>, ...<b>)</b>
?readPLINK <b><span style='color:#54ffff;'># See it for available options</span></b>
<b><span style='color:#54ffff;'># Extract SNP from FASTA alignment</span></b>
usflu.genlight <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>fasta2genlight</span></b><b>(</b><b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;http://adegenet.r-forge.r-project.org/files/usflu.fasta&quot;</span></b>, <b><span style='color:#54ff54;'>quiet=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>saveNbAlleles=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># If it crashes (on Windows), try add parameter &quot;parallel=FALSE&quot;</span></b>
<b><span style='color:#54ffff;'># Function has several options to speed up reading</span></b>
?fasta2genlight

<b><span style='color:#54ffff;'># Read SNP in Adegent format - check Adegenet tutorial first</span></b>
?read.snp

<b><span style='color:#54ffff;'># Checking SNPs</span></b>
<b><span style='color:#54ffff;'># Position of polymorphism within alignment - snpposi.plot() requires input data in form of matrix</span></b>
<b><span style='color:#dede49;'>snpposi.plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>meles.dna<b>)</b>, <b><span style='color:#54ff54;'>codon=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Position of polymorphism within alignment - differentiating codons</span></b>
<b><span style='color:#dede49;'>snpposi.plot</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>meles.dna<b>))</b>
<b><span style='color:#54ffff;'># When converting to genind object, only polymorphic loci are kept - threshold for polymorphism can be arbitrary (polyThres=...)</span></b>
meles.genind <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>DNAbin2genind</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.dna, <b><span style='color:#54ff54;'>polyThres=</span></b><b><span style='color:#ff54ff;'>0.01</span></b><b>)</b>
meles.genind <b><span style='color:#54ffff;'># See it</span></b>
<b><span style='color:#54ffff;'># Test is distribution of SNP is random (1000 permutations)</span></b>
<b><span style='color:#dede49;'>snpposi.test</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>meles.dna<b>))</b>

<b><span style='color:#54ffff;'># Check sequences</span></b>
<b><span style='color:#54ffff;'># Nucleotide diversity</span></b>
pegas<b>::</b><b><span style='color:#dede49;'>nuc.div</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.dna<b>)</b>
<b><span style='color:#54ffff;'># Base frequencies</span></b>
ape<b>::</b><b><span style='color:#dede49;'>base.freq</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.dna<b>)</b>
<b><span style='color:#54ffff;'># GC content</span></b>
ape<b>::</b><b><span style='color:#dede49;'>GC.content</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.dna<b>)</b>
<b><span style='color:#54ffff;'># Number of times any dimer/trimer/etc oligomers occur in a sequence</span></b>
seqinr<b>::</b><b><span style='color:#dede49;'>count</span></b><b>(</b><b><span style='color:#54ff54;'>seq=</span></b><b><span style='color:#dede49;'>as.character.DNAbin</span></b><b>(</b>meles.dna<b>[[</b><b><span style='color:#ff54ff;'>&quot;KJ161328.1&quot;</span></b><b>]])</b>, <b><span style='color:#54ff54;'>wordsize=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#54ffff;'># View sequences - all must be of the same length</span></b>
<b><span style='color:#dede49;'>image</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.dna, <b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;a&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;t&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;c&quot;</span></b> ,<b><span style='color:#ff54ff;'>&quot;g&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;n&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>rainbow</span></b><b>(</b><b><span style='color:#ff54ff;'>5</span></b><b>))</b>
<b><span style='color:#54ffff;'># Function &quot;image&quot; requires as input matrix, so that sequences must be of same length</span></b>
<b><span style='color:#dede49;'>image</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>meles.dna<b>)</b>, <b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;a&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;t&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;c&quot;</span></b> ,<b><span style='color:#ff54ff;'>&quot;g&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;n&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>rainbow</span></b><b>(</b><b><span style='color:#ff54ff;'>5</span></b><b>))</b>
<b><span style='color:#54ffff;'># Direct function to display the sequences</span></b>
<b><span style='color:#dede49;'>image.DNAbin</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.dna<b>)</b>
<b><span style='color:#dede49;'>image.DNAbin</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>meles.dna<b>))</b>

<b><span style='color:#54ffff;'># VCF</span></b>
<b><span style='color:#54ffff;'># Required library</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>vcfR<b>)</b>
<b><span style='color:#54ffff;'># Read input file</span></b>
<b><span style='color:#54ffff;'># Download input file from https://soubory.trapa.cz/rcourse/arabidopsis.vcf.gz (read.vcfR does not work with network addresses yet)</span></b>
arabidopsis.vcf <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.vcfR</span></b><b>(</b><b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#dede49;'>file.choose</span></b><b>())</b> <b><span style='color:#54ffff;'># Pick up downloaded file 'arabidopsis.vcf.gz' from the disc</span></b>
<b><span style='color:#54ffff;'># It returns object of class vcfR-class</span></b>
?read.vcfR <b><span style='color:#54ffff;'># See more import options</span></b>
<b><span style='color:#54ffff;'># Another option from package pegas returns list of objects loci and data.frame</span></b>
?pegas<b>::</b>read.vcf

<b><span style='color:#54ffff;'># Explore VCF</span></b>
arabidopsis.vcf
<b><span style='color:#dede49;'>head</span></b><b>(</b>arabidopsis.vcf<b>)</b>
arabidopsis.vcf<b>@</b>fix<b>[</b><b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>10</span></b>,<b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>5</span></b><b>]</b>
<b><span style='color:#54ffff;'># Get information about depth of coverage (DP)</span></b>
<b><span style='color:#54ffff;'># Se description of DP slot</span></b>
<b><span style='color:#dede49;'>strwrap</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>grep</span></b><b>(</b><b><span style='color:#54ff54;'>pattern=</span></b><b><span style='color:#ff54ff;'>&quot;ID=DP,&quot;</span></b>, <b><span style='color:#54ff54;'>x=</span></b>arabidopsis.vcf<b>@</b>meta, <b><span style='color:#54ff54;'>value=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>))</b>
<b><span style='color:#54ffff;'># Extract the DP</span></b>
arabidopsis.vcf.dp <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>extract.gt</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>arabidopsis.vcf, <b><span style='color:#54ff54;'>element=</span></b><b><span style='color:#ff54ff;'>&quot;DP&quot;</span></b>, <b><span style='color:#54ff54;'>as.numeric=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># See it</span></b>
<b><span style='color:#dede49;'>dim</span></b><b>(</b>arabidopsis.vcf.dp<b>)</b>
<b><span style='color:#dede49;'>head</span></b><b>(</b>arabidopsis.vcf.dp<b>)</b>
<b><span style='color:#54ffff;'># Boxplot of DP</span></b>
<b><span style='color:#dede49;'>boxplot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>arabidopsis.vcf.dp, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;#808080&quot;</span></b>, <b><span style='color:#54ff54;'>ylab=</span></b><b><span style='color:#ff54ff;'>&quot;Depth of coverage&quot;</span></b>, <b><span style='color:#54ff54;'>las=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;DP per specimen&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>abline</span></b><b>(</b><b><span style='color:#54ff54;'>h=</span></b><b><span style='color:#dede49;'>seq</span></b><b>(</b><b><span style='color:#54ff54;'>from=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>to=</span></b><b><span style='color:#ff54ff;'>90</span></b>, <b><span style='color:#54ff54;'>by=</span></b><b><span style='color:#ff54ff;'>10</span></b><b>)</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;#b3b3b3&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Bar plot of mean DP</span></b>
<b><span style='color:#dede49;'>barplot</span></b><b>(</b><b><span style='color:#dede49;'>apply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>arabidopsis.vcf.dp, <b><span style='color:#54ff54;'>MARGIN=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>FUN=</span></b>mean, <b><span style='color:#54ff54;'>na.rm=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>, <b><span style='color:#54ff54;'>las=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Mean DP per specimen&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>abline</span></b><b>(</b><b><span style='color:#54ff54;'>h=</span></b><b><span style='color:#dede49;'>seq</span></b><b>(</b><b><span style='color:#54ff54;'>from=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>to=</span></b><b><span style='color:#ff54ff;'>60</span></b>, <b><span style='color:#54ff54;'>by=</span></b><b><span style='color:#ff54ff;'>10</span></b><b>)</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;#b3b3b3&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Heatmap of DB (subset)</span></b>
<b><span style='color:#dede49;'>heatmap.bp</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>arabidopsis.vcf.dp<b>[</b><b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>100</span></b>,<b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>100</span></b><b>]</b>, <b><span style='color:#54ff54;'>col.ramp=</span></b><b><span style='color:#dede49;'>rainbow</span></b><b>(</b><b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>100</span></b>, <b><span style='color:#54ff54;'>start=</span></b><b><span style='color:#ff54ff;'>0.1</span></b><b>))</b> <b><span style='color:#54ffff;'># Subset - only first 100 loci and individuals</span></b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;DP per specimens and loci&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># Convert VCF into various objects for later processing</span></b>
<b><span style='color:#54ffff;'># Genind</span></b>
arabidopsis.genind <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>vcfR2genind</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>arabidopsis.vcf<b>)</b>
<b><span style='color:#54ffff;'># Check it</span></b>
arabidopsis.genind
<b><span style='color:#dede49;'>nInd</span></b><b>(</b>arabidopsis.genind<b>)</b>
<b><span style='color:#dede49;'>indNames</span></b><b>(</b>arabidopsis.genind<b>)</b>
<b><span style='color:#dede49;'>nLoc</span></b><b>(</b>arabidopsis.genind<b>)</b>
<b><span style='color:#dede49;'>locNames</span></b><b>(</b>arabidopsis.genind<b>)</b>
<b><span style='color:#54ffff;'># Genlight (suitable for huge data, not required now)</span></b>
arabidopsis.genlight <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>vcfR2genlight</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>arabidopsis.vcf, <b><span style='color:#54ff54;'>n.cores=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b> <b><span style='color:#54ffff;'># On Linux/Mac and with large data use higher n.cores</span></b>
<b><span style='color:#54ffff;'># Check it</span></b>
arabidopsis.genlight
<b><span style='color:#54ffff;'># Loci</span></b>
arabidopsis.loci <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>vcfR2loci</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>arabidopsis.vcf<b>)</b>
<b><span style='color:#54ffff;'># Check it</span></b>
arabidopsis.loci
<b><span style='color:#dede49;'>print</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>arabidopsis.loci, <b><span style='color:#54ff54;'>details=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># DNAbin</span></b>
?vcfR2DNAbin <b><span style='color:#54ffff;'># There are various options how to process variants in VCF</span></b>
arabidopsis.dnabin <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>vcfR2DNAbin</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>arabidopsis.vcf, <b><span style='color:#54ff54;'>consensus=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>extract.haps=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>unphased_as_NA=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Check it</span></b>
arabidopsis.dnabin
<b><span style='color:#dede49;'>dim</span></b><b>(</b>arabidopsis.dnabin<b>)</b>
<b><span style='color:#dede49;'>as.character.DNAbin</span></b><b>(</b>arabidopsis.dnabin<b>[</b><b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>15</span></b>,<b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>12</span></b><b>])</b>
<b><span style='color:#dede49;'>image.DNAbin</span></b><b>(</b>arabidopsis.dnabin<b>)</b>
<b><span style='color:#dede49;'>snpposi.plot.DNAbin</span></b><b>(</b>arabidopsis.dnabin<b>)</b>
<b><span style='color:#dede49;'>snpposi.test.DNAbin</span></b><b>(</b>arabidopsis.dnabin<b>)</b>

<b><span style='color:#54ffff;'># Remove non-biallelic loci</span></b>
<b><span style='color:#54ffff;'># Create vector of loci to keep</span></b>
arabidopsis.genind.keep <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>nAll</span></b><b>(</b>arabidopsis.genind<b>)</b> <b>&lt;</b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>3</span></b>
<b><span style='color:#54ffff;'># See the vector of loci</span></b>
arabidopsis.genind.keep
<b><span style='color:#54ffff;'># Keep only passing loci</span></b>
arabidopsis.genind.bial <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>arabidopsis.genind<b>[</b>loc<b><span style='color:#336366;'>=</span></b>arabidopsis.genind.keep<b>]</b>
<b><span style='color:#54ffff;'># See result</span></b>
arabidopsis.genind.bial
<b><span style='color:#54ffff;'># Filtration of missing data</span></b>
?poppr<b>::</b>missingno <b><span style='color:#54ffff;'># Various options...</span></b>

<b><span style='color:#b00000;'>## Exporting data</span></b>
<b><span style='color:#54ffff;'># Convert genind into DF using genind2df()</span></b>
hauss.df <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>genind2df</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind, <b><span style='color:#54ff54;'>pop=</span></b><b><span style='color:#00ff00;'>NULL</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;/&quot;</span></b>, <b><span style='color:#54ff54;'>usepop=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>oneColPerAll=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Save microsatellites to disk - check settings of write.table</span></b>
<b><span style='color:#dede49;'>write.table</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.df, <b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;haussdata.txt&quot;</span></b>, <b><span style='color:#54ff54;'>quote=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;</span></b><b><span style='color:#ff54ff;'>\t</span></b><b><span style='color:#ff54ff;'>&quot;</span></b>, <b><span style='color:#54ff54;'>na=</span></b><b><span style='color:#ff54ff;'>&quot;NA&quot;</span></b>, <b><span style='color:#54ff54;'>dec=</span></b><b><span style='color:#ff54ff;'>&quot;.&quot;</span></b>, <b><span style='color:#54ff54;'>row.names=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>col.names=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Export of DNA sequences into FASTA format</span></b>
<b><span style='color:#dede49;'>write.dna</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.dna, <b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;usflu.fasta&quot;</span></b>, <b><span style='color:#54ff54;'>format=</span></b><b><span style='color:#ff54ff;'>&quot;fasta&quot;</span></b>, <b><span style='color:#54ff54;'>append=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>nbcol=</span></b><b><span style='color:#ff54ff;'>6</span></b><b>)</b>
seqinr<b>::</b><b><span style='color:#dede49;'>write.fasta</span></b><b>(</b><b><span style='color:#54ff54;'>sequences=</span></b>meles.dna, <b><span style='color:#54ff54;'>names=</span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>meles.dna<b>)</b>, <b><span style='color:#54ff54;'>file.out=</span></b><b><span style='color:#ff54ff;'>&quot;meles.fasta&quot;</span></b>, <b><span style='color:#54ff54;'>open=</span></b><b><span style='color:#ff54ff;'>&quot;w&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Export DNA sequnces as NEXUS</span></b>
<b><span style='color:#dede49;'>write.nexus.data</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.dna, <b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;meles.nexus&quot;</span></b>, <b><span style='color:#54ff54;'>format=</span></b><b><span style='color:#ff54ff;'>&quot;dna&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Export VCF</span></b>
<b><span style='color:#dede49;'>write.vcf</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>arabidopsis.vcf, <b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;arabidopsis.vcf.gz&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Export trees (objects of class phylo)</span></b>
<b><span style='color:#dede49;'>write.tree</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>hauss.nj.bruvo, <b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;haussknechtii.nwk&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Writes tree(s) in NEWICK format</span></b>
<b><span style='color:#dede49;'>write.nexus</span></b><b>(</b>hauss.nj.bruvo, <b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;haussknechtii.nexus&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Writes tree(s) in NEXUS format</span></b>

<b><span style='color:#b00000;'>## Descriptive statistics</span></b>

<b><span style='color:#54ffff;'># Get summary - names and sizes of populations, heterozygosity, some info about loci</span></b>
hauss.summ <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>summary</span></b><b>(</b>hauss.genind<b>)</b>
hauss.summ
<b><span style='color:#54ffff;'># Plot expected vs. observed heterozygosity - it looks like big difference</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.summ<b>$</b>Hexp, <b><span style='color:#54ff54;'>y=</span></b>hauss.summ<b>$</b>Hobs, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Observed vs expected heterozygosity&quot;</span></b>, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#ff54ff;'>&quot;Expected heterozygosity&quot;</span></b>, <b><span style='color:#54ff54;'>ylab=</span></b><b><span style='color:#ff54ff;'>&quot;Observed heterozygosity&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>abline</span></b><b>(</b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Bartlett's K-squared test of difference between observed and expected heterozygosity - not significant</span></b>
<b><span style='color:#dede49;'>bartlett.test</span></b><b>(</b><b><span style='color:#dede49;'>list</span></b><b>(</b>hauss.summ<b>$</b>Hexp, hauss.summ<b>$</b>Hobs<b>))</b>
<b><span style='color:#54ffff;'># T-test of difference between observed and expected heterozygosity - strongly significant</span></b>
<b><span style='color:#dede49;'>t.test</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.summ<b>$</b>Hexp, <b><span style='color:#54ff54;'>y=</span></b>hauss.summ<b>$</b>Hobs, <b><span style='color:#54ff54;'>paired=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>var.equal=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Create pane with some information</span></b>
<b><span style='color:#dede49;'>par</span></b><b>(</b><b><span style='color:#54ff54;'>mfrow=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#ff54ff;'>2</span></b><b>))</b> <b><span style='color:#54ffff;'># Divide graphical devices into 4 smaller spaces</span></b>
<b><span style='color:#54ffff;'># Plot alleles number vs. population sizes</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.summ<b>$</b>n.by.pop, <b><span style='color:#54ff54;'>y=</span></b>hauss.summ<b>$</b>pop.nall, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#ff54ff;'>&quot;Populations sample size&quot;</span></b>, <b><span style='color:#54ff54;'>ylab=</span></b><b><span style='color:#ff54ff;'>&quot;Number of alleles&quot;</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Alleles numbers and sample sizes&quot;</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>20</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add text description to the point</span></b>
<b><span style='color:#dede49;'>text</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.summ<b>$</b>n.by.pop, <b><span style='color:#54ff54;'>y=</span></b>hauss.summ<b>$</b>pop.nall, <b><span style='color:#54ff54;'>lab=</span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>hauss.summ<b>$</b>n.by.pop<b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#54ffff;'># Barplots of various data</span></b>
<b><span style='color:#dede49;'>barplot</span></b><b>(</b><b><span style='color:#54ff54;'>height=</span></b>hauss.summ<b>$</b>loc.n.all, <b><span style='color:#54ff54;'>ylab=</span></b><b><span style='color:#ff54ff;'>&quot;Number of alleles&quot;</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Number of alleles per locus&quot;</span></b>, <b><span style='color:#54ff54;'>las=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#dede49;'>barplot</span></b><b>(</b><b><span style='color:#54ff54;'>height=</span></b>hauss.summ<b>$</b>Hexp<b>-</b>hauss.summ<b>$</b>Hobs, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Heterozygosity: expected-observed&quot;</span></b>, <b><span style='color:#54ff54;'>ylab=</span></b><b><span style='color:#ff54ff;'>&quot;Hexp - Hobs&quot;</span></b>, <b><span style='color:#54ff54;'>las=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#dede49;'>barplot</span></b><b>(</b><b><span style='color:#54ff54;'>height=</span></b>hauss.summ<b>[[</b><b><span style='color:#ff54ff;'>&quot;n.by.pop&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Sample sizes per population&quot;</span></b>, <b><span style='color:#54ff54;'>ylab=</span></b><b><span style='color:#ff54ff;'>&quot;Number of genotypes&quot;</span></b>, <b><span style='color:#54ff54;'>las=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#dede49;'>dev.off</span></b><b>()</b> <b><span style='color:#54ffff;'># Closes graphical device - otherwise following graphs would still be divided into 4 parts</span></b>

<b><span style='color:#54ffff;'># poppr returns various population statistics for populations</span></b>
?poppr <b><span style='color:#54ffff;'># See details</span></b>
<b><span style='color:#dede49;'>poppr</span></b><b>(</b><b><span style='color:#54ff54;'>dat=</span></b>hauss.genind, <b><span style='color:#54ff54;'>total=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>sample=</span></b><b><span style='color:#ff54ff;'>1000</span></b>, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>4</span></b>, <b><span style='color:#54ff54;'>missing=</span></b><b><span style='color:#ff54ff;'>&quot;geno&quot;</span></b>, <b><span style='color:#54ff54;'>cutoff=</span></b><b><span style='color:#ff54ff;'>0.15</span></b>, <b><span style='color:#54ff54;'>quiet=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>clonecorrect=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>index=</span></b><b><span style='color:#ff54ff;'>&quot;rbarD&quot;</span></b>, <b><span style='color:#54ff54;'>minsamp=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>legend=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Algorithms and equations utilized in poppr</span></b>
<b><span style='color:#dede49;'>vignette</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;algo&quot;</span></b>, <b><span style='color:#54ff54;'>package=</span></b><b><span style='color:#ff54ff;'>&quot;poppr&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># Departure from HWE</span></b>
<b><span style='color:#54ffff;'># According to loci</span></b>
hauss.hwe.test <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>hw.test</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.loci, <b><span style='color:#54ff54;'>B=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>)</b>
hauss.hwe.test
<b><span style='color:#54ffff;'># According to populations</span></b>
<b><span style='color:#54ffff;'># Separate genind object into list of genind objects for individual populations</span></b>
hauss.pops <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>seppop</span></b><b>(</b>hauss.genind<b>)</b>
hauss.pops
<b><span style='color:#54ffff;'># Convert genind back to loci (list of loci objects according to populations)</span></b>
hauss.pops.loci <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>hauss.pops, <b><span style='color:#54ff54;'>FUN=</span></b>genind2loci<b>)</b>
<b><span style='color:#54ffff;'># Calculate the results per populations</span></b>
<b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>hauss.pops.loci, <b><span style='color:#54ff54;'>FUN=</span></b>hw.test, <b><span style='color:#54ff54;'>B=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>)</b>

<b><span style='color:#54ffff;'># FST</span></b>
<b><span style='color:#54ffff;'># Fit, Fst and Fis for each locus</span></b>
<b><span style='color:#54ffff;'># For Fst, fstat and theta.msat the loci object must contain population column</span></b>
<b><span style='color:#dede49;'>Fst</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.loci, <b><span style='color:#54ff54;'>pop=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>
<b><span style='color:#54ffff;'># multilocus estimators of variance components and F-statistics, alternative to Fst</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>hierfstat<b>)</b>
<b><span style='color:#dede49;'>fstat</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind, <b><span style='color:#54ff54;'>pop=</span></b><b><span style='color:#00ff00;'>NULL</span></b>, <b><span style='color:#54ff54;'>fstonly=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Nei's pairwise Fst between all pairs of populations. Difference in res.type=&quot;dist&quot;/&quot;matrix&quot; is only in format of output</span></b>
<b><span style='color:#dede49;'>pairwise.fst</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind, <b><span style='color:#54ff54;'>pop=</span></b><b><span style='color:#00ff00;'>NULL</span></b>, <b><span style='color:#54ff54;'>res.type=</span></b><b><span style='color:#ff54ff;'>&quot;matrix&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># For mixed ploidy data sets</span></b>
<b><span style='color:#54ffff;'># stamppFst requires population factor in genlight (here, population code consists of first three letters of individual's name)</span></b>
<b><span style='color:#dede49;'>indNames</span></b><b>(</b>arabidopsis.genlight<b>)</b>
<b><span style='color:#dede49;'>pop</span></b><b>(</b>arabidopsis.genlight<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>substr</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>indNames</span></b><b>(</b>arabidopsis.genlight<b>)</b>, <b><span style='color:#54ff54;'>start=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>stop=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#dede49;'>pop</span></b><b>(</b>arabidopsis.genlight<b>)</b> <b><span style='color:#54ffff;'># Check it</span></b>
<b><span style='color:#dede49;'>popNames</span></b><b>(</b>arabidopsis.genlight<b>)</b>
arabidopsis.fst <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>StAMPP<b>::</b><b><span style='color:#dede49;'>stamppFst</span></b><b>(</b><b><span style='color:#54ff54;'>geno=</span></b>arabidopsis.genlight, <b><span style='color:#54ff54;'>nboots=</span></b><b><span style='color:#ff54ff;'>100</span></b>, <b><span style='color:#54ff54;'>percent=</span></b><b><span style='color:#ff54ff;'>95</span></b>, <b><span style='color:#54ff54;'>nclusters=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>
<b><span style='color:#54ffff;'># For large data use higher nclusters to parallelize calculations</span></b>
?StAMPP<b>::</b>stamppFst <b><span style='color:#54ffff;'># See details</span></b>
<b><span style='color:#54ffff;'># Matrix of Fst among populations</span></b>
arabidopsis.fst<b>[[</b><b><span style='color:#ff54ff;'>&quot;Fsts&quot;</span></b><b>]]</b>
<b><span style='color:#54ffff;'># Matrix of P values</span></b>
arabidopsis.fst<b>[[</b><b><span style='color:#ff54ff;'>&quot;Pvalues&quot;</span></b><b>]]</b>

<b><span style='color:#b00000;'>## Multi locus genotypes</span></b>
<b><span style='color:#54ffff;'># Total number of MLGs (simple value)</span></b>
<b><span style='color:#dede49;'>mlg</span></b><b>(</b><b><span style='color:#54ff54;'>gid=</span></b>hauss.genind, <b><span style='color:#54ff54;'>quiet=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># MLGs shared among populations (a list)</span></b>
<b><span style='color:#dede49;'>mlg.crosspop</span></b><b>(</b><b><span style='color:#54ff54;'>gid=</span></b>hauss.genind, <b><span style='color:#54ff54;'>df=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>quiet=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Detailed view on distribution of MLGs into populations (table and/or plot)</span></b>
<b><span style='color:#dede49;'>mlg.table</span></b><b>(</b><b><span style='color:#54ff54;'>gid=</span></b>hauss.genind, <b><span style='color:#54ff54;'>bar=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>total=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>quiet=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Which individual belong to which vector (two ways of display)</span></b>
<b><span style='color:#dede49;'>mlg.vector</span></b><b>(</b>hauss.genind<b>)</b>
<b><span style='color:#dede49;'>mlg.id</span></b><b>(</b>hauss.genind<b>)</b>

<b><span style='color:#b00000;'>## Inbreeding</span></b>
<b><span style='color:#54ffff;'># Load training data</span></b>
<b><span style='color:#dede49;'>data</span></b><b>(</b>microbov<b>)</b>
<b><span style='color:#54ffff;'># Separate populations of Salers cattle</span></b>
microbov.pops <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>seppop</span></b><b>(</b>microbov<b>)[[</b><b><span style='color:#ff54ff;'>&quot;Salers&quot;</span></b><b>]]</b>
microbov.pops <b><span style='color:#54ffff;'># See it</span></b>
<b><span style='color:#54ffff;'># Calculate the inbreeding</span></b>
?inbreeding <b><span style='color:#54ffff;'># Check for more settings</span></b>
microbov.inbr <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>inbreeding</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>microbov.pops, <b><span style='color:#54ff54;'>N=</span></b><b><span style='color:#ff54ff;'>100</span></b><b>)</b>
<b><span style='color:#54ffff;'># Prepare population means for plotting</span></b>
microbov.bar <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>sapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>microbov.inbr, <b><span style='color:#54ff54;'>FUN=</span></b>mean<b>)</b>
<b><span style='color:#54ffff;'># Plot it</span></b>
<b><span style='color:#dede49;'>hist</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>microbov.bar, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;firebrick&quot;</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Average inbreeding in Salers cattle&quot;</span></b><b>)</b>

<b><span style='color:#b00000;'>## Genetic distances</span></b>

<b><span style='color:#54ffff;'># See ?dist.gene for details about methods of this distance constructions</span></b>
hauss.dist.g <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dist.gene</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind<b>@</b>tab, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;pairwise&quot;</span></b><b>)</b>
hauss.dist.g

<b><span style='color:#54ffff;'># Euclidean distance for individuals (plain ordinary distance matrix)</span></b>
hauss.dist <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dist</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;euclidean&quot;</span></b>, <b><span style='color:#54ff54;'>diag=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>upper=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
hauss.dist

<b><span style='color:#54ffff;'># Nei's distance (not Euclidean) for populations (other methods are available, see ?dist.genpop)</span></b>
hauss.dist.pop <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dist.genpop</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genpop, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>diag=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>upper=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
hauss.dist.pop
<b><span style='color:#54ffff;'># Test if it is Euclidean</span></b>
<b><span style='color:#dede49;'>is.euclid</span></b><b>(</b>hauss.dist.pop, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>print=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>tol=</span></b><b><span style='color:#ff54ff;'>1e-10</span></b><b>)</b>
<b><span style='color:#54ffff;'># Turns to be Euclidean</span></b>
hauss.dist.pop <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>cailliez</span></b><b>(</b><b><span style='color:#54ff54;'>distmat=</span></b>hauss.dist.pop, <b><span style='color:#54ff54;'>print=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>tol=</span></b><b><span style='color:#ff54ff;'>1e-07</span></b>, <b><span style='color:#54ff54;'>cor.zero=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Test if it is Euclidean</span></b>
<b><span style='color:#dede49;'>is.euclid</span></b><b>(</b>hauss.dist.pop, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>print=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>tol=</span></b><b><span style='color:#ff54ff;'>1e-10</span></b><b>)</b>
<b><span style='color:#54ffff;'># Show it</span></b>
hauss.dist.pop

<b><span style='color:#54ffff;'># Bruvo's distances weighting SSRs repeats - take care about replen parameter - requires repetition length for every SSRs locus</span></b>
hauss.dist.bruvo <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>bruvo.dist</span></b><b>(</b><b><span style='color:#54ff54;'>pop=</span></b>hauss.genind, <b><span style='color:#54ff54;'>replen=</span></b><b><span style='color:#dede49;'>rep</span></b><b>(</b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#ff54ff;'>12</span></b><b>)</b>, <b><span style='color:#54ff54;'>loss=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Test if it is Euclidean</span></b>
<b><span style='color:#dede49;'>is.euclid</span></b><b>(</b>hauss.dist.bruvo, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>print=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>tol=</span></b><b><span style='color:#ff54ff;'>1e-10</span></b><b>)</b>
<b><span style='color:#54ffff;'># Turn to be Euclidean</span></b>
hauss.dist.bruvo <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>cailliez</span></b><b>(</b><b><span style='color:#54ff54;'>distmat=</span></b>hauss.dist.bruvo, <b><span style='color:#54ff54;'>print=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>tol=</span></b><b><span style='color:#ff54ff;'>1e-07</span></b>, <b><span style='color:#54ff54;'>cor.zero=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Test if it is Euclidean</span></b>
<b><span style='color:#dede49;'>is.euclid</span></b><b>(</b>hauss.dist.bruvo, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>print=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>tol=</span></b><b><span style='color:#ff54ff;'>1e-10</span></b><b>)</b>
<b><span style='color:#54ffff;'># Show it</span></b>
hauss.dist.bruvo

<b><span style='color:#54ffff;'># Nei's distance (not Euclidean) for individuals (other methods are available, see ?nei.dist from poppr package)</span></b>
hauss.dist.nei <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>poppr<b>::</b><b><span style='color:#dede49;'>nei.dist</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind, <b><span style='color:#54ff54;'>warning=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Test if it is Euclidean</span></b>
<b><span style='color:#dede49;'>is.euclid</span></b><b>(</b><b><span style='color:#54ff54;'>distmat=</span></b>hauss.dist.nei, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>print=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>tol=</span></b><b><span style='color:#ff54ff;'>1e-10</span></b><b>)</b>
<b><span style='color:#54ffff;'># Show it</span></b>
hauss.dist.nei

<b><span style='color:#54ffff;'># Dissimilarity matrix returns a distance reflecting the number of allelic differences between two individuals</span></b>
hauss.dist.diss <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>diss.dist</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind, <b><span style='color:#54ff54;'>percent=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>mat=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Test if it is Euclidean</span></b>
<b><span style='color:#dede49;'>is.euclid</span></b><b>(</b><b><span style='color:#54ff54;'>distmat=</span></b><b><span style='color:#dede49;'>as.dist</span></b><b>(</b><b><span style='color:#54ff54;'>m=</span></b>hauss.dist.diss<b>)</b>, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>print=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>tol=</span></b><b><span style='color:#ff54ff;'>1e-10</span></b><b>)</b>
<b><span style='color:#54ffff;'># Show it</span></b>
hauss.dist.diss

<b><span style='color:#54ffff;'># Import custom distance matrix</span></b>
MyDistance <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.csv</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;distances.txt&quot;</span></b>, <b><span style='color:#54ff54;'>header=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;</span></b><b><span style='color:#ff54ff;'>\t</span></b><b><span style='color:#ff54ff;'>&quot;</span></b>, <b><span style='color:#54ff54;'>dec=</span></b><b><span style='color:#ff54ff;'>&quot;.&quot;</span></b>, <b><span style='color:#54ff54;'>row.names=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>
MyDistance <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>as.dist</span></b><b>(</b>MyDistance<b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>MyDistance<b>)</b>
<b><span style='color:#dede49;'>is.euclid</span></b><b>(</b>MyDistance, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>print=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>tol =</span></b> <b><span style='color:#ff54ff;'>1e-10</span></b><b>)</b>
<b><span style='color:#dede49;'>dim</span></b><b>(</b>MyDistance<b>)</b>
MyDistance
MyDistance <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>cailliez</span></b><b>(</b>MyDistance, <b><span style='color:#54ff54;'>print=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>tol =</span></b> <b><span style='color:#ff54ff;'>1e-10</span></b>, <b><span style='color:#54ff54;'>cor.zero=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>is.euclid</span></b><b>(</b>MyDistance, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>print=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>tol =</span></b> <b><span style='color:#ff54ff;'>1e-10</span></b><b>)</b>
MyDistance

<b><span style='color:#54ffff;'># Compare different distance matrices</span></b>
<b><span style='color:#54ffff;'># List of functions to be parsed to respective dist.* function</span></b>
distances <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Nei&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Rogers&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Edwards&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Reynolds&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Prevosti&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Calculate the distance matrices</span></b>
dists <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>lapply</span></b><b>(</b>distances, <b><span style='color:#ffff54;'>function</span></b><b>(</b>x<b>)</b> <b>{</b>
	DISTFUN <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>match.fun</span></b><b>(</b><b><span style='color:#dede49;'>paste</span></b><b>(</b><b><span style='color:#dede49;'>tolower</span></b><b>(</b>x<b>)</b>, <b><span style='color:#ff54ff;'>&quot;dist&quot;</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;.&quot;</span></b><b>))</b>
	<b><span style='color:#dede49;'>DISTFUN</span></b><b>(</b>hauss.genind.cor<b>)</b>
	<b>}</b><b>)</b>
<b><span style='color:#54ffff;'># Add names for the distance names</span></b>
<b><span style='color:#dede49;'>names</span></b><b>(</b>dists<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>distances
<b><span style='color:#54ffff;'># Add Bruvo distance</span></b>
dists<b>[[</b><b><span style='color:#ff54ff;'>&quot;Bruvo&quot;</span></b><b>]]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>hauss.dist.bruvo
dists
<b><span style='color:#54ffff;'># Split graphical device into 2 lines, 3 panes each</span></b>
<b><span style='color:#dede49;'>par</span></b><b>(</b><b><span style='color:#54ff54;'>mfrow=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#ff54ff;'>3</span></b><b>))</b>
<b><span style='color:#54ffff;'># Calculate NJ and plot all trees</span></b>
x <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#dede49;'>names</span></b><b>(</b>dists<b>)</b>, <b><span style='color:#ffff54;'>function</span></b><b>(</b>x<b>)</b> <b>{</b>
	<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#dede49;'>njs</span></b><b>(</b>dists<b>[[</b>x<b>]])</b>, <b><span style='color:#54ff54;'>main=</span></b>x, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;unrooted&quot;</span></b><b>)</b>
	<b><span style='color:#dede49;'>add.scale.bar</span></b><b>(</b><b><span style='color:#54ff54;'>lcol=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>length=</span></b><b><span style='color:#ff54ff;'>0.1</span></b><b>)</b>
	<b>}</b><b>)</b>
<b><span style='color:#dede49;'>dev.off</span></b><b>()</b> <b><span style='color:#54ffff;'># Close graphical device to reset settings</span></b>
<b><span style='color:#54ffff;'># See details of distance methods in package poppr</span></b>
<b><span style='color:#dede49;'>vignette</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;algo&quot;</span></b>, <b><span style='color:#54ff54;'>package=</span></b><b><span style='color:#ff54ff;'>&quot;poppr&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># DNA distances - there are various models available</span></b>
?dist.dna
usflu.dist <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dist.dna</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.dna, <b><span style='color:#54ff54;'>model=</span></b><b><span style='color:#ff54ff;'>&quot;TN93&quot;</span></b><b>)</b>
usflu.dist
<b><span style='color:#dede49;'>class</span></b><b>(</b>usflu.dist<b>)</b>
<b><span style='color:#dede49;'>dim</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>usflu.dist<b>))</b>
meles.dist <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dist.dna</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.dna, <b><span style='color:#54ff54;'>model=</span></b><b><span style='color:#ff54ff;'>&quot;F81&quot;</span></b><b>)</b>
meles.dist
<b><span style='color:#dede49;'>class</span></b><b>(</b>meles.dist<b>)</b>
<b><span style='color:#dede49;'>dim</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>meles.dist<b>))</b>

<b><span style='color:#54ffff;'># Distances and genlight object. Pairwise genetic distances for each data block (genlight objects with whole genome data) - sensitive to missing data</span></b>
usflu.dists.l <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>seploc</span></b><b>(</b>usflu.genlight, <b><span style='color:#54ff54;'>n.block=</span></b><b><span style='color:#ff54ff;'>10</span></b>, <b><span style='color:#54ff54;'>parallel=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>usflu.dists.l<b>)</b>
usflu.dists <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>usflu.dists.l, <b><span style='color:#54ff54;'>FUN=</span></b><b><span style='color:#ffff54;'>function</span></b><b>(</b>DDD<b>)</b> <b><span style='color:#dede49;'>dist</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>DDD<b>)))</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>usflu.dists<b>)</b>
<b><span style='color:#dede49;'>names</span></b><b>(</b>usflu.dists<b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>usflu.dists<b>[[</b><b><span style='color:#ff54ff;'>1</span></b><b>]])</b>
usflu.distr <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>Reduce</span></b><b>(</b><b><span style='color:#54ff54;'>f=</span></b><b><span style='color:#ff54ff;'>&quot;+&quot;</span></b>, <b><span style='color:#54ff54;'>x=</span></b>usflu.dists<b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>usflu.distr<b>)</b>
usflu.distr
<b><span style='color:#54ffff;'># It is possible to use just basic dist function on whole genlight object (might require a lot of RAM)</span></b>
usflu.distg <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dist</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>usflu.genlight<b>))</b>

<b><span style='color:#54ffff;'># Distances in mixed-ploidy data sets</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>StAMPP<b>)</b> <b><span style='color:#54ffff;'># Load required library</span></b>
<b><span style='color:#54ffff;'># stamppNeisD requires population factor in genlight</span></b>
<b><span style='color:#54ffff;'># Nei's 1972 distance between individuals (use pop=TRUE to calculate among populations)</span></b>
arabidopsis.dist <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>stamppNeisD</span></b><b>(</b><b><span style='color:#54ff54;'>geno=</span></b>arabidopsis.genlight, <b><span style='color:#54ff54;'>pop=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Check it</span></b>
<b><span style='color:#dede49;'>head</span></b><b>(</b>arabidopsis.dist<b>)</b>
<b><span style='color:#dede49;'>dim</span></b><b>(</b>arabidopsis.dist<b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>arabidopsis.dist<b>)</b>
<b><span style='color:#54ffff;'># The same on population level</span></b>
arabidopsis.dist.pop <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>stamppNeisD</span></b><b>(</b><b><span style='color:#54ff54;'>geno=</span></b>arabidopsis.genlight, <b><span style='color:#54ff54;'>pop=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Check it</span></b>
<b><span style='color:#dede49;'>head</span></b><b>(</b>arabidopsis.dist.pop<b>)</b>
<b><span style='color:#dede49;'>dim</span></b><b>(</b>arabidopsis.dist.pop<b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>arabidopsis.dist.pop<b>)</b>
<b><span style='color:#54ffff;'># Export the distance matrix as Phylip format for usage in external software (e.g. SplitsTree)</span></b>
<b><span style='color:#dede49;'>stamppPhylip</span></b><b>(</b><b><span style='color:#54ff54;'>distance.mat=</span></b>arabidopsis.dist, <b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;arabidopsis_dist.txt&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Genomic relationship matrix</span></b>
?stamppGmatrix <b><span style='color:#54ffff;'># Method details</span></b>
arabidopsis.genomat <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>stamppGmatrix</span></b><b>(</b><b><span style='color:#54ff54;'>geno=</span></b>arabidopsis.genlight<b>)</b>
<b><span style='color:#54ffff;'># Check it</span></b>
<b><span style='color:#dede49;'>head</span></b><b>(</b>arabidopsis.genomat<b>)</b>
<b><span style='color:#dede49;'>dim</span></b><b>(</b>arabidopsis.genomat<b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>arabidopsis.genomat<b>)</b>

<b><span style='color:#54ffff;'># Over 40 distances from philentropy package</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>philentropy<b>)</b> <b><span style='color:#54ffff;'># Load the library</span></b>
<b><span style='color:#dede49;'>getDistMethods</span></b><b>()</b> <b><span style='color:#54ffff;'># See available distances</span></b>
?distance <b><span style='color:#54ffff;'># See details of distances</span></b>
<b><span style='color:#54ffff;'># Calculate e.g. Jaccard index for AFLP data</span></b>
amara.jac <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>distance</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>amara.aflp<b>[</b>,<b><span style='color:#ff54ff;'>2</span></b><b>:</b><b><span style='color:#ff54ff;'>30</span></b><b>]</b>, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;jaccard&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># amara.aflp has 30 columns (see dim(amara.aflp), column 1 contains names (head(amara.aflp)))</span></b>
<b><span style='color:#54ffff;'># See result</span></b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>amara.jac<b>)</b>
amara.jac
<b><span style='color:#54ffff;'># Make it distance matrix</span></b>
amara.jac <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>as.dist</span></b><b>(</b><b><span style='color:#54ff54;'>m=</span></b>amara.jac, <b><span style='color:#54ff54;'>diag=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>upper=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
amara.jac

<b><span style='color:#54ffff;'># Visualize pairwise genetic similarities</span></b>
<b><span style='color:#54ffff;'># table.paint() requires data frame, dist can't be directly converted to DF</span></b>
<b><span style='color:#dede49;'>table.paint</span></b><b>(</b><b><span style='color:#54ff54;'>df=</span></b><b><span style='color:#dede49;'>as.data.frame</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>usflu.dist<b>))</b>, <b><span style='color:#54ff54;'>cleg=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>clabel.row=</span></b><b><span style='color:#ff54ff;'>0.5</span></b>, <b><span style='color:#54ff54;'>clabel.col=</span></b><b><span style='color:#ff54ff;'>0.5</span></b><b>)</b>
<b><span style='color:#54ffff;'># Same visualization, colored</span></b>
<b><span style='color:#54ffff;'># heatmap() reorders values</span></b>
<b><span style='color:#dede49;'>heatmap</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>usflu.dist<b>)</b>, <b><span style='color:#54ff54;'>Rowv=</span></b><b><span style='color:#00ff00;'>NA</span></b>, <b><span style='color:#54ff54;'>Colv=</span></b><b><span style='color:#00ff00;'>NA</span></b>, <b><span style='color:#54ff54;'>symm=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Another possibility is to use function corrplot() from package corrplot</span></b>

<b><span style='color:#b00000;'>## Heatmaps</span></b>
<b><span style='color:#54ffff;'># Based on various distances</span></b>
<b><span style='color:#dede49;'>heatmap</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>hauss.dist<b>)</b>, <b><span style='color:#54ff54;'>symm=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>labRow=</span></b><b><span style='color:#dede49;'>rownames</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>hauss.dist.bruvo<b>))</b>, <b><span style='color:#54ff54;'>labCol=</span></b><b><span style='color:#dede49;'>colnames</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>hauss.dist.bruvo<b>)))</b> <b><span style='color:#54ffff;'># hauss.dist doesn't contain names of individuals - we have to add them here</span></b>
<b><span style='color:#dede49;'>heatmap</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>hauss.dist.pop<b>)</b>, <b><span style='color:#54ff54;'>symm=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>heatmap</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>hauss.dist.bruvo<b>)</b>, <b><span style='color:#54ff54;'>symm=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>heatmap</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>hauss.dist.diss<b>)</b>, <b><span style='color:#54ff54;'>symm=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
?heatmap <b><span style='color:#54ffff;'># Other options</span></b>

<b><span style='color:#b00000;'>## Hierarchical clustering</span></b>
<b><span style='color:#54ffff;'># According to distance used - see ?hclust for available methods</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#dede49;'>hclust</span></b><b>(</b><b><span style='color:#54ff54;'>d=</span></b>hauss.dist, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;complete&quot;</span></b><b>))</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#dede49;'>hclust</span></b><b>(</b><b><span style='color:#54ff54;'>d=</span></b>hauss.dist.pop, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;complete&quot;</span></b><b>))</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#dede49;'>hclust</span></b><b>(</b><b><span style='color:#54ff54;'>d=</span></b>hauss.dist.bruvo, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;complete&quot;</span></b><b>))</b>

<b><span style='color:#b00000;'>## UPGMA</span></b>
<b><span style='color:#54ffff;'># Calculate it</span></b>
<b><span style='color:#54ffff;'># Saving as phylo object (and not hclust) gives more possibilities for further plotting and manipulations</span></b>
usflu.upgma <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>as.phylo</span></b><b>(</b><b><span style='color:#dede49;'>hclust</span></b><b>(</b><b><span style='color:#54ff54;'>d=</span></b>usflu.dist, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;average&quot;</span></b><b>))</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.upgma, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>0.75</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;UPGMA tree&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Test quality - tests correlation of original distance in the matrix and reconstructed distance from hclust object</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b>usflu.dist<b>)</b>, <b><span style='color:#54ff54;'>y=</span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b><b><span style='color:#dede49;'>as.dist</span></b><b>(</b><b><span style='color:#dede49;'>cophenetic</span></b><b>(</b>usflu.upgma<b>)))</b>, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#ff54ff;'>&quot;Original pairwise distances&quot;</span></b>, <b><span style='color:#54ff54;'>ylab=</span></b><b><span style='color:#ff54ff;'>&quot;Pairwise distances on the tree&quot;</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Is UPGMA appropriate?&quot;</span></b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>20</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>transp</span></b><b>(</b><b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;black&quot;</span></b>, <b><span style='color:#54ff54;'>alpha=</span></b><b><span style='color:#ff54ff;'>0.1</span></b><b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>abline</span></b><b>(</b><b><span style='color:#dede49;'>lm</span></b><b>(</b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b><b><span style='color:#dede49;'>as.dist</span></b><b>(</b><b><span style='color:#dede49;'>cophenetic</span></b><b>(</b>usflu.upgma<b>)))</b><b>~</b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b>usflu.dist<b>))</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b><b>)</b>

<b><span style='color:#b00000;'>## AMOVA</span></b>
<b><span style='color:#54ffff;'># From package pegas (doesn't directly show percentage of variance)</span></b>
hauss.pop <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>pop</span></b><b>(</b>hauss.genind<b>)</b>
hauss.amova <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>pegas<b>::</b><b><span style='color:#dede49;'>amova</span></b><b>(</b>hauss.dist<b>~</b>hauss.pop, <b><span style='color:#54ff54;'>data=</span></b><b><span style='color:#00ff00;'>NULL</span></b>, <b><span style='color:#54ff54;'>nperm=</span></b><b><span style='color:#ff54ff;'>1000</span></b>, <b><span style='color:#54ff54;'>is.squared=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># See results</span></b>
hauss.amova
<b><span style='color:#54ffff;'># For more complicated hierarchy</span></b>
?poppr<b>::</b>poppr.amova
<b><span style='color:#54ffff;'># For mixed-ploidy dat sets</span></b>
?StAMPP<b>::</b>stamppAmova

<b><span style='color:#b00000;'>## MSN based on Bruvo's distance</span></b>
<b><span style='color:#dede49;'>bruvo.msn</span></b><b>(</b><b><span style='color:#54ff54;'>gid=</span></b>hauss.genind, <b><span style='color:#54ff54;'>replen=</span></b><b><span style='color:#dede49;'>rep</span></b><b>(</b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#ff54ff;'>12</span></b><b>)</b>, <b><span style='color:#54ff54;'>loss=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>palette=</span></b>rainbow, <b><span style='color:#54ff54;'>vertex.label=</span></b><b><span style='color:#ff54ff;'>&quot;inds&quot;</span></b>, <b><span style='color:#54ff54;'>gscale=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>wscale=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>showplot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
?bruvo.msn <b><span style='color:#54ffff;'># See details...</span></b>
?msn.poppr <b><span style='color:#54ffff;'># For another data types</span></b>
?imsn <b><span style='color:#54ffff;'># Interactive creation of MSN</span></b>
<b><span style='color:#dede49;'>imsn</span></b><b>()</b> <b><span style='color:#54ffff;'># Try it</span></b>

<b><span style='color:#b00000;'>## NJ</span></b>

<b><span style='color:#54ffff;'># Calculates the tree (try with various distances)</span></b>
hauss.nj <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>nj</span></b><b>(</b>hauss.dist<b>)</b>

<b><span style='color:#54ffff;'># Test tree quality - plot original vs. reconstructed distance</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b>hauss.dist<b>)</b>, <b><span style='color:#dede49;'>as.vector</span></b><b>(</b><b><span style='color:#dede49;'>as.dist</span></b><b>(</b><b><span style='color:#dede49;'>cophenetic</span></b><b>(</b>hauss.nj<b>)))</b>, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#ff54ff;'>&quot;Original distance&quot;</span></b>, <b><span style='color:#54ff54;'>ylab=</span></b><b><span style='color:#ff54ff;'>&quot;Reconstructed distance&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>abline</span></b><b>(</b><b><span style='color:#dede49;'>lm</span></b><b>(</b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b>hauss.dist<b>)</b> <b>~</b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b><b><span style='color:#dede49;'>as.dist</span></b><b>(</b><b><span style='color:#dede49;'>cophenetic</span></b><b>(</b>hauss.nj<b>))))</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>summary</span></b><b>(</b><b><span style='color:#dede49;'>lm</span></b><b>(</b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b>hauss.dist<b>)</b> <b>~</b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b><b><span style='color:#dede49;'>as.dist</span></b><b>(</b><b><span style='color:#dede49;'>cophenetic</span></b><b>(</b>hauss.nj<b>)))))</b> <b><span style='color:#54ffff;'># Prints summary text</span></b>

<b><span style='color:#54ffff;'># Plot a basic tree - see ?plot.phylo for details</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;phylogram&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;cladogram&quot;</span></b>, <b><span style='color:#54ff54;'>edge.width=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;fan&quot;</span></b>, <b><span style='color:#54ff54;'>edge.width=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>edge.lty=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;radial&quot;</span></b>, <b><span style='color:#54ff54;'>edge.color=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>edge.width=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>edge.lty=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>

<b><span style='color:#54ffff;'># Bootstrap</span></b>

<b><span style='color:#54ffff;'># boot.phylo() resamples all columns - remove population column first</span></b>
hauss.loci.nopop <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>hauss.loci
hauss.loci.nopop<b>[[</b><b><span style='color:#ff54ff;'>&quot;population&quot;</span></b><b>]]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#00ff00;'>NULL</span></b>
hauss.boot <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>boot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>hauss.nj, <b><span style='color:#54ff54;'>x=</span></b>hauss.loci.nopop, <b><span style='color:#54ff54;'>FUN=</span></b><b><span style='color:#ffff54;'>function</span></b><b>(</b>XXX<b>)</b> <b><span style='color:#dede49;'>nj</span></b><b>(</b><b><span style='color:#dede49;'>dist</span></b><b>(</b><b><span style='color:#dede49;'>loci2genind</span></b><b>(</b>XXX<b>)))</b>, <b><span style='color:#54ff54;'>B=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>)</b>
<b><span style='color:#54ffff;'># boot.phylo returns NUMBER of replicates - NO PERCENTAGE</span></b>
<b><span style='color:#54ffff;'># Plot the tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;unrooted&quot;</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Neighbour-Joining tree&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Labels for nodes - bootstrap - see ?nodelabels for graphical settings</span></b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b><b><span style='color:#54ff54;'>text=</span></b><b><span style='color:#dede49;'>round</span></b><b>(</b>hauss.boot<b>/</b><b><span style='color:#ff54ff;'>10</span></b><b>))</b>

<b><span style='color:#54ffff;'># Another bootstrap possibility</span></b>
hauss.aboot <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>aboot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind, <b><span style='color:#54ff54;'>tree=</span></b>nj, <b><span style='color:#54ff54;'>distance=</span></b>nei.dist, <b><span style='color:#54ff54;'>sample=</span></b><b><span style='color:#ff54ff;'>100</span></b><b>)</b> <b><span style='color:#54ffff;'># Bootstrap values are in slot node.label</span></b>
<b><span style='color:#54ffff;'># Plot the tree, explicitly display node labels</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.aboot, <b><span style='color:#54ff54;'>show.node.label=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
?aboot <b><span style='color:#54ffff;'># See details...</span></b>
?plot.phylo

<b><span style='color:#54ffff;'># Plot a nice tree with colored tips</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;unrooted&quot;</span></b>, <b><span style='color:#54ff54;'>show.tip=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Neighbour-Joining tree&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Labels for nodes - bootstrap - see ?nodelabels for graphical settings</span></b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b><b><span style='color:#54ff54;'>text=</span></b><b><span style='color:#dede49;'>round</span></b><b>(</b>hauss.boot<b>/</b><b><span style='color:#ff54ff;'>10</span></b><b>))</b>
<b><span style='color:#54ffff;'># Coloured labels - creates vector of colors according to population information in genind object</span></b>
nj.rainbow <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>colorRampPalette</span></b><b>(</b><b><span style='color:#dede49;'>rainbow</span></b><b>(</b><b><span style='color:#dede49;'>length</span></b><b>(</b><b><span style='color:#dede49;'>levels</span></b><b>(</b><b><span style='color:#dede49;'>pop</span></b><b>(</b>hauss.genind<b>)))))</b>
<b><span style='color:#54ffff;'># Colored tips</span></b>
<b><span style='color:#dede49;'>tiplabels</span></b><b>(</b><b><span style='color:#54ff54;'>text=</span></b><b><span style='color:#dede49;'>indNames</span></b><b>(</b>hauss.genind<b>)</b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#dede49;'>fac2col</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>pop</span></b><b>(</b>hauss.genind<b>)</b>, <b><span style='color:#54ff54;'>col.pal=</span></b>nj.rainbow<b>))</b>

<b><span style='color:#54ffff;'># Plot BW tree with tip symbols and legend</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;cladogram&quot;</span></b>, <b><span style='color:#54ff54;'>show.tip=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Neighbour-Joining tree&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add axis with distances</span></b>
<b><span style='color:#dede49;'>axisPhylo</span></b><b>()</b>
<b><span style='color:#54ffff;'># From node labels let's remove unneeded frame</span></b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b><b><span style='color:#54ff54;'>text=</span></b><b><span style='color:#dede49;'>round</span></b><b>(</b>hauss.boot<b>/</b><b><span style='color:#ff54ff;'>10</span></b><b>)</b>, <b><span style='color:#54ff54;'>frame=</span></b><b><span style='color:#ff54ff;'>&quot;none&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># As tip label we use only symbols - see ?points for graphical details</span></b>
<b><span style='color:#dede49;'>tiplabels</span></b><b>(</b><b><span style='color:#54ff54;'>frame=</span></b><b><span style='color:#ff54ff;'>&quot;none&quot;</span></b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#dede49;'>rep</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>0</span></b><b>:</b><b><span style='color:#ff54ff;'>4</span></b>, <b><span style='color:#54ff54;'>times=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>13</span></b>, <b><span style='color:#ff54ff;'>17</span></b>,  <b><span style='color:#ff54ff;'>2</span></b>,  <b><span style='color:#ff54ff;'>6</span></b>,  <b><span style='color:#ff54ff;'>9</span></b><b>))</b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#54ffff;'># Plot a legend explainindg symbols</span></b>
<b><span style='color:#dede49;'>legend</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>&quot;topleft&quot;</span></b>, <b><span style='color:#54ff54;'>legend=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;He&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Oh&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Pr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Ne&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Sk&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>border=</span></b><b><span style='color:#ff54ff;'>&quot;black&quot;</span></b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>0</span></b><b>:</b><b><span style='color:#ff54ff;'>4</span></b>, <b><span style='color:#54ff54;'>pt.lwd=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>pt.cex=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>bty=</span></b><b><span style='color:#ff54ff;'>&quot;o&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;lightgrey&quot;</span></b>, <b><span style='color:#54ff54;'>box.lwd=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.2</span></b>, <b><span style='color:#54ff54;'>title=</span></b><b><span style='color:#ff54ff;'>&quot;Populations&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># Bruvo's distance - NJ</span></b>
hauss.nj.bruvo <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>bruvo.boot</span></b><b>(</b><b><span style='color:#54ff54;'>pop=</span></b>hauss.genind, <b><span style='color:#54ff54;'>replen=</span></b><b><span style='color:#dede49;'>rep</span></b><b>(</b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#ff54ff;'>12</span></b><b>)</b>, <b><span style='color:#54ff54;'>sample=</span></b><b><span style='color:#ff54ff;'>1000</span></b>, <b><span style='color:#54ff54;'>tree=</span></b><b><span style='color:#ff54ff;'>&quot;nj&quot;</span></b>, <b><span style='color:#54ff54;'>showtree=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>cutoff=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>quiet=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj.bruvo, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;unrooted&quot;</span></b>, <b><span style='color:#54ff54;'>show.tip=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Neighbor-Joining tree&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># bruvo.boot() writes all needed information into resulting object, so there is no need for external bootstrap function</span></b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b>hauss.nj.bruvo<b>[[</b><b><span style='color:#ff54ff;'>&quot;node.labels&quot;</span></b><b>]])</b> <b><span style='color:#54ffff;'># Note you can call node labels from phylo object as phylo$node.labels or phylo[[&quot;node.labels&quot;]]</span></b>
<b><span style='color:#dede49;'>tiplabels</span></b><b>(</b>hauss.nj.bruvo<b>[[</b><b><span style='color:#ff54ff;'>&quot;tip.label&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#dede49;'>fac2col</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind<b>$</b>pop, <b><span style='color:#54ff54;'>col.pal=</span></b>nj.rainbow<b>))</b>

<b><span style='color:#54ffff;'># Bruvo's distance - UPGMA</span></b>
hauss.upgma <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>bruvo.boot</span></b><b>(</b><b><span style='color:#54ff54;'>pop=</span></b>hauss.genind, <b><span style='color:#54ff54;'>replen=</span></b><b><span style='color:#dede49;'>rep</span></b><b>(</b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#ff54ff;'>12</span></b><b>)</b>, <b><span style='color:#54ff54;'>sample=</span></b><b><span style='color:#ff54ff;'>1000</span></b>, <b><span style='color:#54ff54;'>tree=</span></b><b><span style='color:#ff54ff;'>&quot;upgma&quot;</span></b>, <b><span style='color:#54ff54;'>showtree=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>cutoff=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>quiet=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.upgma, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;unrooted&quot;</span></b>, <b><span style='color:#54ff54;'>show.tip=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;UPGMA tree&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b>hauss.upgma<b>[[</b><b><span style='color:#ff54ff;'>&quot;node.labels&quot;</span></b><b>]])</b>
<b><span style='color:#dede49;'>tiplabels</span></b><b>(</b>hauss.upgma<b>[[</b><b><span style='color:#ff54ff;'>&quot;tip.label&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#dede49;'>fac2col</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind<b>@</b>pop, <b><span style='color:#54ff54;'>col.pal=</span></b>nj.rainbow<b>))</b>

<b><span style='color:#54ffff;'># Populations</span></b>
<b><span style='color:#54ffff;'># aboot() can use distances implemented in poppr:</span></b>
?poppr<b>::</b>aboot
?poppr<b>::</b>nei.dist
<b><span style='color:#54ffff;'># Calculations</span></b>
hauss.nj.pop <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>aboot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genpop, <b><span style='color:#54ff54;'>tree=</span></b><b><span style='color:#ff54ff;'>&quot;nj&quot;</span></b>, <b><span style='color:#54ff54;'>distance=</span></b><b><span style='color:#ff54ff;'>&quot;nei.dist&quot;</span></b>, <b><span style='color:#54ff54;'>sample=</span></b><b><span style='color:#ff54ff;'>1000</span></b>, <b><span style='color:#54ff54;'>showtree=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Information about result</span></b>
<b><span style='color:#dede49;'>print.phylo</span></b><b>(</b>hauss.nj.pop<b>)</b>
<b><span style='color:#54ffff;'># Plot a tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj.pop, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;radial&quot;</span></b>, <b><span style='color:#54ff54;'>show.node.label=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.2</span></b>, <b><span style='color:#54ff54;'>edge.width=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Neighbor-Joining tree of populations&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># Make a tree based on DNA sequences</span></b>
usflu.tree <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>nj</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>usflu.dist<b>)</b>
<b><span style='color:#54ffff;'># Plot it</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.tree, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;unrooted&quot;</span></b>, <b><span style='color:#54ff54;'>show.tip=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Unrooted NJ tree&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Coloured tips</span></b>
usflu.pal <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>colorRampPalette</span></b><b>(</b><b><span style='color:#dede49;'>topo.colors</span></b><b>(</b><b><span style='color:#dede49;'>length</span></b><b>(</b><b><span style='color:#dede49;'>levels</span></b><b>(</b><b><span style='color:#dede49;'>as.factor</span></b><b>(</b>usflu.annot<b>[[</b><b><span style='color:#ff54ff;'>&quot;year&quot;</span></b><b>]])))))</b>
<b><span style='color:#54ffff;'># Tip labels</span></b>
<b><span style='color:#dede49;'>tiplabels</span></b><b>(</b><b><span style='color:#54ff54;'>text=</span></b>usflu.annot<b>$</b>year, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#dede49;'>num2col</span></b><b>(</b>usflu.annot<b>$</b>year, <b><span style='color:#54ff54;'>col.pal=</span></b>usflu.pal<b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>0.75</span></b><b>)</b>
<b><span style='color:#54ffff;'># Legend - describing years - pretty() automatically shows best values from given range, num2col() selects colors from color scale</span></b>
<b><span style='color:#dede49;'>legend</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>&quot;bottomright&quot;</span></b>, <b><span style='color:#54ff54;'>fill=</span></b><b><span style='color:#dede49;'>num2col</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>pretty</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>1993</span></b><b>:</b><b><span style='color:#ff54ff;'>2008</span></b>, <b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b>, <b><span style='color:#54ff54;'>col.pal=</span></b>usflu.pal<b>)</b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#dede49;'>pretty</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>1993</span></b><b>:</b><b><span style='color:#ff54ff;'>2008</span></b>, <b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b>, <b><span style='color:#54ff54;'>ncol=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>

<b><span style='color:#54ffff;'># Root the tree - &quot;outgroup&quot; is name of accession (in quotation marks) or number (position within phy object)</span></b>
usflu.tree.rooted <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>root</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>usflu.tree, <b><span style='color:#54ff54;'>outgroup=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>
<b><span style='color:#54ffff;'># Plot it</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.tree.rooted, <b><span style='color:#54ff54;'>show.tip=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>edge.width=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Rooted NJ tree&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Labeling of tips</span></b>
<b><span style='color:#dede49;'>tiplabels</span></b><b>(</b><b><span style='color:#54ff54;'>text=</span></b>usflu.annot<b>$</b>year, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#dede49;'>transp</span></b><b>(</b><b><span style='color:#dede49;'>num2col</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.annot<b>$</b>year, <b><span style='color:#54ff54;'>col.pal=</span></b>usflu.pal<b>)</b>, <b><span style='color:#54ff54;'>alpha=</span></b><b><span style='color:#ff54ff;'>0.7</span></b><b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>0.75</span></b>, <b><span style='color:#54ff54;'>fg=</span></b><b><span style='color:#ff54ff;'>&quot;transparent&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add axis with phylogenetic distance</span></b>
<b><span style='color:#dede49;'>axisPhylo</span></b><b>()</b>
<b><span style='color:#54ffff;'># Legend - describing years - pretty() automatically shows best values from given range, num2col() selects colors from color scale</span></b>
<b><span style='color:#dede49;'>legend</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>&quot;topright&quot;</span></b>, <b><span style='color:#54ff54;'>fill=</span></b><b><span style='color:#dede49;'>num2col</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>pretty</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>1993</span></b><b>:</b><b><span style='color:#ff54ff;'>2008</span></b>, <b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b>, <b><span style='color:#54ff54;'>col.pal=</span></b>usflu.pal<b>)</b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#dede49;'>pretty</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>1993</span></b><b>:</b><b><span style='color:#ff54ff;'>2008</span></b>, <b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b>, <b><span style='color:#54ff54;'>ncol=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>

<b><span style='color:#54ffff;'># Bootstrap rooted tree</span></b>
<b><span style='color:#54ffff;'># Calculate it</span></b>
usflu.boot <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>boot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>usflu.tree.rooted, <b><span style='color:#54ff54;'>x=</span></b>usflu.dna, <b><span style='color:#54ff54;'>FUN=</span></b><b><span style='color:#ffff54;'>function</span></b><b>(</b>EEE<b>)</b> <b><span style='color:#dede49;'>root</span></b><b>(</b><b><span style='color:#dede49;'>nj</span></b><b>(</b><b><span style='color:#dede49;'>dist.dna</span></b><b>(</b>EEE, <b><span style='color:#54ff54;'>model=</span></b><b><span style='color:#ff54ff;'>&quot;TN93&quot;</span></b><b>))</b>, <b><span style='color:#54ff54;'>outgroup=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>, <b><span style='color:#54ff54;'>B=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>)</b>
<b><span style='color:#54ffff;'># Plot the tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.tree.rooted, <b><span style='color:#54ff54;'>show.tip=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>edge.width=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;NJ tree + bootstrap values&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>tiplabels</span></b><b>(</b><b><span style='color:#54ff54;'>frame=</span></b><b><span style='color:#ff54ff;'>&quot;none&quot;</span></b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>20</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>transp</span></b><b>(</b><b><span style='color:#dede49;'>num2col</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.annot<b>$</b>year, <b><span style='color:#54ff54;'>col.pal=</span></b>usflu.pal<b>)</b>, <b><span style='color:#54ff54;'>alpha=</span></b><b><span style='color:#ff54ff;'>0.7</span></b><b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>3.5</span></b>, <b><span style='color:#54ff54;'>fg=</span></b><b><span style='color:#ff54ff;'>&quot;transparent&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>axisPhylo</span></b><b>()</b>
<b><span style='color:#54ffff;'># Legend - describing years - pretty() automatically shows best values from given range, num2col() selects colors from color scale</span></b>
<b><span style='color:#dede49;'>legend</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>&quot;topright&quot;</span></b>, <b><span style='color:#54ff54;'>fill=</span></b><b><span style='color:#dede49;'>num2col</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>pretty</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>1993</span></b><b>:</b><b><span style='color:#ff54ff;'>2008</span></b>, <b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b>, <b><span style='color:#54ff54;'>col.pal=</span></b>usflu.pal<b>)</b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#dede49;'>pretty</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>1993</span></b><b>:</b><b><span style='color:#ff54ff;'>2008</span></b>, <b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b>, <b><span style='color:#54ff54;'>ncol=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>
<b><span style='color:#54ffff;'># Plots bootstrap support - note usflu.boot contains raw numbers - transform it into percent</span></b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b><b><span style='color:#54ff54;'>text=</span></b><b><span style='color:#dede49;'>round</span></b><b>(</b>usflu.boot<b>/</b><b><span style='color:#ff54ff;'>10</span></b><b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>0.75</span></b><b>)</b>

<b><span style='color:#54ffff;'># Collapse branches with low bootstrap support</span></b>
usflu.tree.usflu.na.density <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>usflu.tree.rooted
<b><span style='color:#54ffff;'># Determine branches with low support - note BS values are in raw numbers - use desired percentage with respect to number of bootstraps</span></b>
usflu.tocollapse <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>match</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>which</span></b><b>(</b>usflu.boot <b>&lt;</b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>700</span></b><b>)</b><b>+</b><b><span style='color:#dede49;'>length</span></b><b>(</b>usflu.tree.rooted<b>$</b>tip.label<b>)</b>, <b><span style='color:#54ff54;'>table=</span></b>usflu.tree.usflu.na.density<b>$</b>edge<b>[</b>,<b><span style='color:#ff54ff;'>2</span></b><b>])</b>
<b><span style='color:#54ffff;'># Set length of bad branches to zero</span></b>
usflu.tree.usflu.na.density<b>$</b>edge.length<b>[</b>usflu.tocollapse<b>]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>0</span></b>
<b><span style='color:#54ffff;'># Create new tree</span></b>
usflu.tree.collapsed <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>di2multi</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>usflu.tree.usflu.na.density, <b><span style='color:#54ff54;'>tol=</span></b><b><span style='color:#ff54ff;'>0.00001</span></b><b>)</b>
<b><span style='color:#54ffff;'># Plot the consensus tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.tree.collapsed, <b><span style='color:#54ff54;'>show.tip=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>edge.width=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;NJ tree after collapsing weak nodes&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>tiplabels</span></b><b>(</b><b><span style='color:#54ff54;'>text=</span></b>usflu.annot<b>$</b>year, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#dede49;'>transp</span></b><b>(</b><b><span style='color:#dede49;'>num2col</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.annot<b>$</b>year, <b><span style='color:#54ff54;'>col.pal=</span></b>usflu.pal<b>)</b>, <b><span style='color:#54ff54;'>alpha=</span></b><b><span style='color:#ff54ff;'>0.7</span></b><b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>0.5</span></b>, <b><span style='color:#54ff54;'>fg=</span></b><b><span style='color:#ff54ff;'>&quot;transparent&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>axisPhylo</span></b><b>()</b>
<b><span style='color:#dede49;'>legend</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>&quot;topright&quot;</span></b>, <b><span style='color:#54ff54;'>fill=</span></b><b><span style='color:#dede49;'>num2col</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>pretty</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>1993</span></b><b>:</b><b><span style='color:#ff54ff;'>2008</span></b>, <b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b>, <b><span style='color:#54ff54;'>col.pal=</span></b>usflu.pal<b>)</b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#dede49;'>pretty</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>1993</span></b><b>:</b><b><span style='color:#ff54ff;'>2008</span></b>, <b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b>, <b><span style='color:#54ff54;'>ncol=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>

<b><span style='color:#b00000;'>## Replacements of NJ</span></b>
<b><span style='color:#54ffff;'># Usage is same as nj() Check</span></b>
?njs
?bionj
?bionjs
?NJ
?UNJ
?fastme

<b><span style='color:#b00000;'>## PCoA</span></b>

<b><span style='color:#54ffff;'># Calculate it - based on various distance matrix</span></b>
hauss.pcoa <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dudi.pco</span></b><b>(</b><b><span style='color:#54ff54;'>d=</span></b><b><span style='color:#dede49;'>dist</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>scaleGen</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind, <b><span style='color:#54ff54;'>center=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>scale=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>truenames=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;euclidean&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>scannf=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>nf=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#54ffff;'># Basic display</span></b>
<b><span style='color:#dede49;'>s.label</span></b><b>(</b><b><span style='color:#54ff54;'>dfxy=</span></b>hauss.pcoa<b>$</b>li, <b><span style='color:#54ff54;'>clabel=</span></b><b><span style='color:#ff54ff;'>0.75</span></b><b>)</b>
<b><span style='color:#54ffff;'># To plot different axes use for example &quot;...dfxy=hauss.pcoa$li[c(2, 3)]...&quot;</span></b>
<b><span style='color:#54ffff;'># Add kernel density</span></b>
<b><span style='color:#dede49;'>s.kde2d</span></b><b>(</b><b><span style='color:#54ff54;'>dfxy=</span></b>hauss.pcoa<b>$</b>li, <b><span style='color:#54ff54;'>cpoint=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>add.plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add histogram of Eigenvalues</span></b>
<b><span style='color:#dede49;'>add.scatter.eig</span></b><b>(</b><b><span style='color:#54ff54;'>w=</span></b>hauss.pcoa<b>$</b>eig, <b><span style='color:#54ff54;'>nf=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>xax=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>yax=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>posi=</span></b><b><span style='color:#ff54ff;'>&quot;bottomleft&quot;</span></b>, <b><span style='color:#54ff54;'>sub=</span></b><b><span style='color:#ff54ff;'>&quot;Eigenvalues&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># Colored display according to populations</span></b>
hauss.pcoa.col <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>rainbow</span></b><b>(</b><b><span style='color:#dede49;'>length</span></b><b>(</b><b><span style='color:#dede49;'>levels</span></b><b>(</b><b><span style='color:#dede49;'>pop</span></b><b>(</b>hauss.genind<b>))))</b> <b><span style='color:#54ffff;'># Creates vector of colors according to populations</span></b>
<b><span style='color:#dede49;'>s.class</span></b><b>(</b><b><span style='color:#54ff54;'>dfxy=</span></b>hauss.pcoa<b>$</b>li, <b><span style='color:#54ff54;'>fac=</span></b><b><span style='color:#dede49;'>pop</span></b><b>(</b>hauss.genind<b>)</b>, <b><span style='color:#54ff54;'>col=</span></b>hauss.pcoa.col<b>)</b>
<b><span style='color:#dede49;'>add.scatter.eig</span></b><b>(</b><b><span style='color:#54ff54;'>w=</span></b>hauss.pcoa<b>$</b>eig, <b><span style='color:#54ff54;'>nf=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>xax=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>yax=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>posi=</span></b><b><span style='color:#ff54ff;'>&quot;bottomleft&quot;</span></b>, <b><span style='color:#54ff54;'>sub=</span></b><b><span style='color:#ff54ff;'>&quot;Eigenvalues&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Principal Coordinates Analysis&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Adds title to the graph</span></b>

<b><span style='color:#54ffff;'># Based on Bruvo's distance</span></b>
hauss.pcoa.bruvo <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dudi.pco</span></b><b>(</b><b><span style='color:#54ff54;'>d=</span></b><b><span style='color:#dede49;'>bruvo.dist</span></b><b>(</b><b><span style='color:#54ff54;'>pop=</span></b>hauss.genind, <b><span style='color:#54ff54;'>replen=</span></b><b><span style='color:#dede49;'>rep</span></b><b>(</b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#ff54ff;'>12</span></b><b>))</b>, <b><span style='color:#54ff54;'>scannf=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>nf=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#dede49;'>s.class</span></b><b>(</b><b><span style='color:#54ff54;'>dfxy=</span></b>hauss.pcoa.bruvo<b>$</b>li, <b><span style='color:#54ff54;'>fac=</span></b><b><span style='color:#dede49;'>pop</span></b><b>(</b>hauss.genind<b>)</b>, <b><span style='color:#54ff54;'>col=</span></b>hauss.pcoa.col<b>)</b>
<b><span style='color:#dede49;'>add.scatter.eig</span></b><b>(</b>hauss.pcoa.bruvo<b>$</b>eig, <b><span style='color:#54ff54;'>posi=</span></b><b><span style='color:#ff54ff;'>&quot;bottomright&quot;</span></b>, <b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#54ffff;'># Another possibility for colored plot (see ?colorplot for details)</span></b>
<b><span style='color:#dede49;'>colorplot</span></b><b>(</b><b><span style='color:#54ff54;'>xy=</span></b>hauss.pcoa<b>$</b>li<b>[</b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#ff54ff;'>2</span></b><b>)]</b>, <b><span style='color:#54ff54;'>X=</span></b>hauss.pcoa<b>$</b>li, <b><span style='color:#54ff54;'>transp=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#ff54ff;'>&quot;PC 1&quot;</span></b>, <b><span style='color:#54ff54;'>ylab=</span></b><b><span style='color:#ff54ff;'>&quot;PC 2&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;PCoA, axes 1 and 3&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>abline</span></b><b>(</b><b><span style='color:#54ff54;'>v=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>h=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;grey&quot;</span></b>, <b><span style='color:#54ff54;'>lty=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>

<b><span style='color:#b00000;'>## Clustering analysis</span></b>

<b><span style='color:#54ffff;'># Graphical web interface for DAPC</span></b>
<b><span style='color:#dede49;'>adegenetServer</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;DAPC&quot;</span></b><b>)</b>

<b><span style='color:#b00000;'>## K-find</span></b>
<b><span style='color:#54ffff;'># Bayesian K-means clustering</span></b>
<b><span style='color:#54ffff;'># Retain all informative PC (here about 35). According to second graph select best K (here 2 or 3).</span></b>
hauss.kfind <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>find.clusters</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind, <b><span style='color:#54ff54;'>stat=</span></b><b><span style='color:#ff54ff;'>&quot;BIC&quot;</span></b>, <b><span style='color:#54ff54;'>choose.n.clust=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>max.n.clust=</span></b><b><span style='color:#ff54ff;'>10</span></b>, <b><span style='color:#54ff54;'>n.iter=</span></b><b><span style='color:#ff54ff;'>100000</span></b>, <b><span style='color:#54ff54;'>n.start=</span></b><b><span style='color:#ff54ff;'>100</span></b>, <b><span style='color:#54ff54;'>scale=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>truenames=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># See results as text</span></b>
<b><span style='color:#dede49;'>table</span></b><b>(</b><b><span style='color:#dede49;'>pop</span></b><b>(</b>hauss.genind<b>)</b>, hauss.kfind<b>$</b>grp<b>)</b>
hauss.kfind
<b><span style='color:#54ffff;'># Graph showing table of original and inferred populations and assignment of individuals</span></b>
<b><span style='color:#dede49;'>table.value</span></b><b>(</b><b><span style='color:#54ff54;'>df=</span></b><b><span style='color:#dede49;'>table</span></b><b>(</b><b><span style='color:#dede49;'>pop</span></b><b>(</b>hauss.genind<b>)</b>, hauss.kfind<b>$</b>grp<b>)</b>, <b><span style='color:#54ff54;'>col.lab=</span></b><b><span style='color:#dede49;'>paste</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Inferred</span></b><b><span style='color:#ff54ff;'>\n</span></b><b><span style='color:#ff54ff;'>cluster&quot;</span></b>, <b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#dede49;'>length</span></b><b>(</b>hauss.kfind<b>$</b>size<b>))</b>, <b><span style='color:#54ff54;'>grid=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># For K=3 - note parameters n.pca and n.clust - we just rerun the analysis and when results are stable, no problem here</span></b>
hauss.kfind3 <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>find.clusters</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind, <b><span style='color:#54ff54;'>n.pca=</span></b><b><span style='color:#ff54ff;'>35</span></b>, <b><span style='color:#54ff54;'>n.clust=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>stat=</span></b><b><span style='color:#ff54ff;'>&quot;BIC&quot;</span></b>, <b><span style='color:#54ff54;'>choose.n.clust=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>max.n.clust=</span></b><b><span style='color:#ff54ff;'>10</span></b>, <b><span style='color:#54ff54;'>n.iter=</span></b><b><span style='color:#ff54ff;'>100000</span></b>, <b><span style='color:#54ff54;'>n.start=</span></b><b><span style='color:#ff54ff;'>100</span></b>, <b><span style='color:#54ff54;'>scale=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>truenames=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># See results as text</span></b>
<b><span style='color:#dede49;'>table</span></b><b>(</b><b><span style='color:#dede49;'>pop</span></b><b>(</b>hauss.genind<b>)</b>, hauss.kfind3<b>$</b>grp<b>)</b>
hauss.kfind3
<b><span style='color:#54ffff;'># Graph showing table of original and inferred populations and assignment of individuals</span></b>
<b><span style='color:#dede49;'>table.value</span></b><b>(</b><b><span style='color:#54ff54;'>df=</span></b><b><span style='color:#dede49;'>table</span></b><b>(</b><b><span style='color:#dede49;'>pop</span></b><b>(</b>hauss.genind<b>)</b>, hauss.kfind3<b>$</b>grp<b>)</b>, <b><span style='color:#54ff54;'>col.lab=</span></b><b><span style='color:#dede49;'>paste</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Inferred</span></b><b><span style='color:#ff54ff;'>\n</span></b><b><span style='color:#ff54ff;'>cluster&quot;</span></b>, <b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#dede49;'>length</span></b><b>(</b>hauss.kfind3<b>$</b>size<b>))</b>, <b><span style='color:#54ff54;'>grid=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>

<b><span style='color:#b00000;'>## DAPC</span></b>
<b><span style='color:#54ffff;'># K=2</span></b>
<b><span style='color:#54ffff;'># Create DAPC</span></b>
<b><span style='color:#54ffff;'># Number of informative PC (Here 15, adegenet recommends &lt; N/3). Select number of informative DA (here 1 - only one is available - it won't produce PCA graph).</span></b>
hauss.dapc <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dapc</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind, <b><span style='color:#54ff54;'>pop=</span></b>hauss.kfind<b>$</b>grp, <b><span style='color:#54ff54;'>center=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>scale=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>var.contrib=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>pca.info=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>truenames=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Information</span></b>
hauss.dapc
<b><span style='color:#54ffff;'># Density functions</span></b>
<b><span style='color:#dede49;'>scatter</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.dapc, <b><span style='color:#54ff54;'>xax=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>yax=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;DAPC&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b>, <b><span style='color:#54ff54;'>solid=</span></b><b><span style='color:#ff54ff;'>0.5</span></b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>txt.leg=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Group 1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Group 2&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>posi.leg=</span></b><b><span style='color:#ff54ff;'>&quot;topright&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Assignment of individuals to clusters</span></b>
<b><span style='color:#dede49;'>assignplot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.dapc<b>)</b>
<b><span style='color:#54ffff;'># Structure-like plot</span></b>
<b><span style='color:#dede49;'>compoplot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.dapc, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#ff54ff;'>&quot;Individuals&quot;</span></b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Loadingplot - alleles the most adding to separation of individuals</span></b>
<b><span style='color:#dede49;'>loadingplot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.dapc<b>$</b>var.contr<b>)</b>
<b><span style='color:#54ffff;'># alfa-score - according to number of PC axis</span></b>
<b><span style='color:#dede49;'>optim.a.score</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.dapc<b>)</b>

<b><span style='color:#54ffff;'># K=3</span></b>
<b><span style='color:#54ffff;'># Create DAPC</span></b>
<b><span style='color:#54ffff;'># Number of informative PC (Here 15, adegenet recommends &lt; N/3). Select number of informative DA (here 2).</span></b>
hauss.dapc3 <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dapc</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind, <b><span style='color:#54ff54;'>pop=</span></b>hauss.kfind3<b>$</b>grp, <b><span style='color:#54ff54;'>center=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>scale=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>var.contrib=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>pca.info=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>truenames=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Information</span></b>
hauss.dapc
<b><span style='color:#54ffff;'># A la PCA graph</span></b>
<b><span style='color:#dede49;'>scatter</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.dapc3, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;DAPC, Taraxacum haussknechtii&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>clab=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>rainbow</span></b><b>(</b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>, <b><span style='color:#54ff54;'>posi.da=</span></b><b><span style='color:#ff54ff;'>&quot;bottomleft&quot;</span></b>, <b><span style='color:#54ff54;'>scree.pca=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>posi.pca=</span></b><b><span style='color:#ff54ff;'>&quot;bottomright&quot;</span></b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>txt.leg=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Group 1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Group 2&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Group 3&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>posi.leg=</span></b><b><span style='color:#ff54ff;'>&quot;topleft&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Same in BW</span></b>
<b><span style='color:#dede49;'>scatter</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.dapc3, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;DAPC, Taraxacum haussknechtii&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>15</span></b><b>:</b><b><span style='color:#ff54ff;'>17</span></b><b>)</b>, <b><span style='color:#54ff54;'>cell=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>cstar=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>solid=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2.5</span></b>, <b><span style='color:#54ff54;'>clab=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>grey.colors</span></b><b>(</b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>start=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>end=</span></b><b><span style='color:#ff54ff;'>0.8</span></b>, <b><span style='color:#54ff54;'>gamma=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>alpha=</span></b><b><span style='color:#ff54ff;'>0</span></b><b>)</b>, <b><span style='color:#54ff54;'>posi.da=</span></b><b><span style='color:#ff54ff;'>&quot;bottomleft&quot;</span></b>, <b><span style='color:#54ff54;'>scree.pca=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>posi.pca=</span></b><b><span style='color:#ff54ff;'>&quot;bottomright&quot;</span></b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>txt.leg=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Group 1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Group 2&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Group 3&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>posi.leg=</span></b><b><span style='color:#ff54ff;'>&quot;topleft&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Density functions - only for first axis here!</span></b>
<b><span style='color:#dede49;'>scatter</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.dapc3, <b><span style='color:#54ff54;'>xax=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>yax=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;DAPC&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b>, <b><span style='color:#54ff54;'>solid=</span></b><b><span style='color:#ff54ff;'>0.5</span></b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>txt.leg=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Group 1&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Group 2&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Group 3&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>posi.leg=</span></b><b><span style='color:#ff54ff;'>&quot;topleft&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Assignment of individuals to clusters</span></b>
<b><span style='color:#dede49;'>assignplot</span></b><b>(</b>hauss.dapc3<b>)</b>
<b><span style='color:#54ffff;'># Structure-like plot</span></b>
<b><span style='color:#dede49;'>compoplot</span></b><b>(</b>hauss.dapc3, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#ff54ff;'>&quot;Individuals&quot;</span></b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Loadingplot - alleles the most adding to separation of individuals</span></b>
<b><span style='color:#dede49;'>loadingplot</span></b><b>(</b>hauss.dapc3<b>$</b>var.contr<b>)</b>
<b><span style='color:#54ffff;'># alfa-score - according to number of PC axis</span></b>
<b><span style='color:#dede49;'>optim.a.score</span></b><b>(</b>hauss.dapc3<b>)</b>

<b><span style='color:#b00000;'>## SNP</span></b>

<b><span style='color:#54ffff;'># Plot of missing data (white) and number of 2nd alleles</span></b>
<b><span style='color:#dede49;'>glPlot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.genlight, <b><span style='color:#54ff54;'>legend=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>posi=</span></b><b><span style='color:#ff54ff;'>&quot;topleft&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Sum of the number of second allele in each SNP</span></b>
usflu.freq <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>glSum</span></b><b>(</b>usflu.genlight<b>)</b>
<b><span style='color:#54ffff;'># Plot distribution of (second) allele frequencies</span></b>
<b><span style='color:#dede49;'>hist</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.freq, <b><span style='color:#54ff54;'>proba=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;gold&quot;</span></b>, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#ff54ff;'>&quot;Allele frequencies&quot;</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Distribution of (second) allele frequencies&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>lines</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>density</span></b><b>(</b>usflu.freq<b>)</b><b>$</b>x, <b><span style='color:#54ff54;'>y=</span></b><b><span style='color:#dede49;'>density</span></b><b>(</b>usflu.freq<b>)</b><b>$</b>y<b>*</b><b><span style='color:#ff54ff;'>1.5</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>3</span></b> <b>)</b>
<b><span style='color:#54ffff;'># Mean number of second allele in each SNP</span></b>
usflu.mean <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>glMean</span></b><b>(</b>usflu.genlight<b>)</b>
usflu.mean <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>c</span></b><b>(</b>usflu.mean, <b><span style='color:#ff54ff;'>1</span></b><b>-</b>usflu.mean<b>)</b>
<b><span style='color:#54ffff;'># Plot distribution of allele frequencies - play with parameters to get nice image</span></b>
<b><span style='color:#dede49;'>hist</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.mean, <b><span style='color:#54ff54;'>proba=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;darkseagreen3&quot;</span></b>, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#ff54ff;'>&quot;Allele frequencies&quot;</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Distribution of allele frequencies&quot;</span></b>, <b><span style='color:#54ff54;'>nclass=</span></b><b><span style='color:#ff54ff;'>20</span></b><b>)</b>
<b><span style='color:#dede49;'>lines</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>density</span></b><b>(</b>usflu.mean, <b><span style='color:#54ff54;'>bw=</span></b><b><span style='color:#ff54ff;'>0.05</span></b><b>)</b><b>$</b>x, <b><span style='color:#54ff54;'>y=</span></b><b><span style='color:#dede49;'>density</span></b><b>(</b>usflu.mean, <b><span style='color:#54ff54;'>bw=</span></b><b><span style='color:#ff54ff;'>0.05</span></b><b>)</b><b>$</b>y<b>*</b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>

<b><span style='color:#54ffff;'># Number of missing values in each locus</span></b>
usflu.na.density <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>density</span></b><b>(</b><b><span style='color:#dede49;'>glNA</span></b><b>(</b>usflu.genlight<b>)</b>, <b><span style='color:#54ff54;'>bw=</span></b><b><span style='color:#ff54ff;'>10</span></b><b>)</b> <b><span style='color:#54ffff;'># Play with bw parameter to get optimal image</span></b>
<b><span style='color:#54ffff;'># Set range of xlim parameter from 0 to the length of original alignment</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.na.density, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;n&quot;</span></b>, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#ff54ff;'>&quot;Position in the alignment&quot;</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Location of the missing values (NA)&quot;</span></b>, <b><span style='color:#54ff54;'>xlim=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#ff54ff;'>1701</span></b><b>))</b>
<b><span style='color:#dede49;'>polygon</span></b><b>(</b><b><span style='color:#dede49;'>c</span></b><b>(</b>usflu.na.density<b>$</b>x, <b><span style='color:#dede49;'>rev</span></b><b>(</b>usflu.na.density<b>$</b>x<b>))</b>, <b><span style='color:#dede49;'>c</span></b><b>(</b>usflu.na.density<b>$</b>y, <b><span style='color:#dede49;'>rep</span></b><b>(</b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#dede49;'>length</span></b><b>(</b>usflu.na.density<b>$</b>x<b>)))</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>transp</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;blue&quot;</span></b>, <b><span style='color:#54ff54;'>alpha=</span></b><b><span style='color:#ff54ff;'>0.3</span></b><b>))</b>
<b><span style='color:#dede49;'>points</span></b><b>(</b><b><span style='color:#dede49;'>glNA</span></b><b>(</b>usflu.genlight<b>)</b>, <b><span style='color:#dede49;'>rep</span></b><b>(</b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#dede49;'>nLoc</span></b><b>(</b>usflu.genlight<b>))</b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>&quot;|&quot;</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;blue&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># PCA - select number of retained PC axes, about 10 here</span></b>
usflu.pca <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>glPca</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.genlight, <b><span style='color:#54ff54;'>center=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>scale=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>loadings=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Plot PCA</span></b>
<b><span style='color:#dede49;'>scatter.glPca</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.pca, <b><span style='color:#54ff54;'>posi=</span></b><b><span style='color:#ff54ff;'>&quot;bottomright&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;PCA of the US influenza data&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Coloured plot</span></b>
<b><span style='color:#dede49;'>colorplot</span></b><b>(</b>usflu.pca<b>$</b>scores, usflu.pca<b>$</b>scores, <b><span style='color:#54ff54;'>transp=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>4</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;PCA of the US influenza data&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>abline</span></b><b>(</b><b><span style='color:#54ff54;'>h=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>v=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;grey&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>add.scatter.eig</span></b><b>(</b>usflu.pca<b>$</b>eig<b>[</b><b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>40</span></b><b>]</b>, <b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>posi=</span></b><b><span style='color:#ff54ff;'>&quot;topright&quot;</span></b>, <b><span style='color:#54ff54;'>inset=</span></b><b><span style='color:#ff54ff;'>0.05</span></b>, <b><span style='color:#54ff54;'>ratio=</span></b><b><span style='color:#ff54ff;'>0.3</span></b><b>)</b>
<b><span style='color:#54ffff;'># Calculate phylogenetic tree</span></b>
usflu.tree.genlight <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>nj</span></b><b>(</b><b><span style='color:#dede49;'>dist</span></b><b>(</b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>usflu.genlight<b>)))</b>
<b><span style='color:#54ffff;'># Plot colored phylogenetic tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.tree.genlight, <b><span style='color:#54ff54;'>typ=</span></b><b><span style='color:#ff54ff;'>&quot;fan&quot;</span></b>, <b><span style='color:#54ff54;'>show.tip=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#dede49;'>tiplabels</span></b><b>(</b><b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>20</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>num2col</span></b><b>(</b>usflu.annot<b>[[</b><b><span style='color:#ff54ff;'>&quot;year&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>col.pal=</span></b>usflu.pal<b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>4</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;NJ tree of the US influenza data&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add legend</span></b>
<b><span style='color:#dede49;'>legend</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>&quot;topright&quot;</span></b>, <b><span style='color:#54ff54;'>fill=</span></b><b><span style='color:#dede49;'>num2col</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>pretty</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>1993</span></b><b>:</b><b><span style='color:#ff54ff;'>2008</span></b>, <b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b>, <b><span style='color:#54ff54;'>col.pal=</span></b>usflu.pal<b>)</b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#dede49;'>pretty</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>1993</span></b><b>:</b><b><span style='color:#ff54ff;'>2008</span></b>, <b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b>, <b><span style='color:#54ff54;'>ncol=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>

<b><span style='color:#b00000;'>## Moran's I</span></b>
<b><span style='color:#54ffff;'># Load required library</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>spdep<b>)</b>
<b><span style='color:#54ffff;'># Creates connection network</span></b>
hauss.connectivity <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>chooseCN</span></b><b>(</b><b><span style='color:#54ff54;'>xy=</span></b>hauss.genind<b>$</b>other<b>$</b>xy, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>5</span></b>, <b><span style='color:#54ff54;'>d1=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>d2=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>plot.nb=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>result.type=</span></b><b><span style='color:#ff54ff;'>&quot;listw&quot;</span></b>, <b><span style='color:#54ff54;'>edit.nb=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
hauss.connectivity
<b><span style='color:#54ffff;'># Test of Moran's I for 1st PCoA axis</span></b>
<b><span style='color:#dede49;'>moran.test</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.pcoa<b>$</b>li<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>]</b>, <b><span style='color:#54ff54;'>listw=</span></b>hauss.connectivity, <b><span style='color:#54ff54;'>alternative=</span></b><b><span style='color:#ff54ff;'>&quot;greater&quot;</span></b>, <b><span style='color:#54ff54;'>randomisation=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
hauss.pcoa1.mctest <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>moran.mc</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.pcoa<b>[[</b><b><span style='color:#ff54ff;'>&quot;li&quot;</span></b><b>]][</b>,<b><span style='color:#ff54ff;'>1</span></b><b>]</b>, <b><span style='color:#54ff54;'>listw=</span></b>hauss.connectivity, <b><span style='color:#54ff54;'>alternative=</span></b><b><span style='color:#ff54ff;'>&quot;greater&quot;</span></b>, <b><span style='color:#54ff54;'>nsim=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>)</b>
hauss.pcoa1.mctest
<b><span style='color:#54ffff;'># Plot the results</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>hauss.pcoa1.mctest<b>)</b> <b><span style='color:#54ffff;'># Plot of densitiy of permutations</span></b>
<b><span style='color:#dede49;'>moran.plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.pcoa<b>$</b>li<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>]</b>, <b><span style='color:#54ff54;'>listw=</span></b>hauss.connectivity<b>)</b> <b><span style='color:#54ffff;'># PC plot</span></b>
<b><span style='color:#54ffff;'># Test of Moran's I for 2nd PCoA axis</span></b>
<b><span style='color:#dede49;'>moran.test</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.pcoa<b>$</b>li<b>[</b>,<b><span style='color:#ff54ff;'>2</span></b><b>]</b>, <b><span style='color:#54ff54;'>listw=</span></b>hauss.connectivity, <b><span style='color:#54ff54;'>alternative=</span></b><b><span style='color:#ff54ff;'>&quot;greater&quot;</span></b>, <b><span style='color:#54ff54;'>randomisation=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
hauss.pcoa2.mctest <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>moran.mc</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.pcoa<b>$</b>li<b>[</b>,<b><span style='color:#ff54ff;'>2</span></b><b>]</b>, <b><span style='color:#54ff54;'>listw=</span></b>hauss.connectivity, <b><span style='color:#54ff54;'>alternative=</span></b><b><span style='color:#ff54ff;'>&quot;greater&quot;</span></b>, <b><span style='color:#54ff54;'>nsim=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>)</b>
hauss.pcoa2.mctest
<b><span style='color:#dede49;'>plot</span></b><b>(</b>hauss.pcoa2.mctest<b>)</b> <b><span style='color:#54ffff;'># Plot of densitiy of permutations</span></b>
<b><span style='color:#dede49;'>moran.plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.pcoa<b>[[</b><b><span style='color:#ff54ff;'>&quot;li&quot;</span></b><b>]][</b>,<b><span style='color:#ff54ff;'>2</span></b><b>]</b>, <b><span style='color:#54ff54;'>listw=</span></b>hauss.connectivity<b>)</b> <b><span style='color:#54ffff;'># PC plot</span></b>

<b><span style='color:#b00000;'>## sPCA</span></b>

<b><span style='color:#54ffff;'># Explore various networks</span></b>
<b><span style='color:#dede49;'>data</span></b><b>(</b>rupica<b>)</b>
<b><span style='color:#54ffff;'># Try various settings for chooseCN (type=X) - type 1-4 as there are identical coordinates (multiple sampling from same locality)</span></b>
<b><span style='color:#dede49;'>chooseCN</span></b><b>(</b><b><span style='color:#54ff54;'>xy=</span></b>rupica<b>$</b>other<b>$</b>xy, <b><span style='color:#54ff54;'>ask=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>5</span></b><b>/</b><b><span style='color:#ff54ff;'>6</span></b><b>/</b><b><span style='color:#ff54ff;'>7</span></b>, <b><span style='color:#54ff54;'>plot.nb=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>edit.nb=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, ...<b>)</b> <b><span style='color:#54ffff;'># Play with settings little bit...</span></b>
?chooseCN <b><span style='color:#54ffff;'># See for more details - select the best &quot;type&quot; for your data</span></b>

<b><span style='color:#54ffff;'># Calculates sPCA - here is only 1 positive and 3 negative factors</span></b>
hauss.spca <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>spca</span></b><b>(</b><b><span style='color:#54ff54;'>obj=</span></b>hauss.genind, <b><span style='color:#54ff54;'>cn=</span></b>hauss.connectivity, <b><span style='color:#54ff54;'>scale=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>scannf=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Plot eigenvalues of sPCA - gobal vs. local structure</span></b>
<b><span style='color:#dede49;'>barplot</span></b><b>(</b><b><span style='color:#54ff54;'>height=</span></b>hauss.spca<b>$</b>eig, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Eigenvalues of sPCA&quot;</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>spectral</span></b><b>(</b><b><span style='color:#dede49;'>length</span></b><b>(</b>hauss.spca<b>[[</b><b><span style='color:#ff54ff;'>&quot;eig&quot;</span></b><b>]])))</b>
<b><span style='color:#dede49;'>legend</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;topright&quot;</span></b>, <b><span style='color:#54ff54;'>fill=</span></b><b><span style='color:#dede49;'>spectral</span></b><b>(</b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Global structures&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Local structures&quot;</span></b><b>))</b> <b><span style='color:#54ffff;'># Add legend</span></b>
<b><span style='color:#dede49;'>abline</span></b><b>(</b><b><span style='color:#54ff54;'>h=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;grey&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Add line showing zero</span></b>
<b><span style='color:#54ffff;'># Information about sPCA</span></b>
<b><span style='color:#dede49;'>print.spca</span></b><b>(</b>hauss.spca<b>)</b>
<b><span style='color:#54ffff;'># Summary of sPCA results</span></b>
<b><span style='color:#dede49;'>summary.spca</span></b><b>(</b>hauss.spca<b>)</b>
<b><span style='color:#54ffff;'># Shows connectivity network, 3 different scores, barplot of eigenvalues and eigenvalues decomposition</span></b>
<b><span style='color:#dede49;'>plot.spca</span></b><b>(</b>hauss.spca<b>)</b>
<b><span style='color:#54ffff;'># Display of scores in color canals (two analogous variants)</span></b>
<b><span style='color:#dede49;'>colorplot.spca</span></b><b>(</b>hauss.spca, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;sPCA - colorplot of PC 1 and 2 (lagged scores)&quot;</span></b>, <b><span style='color:#54ff54;'>line=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#dede49;'>colorplot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind<b>$</b>other<b>$</b>xy, hauss.spca<b>$</b>ls, <b><span style='color:#54ff54;'>axes=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#dede49;'>ncol</span></b><b>(</b>hauss.spca<b>$</b>li<b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;sPCA - colorplot of PC 1 and 2 (lagged scores)&quot;</span></b>, <b><span style='color:#54ff54;'>line=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#54ffff;'># Spatial and variance components of the eigenvalues</span></b>
<b><span style='color:#dede49;'>screeplot.spca</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.spca, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#00ff00;'>NULL</span></b><b>)</b>

<b><span style='color:#54ffff;'># Test if global/local structure is significant</span></b>
hauss.spca.glo <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>global.rtest</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>hauss.genind<b>$</b>tab, <b><span style='color:#54ff54;'>listw=</span></b>hauss.spca<b>$</b>lw, <b><span style='color:#54ff54;'>nperm=</span></b><b><span style='color:#ff54ff;'>999</span></b><b>)</b>
hauss.spca.glo
<b><span style='color:#dede49;'>plot</span></b><b>(</b>hauss.spca.glo<b>)</b>
hauss.spca.loc <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>local.rtest</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>hauss.genind<b>$</b>tab, <b><span style='color:#54ff54;'>listw=</span></b>hauss.spca<b>$</b>lw, <b><span style='color:#54ff54;'>nperm=</span></b><b><span style='color:#ff54ff;'>999</span></b><b>)</b>
hauss.spca.loc
<b><span style='color:#dede49;'>plot</span></b><b>(</b>hauss.spca.loc<b>)</b>

<b><span style='color:#54ffff;'># Map of genetic clines</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>akima<b>)</b> <b><span style='color:#54ffff;'># This library is needed for some manipulation with coordinates</span></b>
hauss.spca.temp <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>interp</span></b><b>(</b><b><span style='color:#dede49;'>other</span></b><b>(</b>hauss.genind<b>)</b><b>$</b>xy<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>]</b>, <b><span style='color:#dede49;'>other</span></b><b>(</b>hauss.genind<b>)</b><b>$</b>xy<b>[</b>,<b><span style='color:#ff54ff;'>2</span></b><b>]</b>, hauss.spca<b>$</b>ls<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>]</b>, <b><span style='color:#54ff54;'>xo=</span></b><b><span style='color:#dede49;'>seq</span></b><b>(</b><b><span style='color:#dede49;'>min</span></b><b>(</b><b><span style='color:#dede49;'>other</span></b><b>(</b>hauss.genind<b>)</b><b>$</b>xy<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>])</b>, <b><span style='color:#dede49;'>max</span></b><b>(</b><b><span style='color:#dede49;'>other</span></b><b>(</b>hauss.genind<b>)</b><b>$</b>xy<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>])</b>, <b><span style='color:#54ff54;'>le=</span></b><b><span style='color:#ff54ff;'>200</span></b><b>)</b>, <b><span style='color:#54ff54;'>yo=</span></b><b><span style='color:#dede49;'>seq</span></b><b>(</b><b><span style='color:#dede49;'>min</span></b><b>(</b><b><span style='color:#dede49;'>other</span></b><b>(</b>hauss.genind<b>)</b><b>$</b>xy<b>[</b>,<b><span style='color:#ff54ff;'>2</span></b><b>])</b>, <b><span style='color:#dede49;'>max</span></b><b>(</b><b><span style='color:#dede49;'>other</span></b><b>(</b>hauss.genind<b>)</b><b>$</b>xy<b>[</b>,<b><span style='color:#ff54ff;'>2</span></b><b>])</b>, <b><span style='color:#54ff54;'>le=</span></b><b><span style='color:#ff54ff;'>200</span></b><b>)</b>, <b><span style='color:#54ff54;'>duplicate=</span></b><b><span style='color:#ff54ff;'>&quot;median&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># For 1st axis</span></b>
<b><span style='color:#dede49;'>image</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.spca.temp, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>spectral</span></b><b>(</b><b><span style='color:#ff54ff;'>100</span></b><b>))</b>
<b><span style='color:#dede49;'>s.value</span></b><b>(</b><b><span style='color:#54ff54;'>dfxy=</span></b>hauss.genind<b>$</b>other<b>$</b>xy, <b><span style='color:#54ff54;'>z=</span></b>hauss.pcoa<b>$</b>li<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>]</b>, <b><span style='color:#54ff54;'>add.p=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>csize=</span></b><b><span style='color:#ff54ff;'>0.5</span></b>, <b><span style='color:#54ff54;'>sub=</span></b><b><span style='color:#ff54ff;'>&quot;PCoA - first PC&quot;</span></b>, <b><span style='color:#54ff54;'>csub=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>possub=</span></b><b><span style='color:#ff54ff;'>&quot;topleft&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># For 2nd axis</span></b>
<b><span style='color:#dede49;'>image</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.spca.temp, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>spectral</span></b><b>(</b><b><span style='color:#ff54ff;'>100</span></b><b>))</b>
<b><span style='color:#dede49;'>s.value</span></b><b>(</b><b><span style='color:#54ff54;'>dfxy=</span></b>hauss.genind<b>$</b>other<b>$</b>xy, <b><span style='color:#54ff54;'>z=</span></b>hauss.pcoa<b>[[</b><b><span style='color:#ff54ff;'>&quot;li&quot;</span></b><b>]][</b>,<b><span style='color:#ff54ff;'>2</span></b><b>]</b>, <b><span style='color:#54ff54;'>add.p=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>csize=</span></b><b><span style='color:#ff54ff;'>0.5</span></b>, <b><span style='color:#54ff54;'>sub=</span></b><b><span style='color:#ff54ff;'>&quot;PCoA - second PC&quot;</span></b>, <b><span style='color:#54ff54;'>csub=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>possub=</span></b><b><span style='color:#ff54ff;'>&quot;topleft&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Interpolated lagged score on a map</span></b>
hauss.spca.annot <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ffff54;'>function</span></b><b>()</b> <b>{</b>
	<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;sPCA - interpolated map of individual scores&quot;</span></b><b>)</b>
	<b><span style='color:#dede49;'>points</span></b><b>(</b><b><span style='color:#dede49;'>other</span></b><b>(</b>hauss.genind<b>)</b><b>$</b>xy<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>]</b>, <b><span style='color:#dede49;'>other</span></b><b>(</b>hauss.genind<b>)</b><b>$</b>xy<b>[</b>,<b><span style='color:#ff54ff;'>2</span></b><b>])</b>
	<b>}</b>
<b><span style='color:#dede49;'>filled.contour</span></b><b>(</b>hauss.spca.temp, <b><span style='color:#54ff54;'>color.pal=</span></b><b><span style='color:#dede49;'>colorRampPalette</span></b><b>(</b><b><span style='color:#dede49;'>lightseasun</span></b><b>(</b><b><span style='color:#ff54ff;'>100</span></b><b>))</b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>20</span></b>, <b><span style='color:#54ff54;'>nlevels=</span></b><b><span style='color:#ff54ff;'>100</span></b>, <b><span style='color:#54ff54;'>key.title=</span></b><b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Lagged</span></b><b><span style='color:#ff54ff;'>\n</span></b><b><span style='color:#ff54ff;'>score 1&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>plot.title=</span></b><b><span style='color:#dede49;'>hauss.spca.annot</span></b><b>())</b>

<b><span style='color:#54ffff;'># Loading plots - which alleles contribute the most?</span></b>
hauss.spca.loadings <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>hauss.spca<b>$</b>c1<b>[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>]</b><b>^</b><b><span style='color:#ff54ff;'>2</span></b>
<b><span style='color:#dede49;'>names</span></b><b>(</b>hauss.spca.loadings<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>rownames</span></b><b>(</b>hauss.spca<b>$</b>c1<b>)</b>
<b><span style='color:#dede49;'>loadingplot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.spca.loadings, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#ff54ff;'>&quot;Alleles&quot;</span></b>, <b><span style='color:#54ff54;'>ylab=</span></b><b><span style='color:#ff54ff;'>&quot;Weight of the alleles&quot;</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Contribution of alleles to the first sPCA axis&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>boxplot</span></b><b>(</b><b><span style='color:#54ff54;'>formula=</span></b>hauss.spca.loadings<b>~</b>hauss.genind<b>$</b>loc.fac, <b><span style='color:#54ff54;'>las=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>ylab=</span></b><b><span style='color:#ff54ff;'>&quot;Contribution&quot;</span></b>, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#ff54ff;'>&quot;Marker&quot;</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Contribution by markers into the first global score&quot;</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;grey&quot;</span></b><b>)</b>

<b><span style='color:#b00000;'>## Monmonier - genetic boundaries</span></b>
<b><span style='color:#54ffff;'># It requires every point to have unique coordinates (one can use jitter() or difference in scale of meters). Example here is on population level, which is not ideal.</span></b>
<b><span style='color:#54ffff;'># Calculates Monmonier's function (for threshold use 'd')</span></b>
hauss.monmonier <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>monmonier</span></b><b>(</b><b><span style='color:#54ff54;'>xy=</span></b>hauss.genpop<b>$</b>other<b>$</b>xy, <b><span style='color:#54ff54;'>dist=</span></b><b><span style='color:#dede49;'>dist</span></b><b>(</b>hauss.genpop<b>$</b>tab<b>)</b>, <b><span style='color:#54ff54;'>cn=</span></b><b><span style='color:#dede49;'>chooseCN</span></b><b>(</b>hauss.genpop<b>$</b>other<b>$</b>xy, <b><span style='color:#54ff54;'>ask=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>plot.nb=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>edit.nb=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>, <b><span style='color:#54ff54;'>nrun=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>
<b><span style='color:#dede49;'>coords.monmonier</span></b><b>(</b>hauss.monmonier<b>)</b> <b><span style='color:#54ffff;'># See result as text</span></b>
<b><span style='color:#54ffff;'># Plot genetic boundaries</span></b>
<b><span style='color:#dede49;'>plot.monmonier</span></b><b>(</b>hauss.monmonier, <b><span style='color:#54ff54;'>add.arrows=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>bwd=</span></b><b><span style='color:#ff54ff;'>10</span></b>, <b><span style='color:#54ff54;'>sub=</span></b><b><span style='color:#ff54ff;'>&quot;Monmonier plot&quot;</span></b>, <b><span style='color:#54ff54;'>csub=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>points</span></b><b>(</b>hauss.genpop<b>$</b>other<b>$</b>xy, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2.5</span></b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>20</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>text</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genpop<b>$</b>other<b>$</b>xy<b>$</b>lon, <b><span style='color:#54ff54;'>y=</span></b>hauss.genpop<b>$</b>other<b>$</b>xy<b>$</b>lat, <b><span style='color:#54ff54;'>labels=</span></b><b><span style='color:#dede49;'>popNames</span></b><b>(</b>hauss.genpop<b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#dede49;'>legend</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;bottomright&quot;</span></b>, <b><span style='color:#54ff54;'>leg=</span></b><b><span style='color:#ff54ff;'>&quot;Genetic boundaries</span></b><b><span style='color:#ff54ff;'>\n</span></b><b><span style='color:#ff54ff;'>among populations&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># For plotting see</span></b>
?points
?text
?legend

<b><span style='color:#b00000;'>## Mantel test - isolation by distance</span></b>
<b><span style='color:#54ffff;'># Geographical distance</span></b>
hauss.gdist <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dist</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genind<b>$</b>other<b>$</b>xy, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;euclidean&quot;</span></b>, <b><span style='color:#54ff54;'>diag=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>upper=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Mantel test</span></b>
hauss.mantel <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>mantel.randtest</span></b><b>(</b><b><span style='color:#54ff54;'>m1=</span></b>hauss.dist, <b><span style='color:#54ff54;'>m2=</span></b>hauss.gdist, <b><span style='color:#54ff54;'>nrepet=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>)</b>
hauss.mantel <b><span style='color:#54ffff;'># See text output</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>hauss.mantel, <b><span style='color:#54ff54;'>nclass=</span></b><b><span style='color:#ff54ff;'>30</span></b><b>)</b>
<b><span style='color:#54ffff;'># Libraries required by mantel.correlog:</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>permute<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>lattice<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>vegan<b>)</b>
<b><span style='color:#54ffff;'># Different implementation of Mantel test testing distance classes</span></b>
hauss.mantel.cor <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>mantel.correlog</span></b><b>(</b><b><span style='color:#54ff54;'>D.eco=</span></b>hauss.dist, <b><span style='color:#54ff54;'>D.geo=</span></b>hauss.gdist, <b><span style='color:#54ff54;'>XY=</span></b><b><span style='color:#00ff00;'>NULL</span></b>, <b><span style='color:#54ff54;'>n.class=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>break.pts=</span></b><b><span style='color:#00ff00;'>NULL</span></b>, <b><span style='color:#54ff54;'>cutoff=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>r.type=</span></b><b><span style='color:#ff54ff;'>&quot;pearson&quot;</span></b>, <b><span style='color:#54ff54;'>nperm=</span></b><b><span style='color:#ff54ff;'>1000</span></b>, <b><span style='color:#54ff54;'>mult=</span></b><b><span style='color:#ff54ff;'>&quot;holm&quot;</span></b>, <b><span style='color:#54ff54;'>progressive=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># See results for respective classes</span></b>
hauss.mantel.cor
<b><span style='color:#dede49;'>summary</span></b><b>(</b>hauss.mantel.cor<b>)</b>
<b><span style='color:#54ffff;'># Plot it</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>hauss.mantel.cor<b>)</b>

<b><span style='color:#b00000;'>## Geneland</span></b>
<b><span style='color:#54ffff;'># Haploid and diploid codominant markers (microsattelites or SNPs)</span></b>
<b><span style='color:#54ffff;'># Load needed libraries</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>PBSmapping<b>)</b> <b><span style='color:#54ffff;'># Required to transform coordinates</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>Geneland<b>)</b>
<b><span style='color:#54ffff;'># Graphical interface available for Geneland - we will use only command line</span></b>
<b><span style='color:#dede49;'>Geneland.GUI</span></b><b>()</b>
<b><span style='color:#54ffff;'># Loading and conversions of coordinates - Geneland requires specific coordinate space</span></b>
hauss.geneland.coord <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b>hauss.coord<b>)</b> <b><span style='color:#54ffff;'># hauss.cord is DF, we need just plain matrix</span></b>
<b><span style='color:#dede49;'>colnames</span></b><b>(</b>hauss.geneland.coord<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;X&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Y&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>attr</span></b><b>(</b>hauss.geneland.coord, <b><span style='color:#ff54ff;'>&quot;projection&quot;</span></b><b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> &quot;LL&quot;</span></b>
<b><span style='color:#dede49;'>attr</span></b><b>(</b>hauss.geneland.coord, <b><span style='color:#ff54ff;'>&quot;zone&quot;</span></b><b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#00ff00;'>NA</span></b>
hauss.geneland.coord.utm <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>convUL</span></b><b>(</b>hauss.geneland.coord<b>)</b>
<b><span style='color:#dede49;'>dim</span></b><b>(</b>hauss.geneland.coord<b>)</b>
hauss.geneland.coord
<b><span style='color:#dede49;'>dim</span></b><b>(</b>hauss.geneland.coord.utm<b>)</b>
hauss.geneland.coord.utm <b><span style='color:#54ffff;'># Final coordinates</span></b>
<b><span style='color:#54ffff;'># Load data (only haploid or diploid data are supported) - only plain table with alleles</span></b>
hauss.geneland.data <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.table</span></b><b>(</b><b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;https://soubory.trapa.cz/rcourse/haussknechtii_geneland.txt&quot;</span></b>, <b><span style='color:#54ff54;'>na.string=</span></b><b><span style='color:#ff54ff;'>&quot;-999&quot;</span></b>, <b><span style='color:#54ff54;'>header=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;</span></b><b><span style='color:#ff54ff;'>\t</span></b><b><span style='color:#ff54ff;'>&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>dim</span></b><b>(</b>hauss.geneland.data<b>)</b>
hauss.geneland.data
<b><span style='color:#54ffff;'># Set number of independent runs</span></b>
hauss.geneland.nrun <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>5</span></b>
<b><span style='color:#54ffff;'># Set length of burnin chain</span></b>
hauss.geneland.burnin <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>100</span></b>
<b><span style='color:#54ffff;'># Set maximal K (number of populations)</span></b>
hauss.geneland.maxpop <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>10</span></b>
<b><span style='color:#54ffff;'># FOR loop will run several independent runs and produce output maps of genetic clusters - outputs are written into subdirectory within geneland directory</span></b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>hauss.geneland.irun <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b>hauss.geneland.nrun<b>)</b> <b>{</b>
	hauss.geneland.path.mcmc <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>paste</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;geneland/&quot;</span></b>, hauss.geneland.irun, <b><span style='color:#ff54ff;'>&quot;/&quot;</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># paste is good especially for joining several text chains into one</span></b>
	<b><span style='color:#54ffff;'># On Windows, remove following line and create subdirectories from 1 to max K manually (creating subdirs in Windows in R is complicated)</span></b>
	<b><span style='color:#dede49;'>system</span></b><b>(</b><b><span style='color:#dede49;'>paste</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;mkdir &quot;</span></b>, hauss.geneland.path.mcmc<b>))</b> <b><span style='color:#54ffff;'># Creates subdirectory</span></b>
	<b><span style='color:#54ffff;'># Inferrence - MCMC chain - see ?MCMC for details</span></b>
	<b><span style='color:#54ffff;'># In practice set much higher number of iterations (nit, millions), appropriate sampling (thinning, thousands) and longer burnin</span></b>
	<b><span style='color:#dede49;'>MCMC</span></b><b>(</b><b><span style='color:#54ff54;'>coordinates=</span></b>hauss.geneland.coord.utm, <b><span style='color:#54ff54;'>geno.dip.codom=</span></b>hauss.geneland.data, <b><span style='color:#54ff54;'>path.mcmc=</span></b>hauss.geneland.path.mcmc, <b><span style='color:#54ff54;'>delta.coord=</span></b><b><span style='color:#ff54ff;'>0.001</span></b>, <b><span style='color:#54ff54;'>varnpop=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>npopmin=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>npopmax=</span></b>hauss.geneland.maxpop, <b><span style='color:#54ff54;'>nit=</span></b><b><span style='color:#ff54ff;'>10000</span></b>, <b><span style='color:#54ff54;'>thinning=</span></b><b><span style='color:#ff54ff;'>10</span></b>, <b><span style='color:#54ff54;'>freq.model=</span></b><b><span style='color:#ff54ff;'>&quot;Uncorrelated&quot;</span></b>, <b><span style='color:#54ff54;'>spatial=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
	<b><span style='color:#54ffff;'># Post-process chains</span></b>
	<b><span style='color:#dede49;'>PostProcessChain</span></b><b>(</b><b><span style='color:#54ff54;'>coordinates=</span></b>hauss.geneland.coord.utm, <b><span style='color:#54ff54;'>path.mcmc=</span></b>hauss.geneland.path.mcmc, <b><span style='color:#54ff54;'>nxdom=</span></b><b><span style='color:#ff54ff;'>500</span></b>, <b><span style='color:#54ff54;'>nydom=</span></b><b><span style='color:#ff54ff;'>500</span></b>, <b><span style='color:#54ff54;'>burnin=</span></b>hauss.geneland.burnin<b>)</b>
	<b><span style='color:#54ffff;'># Output</span></b>
	<b><span style='color:#54ffff;'># Simulated number of populations</span></b>
	<b><span style='color:#dede49;'>Plotnpop</span></b><b>(</b><b><span style='color:#54ff54;'>path.mcmc=</span></b>hauss.geneland.path.mcmc, <b><span style='color:#54ff54;'>printit=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#dede49;'>paste</span></b><b>(</b>hauss.geneland.path.mcmc, <b><span style='color:#ff54ff;'>&quot;/geneland-number_of_clusters.pdf&quot;</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>format=</span></b><b><span style='color:#ff54ff;'>&quot;pdf&quot;</span></b>, <b><span style='color:#54ff54;'>burnin=</span></b>hauss.geneland.burnin<b>)</b>
	<b><span style='color:#dede49;'>dev.off</span></b><b>()</b> <b><span style='color:#54ffff;'># We must close graphical device manually</span></b>
	<b><span style='color:#54ffff;'># Map of estimated population membership</span></b>
	<b><span style='color:#dede49;'>PosteriorMode</span></b><b>(</b><b><span style='color:#54ff54;'>coordinates=</span></b>hauss.geneland.coord.utm, <b><span style='color:#54ff54;'>path.mcmc=</span></b>hauss.geneland.path.mcmc, <b><span style='color:#54ff54;'>printit=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>format=</span></b><b><span style='color:#ff54ff;'>&quot;pdf&quot;</span></b>, <b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#dede49;'>paste</span></b><b>(</b>hauss.geneland.path.mcmc,<b><span style='color:#ff54ff;'>&quot;/geneland-map.pdf&quot;</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b><b>))</b>
	<b><span style='color:#dede49;'>dev.off</span></b><b>()</b> <b><span style='color:#54ffff;'># We must close graphical device manually</span></b>
	<b>}</b>
<b><span style='color:#54ffff;'># Prepare list to record values of Fst for all runs</span></b>
hauss.geneland.fstat <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>list</span></b><b>()</b>
<b><span style='color:#54ffff;'># Estimate Fst</span></b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>hauss.geneland.irun <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b>hauss.geneland.nrun<b>)</b> <b>{</b>
	hauss.geneland.path.mcmc <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>paste</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;geneland/&quot;</span></b>, hauss.geneland.irun, <b><span style='color:#ff54ff;'>&quot;/&quot;</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b><b>)</b>
	<b><span style='color:#54ffff;'># F-statistics - Fis and Fst</span></b>
	hauss.geneland.fstat<b>[[</b>hauss.geneland.irun<b>]]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>Fstat.output</span></b><b>(</b><b><span style='color:#54ff54;'>coordinates=</span></b>hauss.geneland.coord.utm, <b><span style='color:#54ff54;'>genotypes=</span></b>hauss.geneland.data, <b><span style='color:#54ff54;'>burnin=</span></b>hauss.geneland.burnin, <b><span style='color:#54ff54;'>ploidy=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>path.mcmc=</span></b>hauss.geneland.path.mcmc<b>)</b>
	<b>}</b>
<b><span style='color:#54ffff;'># Print Fst output</span></b>
hauss.geneland.fstat
<b><span style='color:#54ffff;'># MCMC inference under the admixture model</span></b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>hauss.geneland.irun <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b>hauss.geneland.nrun<b>)</b> <b>{</b>
	hauss.geneland.path.mcmc <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>paste</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;geneland/&quot;</span></b>, hauss.geneland.irun, <b><span style='color:#ff54ff;'>&quot;/&quot;</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b><b>)</b>
	hauss.geneland.path.mcmc.adm <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>paste</span></b><b>(</b>hauss.geneland.path.mcmc, <b><span style='color:#ff54ff;'>&quot;admixture&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;/&quot;</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b><b>)</b>
	<b><span style='color:#54ffff;'># On Windows, remove following line of code and create in each result directory (from 1 to max K) new subdirectory &quot;admixture&quot; (creating subdirs in Windows in R is complicated)</span></b>
	<b><span style='color:#dede49;'>system</span></b><b>(</b><b><span style='color:#dede49;'>paste</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;mkdir &quot;</span></b>, hauss.geneland.path.mcmc.adm<b>))</b>
	<b><span style='color:#dede49;'>HZ</span></b><b>(</b><b><span style='color:#54ff54;'>coordinates=</span></b>hauss.geneland.coord.utm, <b><span style='color:#54ff54;'>geno.dip.codom=</span></b>hauss.geneland.data, <b><span style='color:#54ff54;'>path.mcmc.noadm=</span></b>hauss.geneland.path.mcmc, <b><span style='color:#54ff54;'>nit=</span></b><b><span style='color:#ff54ff;'>10000</span></b>, <b><span style='color:#54ff54;'>thinning=</span></b><b><span style='color:#ff54ff;'>10</span></b>, <b><span style='color:#54ff54;'>path.mcmc.adm=</span></b>hauss.geneland.path.mcmc.adm<b>)</b>
	<b>}</b>
<b><span style='color:#54ffff;'># Produce maps of respective inferred clusters</span></b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>hauss.geneland.irun <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b>hauss.geneland.nrun<b>)</b> <b>{</b>
	hauss.geneland.path.mcmc <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>paste</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;geneland/&quot;</span></b>, hauss.geneland.irun, <b><span style='color:#ff54ff;'>&quot;/&quot;</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b><b>)</b>
	<b><span style='color:#54ffff;'># Maps - tesselations</span></b>
	<b><span style='color:#dede49;'>PlotTessellation</span></b><b>(</b><b><span style='color:#54ff54;'>coordinates=</span></b>hauss.geneland.coord.utm, <b><span style='color:#54ff54;'>path.mcmc=</span></b>hauss.geneland.path.mcmc, <b><span style='color:#54ff54;'>printit=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>path=</span></b>hauss.geneland.path.mcmc<b>)</b>
	<b><span style='color:#ffff54;'>for</span></b> <b>(</b>hauss.geneland.irun.img <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b>hauss.geneland.maxpop<b>)</b> <b>{</b> <b><span style='color:#dede49;'>dev.off</span></b><b>()</b> <b>}</b> <b><span style='color:#54ffff;'># We must close graphical device manually</span></b>
	<b>}</b>
<b><span style='color:#54ffff;'># Estimate frequencies of null alleles</span></b>
hauss.geneland.fna <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>list</span></b><b>()</b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>hauss.geneland.irun <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b>hauss.geneland.nrun<b>)</b> <b>{</b>
	hauss.geneland.path.mcmc <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>paste</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;geneland/&quot;</span></b>, hauss.geneland.irun, <b><span style='color:#ff54ff;'>&quot;/&quot;</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b><b>)</b>
	<b><span style='color:#54ffff;'># Estimation</span></b>
	hauss.geneland.fna<b>[[</b>hauss.geneland.irun<b>]]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>EstimateFreqNA</span></b><b>(</b><b><span style='color:#54ff54;'>path.mcmc=</span></b>hauss.geneland.path.mcmc<b>)</b>
	<b>}</b>
<b><span style='color:#54ffff;'># See output</span></b>
hauss.geneland.fna
<b><span style='color:#54ffff;'># Calculate average posterior probability</span></b>
hauss.geneland.lpd <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>rep</span></b><b>(</b><b><span style='color:#00ff00;'>NA</span></b>, hauss.geneland.nrun<b>)</b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>hauss.geneland.irun <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b>hauss.geneland.nrun<b>)</b> <b>{</b>
	hauss.geneland.path.mcmc <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>paste</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;geneland/&quot;</span></b>, hauss.geneland.irun, <b><span style='color:#ff54ff;'>&quot;/&quot;</span></b>, <b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b><b>)</b>
	hauss.geneland.path.lpd <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>paste</span></b><b>(</b>hauss.geneland.path.mcmc,<b><span style='color:#ff54ff;'>&quot;log.posterior.density.txt&quot;</span></b>,<b><span style='color:#54ff54;'>sep=</span></b><b><span style='color:#ff54ff;'>&quot;&quot;</span></b><b>)</b>
	hauss.geneland.lpd<b>[</b>hauss.geneland.irun<b>]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>mean</span></b><b>(</b><b><span style='color:#dede49;'>scan</span></b><b>(</b>hauss.geneland.path.lpd<b>)[</b><b>-</b><b>(</b><b><span style='color:#ff54ff;'>1</span></b><b>:</b>hauss.geneland.burnin<b>)])</b>
	<b>}</b>
<b><span style='color:#54ffff;'># Sorts runs according to decreasing posterior probability - the first one is the best</span></b>
<b><span style='color:#dede49;'>order</span></b><b>(</b>hauss.geneland.lpd, <b><span style='color:#54ff54;'>decreasing=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
hauss.geneland.lpd

<b><span style='color:#b00000;'>## Maps</span></b>

<b><span style='color:#54ffff;'># Load libraries</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>sp<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>rworldmap<b>)</b> <b><span style='color:#54ffff;'># Basic world maps</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>TeachingDemos<b>)</b> <b><span style='color:#54ffff;'># To be able to move text little bit</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>RgoogleMaps<b>)</b> <b><span style='color:#54ffff;'># Google and OpenStreetMaps</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>mapplots<b>)</b> <b><span style='color:#54ffff;'># Plot pie charts</span></b>
<b><span style='color:#54ffff;'># Plot basic map with state boundaries within selected range</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>getMap</span></b><b>(</b><b><span style='color:#54ff54;'>resolution=</span></b><b><span style='color:#ff54ff;'>&quot;high&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>xlim=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>19</span></b>, <b><span style='color:#ff54ff;'>24</span></b><b>)</b>, <b><span style='color:#54ff54;'>ylim=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>39</span></b>, <b><span style='color:#ff54ff;'>44</span></b><b>)</b>, <b><span style='color:#54ff54;'>asp=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#dede49;'>box</span></b><b>()</b> <b><span style='color:#54ffff;'># Add frame around the map</span></b>
<b><span style='color:#54ffff;'># Plot location points</span></b>
<b><span style='color:#dede49;'>points</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lon&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>y=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lat&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>15</span></b><b>:</b><b><span style='color:#ff54ff;'>19</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>4</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add text descriptions for points. Text is aside and with background</span></b>
<b><span style='color:#dede49;'>shadowtext</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lon&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>y=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lat&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>labels=</span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b><b><span style='color:#dede49;'>popNames</span></b><b>(</b>hauss.genind<b>))</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;black&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b>, <b><span style='color:#54ff54;'>theta=</span></b><b><span style='color:#dede49;'>seq</span></b><b>(</b>pi<b>/</b><b><span style='color:#ff54ff;'>4</span></b>, <b><span style='color:#ff54ff;'>2</span></b><b>*</b>pi, <b><span style='color:#54ff54;'>length.out=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b>, <b><span style='color:#54ff54;'>r=</span></b><b><span style='color:#ff54ff;'>0.15</span></b>, <b><span style='color:#54ff54;'>pos=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#ff54ff;'>4</span></b>, <b><span style='color:#ff54ff;'>4</span></b><b>)</b>, <b><span style='color:#54ff54;'>offset=</span></b><b><span style='color:#ff54ff;'>0.75</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#54ffff;'># Insert legend</span></b>
<b><span style='color:#dede49;'>legend</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>&quot;topright&quot;</span></b>, <b><span style='color:#54ff54;'>inset=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>/</b><b><span style='color:#ff54ff;'>50</span></b>, <b><span style='color:#54ff54;'>legend=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;He&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Oh&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Pr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Ne&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Sk&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>border=</span></b><b><span style='color:#ff54ff;'>&quot;black&quot;</span></b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>15</span></b><b>:</b><b><span style='color:#ff54ff;'>19</span></b>, <b><span style='color:#54ff54;'>pt.cex=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>bty=</span></b><b><span style='color:#ff54ff;'>&quot;o&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;lightgrey&quot;</span></b>, <b><span style='color:#54ff54;'>box.lwd=</span></b><b><span style='color:#ff54ff;'>1.5</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b>, <b><span style='color:#54ff54;'>title=</span></b><b><span style='color:#ff54ff;'>&quot;Populations&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># Google map is produced into a file. Parametr markers contain data frame with coordinates and possibly with another information</span></b>
hauss.gmap <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>GetMap</span></b><b>(</b><b><span style='color:#54ff54;'>center=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#54ff54;'>lat=</span></b><b><span style='color:#ff54ff;'>41</span></b>, <b><span style='color:#54ff54;'>lon=</span></b><b><span style='color:#ff54ff;'>21</span></b><b>)</b>, <b><span style='color:#54ff54;'>size=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>640</span></b>, <b><span style='color:#ff54ff;'>640</span></b><b>)</b>, <b><span style='color:#54ff54;'>destfile=</span></b><b><span style='color:#ff54ff;'>&quot;gmap.png&quot;</span></b>, <b><span style='color:#54ff54;'>zoom=</span></b><b><span style='color:#ff54ff;'>8</span></b>, <b><span style='color:#54ff54;'>markers=</span></b>hauss.coord, <b><span style='color:#54ff54;'>maptype=</span></b><b><span style='color:#ff54ff;'>&quot;terrain&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Plot saved map</span></b>
<b><span style='color:#dede49;'>PlotOnStaticMap</span></b><b>(</b><b><span style='color:#54ff54;'>MyMap=</span></b>hauss.gmap<b>)</b>
?PlotOnStaticMap <b><span style='color:#54ffff;'># See all options</span></b>
<b><span style='color:#54ffff;'># Other option of adding points and labels</span></b>
hauss.gmap2 <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>GetMap</span></b><b>(</b><b><span style='color:#54ff54;'>center=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#54ff54;'>lat=</span></b><b><span style='color:#ff54ff;'>41</span></b>, <b><span style='color:#54ff54;'>lon=</span></b><b><span style='color:#ff54ff;'>21</span></b><b>)</b>, <b><span style='color:#54ff54;'>size=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>640</span></b>, <b><span style='color:#ff54ff;'>640</span></b><b>)</b>, <b><span style='color:#54ff54;'>destfile=</span></b><b><span style='color:#ff54ff;'>&quot;gmap2.png&quot;</span></b>, <b><span style='color:#54ff54;'>zoom=</span></b><b><span style='color:#ff54ff;'>8</span></b>, <b><span style='color:#54ff54;'>maptype=</span></b><b><span style='color:#ff54ff;'>&quot;satellite&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>PlotOnStaticMap</span></b><b>(</b><b><span style='color:#54ff54;'>MyMap=</span></b>hauss.gmap2, <b><span style='color:#54ff54;'>lat=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lat&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>lon=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lon&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>FUN=</span></b>points, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>19</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;blue&quot;</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#dede49;'>PlotOnStaticMap</span></b><b>(</b><b><span style='color:#54ff54;'>MyMap=</span></b>hauss.gmap2, <b><span style='color:#54ff54;'>lat=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lat&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>lon=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lon&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>add=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>FUN=</span></b>points, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>19</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>
<b><span style='color:#dede49;'>PlotOnStaticMap</span></b><b>(</b><b><span style='color:#54ff54;'>MyMap=</span></b>hauss.gmap2, <b><span style='color:#54ff54;'>lat=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lat&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>lon=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lon&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>add=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>FUN=</span></b>text, <b><span style='color:#54ff54;'>labels=</span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b><b><span style='color:#dede49;'>popNames</span></b><b>(</b>hauss.genind<b>))</b>, <b><span style='color:#54ff54;'>pos=</span></b><b><span style='color:#ff54ff;'>4</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Google maps have their own internal scaling, adding of points by standard functions will not work correctly</span></b>

<b><span style='color:#54ffff;'># Pie charts on a map</span></b>
<b><span style='color:#54ffff;'># Prepare matrix with some data (exemplary distribution of haplotypes)</span></b>
hauss.pie <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>cbind</span></b><b>(</b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>20</span></b>, <b><span style='color:#ff54ff;'>30</span></b>, <b><span style='color:#ff54ff;'>15</span></b>, <b><span style='color:#ff54ff;'>40</span></b>, <b><span style='color:#ff54ff;'>10</span></b><b>)</b>, <b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>30</span></b>, <b><span style='color:#ff54ff;'>10</span></b>, <b><span style='color:#ff54ff;'>25</span></b>, <b><span style='color:#ff54ff;'>5</span></b>, <b><span style='color:#ff54ff;'>45</span></b><b>)</b>, <b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>10</span></b>, <b><span style='color:#ff54ff;'>20</span></b>, <b><span style='color:#ff54ff;'>20</span></b>, <b><span style='color:#ff54ff;'>15</span></b>, <b><span style='color:#ff54ff;'>5</span></b><b>))</b>
<b><span style='color:#dede49;'>rownames</span></b><b>(</b>hauss.pie<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>popNames</span></b><b>(</b>hauss.genpop<b>)</b> <b><span style='color:#54ffff;'># Add row names according to populations</span></b>
<b><span style='color:#dede49;'>colnames</span></b><b>(</b>hauss.pie<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;HapA&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;HapB&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;HapC&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Add column names according to data displayed</span></b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>hauss.pie<b>)</b> <b><span style='color:#54ffff;'># Check it is matrix</span></b>
hauss.pie <b><span style='color:#54ffff;'># See resulting matrix</span></b>
<b><span style='color:#54ffff;'># Plot basic map with state boundaries within selected range</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>getMap</span></b><b>(</b><b><span style='color:#54ff54;'>resolution=</span></b><b><span style='color:#ff54ff;'>&quot;high&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>xlim=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>20</span></b>, <b><span style='color:#ff54ff;'>23</span></b><b>)</b>, <b><span style='color:#54ff54;'>ylim=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>41</span></b>, <b><span style='color:#ff54ff;'>42</span></b><b>)</b>, <b><span style='color:#54ff54;'>asp=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#dede49;'>box</span></b><b>()</b> <b><span style='color:#54ffff;'># Add frame around the map</span></b>
<b><span style='color:#54ffff;'># Plot the pie charts</span></b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>L <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>5</span></b><b>)</b> <b>{</b> <b><span style='color:#dede49;'>add.pie</span></b><b>(</b><b><span style='color:#54ff54;'>z=</span></b>hauss.pie<b>[</b>L,<b>]</b>, <b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lon&quot;</span></b><b>]])[</b>L<b>]</b>, <b><span style='color:#54ff54;'>y=</span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lat&quot;</span></b><b>]])[</b>L<b>]</b>, <b><span style='color:#54ff54;'>labels=</span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>hauss.pie<b>[</b>L,<b>])</b>, <b><span style='color:#54ff54;'>radius=</span></b><b><span style='color:#ff54ff;'>0.1</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>topo.colors</span></b><b>(</b><b><span style='color:#ff54ff;'>3</span></b><b>))</b> <b>}</b>
?add.pie <b><span style='color:#54ffff;'># See more options</span></b>
<b><span style='color:#54ffff;'># Add population labels</span></b>
<b><span style='color:#dede49;'>text</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lon&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>y=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lat&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>labels=</span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b><b><span style='color:#dede49;'>popNames</span></b><b>(</b>hauss.genind<b>))</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>

<b><span style='color:#54ffff;'># Pie charts on Google map</span></b>
<b><span style='color:#54ffff;'># Prepare list to store recalculated coordinates</span></b>
hauss.gmap2.coord <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>list</span></b><b>()</b>
<b><span style='color:#54ffff;'># Calcualtion of coordinates to form required by Google Maps</span></b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>LC <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>5</span></b><b>)</b> <b>{</b> hauss.gmap2.coord<b>[[</b>LC<b>]]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>LatLon2XY.centered</span></b><b>(</b><b><span style='color:#54ff54;'>MyMap=</span></b>hauss.gmap2, <b><span style='color:#54ff54;'>lat=</span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lat&quot;</span></b><b>]])[</b>LC<b>]</b>, <b><span style='color:#54ff54;'>lon=</span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lon&quot;</span></b><b>]])[</b>LC<b>]</b>, <b><span style='color:#54ff54;'>zoom=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b> <b>}</b>
hauss.gmap2.coord <b><span style='color:#54ffff;'># See result</span></b>
<b><span style='color:#54ffff;'># Plot plain map</span></b>
<b><span style='color:#dede49;'>PlotOnStaticMap</span></b><b>(</b><b><span style='color:#54ff54;'>MyMap=</span></b>hauss.gmap2<b>)</b>
<b><span style='color:#54ffff;'># Plot pie charts</span></b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>LP <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>5</span></b><b>)</b> <b>{</b> <b><span style='color:#dede49;'>add.pie</span></b><b>(</b><b><span style='color:#54ff54;'>z=</span></b>hauss.pie<b>[</b>LP,<b>]</b>, <b><span style='color:#54ff54;'>x=</span></b>hauss.gmap2.coord<b>[[</b>LP<b>]]</b><b>$</b>newX, <b><span style='color:#54ff54;'>y=</span></b>hauss.gmap2.coord<b>[[</b>LP<b>]]</b><b>$</b>newY, <b><span style='color:#54ff54;'>labels=</span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>hauss.pie<b>[</b>LP,<b>])</b>, <b><span style='color:#54ff54;'>radius=</span></b><b><span style='color:#ff54ff;'>25</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>topo.colors</span></b><b>(</b><b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>alpha=</span></b><b><span style='color:#ff54ff;'>0.7</span></b><b>))</b> <b>}</b>
<b><span style='color:#54ffff;'># Alternative option to plot pie charts</span></b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>LF <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>5</span></b><b>)</b> <b>{</b> plotrix<b>::</b><b><span style='color:#dede49;'>floating.pie</span></b><b>(</b><b><span style='color:#54ff54;'>xpos=</span></b>hauss.gmap2.coord<b>[[</b>LF<b>]]</b><b>$</b>newX, <b><span style='color:#54ff54;'>ypos=</span></b>hauss.gmap2.coord<b>[[</b>LF<b>]]</b><b>$</b>newY, <b><span style='color:#54ff54;'>x=</span></b>hauss.pie<b>[</b>LF,<b>]</b>, <b><span style='color:#54ff54;'>radius=</span></b><b><span style='color:#ff54ff;'>30</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>heat.colors</span></b><b>(</b><b><span style='color:#54ff54;'>n=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>alpha=</span></b><b><span style='color:#ff54ff;'>0.5</span></b><b>)</b> <b>)</b> <b>}</b>
<b><span style='color:#54ffff;'># Add population text labels</span></b>
<b><span style='color:#dede49;'>PlotOnStaticMap</span></b><b>(</b><b><span style='color:#54ff54;'>MyMap=</span></b>hauss.gmap2, <b><span style='color:#54ff54;'>lat=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lat&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>lon=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lon&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>add=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>FUN=</span></b>text, <b><span style='color:#54ff54;'>labels=</span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b><b><span style='color:#dede49;'>popNames</span></b><b>(</b>hauss.genind<b>))</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2.5</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># Plot on OpenStreeMap - server is commonly overloaded and doesn't respond correctly</span></b>
<b><span style='color:#dede49;'>GetOsmMap</span></b><b>(</b><b><span style='color:#54ff54;'>lonR=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>18</span></b>, <b><span style='color:#ff54ff;'>24</span></b><b>)</b>, <b><span style='color:#54ff54;'>latR=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>39</span></b>, <b><span style='color:#ff54ff;'>44</span></b><b>)</b>, <b><span style='color:#54ff54;'>scale=</span></b><b><span style='color:#ff54ff;'>200000</span></b>, <b><span style='color:#54ff54;'>destfile=</span></b><b><span style='color:#ff54ff;'>&quot;osmmap.png&quot;</span></b>, <b><span style='color:#54ff54;'>format=</span></b><b><span style='color:#ff54ff;'>&quot;png&quot;</span></b>, <b><span style='color:#54ff54;'>RETURNIMAGE=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>GRAYSCALE=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>NEWMAP=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>verbose=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>

<b><span style='color:#dede49;'>library</span></b><b>(</b>maps<b>)</b> <b><span style='color:#54ffff;'># Various mapping tools (plotting, ...)</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>mapdata<b>)</b> <b><span style='color:#54ffff;'># More detailed maps, but political boundaries often outdated, see http://cran.r-project.org/web/packages/mapdata/</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>mapproj<b>)</b> <b><span style='color:#54ffff;'># Converts latitude/longitude into projected coordinates</span></b>
<b><span style='color:#54ffff;'># Plot a map, check parameters, among others projection and ?mapproject for its details</span></b>
<b><span style='color:#dede49;'>map</span></b><b>(</b><b><span style='color:#54ff54;'>database=</span></b><b><span style='color:#ff54ff;'>&quot;worldHires&quot;</span></b>, <b><span style='color:#54ff54;'>boundary=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>interior=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>fill=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;lightgrey&quot;</span></b>, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>xlim=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>16</span></b>, <b><span style='color:#ff54ff;'>27</span></b><b>)</b>, <b><span style='color:#54ff54;'>ylim=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>37</span></b>, <b><span style='color:#ff54ff;'>46</span></b><b>))</b>
<b><span style='color:#54ffff;'># If you'd use projection, use - mapproject() to convert also coordinates! See ?mapproject for details</span></b>
<b><span style='color:#dede49;'>points</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lon&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>y=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lat&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>15</span></b><b>:</b><b><span style='color:#ff54ff;'>19</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>

<b><span style='color:#54ffff;'># Plotting on SHP files</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>maptools<b>)</b>
<b><span style='color:#54ffff;'># Load SHP file</span></b>
<b><span style='color:#54ffff;'># Data from http://download.geofabrik.de/europe/macedonia.html</span></b>
<b><span style='color:#54ffff;'># R working directory has to contain also respective DBF and SHX files (same name, only different suffix)</span></b>
<b><span style='color:#dede49;'>dir</span></b><b>()</b> <b><span style='color:#54ffff;'># Verify required files are unpacked in the working directory</span></b>
<b><span style='color:#54ffff;'># Get from https://soubory.trapa.cz/rcourse/macedonia.zip</span></b>
<b><span style='color:#54ffff;'># There are several functions readShape* - select appropriate according to data stored in respective SHP file</span></b>
<b><span style='color:#54ffff;'># Check correct import by plotting all layers</span></b>
macedonia_building <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>readShapeLines</span></b><b>(</b><b><span style='color:#54ff54;'>fn=</span></b><b><span style='color:#ff54ff;'>&quot;macedonia_buildings.shp&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>macedonia_building<b>)</b>
macedonia_landuse <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>readShapeLines</span></b><b>(</b><b><span style='color:#54ff54;'>fn=</span></b><b><span style='color:#ff54ff;'>&quot;macedonia_landuse.shp&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>macedonia_landuse<b>)</b>
macedonia_natural <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>readShapeLines</span></b><b>(</b><b><span style='color:#54ff54;'>fn=</span></b><b><span style='color:#ff54ff;'>&quot;macedonia_natural.shp&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>macedonia_natural<b>)</b>
macedonia_railways <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>readShapeLines</span></b><b>(</b><b><span style='color:#54ff54;'>fn=</span></b><b><span style='color:#ff54ff;'>&quot;macedonia_railways.shp&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>macedonia_railways<b>)</b>
macedonia_roads <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>readShapeLines</span></b><b>(</b><b><span style='color:#54ff54;'>fn=</span></b><b><span style='color:#ff54ff;'>&quot;macedonia_roads.shp&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>macedonia_roads<b>)</b>
macedonia_waterways <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>readShapeLines</span></b><b>(</b><b><span style='color:#54ff54;'>fn=</span></b><b><span style='color:#ff54ff;'>&quot;macedonia_waterways.shp&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>macedonia_waterways<b>)</b>
<b><span style='color:#54ffff;'># Plot all layers into single image, add more information</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>macedonia_building<b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>macedonia_landuse, <b><span style='color:#54ff54;'>add=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;darkgreen&quot;</span></b>, <b><span style='color:#54ff54;'>fill=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>macedonia_natural, <b><span style='color:#54ff54;'>add=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;green&quot;</span></b>, <b><span style='color:#54ff54;'>fill=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>macedonia_railways, <b><span style='color:#54ff54;'>add=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;brown&quot;</span></b>, <b><span style='color:#54ff54;'>lty=</span></b><b><span style='color:#ff54ff;'>&quot;dotted&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>macedonia_roads, <b><span style='color:#54ff54;'>add=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;orange&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>macedonia_waterways, <b><span style='color:#54ff54;'>add=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;blue&quot;</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add state boundaries</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>getMap</span></b><b>(</b><b><span style='color:#54ff54;'>resolution=</span></b><b><span style='color:#ff54ff;'>&quot;high&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>xlim=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>19</span></b>, <b><span style='color:#ff54ff;'>24</span></b><b>)</b>, <b><span style='color:#54ff54;'>ylim=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>39</span></b>, <b><span style='color:#ff54ff;'>44</span></b><b>)</b>, <b><span style='color:#54ff54;'>asp=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>5</span></b>, <b><span style='color:#54ff54;'>add=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b> <b><span style='color:#54ffff;'># Or e.g.</span></b>
<b><span style='color:#dede49;'>map</span></b><b>(</b><b><span style='color:#54ff54;'>database=</span></b><b><span style='color:#ff54ff;'>&quot;worldHires&quot;</span></b>, <b><span style='color:#54ff54;'>boundary=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>interior=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>fill=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>add=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>xlim=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>16</span></b>, <b><span style='color:#ff54ff;'>27</span></b><b>)</b>, <b><span style='color:#54ff54;'>ylim=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>37</span></b>, <b><span style='color:#ff54ff;'>46</span></b><b>)</b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add sampling points</span></b>
<b><span style='color:#dede49;'>points</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lon&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>y=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lat&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>15</span></b><b>:</b><b><span style='color:#ff54ff;'>19</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>4</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add description of psampling points</span></b>
<b><span style='color:#dede49;'>shadowtext</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lon&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>y=</span></b>hauss.genpop<b>@</b>other<b>$</b>xy<b>[[</b><b><span style='color:#ff54ff;'>&quot;lat&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>labels=</span></b><b><span style='color:#dede49;'>as.vector</span></b><b>(</b><b><span style='color:#dede49;'>popNames</span></b><b>(</b>hauss.genind<b>))</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;black&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b>, <b><span style='color:#54ff54;'>theta=</span></b><b><span style='color:#dede49;'>seq</span></b><b>(</b>pi<b>/</b><b><span style='color:#ff54ff;'>4</span></b>, <b><span style='color:#ff54ff;'>2</span></b><b>*</b>pi, <b><span style='color:#54ff54;'>length.out=</span></b><b><span style='color:#ff54ff;'>8</span></b><b>)</b>, <b><span style='color:#54ff54;'>r=</span></b><b><span style='color:#ff54ff;'>0.15</span></b>, <b><span style='color:#54ff54;'>pos=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#ff54ff;'>4</span></b>, <b><span style='color:#ff54ff;'>4</span></b><b>)</b>, <b><span style='color:#54ff54;'>offset=</span></b><b><span style='color:#ff54ff;'>0.75</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add legend</span></b>
<b><span style='color:#dede49;'>legend</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>&quot;topright&quot;</span></b>, <b><span style='color:#54ff54;'>inset=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>/</b><b><span style='color:#ff54ff;'>50</span></b>, <b><span style='color:#54ff54;'>legend=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;He&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Oh&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Pr&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Ne&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Sk&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>border=</span></b><b><span style='color:#ff54ff;'>&quot;black&quot;</span></b>, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>15</span></b><b>:</b><b><span style='color:#ff54ff;'>19</span></b>, <b><span style='color:#54ff54;'>pt.cex=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>bty=</span></b><b><span style='color:#ff54ff;'>&quot;o&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;lightgrey&quot;</span></b>, <b><span style='color:#54ff54;'>box.lwd=</span></b><b><span style='color:#ff54ff;'>1.5</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b>, <b><span style='color:#54ff54;'>title=</span></b><b><span style='color:#ff54ff;'>&quot;Populations&quot;</span></b><b>)</b>

<b><span style='color:#b00000;'>## Structure</span></b>

<b><span style='color:#54ffff;'># Inspiration: https://www.molecularecologist.com/913/09/using-r-to-run-parallel-analyses-of-population-genetic-data-in-structure-parallelstructure/</span></b>
<b><span style='color:#54ffff;'># Install ParallelStructure, see https://r-forge.r-project.org/R/?group_id=1636 and http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0070651</span></b>
<b><span style='color:#54ffff;'># get input data from https://soubory.trapa.cz/rcourse/hauss_stru.in and joblist https://soubory.trapa.cz/rcourse/joblist.txt</span></b>
<b><span style='color:#54ffff;'># Set working directory</span></b>
<b><span style='color:#dede49;'>setwd</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;~/dokumenty/fakulta/vyuka/r_mol_data/examples/structure/&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>getwd</span></b><b>()</b>
<b><span style='color:#54ffff;'># Load ParallelStructure package</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>ParallelStructure<b>)</b>
<b><span style='color:#54ffff;'># See manual for the R package and Structure http://pritchardlab.stanford.edu/structure.html</span></b>
?parallel_structure
<b><span style='color:#54ffff;'># Run Structure</span></b>
<b><span style='color:#dede49;'>parallel_structure</span></b><b>(</b><b><span style='color:#54ff54;'>joblist=</span></b><b><span style='color:#ff54ff;'>&quot;joblist.txt&quot;</span></b>, <b><span style='color:#54ff54;'>n_cpu=</span></b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>structure_path=</span></b><b><span style='color:#ff54ff;'>&quot;~/bin/&quot;</span></b>, <b><span style='color:#54ff54;'>infile=</span></b><b><span style='color:#ff54ff;'>&quot;hauss_stru.in&quot;</span></b>, <b><span style='color:#54ff54;'>outpath=</span></b><b><span style='color:#ff54ff;'>&quot;results/&quot;</span></b>, <b><span style='color:#54ff54;'>numinds=</span></b><b><span style='color:#ff54ff;'>47</span></b>, <b><span style='color:#54ff54;'>numloci=</span></b><b><span style='color:#ff54ff;'>12</span></b>, <b><span style='color:#54ff54;'>plot_output=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>label=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>popdata=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>popflag=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>phenotypes=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>markernames=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>mapdist=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>onerowperind=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>phaseinfo=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>extracol=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>missing=</span></b><b>-</b><b><span style='color:#ff54ff;'>9</span></b>, <b><span style='color:#54ff54;'>ploidy=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>usepopinfo=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>revert_convert=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>printqhat=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>locdata=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>recessivealleles=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>phased=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>noadmix=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>linkage=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>locprior=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>inferalpha=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>

<b><span style='color:#54ffff;'># When running on Windows, parallel support may be missing - install Rmpi library required by ParallelStructure for parallelisation on Windows</span></b>
<b><span style='color:#dede49;'>install.packages</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Rmpi&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>Rmpi<b>)</b>
<b><span style='color:#54ffff;'># Instead of parallel_structure() use MPI_structure() with same arguments</span></b>
<b><span style='color:#dede49;'>MPI_structure</span></b><b>(</b>...<b>)</b> <b><span style='color:#54ffff;'># Same arguments as on previous slide</span></b>
<b><span style='color:#54ffff;'># If this fails, look for some UNIX machine...</span></b>

<b><span style='color:#54ffff;'># Postprocess results with Structure sum R script by Dorothee Ehrich</span></b>
<b><span style='color:#dede49;'>source</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;https://soubory.trapa.cz/rcourse/structure-sum-2011.r&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Create new directory with result files results_job_*_f and set working directory accordingly</span></b>
<b><span style='color:#dede49;'>setwd</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;/home/vojta/dokumenty/fakulta/vyuka/r_mol_data/examples/structure/structure_sum/&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>dir</span></b><b>()</b>
<b><span style='color:#54ffff;'># Prepare file list_k.txt containing on each line K and name of output &quot;_f&quot; file - get it from https://soubory.trapa.cz/rcourse/list_k.txt</span></b>
<b><span style='color:#54ffff;'># See documentation for details. Functions take as an argument list_k file and number of populations</span></b>
<b><span style='color:#dede49;'>Structure.table</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;list_k.txt&quot;</span></b>, <b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#dede49;'>Structure.simil</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;list_k.txt&quot;</span></b>, <b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#dede49;'>Structure.deltaK</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;list_k.txt&quot;</span></b>, <b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#dede49;'>graphics.off</span></b><b>()</b> <b><span style='color:#54ffff;'># Close graphics</span></b>
<b><span style='color:#dede49;'>Structure.cluster</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;list_k.txt&quot;</span></b>, <b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#54ffff;'># Reordering (&quot;alignment&quot;) of runs to get same clusters in same columns (prepare respective list_k files - one for each K)</span></b>
<b><span style='color:#54ffff;'># Preparing data for CLUMPP</span></b>
<b><span style='color:#dede49;'>Structure.order</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;list_k_02.txt&quot;</span></b>, <b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#dede49;'>Structure.order</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;list_k_03.txt&quot;</span></b>, <b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#dede49;'>Structure.order</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;list_k_04.txt&quot;</span></b>, <b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#dede49;'>Structure.order</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;list_k_05.txt&quot;</span></b>, <b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#dede49;'>Structure.order</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;list_k_06.txt&quot;</span></b>, <b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#dede49;'>Structure.order</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;list_k_07.txt&quot;</span></b>, <b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#54ffff;'># Continue with CLUMPP and distruct</span></b>
<b><span style='color:#54ffff;'># Details: https://trapa.cz/en/structure-r-linux</span></b>

<b><span style='color:#b00000;'>## Multiple sequence alignment</span></b>

<b><span style='color:#54ffff;'># Libraries</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>colorspace<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>XML<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>phyloch<b>)</b> <b><span style='color:#54ffff;'># Alignment with mafft, you can also try package ips</span></b>
<b><span style='color:#54ffff;'># Requires path to MAFFT binary - set it according to your installation</span></b>
<b><span style='color:#54ffff;'># Read ?mafft and mafft's documentation</span></b>
meles.mafft <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>mafft</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.dna, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;localpair&quot;</span></b>, <b><span style='color:#54ff54;'>maxiterate=</span></b><b><span style='color:#ff54ff;'>100</span></b>, <b><span style='color:#54ff54;'>path=</span></b><b><span style='color:#ff54ff;'>&quot;/usr/bin/mafft&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Change &quot;path&quot; to fit your path to mafft!</span></b>
meles.mafft
<b><span style='color:#dede49;'>class</span></b><b>(</b>meles.mafft<b>)</b>

<b><span style='color:#54ffff;'># Multiple sequence alignments using clustal, muscle and t-coffee are available in package ape</span></b>
<b><span style='color:#54ffff;'># read ?clustal and documentation of Clustal and Muscle to set correct parameters</span></b>
meles.clustal <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>ape<b>::</b><b><span style='color:#dede49;'>clustal</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.dna, <b><span style='color:#54ff54;'>pw.gapopen=</span></b><b><span style='color:#ff54ff;'>10</span></b>, <b><span style='color:#54ff54;'>pw.gapext=</span></b><b><span style='color:#ff54ff;'>0.1</span></b>, <b><span style='color:#54ff54;'>gapopen=</span></b><b><span style='color:#ff54ff;'>10</span></b>, <b><span style='color:#54ff54;'>gapext=</span></b><b><span style='color:#ff54ff;'>0.2</span></b>, <b><span style='color:#54ff54;'>exec=</span></b><b><span style='color:#ff54ff;'>&quot;/usr/bin/clustalw2&quot;</span></b>, <b><span style='color:#54ff54;'>quiet=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>original.ordering=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b> <b><span style='color:#54ffff;'># Change &quot;exec&quot; to fit your path to clustal!</span></b>
meles.clustal
<b><span style='color:#dede49;'>class</span></b><b>(</b>meles.clustal<b>)</b>
meles.muscle <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>ape<b>::</b><b><span style='color:#dede49;'>muscle</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.dna, <b><span style='color:#54ff54;'>exec=</span></b><b><span style='color:#ff54ff;'>&quot;muscle&quot;</span></b>, <b><span style='color:#54ff54;'>quiet=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>original.ordering=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b> <b><span style='color:#54ffff;'># Change &quot;exec&quot; to fit your path to muscle!</span></b>
meles.muscle
<b><span style='color:#dede49;'>class</span></b><b>(</b>meles.muscle<b>)</b>

<b><span style='color:#54ffff;'># Remove gaps from alignment - destroy it</span></b>
meles.nogaps <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>del.gaps</span></b><b>(</b>meles.muscle<b>)</b>
?del.gaps <b><span style='color:#54ffff;'># See for details!</span></b>

<b><span style='color:#54ffff;'># Plot the alignment - you can select which bases to plot and/or modify colors</span></b>
<b><span style='color:#dede49;'>image</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.muscle, <b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;a&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;t&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;c&quot;</span></b> ,<b><span style='color:#ff54ff;'>&quot;g&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;n&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>rainbow</span></b><b>(</b><b><span style='color:#ff54ff;'>5</span></b><b>))</b>
<b><span style='color:#54ffff;'># Add grey dotted grid</span></b>
<b><span style='color:#dede49;'>grid</span></b><b>(</b><b><span style='color:#54ff54;'>nx=</span></b><b><span style='color:#dede49;'>ncol</span></b><b>(</b>meles.muscle<b>)</b>, <b><span style='color:#54ff54;'>ny=</span></b><b><span style='color:#dede49;'>nrow</span></b><b>(</b>meles.muscle<b>)</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;lightgrey&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># Align multiple genes</span></b>
<b><span style='color:#54ffff;'># Create a list of DNAbin objects to process</span></b>
multialign <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>list</span></b><b>(</b>meles.dna, usflu.dna, usflu.dna2, usflu.dna3<b>)</b>
<b><span style='color:#54ffff;'># See it</span></b>
multialign
<b><span style='color:#dede49;'>class</span></b><b>(</b>multialign<b>)</b>
<b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>multialign, <b><span style='color:#54ff54;'>FUN=</span></b>class<b>)</b>
<b><span style='color:#54ffff;'># Do the alignment</span></b>
multialign.aln <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>multialign, <b><span style='color:#54ff54;'>FUN=</span></b>phyloch<b>::</b>mafft, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;localpair&quot;</span></b>, <b><span style='color:#54ff54;'>maxiterate=</span></b><b><span style='color:#ff54ff;'>100</span></b>, <b><span style='color:#54ff54;'>path=</span></b><b><span style='color:#ff54ff;'>&quot;/usr/bin/mafft&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Change &quot;path&quot; to fit your path to mafft!</span></b>
<b><span style='color:#54ffff;'># See result</span></b>
multialign.aln
multialign.aln<b>[[</b><b><span style='color:#ff54ff;'>1</span></b><b>]]</b>
<b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>multialign.aln, <b><span style='color:#54ff54;'>FUN=</span></b>class<b>)</b>
<b><span style='color:#54ffff;'># Do the same in parallel (mclapply do the tasks in parallel, not one-by-one like lapply)</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>parallel<b>)</b>
multialign.aln2 <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>mclapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>multialign, <b><span style='color:#54ff54;'>FUN=</span></b>ape<b>::</b>muscle, <b><span style='color:#54ff54;'>exec=</span></b><b><span style='color:#ff54ff;'>&quot;muscle&quot;</span></b>, <b><span style='color:#54ff54;'>quiet=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>original.ordering=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b> <b><span style='color:#54ffff;'># Change &quot;path&quot; to fit your path to muscle!</span></b>
<b><span style='color:#54ffff;'># mclapply() relies on forking and hence is not available on Windows unless &quot;mc.cores=1&quot;</span></b>
<b><span style='color:#54ffff;'># See result</span></b>
multialign.aln2
<b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>multialign.aln2, <b><span style='color:#54ff54;'>FUN=</span></b>class<b>)</b>
?mclapply <b><span style='color:#54ffff;'># See more options</span></b>
?clusterApply <b><span style='color:#54ffff;'># See more options (parLapply should work on Windows)</span></b>

<b><span style='color:#54ffff;'># Shortcut for plotting alignment</span></b>
<b><span style='color:#dede49;'>image.DNAbin</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.mafft<b>)</b>
<b><span style='color:#54ffff;'># Display aligned sequences with gaps</span></b>
<b><span style='color:#dede49;'>image.DNAbin</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.dna<b>)</b>
<b><span style='color:#54ffff;'># Check the alignment</span></b>
<b><span style='color:#dede49;'>checkAlignment</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.dna, <b><span style='color:#54ff54;'>check.gaps=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>what=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>4</span></b><b>)</b>
<b><span style='color:#dede49;'>checkAlignment</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>as.matrix.DNAbin</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.dna<b>)</b>, <b><span style='color:#54ff54;'>check.gaps=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>what=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>4</span></b><b>)</b>
?checkAlignment <b><span style='color:#54ffff;'># See details</span></b>
<b><span style='color:#54ffff;'># DNAbin can be techically list or matrix - some functions require list, some matrix, some can handle both - check manual and if needed, use</span></b>
<b><span style='color:#dede49;'>as.matrix.DNAbin</span></b><b>()</b>
<b><span style='color:#dede49;'>as.list.DNAbin</span></b><b>()</b>
<b><span style='color:#54ffff;'># Matrix makes sense only for alignments, list for any import (sequences do no have to have same lengths)</span></b>

<b><span style='color:#54ffff;'># Delete all columns containing any gap</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>ips<b>)</b>
usflu.dna.ng <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>deleteGaps</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.dna, <b><span style='color:#54ff54;'>nmax=</span></b><b><span style='color:#ff54ff;'>0</span></b><b>)</b>
usflu.dna.ng
<b><span style='color:#54ffff;'># See of settings of &quot;nmax&quot; value - threshold for gap deletion</span></b>
?deleteGaps <b><span style='color:#54ffff;'># &quot;nmax=0&quot; deletes all columns with any gap</span></b>
multialign.aln.ng <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>multialign.aln, <b><span style='color:#54ff54;'>FUN=</span></b>deleteGaps, <b><span style='color:#54ff54;'>nmax=</span></b><b><span style='color:#ff54ff;'>0</span></b><b>)</b>
multialign.aln.ng
<b><span style='color:#54ffff;'># Do not confuse with function delete.gaps() from phyloch package</span></b>
<b><span style='color:#54ffff;'># Display the result</span></b>
<b><span style='color:#dede49;'>image.DNAbin</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>usflu.dna.ng<b>)</b>
<b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>multialign.aln.ng, <b><span style='color:#54ff54;'>FUN=</span></b>image.DNAbin<b>)</b>
<b><span style='color:#54ffff;'># Delete positions in alignment containing only missing data/N (rows as well as columns)</span></b>
?ips<b>::</b>deleteEmptyCells <b><span style='color:#54ffff;'># See help page for details</span></b>
?phyloch<b>::</b>delete.empty.cells <b><span style='color:#54ffff;'># See help page for details</span></b>
<b><span style='color:#54ffff;'># Delete only empty columns/rows (usage as above)</span></b>
?ape<b>::</b>del.colgapsonly <b><span style='color:#54ffff;'># See help page for details</span></b>
?ape<b>::</b>del.rowgapsonly <b><span style='color:#54ffff;'># See help page for details</span></b>

<b><span style='color:#b00000;'>## Tree manipulations</span></b>

<b><span style='color:#54ffff;'># Read trees in NEWICK format - single or multiple tree(s)</span></b>
oxalis.trees <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.tree</span></b><b>(</b><b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;https://soubory.trapa.cz/rcourse/oxalis.nwk&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>summary</span></b><b>(</b>oxalis.trees<b>)</b>
<b><span style='color:#dede49;'>length</span></b><b>(</b>oxalis.trees<b>)</b>
<b><span style='color:#dede49;'>names</span></b><b>(</b>oxalis.trees<b>)</b>
<b><span style='color:#54ffff;'># Export trees in NEWICK format</span></b>
<b><span style='color:#dede49;'>write.tree</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>oxalis.trees, <b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;trees.nwk&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># Drop a tip from multiPhylo</span></b>
<b><span style='color:#dede49;'>plot.multiPhylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>oxalis.trees<b>)</b>
<b><span style='color:#54ffff;'># See tip labels</span></b>
oxalis.trees<b>[[</b><b><span style='color:#ff54ff;'>1</span></b><b>]][[</b><b><span style='color:#ff54ff;'>&quot;tip.label&quot;</span></b><b>]]</b>
oxalis.trees.drop <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>oxalis.trees, <b><span style='color:#54ff54;'>FUN=</span></b>drop.tip, <b><span style='color:#54ff54;'>tip=</span></b><b><span style='color:#ff54ff;'>&quot;O._callosa_S15&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>oxalis.trees.drop<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> &quot;multiPhylo&quot;</span></b>
<b><span style='color:#dede49;'>plot.multiPhylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>oxalis.trees.drop<b>)</b>

<b><span style='color:#54ffff;'># Drop a tip</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj<b>)</b>
hauss.nj<b>[[</b><b><span style='color:#ff54ff;'>&quot;tip.label&quot;</span></b><b>]]</b>
hauss.nj.drop <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>drop.tip</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>hauss.nj, <b><span style='color:#54ff54;'>tip=</span></b><b><span style='color:#ff54ff;'>47</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj.drop<b>)</b>

<b><span style='color:#54ffff;'># Interactively extract tree</span></b>
<b><span style='color:#54ffff;'># Plot source tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj<b>)</b>
<b><span style='color:#54ffff;'># See node labels (numbers) - needed for some tasks</span></b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>()</b>
<b><span style='color:#54ffff;'># Select clade to extract by clicking on it</span></b>
hauss.nj.extracted <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>extract.clade</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>hauss.nj, <b><span style='color:#54ff54;'>interactive=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># See new extracted tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj.extracted<b>)</b>
<b><span style='color:#54ffff;'># Non-interactively extract tree</span></b>
hauss.nj.extracted <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>extract.clade</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>hauss.nj, <b><span style='color:#54ff54;'>node=</span></b><b><span style='color:#ff54ff;'>60</span></b>, <b><span style='color:#54ff54;'>interactive=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># See new extracted tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj.extracted<b>)</b>

<b><span style='color:#54ffff;'># Drop &quot;extinct&quot; tips - those who don't reach end the tree</span></b>
<b><span style='color:#54ffff;'># tolerance is respective to the used metrics</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj<b>)</b>
<b><span style='color:#dede49;'>axisPhylo</span></b><b>()</b>
hauss.nj.fossil <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>drop.fossil</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>hauss.nj, <b><span style='color:#54ff54;'>tol=</span></b><b><span style='color:#ff54ff;'>0.4</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj.fossil<b>)</b>

<b><span style='color:#54ffff;'># Bind two trees into one</span></b>
hauss.nj.bind <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>bind.tree</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj.fossil, <b><span style='color:#54ff54;'>y=</span></b>hauss.nj.extracted, <b><span style='color:#54ff54;'>where=</span></b><b><span style='color:#ff54ff;'>&quot;root&quot;</span></b>, <b><span style='color:#54ff54;'>position=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>interactive=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj.bind<b>)</b>
<b><span style='color:#54ffff;'># Bind two trees interactively</span></b>
<b><span style='color:#54ffff;'># Plot tree receiving the new one</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj.fossil<b>)</b>
<b><span style='color:#54ffff;'># Select where to bind new tree to</span></b>
hauss.nj.bind <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>bind.tree</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj.fossil, <b><span style='color:#54ff54;'>y=</span></b>hauss.nj.extracted, <b><span style='color:#54ff54;'>interactive=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj.bind<b>)</b>

<b><span style='color:#54ffff;'># Rotate tree</span></b>
<b><span style='color:#54ffff;'># plot.phylo plots tree in exact order as it is in the phylo object</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj<b>)</b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>()</b>
hauss.nj.rotated <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>rotate</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>hauss.nj, <b><span style='color:#54ff54;'>node=</span></b><b><span style='color:#ff54ff;'>&quot;70&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj.rotated<b>)</b>

<b><span style='color:#54ffff;'># Ladderize the tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj<b>)</b>
hauss.nj.ladderized <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>ladderize</span></b><b>(</b>hauss.nj<b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj.ladderized<b>)</b>

<b><span style='color:#54ffff;'># Root the tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj<b>)</b>
<b><span style='color:#dede49;'>print.phylo</span></b><b>(</b>hauss.nj<b>)</b>
hauss.nj.rooted <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>root</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>hauss.nj, <b><span style='color:#54ff54;'>resolve.root=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>outgroup=</span></b><b><span style='color:#ff54ff;'>10</span></b><b>)</b> <b><span style='color:#54ffff;'># resolve.root=TRUE ensures root will be bifurcating (without this parameter it sometimes doesn't work)</span></b>
<b><span style='color:#dede49;'>print.phylo</span></b><b>(</b>hauss.nj.rooted<b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj.rooted<b>)</b>
<b><span style='color:#54ffff;'># Root the tree interactive</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj<b>)</b>
hauss.nj.rooted <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>root</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>hauss.nj, <b><span style='color:#54ff54;'>interactive=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>hauss.nj.rooted<b>)</b>
<b><span style='color:#54ffff;'># Unroot the tree</span></b>
<b><span style='color:#dede49;'>unroot</span></b><b>()</b>
<b><span style='color:#54ffff;'># Check if it is rooted</span></b>
<b><span style='color:#dede49;'>is.rooted</span></b><b>()</b>

<b><span style='color:#54ffff;'># Check if the tree is ultrametric - is variance of distances of all tips to node 0? It is required for some analysis</span></b>
<b><span style='color:#dede49;'>is.ultrametric</span></b><b>()</b>
<b><span style='color:#54ffff;'># Make tree ultrametric</span></b>
<b><span style='color:#dede49;'>chronos</span></b><b>()</b>
?chronos <b><span style='color:#54ffff;'># Check it for mode how to calculate the lengths</span></b>

<b><span style='color:#54ffff;'># Compute branch lengths for trees without branch lengths</span></b>
<b><span style='color:#dede49;'>compute.brlen</span></b><b>()</b>
?compute.brlen <b><span style='color:#54ffff;'># Check it for mode how to calculate the lengths</span></b>

<b><span style='color:#54ffff;'># Computes the branch lengths of a tree giving its branching times (aka node ages or heights)</span></b>
<b><span style='color:#dede49;'>compute.brtime</span></b><b>()</b>
?compute.brtime <b><span style='color:#54ffff;'># Check it for mode how to calculate the lengths</span></b>

<b><span style='color:#b00000;'>## Topographical distances among trees</span></b>

<b><span style='color:#dede49;'>library</span></b><b>(</b>gplots<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>corrplot<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>phytools<b>)</b>

<b><span style='color:#54ffff;'># Prepare matrix for distances</span></b>
oxalis.trees.d <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>matrix</span></b><b>(</b><b><span style='color:#54ff54;'>nrow=</span></b><b><span style='color:#dede49;'>length</span></b><b>(</b>oxalis.trees<b>)</b>, <b><span style='color:#54ff54;'>ncol=</span></b><b><span style='color:#dede49;'>length</span></b><b>(</b>oxalis.trees<b>))</b>

<b><span style='color:#54ffff;'># Calculate pairwise topographic distances</span></b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>i <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#dede49;'>length</span></b><b>(</b>oxalis.trees<b>))</b> <b>{</b>
	<b><span style='color:#ffff54;'>for</span></b> <b>(</b>j <b><span style='color:#ffff54;'>in</span></b> i<b>:</b><b><span style='color:#dede49;'>length</span></b><b>(</b>oxalis.trees<b>))</b> <b>{</b>
		<b><span style='color:#dede49;'>print</span></b><b>(</b><b><span style='color:#dede49;'>c</span></b><b>(</b>i,j<b>))</b>
		oxalis.trees.d<b>[</b>i,j<b>]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dist.topo</span></b><b>(</b>oxalis.trees<b>[[</b>i<b>]]</b>, oxalis.trees<b>[[</b>j<b>]])</b>
		<b>}</b>
	<b>}</b>

<b><span style='color:#54ffff;'># Basic information about the distance matrix</span></b>
<b><span style='color:#dede49;'>dim</span></b><b>(</b>oxalis.trees.d<b>)</b>
<b><span style='color:#dede49;'>head.matrix</span></b><b>(</b>oxalis.trees.d<b>)</b>

<b><span style='color:#54ffff;'># Add names of columns and rows</span></b>
<b><span style='color:#dede49;'>colnames</span></b><b>(</b>oxalis.trees.d<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>oxalis.trees<b>)</b>
<b><span style='color:#dede49;'>rownames</span></b><b>(</b>oxalis.trees.d<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>oxalis.trees<b>)</b>

<b><span style='color:#54ffff;'># Make matrix symetric</span></b>
oxalis.trees.d<b>[</b><b><span style='color:#dede49;'>lower.tri</span></b><b>(</b>oxalis.trees.d<b>)]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>t</span></b><b>(</b>oxalis.trees.d<b>)[</b><b><span style='color:#dede49;'>lower.tri</span></b><b>(</b>oxalis.trees.d<b>)]</b>

<b><span style='color:#54ffff;'># Create heatmaps using heatmap.2 function from gplots package</span></b>
<b><span style='color:#dede49;'>heatmap.2</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>oxalis.trees.d, <b><span style='color:#54ff54;'>Rowv=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>Colv=</span></b><b><span style='color:#ff54ff;'>&quot;Rowv&quot;</span></b>, <b><span style='color:#54ff54;'>dendrogram=</span></b><b><span style='color:#ff54ff;'>&quot;none&quot;</span></b>, <b><span style='color:#54ff54;'>symm=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>scale=</span></b><b><span style='color:#ff54ff;'>&quot;none&quot;</span></b>, <b><span style='color:#54ff54;'>na.rm=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>revC=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>rainbow</span></b><b>(</b><b><span style='color:#ff54ff;'>15</span></b><b>)</b>, <b><span style='color:#54ff54;'>cellnote=</span></b>oxalis.trees.d, <b><span style='color:#54ff54;'>notecex=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>notecol=</span></b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b>, <b><span style='color:#54ff54;'>trace=</span></b><b><span style='color:#ff54ff;'>&quot;row&quot;</span></b>, <b><span style='color:#54ff54;'>linecol=</span></b><b><span style='color:#ff54ff;'>&quot;black&quot;</span></b>, <b><span style='color:#54ff54;'>labRow=</span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>oxalis.trees<b>)</b>, <b><span style='color:#54ff54;'>labCol=</span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>oxalis.trees<b>)</b>, <b><span style='color:#54ff54;'>key=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>keysize=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>density.info=</span></b><b><span style='color:#ff54ff;'>&quot;density&quot;</span></b>, <b><span style='color:#54ff54;'>symkey=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Correlation matrix of topographical distances&quot;</span></b>, <b><span style='color:#54ff54;'>xlab=</span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>oxalis.trees<b>)</b>, <b><span style='color:#54ff54;'>ylab=</span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>oxalis.trees<b>))</b>

<b><span style='color:#54ffff;'># Robinsons-Foulds distance</span></b>
oxalis.trees.d.rf <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>multiRF</span></b><b>(</b>oxalis.trees<b>)</b>

<b><span style='color:#54ffff;'># Add names of columns and rows</span></b>
<b><span style='color:#dede49;'>colnames</span></b><b>(</b>oxalis.trees.d.rf<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>oxalis.trees<b>)</b>
<b><span style='color:#dede49;'>rownames</span></b><b>(</b>oxalis.trees.d.rf<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>oxalis.trees<b>)</b>

<b><span style='color:#54ffff;'># Create heatmap using corrplot function from corrplot package</span></b>
<b><span style='color:#dede49;'>corrplot</span></b><b>(</b><b><span style='color:#54ff54;'>corr=</span></b>oxalis.trees.d.rf, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;circle&quot;</span></b>, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;upper&quot;</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>rainbow</span></b><b>(</b><b><span style='color:#ff54ff;'>15</span></b><b>)</b>, <b><span style='color:#54ff54;'>title=</span></b><b><span style='color:#ff54ff;'>&quot;Correlation matrix of topographical distances&quot;</span></b>, <b><span style='color:#54ff54;'>is.corr=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>diag=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>outline=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>order=</span></b><b><span style='color:#ff54ff;'>&quot;alphabet&quot;</span></b>, <b><span style='color:#54ff54;'>tl.pos=</span></b><b><span style='color:#ff54ff;'>&quot;lt&quot;</span></b>, <b><span style='color:#54ff54;'>tl.col=</span></b><b><span style='color:#ff54ff;'>&quot;black&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>corrplot</span></b><b>(</b><b><span style='color:#54ff54;'>corr=</span></b>oxalis.trees.d.rf, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;number&quot;</span></b>, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;lower&quot;</span></b>, <b><span style='color:#54ff54;'>add=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>rainbow</span></b><b>(</b><b><span style='color:#ff54ff;'>15</span></b><b>)</b>, <b><span style='color:#54ff54;'>title=</span></b><b><span style='color:#ff54ff;'>&quot;Correlation matrix of topographical distances&quot;</span></b>, <b><span style='color:#54ff54;'>is.corr=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>diag=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>outline=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>order=</span></b><b><span style='color:#ff54ff;'>&quot;alphabet&quot;</span></b>, <b><span style='color:#54ff54;'>tl.pos=</span></b><b><span style='color:#ff54ff;'>&quot;ld&quot;</span></b>, <b><span style='color:#54ff54;'>cl.pos=</span></b><b><span style='color:#ff54ff;'>&quot;n&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># PCoA from distance matrices of topographical differences among trees</span></b>
<b><span style='color:#54ffff;'># Test if the distance matrix is Euclidean or not</span></b>
<b><span style='color:#dede49;'>is.euclid</span></b><b>(</b><b><span style='color:#54ff54;'>distmat=</span></b><b><span style='color:#dede49;'>as.dist</span></b><b>(</b>oxalis.trees.d<b>)</b>, <b><span style='color:#54ff54;'>plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Calculate the PCoA</span></b>
oxalis.trees.pcoa <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>dudi.pco</span></b><b>(</b><b><span style='color:#54ff54;'>d=</span></b><b><span style='color:#dede49;'>quasieuclid</span></b><b>(</b><b><span style='color:#dede49;'>as.dist</span></b><b>(</b>oxalis.trees.d<b>))</b>, <b><span style='color:#54ff54;'>scannf=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>full=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Plot PCoA</span></b>
<b><span style='color:#dede49;'>s.label</span></b><b>(</b><b><span style='color:#54ff54;'>dfxy=</span></b>oxalis.trees.pcoa<b>$</b>li<b>)</b>
<b><span style='color:#54ffff;'># Add kernel densities</span></b>
<b><span style='color:#dede49;'>s.kde2d</span></b><b>(</b><b><span style='color:#54ff54;'>dfxy=</span></b>oxalis.trees.pcoa<b>$</b>li, <b><span style='color:#54ff54;'>cpoint=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>add.plot=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add histogram of eigenvalues</span></b>
<b><span style='color:#dede49;'>add.scatter.eig</span></b><b>(</b>oxalis.trees.pcoa<b>[[</b><b><span style='color:#ff54ff;'>&quot;eig&quot;</span></b><b>]]</b>, <b><span style='color:#ff54ff;'>3</span></b>,<b><span style='color:#ff54ff;'>1</span></b>,<b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>posi=</span></b><b><span style='color:#ff54ff;'>&quot;topleft&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add title to the plot</span></b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;PCoA of matrix of pairwise trees distances&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Alternative function to plot PCA plot</span></b>
<b><span style='color:#dede49;'>scatter</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>oxalis.trees.pcoa, <b><span style='color:#54ff54;'>posieig=</span></b><b><span style='color:#ff54ff;'>&quot;topleft&quot;</span></b><b>)</b>

<b><span style='color:#b00000;'>## Seeing trees in the forest</span></b>

<b><span style='color:#54ffff;'># Root all trees</span></b>
oxalis.trees.rooted <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>oxalis.trees, <b><span style='color:#54ff54;'>FUN=</span></b>root, <b><span style='color:#ff54ff;'>&quot;O._fibrosa_S159&quot;</span></b>, <b><span style='color:#54ff54;'>resolve.root=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>oxalis.trees.rooted<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> &quot;multiPhylo&quot;</span></b>
<b><span style='color:#54ffff;'># Consenus tree (50 % rule)</span></b>
oxalis.tree.con <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>ape<b>::</b><b><span style='color:#dede49;'>consensus</span></b><b>(</b>oxalis.trees.rooted, <b><span style='color:#54ff54;'>p=</span></b><b><span style='color:#ff54ff;'>0.5</span></b>, <b><span style='color:#54ff54;'>check.labels=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>print.phylo</span></b><b>(</b>oxalis.tree.con<b>)</b>
<b><span style='color:#54ffff;'># Plot the tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>oxalis.tree.con, <b><span style='color:#54ff54;'>edge.width=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>label.offset=</span></b><b><span style='color:#ff54ff;'>0.3</span></b><b>)</b>
<b><span style='color:#dede49;'>axisPhylo</span></b><b>(</b><b><span style='color:#54ff54;'>side=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>

<b><span style='color:#54ffff;'># Species tree</span></b>
<b><span style='color:#54ffff;'># All trees must be ultrametric - chronos scale them</span></b>
oxalis.trees.ultra <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>oxalis.trees.rooted, <b><span style='color:#54ff54;'>FUN=</span></b>chronos, <b><span style='color:#54ff54;'>model=</span></b><b><span style='color:#ff54ff;'>&quot;correlated&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>oxalis.trees.ultra<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> &quot;multiPhylo&quot;</span></b>
<b><span style='color:#54ffff;'># Mean distances</span></b>
oxalis.tree.sp.mean <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>speciesTree</span></b><b>(</b>oxalis.trees.ultra, mean<b>)</b>
<b><span style='color:#dede49;'>print.phylo</span></b><b>(</b>oxalis.tree.sp.mean<b>)</b>
<b><span style='color:#54ffff;'># Plot the tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>oxalis.tree.sp.mean, <b><span style='color:#54ff54;'>edge.width=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>label.offset=</span></b><b><span style='color:#ff54ff;'>0.01</span></b><b>)</b>
<b><span style='color:#dede49;'>edgelabels</span></b><b>(</b><b><span style='color:#54ff54;'>text=</span></b><b><span style='color:#dede49;'>round</span></b><b>(</b>oxalis.tree.sp.mean<b>[[</b><b><span style='color:#ff54ff;'>&quot;edge.length&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>digits=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>, <b><span style='color:#54ff54;'>frame=</span></b><b><span style='color:#ff54ff;'>&quot;none&quot;</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;none&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>axisPhylo</span></b><b>(</b><b><span style='color:#54ff54;'>side=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>

<b><span style='color:#54ffff;'># Parsimony super tree</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>phytools<b>)</b>
oxalis.tree.sp <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>mrp.supertree</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>oxalis.trees.rooted, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;optim.parsimony&quot;</span></b>, <b><span style='color:#54ff54;'>rooted=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>print.phylo</span></b><b>(</b>oxalis.tree.sp<b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>oxalis.tree.sp, <b><span style='color:#54ff54;'>edge.width=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>label.offset=</span></b><b><span style='color:#ff54ff;'>0.01</span></b><b>)</b>
<b><span style='color:#dede49;'>axisPhylo</span></b><b>(</b><b><span style='color:#54ff54;'>side=</span></b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>
?phangorn<b>::</b>superTree <b><span style='color:#54ffff;'># Similar function</span></b>

<b><span style='color:#54ffff;'># Density tree</span></b>
<b><span style='color:#54ffff;'># Prepare list of trees to show</span></b>
hauss.nj.trees <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>list</span></b><b>(</b>hauss.nj, hauss.nj.bruvo, hauss.nj.rooted<b>)</b>
hauss.nj.trees <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>hauss.nj.trees, <b><span style='color:#54ff54;'>FUN=</span></b>compute.brlen<b>)</b>
hauss.nj.trees <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>lapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b>hauss.nj.trees, <b><span style='color:#54ff54;'>FUN=</span></b>chronos<b>)</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b>hauss.nj.trees<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> &quot;multiPhylo&quot;</span></b>
<b><span style='color:#54ffff;'># The trees should be (otherwise plotting works, but may be more ugly)...</span></b>
<b><span style='color:#dede49;'>is.rooted.multiPhylo</span></b><b>(</b>hauss.nj.trees<b>)</b> <b><span style='color:#54ffff;'># rooted,</span></b>
<b><span style='color:#dede49;'>is.ultrametric.multiPhylo</span></b><b>(</b>hauss.nj.trees<b>)</b> <b><span style='color:#54ffff;'># ultrametric and</span></b>
<b><span style='color:#dede49;'>is.binary.multiPhylo</span></b><b>(</b>hauss.nj.trees<b>)</b> <b><span style='color:#54ffff;'># binary bifurcating.</span></b>
<b><span style='color:#54ffff;'># Plotting has various options, play with it</span></b>
phangorn<b>::</b><b><span style='color:#dede49;'>densiTree</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj.trees, <b><span style='color:#54ff54;'>direction=</span></b><b><span style='color:#ff54ff;'>&quot;downwards&quot;</span></b>, <b><span style='color:#54ff54;'>scaleX=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>rainbow</span></b><b>(</b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>, <b><span style='color:#54ff54;'>width=</span></b><b><span style='color:#ff54ff;'>5</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#dede49;'>densiTree</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj.trees, <b><span style='color:#54ff54;'>direction=</span></b><b><span style='color:#ff54ff;'>&quot;upwards&quot;</span></b>, <b><span style='color:#54ff54;'>scaleX=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>width=</span></b><b><span style='color:#ff54ff;'>5</span></b><b>)</b>
<b><span style='color:#dede49;'>densiTree</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj.trees, <b><span style='color:#54ff54;'>scaleX=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>width=</span></b><b><span style='color:#ff54ff;'>5</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#54ffff;'># Compare this option with similar on following slide</span></b>
?phangorn<b>::</b>densiTree
?phytools<b>::</b>densityTree
<b><span style='color:#54ffff;'># Another version for comparing several trees</span></b>
phytools<b>::</b><b><span style='color:#dede49;'>densityTree</span></b><b>(</b><b><span style='color:#54ff54;'>trees=</span></b>oxalis.trees.ultra, <b><span style='color:#54ff54;'>fix.depth=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>use.gradient=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>alpha=</span></b><b><span style='color:#ff54ff;'>0.5</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>4</span></b><b>)</b>
phytools<b>::</b><b><span style='color:#dede49;'>densityTree</span></b><b>(</b><b><span style='color:#54ff54;'>trees=</span></b>oxalis.trees.ultra<b>[</b><b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>3</span></b><b>]</b>, <b><span style='color:#54ff54;'>fix.depth=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>use.gradient=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>alpha=</span></b><b><span style='color:#ff54ff;'>0.5</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>4</span></b><b>)</b>
phytools<b>::</b><b><span style='color:#dede49;'>densityTree</span></b><b>(</b><b><span style='color:#54ff54;'>trees=</span></b>oxalis.trees.ultra<b>[</b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>2</span></b>,<b><span style='color:#ff54ff;'>4</span></b>,<b><span style='color:#ff54ff;'>6</span></b>,<b><span style='color:#ff54ff;'>7</span></b><b>)]</b>, <b><span style='color:#54ff54;'>fix.depth=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>use.gradient=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>alpha=</span></b><b><span style='color:#ff54ff;'>0.5</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>4</span></b><b>)</b>

<b><span style='color:#54ffff;'># Networks</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>phangorn<b>)</b>
oxalis.tree.net <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>consensusNet</span></b><b>(</b>oxalis.trees.rooted, <b><span style='color:#54ff54;'>prob=</span></b><b><span style='color:#ff54ff;'>0.25</span></b><b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>oxalis.tree.net, <b><span style='color:#54ff54;'>planar=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;2D&quot;</span></b>, <b><span style='color:#54ff54;'>use.edge.length=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>show.tip.label=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>show.edge.label=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>show.node.label=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>show.nodes=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>edge.color=</span></b><b><span style='color:#ff54ff;'>&quot;black&quot;</span></b>, <b><span style='color:#54ff54;'>tip.color=</span></b><b><span style='color:#ff54ff;'>&quot;blue&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>oxalis.tree.net, <b><span style='color:#54ff54;'>planar=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;3D&quot;</span></b>, <b><span style='color:#54ff54;'>use.edge.length=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>show.tip.label=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>show.edge.label=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>show.node.label=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>show.nodes=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>edge.color=</span></b><b><span style='color:#ff54ff;'>&quot;black&quot;</span></b>, <b><span style='color:#54ff54;'>tip.color=</span></b><b><span style='color:#ff54ff;'>&quot;blue&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># Plot all trees on same scale</span></b>
<b><span style='color:#dede49;'>kronoviz</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>oxalis.trees.rooted, <b><span style='color:#54ff54;'>layout=</span></b><b><span style='color:#dede49;'>length</span></b><b>(</b>oxalis.trees.rooted<b>)</b>, <b><span style='color:#54ff54;'>horiz=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Close graphical device to cancel division of plotting device</span></b>
<b><span style='color:#dede49;'>dev.off</span></b><b>()</b>

<b><span style='color:#b00000;'>## Maximum parsimony</span></b>
<b><span style='color:#54ffff;'># Conversion to phyDat for phangorn</span></b>
meles.phydat <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>as.phyDat</span></b><b>(</b>meles.dna<b>)</b>
<b><span style='color:#54ffff;'># Prepare starting tree</span></b>
meles.tre.ini <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>nj</span></b><b>(</b><b><span style='color:#dede49;'>dist.dna</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.dna, <b><span style='color:#54ff54;'>model=</span></b><b><span style='color:#ff54ff;'>&quot;raw&quot;</span></b><b>))</b>
<b><span style='color:#54ffff;'># Parsimony</span></b>
?parsimony
<b><span style='color:#54ffff;'># Returns maximum parsimony score</span></b>
<b><span style='color:#dede49;'>parsimony</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>meles.tre.ini, <b><span style='color:#54ff54;'>data=</span></b>meles.phydat<b>)</b>
<b><span style='color:#54ffff;'># Optimisation - returns maximum parsimony tree</span></b>
meles.tre.pars <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>optim.parsimony</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>meles.tre.ini, <b><span style='color:#54ff54;'>data=</span></b>meles.phydat<b>)</b>
<b><span style='color:#54ffff;'># Draw a tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>meles.tre.pars, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;cladogram&quot;</span></b>, <b><span style='color:#54ff54;'>edge.width=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Maximum-parsimony tree of Meles&quot;</span></b><b>)</b>

<b><span style='color:#b00000;'>## Compare two trees</span></b>
<b><span style='color:#54ffff;'># Compare topology of the species trees - basically outputs TRUE/FALSE</span></b>
<b><span style='color:#dede49;'>all.equal.phylo</span></b><b>(</b>oxalis.tree.sp, oxalis.tree.sp.mean, <b><span style='color:#54ff54;'>use.edge.length=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
?all.equal.phylo <b><span style='color:#54ffff;'># Use to see comparison possibilities</span></b>
<b><span style='color:#54ffff;'># Plot two trees with connecting lines</span></b>
<b><span style='color:#54ffff;'># We need 2 column matrix with tip labels</span></b>
tips.labels <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>matrix</span></b><b>(</b><b><span style='color:#54ff54;'>data=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#dede49;'>sort</span></b><b>(</b>oxalis.tree.sp<b>[[</b><b><span style='color:#ff54ff;'>&quot;tip.label&quot;</span></b><b>]])</b>, <b><span style='color:#dede49;'>sort</span></b><b>(</b>oxalis.tree.sp.mean<b>[[</b><b><span style='color:#ff54ff;'>&quot;tip.label&quot;</span></b><b>]]))</b>, <b><span style='color:#54ff54;'>nrow=</span></b><b><span style='color:#dede49;'>length</span></b><b>(</b>oxalis.tree.sp<b>[[</b><b><span style='color:#ff54ff;'>&quot;tip.label&quot;</span></b><b>]])</b>, <b><span style='color:#54ff54;'>ncol=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#54ffff;'># Draw a tree - play with graphical parameters and use rotate=TRUE</span></b>
<b><span style='color:#54ffff;'># to be able to adjust fit manually</span></b>
<b><span style='color:#dede49;'>cophyloplot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>ladderize</span></b><b>(</b>oxalis.tree.sp<b>)</b>, <b><span style='color:#54ff54;'>y=</span></b><b><span style='color:#dede49;'>ladderize</span></b><b>(</b>oxalis.tree.sp.mean<b>)</b>,  <b><span style='color:#54ff54;'>assoc=</span></b>tips.labels, <b><span style='color:#54ff54;'>use.edge.length=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>space=</span></b><b><span style='color:#ff54ff;'>60</span></b>, <b><span style='color:#54ff54;'>length.line=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>gap=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;phylogram&quot;</span></b>, <b><span style='color:#54ff54;'>rotate=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>1.5</span></b>, <b><span style='color:#54ff54;'>lty=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Comparing the trees</span></b><b><span style='color:#ff54ff;'>\n</span></b><b><span style='color:#ff54ff;'>Parsimony super tree</span></b><b><span style='color:#ff54ff;'>\t</span></b><b><span style='color:#ff54ff;'>Species tree&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>legend</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;topleft&quot;</span></b>, <b><span style='color:#54ff54;'>legend=</span></b><b><span style='color:#ff54ff;'>&quot;Red lines</span></b><b><span style='color:#ff54ff;'>\n</span></b><b><span style='color:#ff54ff;'>connect tips&quot;</span></b>, <b><span style='color:#54ff54;'>text.col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>0.75</span></b>, <b><span style='color:#54ff54;'>bty=</span></b><b><span style='color:#ff54ff;'>&quot;n&quot;</span></b>, <b><span style='color:#54ff54;'>x.intersp=</span></b><b>-</b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>y.intersp=</span></b><b>-</b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#54ffff;'># Alternative implementation</span></b>
oxalis.cophylo <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>cophylo</span></b><b>(</b><b><span style='color:#54ff54;'>tr1=</span></b>oxalis.tree.sp, <b><span style='color:#54ff54;'>tr2=</span></b>oxalis.tree.sp.mean, <b><span style='color:#54ff54;'>assoc=</span></b><b>(</b><b><span style='color:#dede49;'>cbind</span></b><b>(</b><b><span style='color:#dede49;'>sort</span></b><b>(</b>oxalis.tree.sp<b>$</b>tip.label<b>)</b>, <b><span style='color:#dede49;'>sort</span></b><b>(</b>oxalis.tree.sp<b>$</b>tip.label<b>)))</b>, <b><span style='color:#54ff54;'>rotate=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.cophylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>oxalis.cophylo, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>link.type=</span></b><b><span style='color:#ff54ff;'>&quot;curved&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>title</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;</span></b><b><span style='color:#ff54ff;'>\n</span></b><b><span style='color:#ff54ff;'>Comparisson of species tree (left) and parsimony supertree (right)&quot;</span></b><b>)</b>

<b><span style='color:#b00000;'>## More about plotting the trees</span></b>
?plot.phylo <b><span style='color:#54ffff;'># check it for various possibilities what to influence</span></b>
?par
?points
<b><span style='color:#dede49;'>par</span></b><b>(</b><b><span style='color:#54ff54;'>mfrow=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#ff54ff;'>2</span></b><b>))</b> <b><span style='color:#54ffff;'># Plot two plots in one row</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;cladogram&quot;</span></b>, <b><span style='color:#54ff54;'>use.edge.length=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>direction=</span></b><b><span style='color:#ff54ff;'>&quot;rightwards&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>hauss.nj, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;cladogram&quot;</span></b>, <b><span style='color:#54ff54;'>use.edge.length=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>direction=</span></b><b><span style='color:#ff54ff;'>&quot;leftwards&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>dev.off</span></b><b>()</b> <b><span style='color:#54ffff;'># Close graphical device to cancel par() settings</span></b>

<b><span style='color:#54ffff;'># Nice tiplabels and higlighted tiplabel</span></b>
<b><span style='color:#54ffff;'># Load tree in text format</span></b>
trape <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.tree</span></b><b>(</b><b><span style='color:#54ff54;'>text =</span></b> <b><span style='color:#ff54ff;'>&quot;((Homo, Pan), Gorilla);&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Plot the tree</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>trape, <b><span style='color:#54ff54;'>show.tip.label=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add colored tip labels</span></b>
<b><span style='color:#dede49;'>tiplabels</span></b><b>(</b>trape<b>[[</b><b><span style='color:#ff54ff;'>&quot;tip.label&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;black&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;white&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;black&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;white&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;black&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add colored node labels</span></b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b><b><span style='color:#54ff54;'>text=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;6.4 Ma&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;5.4 Ma&quot;</span></b><b>)</b>, <b><span style='color:#54ff54;'>frame=</span></b><b><span style='color:#ff54ff;'>&quot;circle&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;yellow&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add scale bar</span></b>
<b><span style='color:#dede49;'>add.scale.bar</span></b><b>()</b>
<b><span style='color:#54ffff;'># Note vectors for tip/nodelabels</span></b>

<b><span style='color:#b00000;'>## PIC</span></b>

<b><span style='color:#54ffff;'># Prepare the data</span></b>
<b><span style='color:#54ffff;'># Body mass of primates</span></b>
primates.body <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>4.09434</span></b>, <b><span style='color:#ff54ff;'>3.61092</span></b>, <b><span style='color:#ff54ff;'>2.37024</span></b>, <b><span style='color:#ff54ff;'>2.02815</span></b>, <b><span style='color:#ff54ff;'>1.46968</span></b><b>)</b>
<b><span style='color:#54ffff;'># Longevity of primates</span></b>
primates.longevity <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>4.74493</span></b>, <b><span style='color:#ff54ff;'>3.3322</span></b>, <b><span style='color:#ff54ff;'>3.3673</span></b>, <b><span style='color:#ff54ff;'>2.89037</span></b>, <b><span style='color:#ff54ff;'>2.30259</span></b><b>)</b>
<b><span style='color:#54ffff;'># Add names to the values</span></b>
<b><span style='color:#dede49;'>names</span></b><b>(</b>primates.body<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>names</span></b><b>(</b>primates.longevity<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Homo&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Pongo&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Macaca&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Ateles&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;Galago&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Create a tree in Newick format</span></b>
primates.tree <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.tree</span></b><b>(</b><b><span style='color:#54ff54;'>text=</span></b><b><span style='color:#ff54ff;'>&quot;((((Homo:0.21, Pongo:0.21):0.28, Macaca:0.49):0.13, Ateles:0.62):0.38, Galago:1.00);&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>primates.tree<b>)</b>

<b><span style='color:#54ffff;'># PIC - phylogenetically independent contrasts</span></b>
primates.pic.body <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>pic</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>primates.body, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree, <b><span style='color:#54ff54;'>scaled=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>var.contrasts=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>rescaled.tree=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
primates.pic.longevity <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>pic</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>primates.longevity, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree, <b><span style='color:#54ff54;'>scaled=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>var.contrasts=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>rescaled.tree=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>

<b><span style='color:#54ffff;'># Plot a tree with PIC values</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>primates.tree, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b><b><span style='color:#dede49;'>round</span></b><b>(</b>primates.pic.body, <b><span style='color:#54ff54;'>digits=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>, <b><span style='color:#54ff54;'>adj=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>0</span></b>, <b>-</b><b><span style='color:#ff54ff;'>0.5</span></b><b>)</b>, <b><span style='color:#54ff54;'>frame=</span></b><b><span style='color:#ff54ff;'>&quot;none&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b><b><span style='color:#dede49;'>round</span></b><b>(</b>primates.pic.longevity, <b><span style='color:#54ff54;'>digits=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>, <b><span style='color:#54ff54;'>adj=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#ff54ff;'>1</span></b><b>)</b>, <b><span style='color:#54ff54;'>frame=</span></b><b><span style='color:#ff54ff;'>&quot;none&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>add.scale.bar</span></b><b>()</b>

<b><span style='color:#54ffff;'># Plot PIC</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>primates.pic.body, <b><span style='color:#54ff54;'>y=</span></b>primates.pic.longevity, <b><span style='color:#54ff54;'>pch=</span></b><b><span style='color:#ff54ff;'>16</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#dede49;'>abline</span></b><b>(</b><b><span style='color:#54ff54;'>a=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>b=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>lty=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b> <b><span style='color:#54ffff;'># x=y line</span></b>

<b><span style='color:#54ffff;'># Correlation of PIC of body mass and longevity</span></b>
<b><span style='color:#dede49;'>cor</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>primates.pic.body, <b><span style='color:#54ff54;'>y=</span></b>primates.pic.longevity, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;pearson&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>lm</span></b><b>(</b><b><span style='color:#54ff54;'>formula=</span></b>primates.pic.longevity<b>~</b>primates.pic.body<b>)</b>
<b><span style='color:#54ffff;'># Because PICs have expected mean zero - such linear regressions should be done through the origin (i.e. the intercept is set to zero)</span></b>
<b><span style='color:#dede49;'>lm</span></b><b>(</b><b><span style='color:#54ff54;'>formula=</span></b>primates.pic.longevity<b>~</b>primates.pic.body<b>-</b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>

<b><span style='color:#54ffff;'># Permutation procedure to test PIC</span></b>
<b><span style='color:#dede49;'>lmorigin</span></b><b>(</b><b><span style='color:#54ff54;'>formula=</span></b>primates.pic.longevity<b>~</b>primates.pic.body, <b><span style='color:#54ff54;'>nperm=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>)</b>

<b><span style='color:#54ffff;'># Intraspecific variation</span></b>
<b><span style='color:#54ffff;'># PIC - orthonormal contrasts using the method</span></b>
primates.pic.ortho <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>pic.ortho</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>list</span></b><b>(</b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>)</b>, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>))[</b><b><span style='color:#ff54ff;'>1</span></b>,<b>]</b>, <b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>)</b>, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>))[</b><b><span style='color:#ff54ff;'>2</span></b>,<b>]</b>, <b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>)</b>, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>))[</b><b><span style='color:#ff54ff;'>3</span></b>,<b>]</b>, <b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>)</b>, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>))[</b><b><span style='color:#ff54ff;'>4</span></b>,<b>]</b>, <b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>)</b>, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>))[</b><b><span style='color:#ff54ff;'>5</span></b>,<b>])</b>, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree, <b><span style='color:#54ff54;'>var.contrasts=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>intra=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
primates.pic.ortho
<b><span style='color:#54ffff;'># Explanation of the cbind trick</span></b>
<b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>)</b>, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>))</b>
<b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>)</b>, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>))[</b><b><span style='color:#ff54ff;'>1</span></b>,<b>]</b>
<b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>)</b>, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>))[</b><b><span style='color:#ff54ff;'>2</span></b>,<b>]</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>)</b>, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>)))</b>
<b><span style='color:#dede49;'>class</span></b><b>(</b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>)</b>, <b><span style='color:#dede49;'>jitter</span></b><b>(</b>primates.body<b>))[</b><b><span style='color:#ff54ff;'>1</span></b>,<b>])</b>
<b><span style='color:#54ffff;'># jitter() adds random noise every time, so that the values differ</span></b>

<b><span style='color:#b00000;'>## Phylogenetic autocorrelation</span></b>

<b><span style='color:#54ffff;'># Autocorrelation coefficient to quantify whether the distribution of a trait among a set of species is affected or not by their phylogenetic relationships</span></b>
<b><span style='color:#54ffff;'># In the absence of phylogenetic autocorrelation, the mean expected value of I and its variance are known - it is thus possible to test the null hypothesis of the absence of dependence among observations</span></b>
<b><span style='color:#54ffff;'># Let's choose weights as wij = 1/dij, where the d’s is the distances measured on the tree - cophenetic() calculates cophenetic distances</span></b>
primates.weights <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>1</span></b><b>/</b><b><span style='color:#dede49;'>cophenetic</span></b><b>(</b>primates.tree<b>)</b> <b><span style='color:#54ffff;'># can be just cophenetic(primates.tree) or some other transformation</span></b>
<b><span style='color:#54ffff;'># See it</span></b>
primates.weights
<b><span style='color:#dede49;'>class</span></b><b>(</b>primates.weights<b>)</b>
<b><span style='color:#54ffff;'># Set diagonal to 0</span></b>
<b><span style='color:#dede49;'>diag</span></b><b>(</b>primates.weights<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>0</span></b>
<b><span style='color:#54ffff;'># Calculate Moran's I</span></b>
<b><span style='color:#dede49;'>Moran.I</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>primates.body, <b><span style='color:#54ff54;'>weight=</span></b>primates.weights, <b><span style='color:#54ff54;'>alternative=</span></b><b><span style='color:#ff54ff;'>&quot;greater&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Slighly significant positive phylogenetic correlation among body mass</span></b>
<b><span style='color:#dede49;'>Moran.I</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>primates.longevity, <b><span style='color:#54ff54;'>weight=</span></b>primates.weights, <b><span style='color:#54ff54;'>alternative=</span></b><b><span style='color:#ff54ff;'>&quot;greater&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Positive, but non-significant</span></b>

<b><span style='color:#54ffff;'># Test of Moran's with randomisation procedure</span></b>
<b><span style='color:#dede49;'>gearymoran</span></b><b>(</b><b><span style='color:#54ff54;'>bilis=</span></b>primates.weights, <b><span style='color:#54ff54;'>X=</span></b><b><span style='color:#dede49;'>data.frame</span></b><b>(</b>primates.body, primates.longevity<b>)</b>, <b><span style='color:#54ff54;'>nrepet=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>)</b> <b><span style='color:#54ffff;'># Body is significant - nonrandom, longevity not (random)</span></b>

<b><span style='color:#54ffff;'># Test of Abouheif designed to detect phylogenetic autocorrelation in a quantitative trait - in fact Moran's I test using a particular phylogenetic proximity between tips</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>adephylo<b>)</b>
<b><span style='color:#dede49;'>abouheif.moran</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, primates.longevity<b>)</b>, <b><span style='color:#54ff54;'>W=</span></b>primates.weights, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;oriAbouheif&quot;</span></b>, <b><span style='color:#54ff54;'>nrepet=</span></b><b><span style='color:#ff54ff;'>1000</span></b>, <b><span style='color:#54ff54;'>alter=</span></b><b><span style='color:#ff54ff;'>&quot;greater&quot;</span></b><b>)</b>

<b><span style='color:#54ffff;'># correlogram can be used to visualize the results of phylogenetic autocorrelative analysis</span></b>
<b><span style='color:#54ffff;'># Loads training data set</span></b>
<b><span style='color:#dede49;'>data</span></b><b>(</b>carnivora<b>)</b>
<b><span style='color:#54ffff;'># Look at the data</span></b>
<b><span style='color:#dede49;'>head</span></b><b>(</b>carnivora<b>)</b>
<b><span style='color:#54ffff;'># Calculate the correlogram</span></b>
carnivora.correlogram <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>correlogram.formula</span></b><b>(</b><b><span style='color:#54ff54;'>formula=</span></b>SW<b>~</b>Order<b>/</b>SuperFamily<b>/</b>Family<b>/</b>Genus, <b><span style='color:#54ff54;'>data=</span></b>carnivora<b>)</b>
<b><span style='color:#54ffff;'># See results</span></b>
carnivora.correlogram
<b><span style='color:#54ffff;'># Calculate the correlogram - test for both body masses</span></b>
carnivora.correlogram2 <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>correlogram.formula</span></b><b>(</b><b><span style='color:#54ff54;'>formula=</span></b>SW<b>+</b>FW<b>~</b>Order<b>/</b>SuperFamily<b>/</b>Family<b>/</b>Genus, <b><span style='color:#54ff54;'>data=</span></b>carnivora<b>)</b>
<b><span style='color:#54ffff;'># See results</span></b>
carnivora.correlogram2
<b><span style='color:#54ffff;'># Plot it</span></b>
<b><span style='color:#dede49;'>plot.correlogram</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>carnivora.correlogram, <b><span style='color:#54ff54;'>legend=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>test.level=</span></b><b><span style='color:#ff54ff;'>0.05</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;black&quot;</span></b><b>))</b>
<b><span style='color:#54ffff;'># Plot it - test for both body masses</span></b>
<b><span style='color:#54ffff;'># Two graphs</span></b>
<b><span style='color:#dede49;'>plot.correlogramList</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>carnivora.correlogram2, <b><span style='color:#54ff54;'>lattice=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>legend=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>test.level=</span></b><b><span style='color:#ff54ff;'>0.05</span></b><b>)</b>
<b><span style='color:#54ffff;'># Only one graph</span></b>
<b><span style='color:#dede49;'>plot.correlogramList</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>carnivora.correlogram2, <b><span style='color:#54ff54;'>lattice=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>legend=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>test.level=</span></b><b><span style='color:#ff54ff;'>0.05</span></b><b>)</b>

<b><span style='color:#b00000;'>## Orthonormal Decomposition - phylogenetic eigenvector regression</span></b>

<b><span style='color:#54ffff;'># Prepare toy data</span></b>
<b><span style='color:#54ffff;'># Load MrBayes tree in NEXUS format</span></b>
apiaceae.tree <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>read.nexus</span></b><b>(</b><b><span style='color:#54ff54;'>file=</span></b><b><span style='color:#ff54ff;'>&quot;https://soubory.trapa.cz/rcourse/apiaceae_mrbayes.nexus&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># See it</span></b>
<b><span style='color:#dede49;'>print.phylo</span></b><b>(</b>apiaceae.tree<b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>apiaceae.tree<b>)</b>
<b><span style='color:#54ffff;'># Root the tree</span></b>
apiaceae.tree <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>root</span></b><b>(</b>apiaceae.tree, <b><span style='color:#ff54ff;'>&quot;Aralia_elata&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Remove &quot;_&quot; from taxa names</span></b>
<b><span style='color:#54ffff;'># plot.phylo() by default omits &quot;_&quot; from tip names</span></b>
apiaceae.tree<b>$</b>tip.label <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>gsub</span></b><b>(</b><b><span style='color:#54ff54;'>pattern=</span></b><b><span style='color:#ff54ff;'>&quot;_&quot;</span></b>, <b><span style='color:#54ff54;'>replacement=</span></b><b><span style='color:#ff54ff;'>&quot; &quot;</span></b>, <b><span style='color:#54ff54;'>x=</span></b>apiaceae.tree<b>$</b>tip.label<b>)</b>
<b><span style='color:#54ffff;'># Drop outgroup (Aralia and Hydrocotyle)</span></b>
<b><span style='color:#54ffff;'># Click on last common ancestor of ingroup desired to be kept</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>apiaceae.tree<b>)</b>
apiaceae.tree <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>extract.clade</span></b><b>(</b>apiaceae.tree, <b><span style='color:#54ff54;'>interactive=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>apiaceae.tree<b>)</b>
<b><span style='color:#54ffff;'># Decomposition of topographical distances (right plot)</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>adephylo<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>phylobase<b>)</b>
<b><span style='color:#dede49;'>table.phylo4d</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>phylo4d</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>apiaceae.tree, <b><span style='color:#54ff54;'>tip.data=</span></b><b><span style='color:#dede49;'>treePart</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>apiaceae.tree, <b><span style='color:#54ff54;'>result=</span></b><b><span style='color:#ff54ff;'>&quot;orthobasis&quot;</span></b><b>))</b>, <b><span style='color:#54ff54;'>treetype=</span></b><b><span style='color:#ff54ff;'>&quot;cladogram&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Generate some random variable</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>geiger<b>)</b>
apiaceae.eco <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>sim.char</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>apiaceae.tree, <b><span style='color:#54ff54;'>par=</span></b><b><span style='color:#ff54ff;'>0.1</span></b>, <b><span style='color:#54ff54;'>nsim=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>model=</span></b><b><span style='color:#ff54ff;'>&quot;BM&quot;</span></b><b>)[</b>,,<b><span style='color:#ff54ff;'>1</span></b><b>]</b>
?sim.char <b><span style='color:#54ffff;'># See it for another possibilities to simulate data</span></b>
<b><span style='color:#54ffff;'># Names for the values</span></b>
<b><span style='color:#dede49;'>names</span></b><b>(</b>apiaceae.eco<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>apiaceae.tree<b>[[</b><b><span style='color:#ff54ff;'>&quot;tip.label&quot;</span></b><b>]]</b>
apiaceae.eco <b><span style='color:#54ffff;'># See it</span></b>

<b><span style='color:#54ffff;'># significant result - significant phylogenetic inertia (phylogenetic effect) - the tendency for traits to resist evolutionary change despite environmental perturbations</span></b>
<b><span style='color:#dede49;'>anova</span></b><b>(</b><b><span style='color:#dede49;'>lm</span></b><b>(</b>apiaceae.eco <b>~</b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>as.matrix</span></b><b>(</b><b><span style='color:#dede49;'>orthobasis.phylo</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>apiaceae.tree, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;patristic&quot;</span></b><b>)[</b>,<b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>2</span></b><b>])))</b>
<b><span style='color:#dede49;'>orthogram</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>apiaceae.eco, <b><span style='color:#54ff54;'>tre=</span></b>apiaceae.tree, <b><span style='color:#54ff54;'>nrepet=</span></b><b><span style='color:#ff54ff;'>1000</span></b>, <b><span style='color:#54ff54;'>alter=</span></b><b><span style='color:#ff54ff;'>&quot;two-sided&quot;</span></b><b>)</b>
?orthogram <b><span style='color:#54ffff;'># See another calculation possibilities</span></b>

<b><span style='color:#b00000;'>## Phylogenetic Generalized Least Squares</span></b>

<b><span style='color:#54ffff;'># Funkce pro correlation: corBlomberg(), corMartins(), corPagel(), corBrownian() - stejné parametry (kromě hodnoty 'value')</span></b>
<b><span style='color:#54ffff;'># Fitting an Ornstein-Uhlenbeck Motion model in PGLS</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>nlme<b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>ape<b>)</b>
<b><span style='color:#dede49;'>summary</span></b><b>(</b><b><span style='color:#dede49;'>gls</span></b><b>(</b><b><span style='color:#54ff54;'>model=</span></b>primates.longevity <b>~</b><b><span style='color:#ff54ff;'> </span></b>primates.body, <b><span style='color:#54ff54;'>data=</span></b><b><span style='color:#dede49;'>as.data.frame</span></b><b>(</b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.longevity, primates.body<b>))</b>, <b><span style='color:#54ff54;'>correlation=</span></b><b><span style='color:#dede49;'>corBrownian</span></b><b>(</b><b><span style='color:#54ff54;'>value=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree<b>)))</b>

<b><span style='color:#54ffff;'># Implementation in caper package</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>caper<b>)</b> <b><span style='color:#54ffff;'># Load needed library</span></b>
<b><span style='color:#dede49;'>data</span></b><b>(</b>shorebird<b>)</b> <b><span style='color:#54ffff;'># Load training data</span></b>
?shorebird.data
shorebird.pgls <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>pgls</span></b><b>(</b><b><span style='color:#54ff54;'>formula=</span></b>shorebird.data<b>[[</b><b><span style='color:#ff54ff;'>&quot;F.Mass&quot;</span></b><b>]]</b> <b>~</b><b><span style='color:#ff54ff;'> </span></b>shorebird.data<b>[[</b><b><span style='color:#ff54ff;'>&quot;Egg.Mass&quot;</span></b><b>]]</b>, <b><span style='color:#54ff54;'>data=</span></b><b><span style='color:#dede49;'>comparative.data</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>shorebird.tree, <b><span style='color:#54ff54;'>data=</span></b><b><span style='color:#dede49;'>as.data.frame</span></b><b>(</b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>shorebird.data<b>[[</b><b><span style='color:#ff54ff;'>&quot;F.Mass&quot;</span></b><b>]]</b>, shorebird.data<b>[[</b><b><span style='color:#ff54ff;'>&quot;Egg.Mass&quot;</span></b><b>]]</b>, shorebird.data<b>[[</b><b><span style='color:#ff54ff;'>&quot;Species&quot;</span></b><b>]]))</b>, <b><span style='color:#54ff54;'>names.col=</span></b>V3, <b><span style='color:#54ff54;'>vcv=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>))</b>
<b><span style='color:#dede49;'>summary</span></b><b>(</b>shorebird.pgls<b>)</b> <b><span style='color:#54ffff;'># See the result</span></b>
<b><span style='color:#54ffff;'># See the plot of observer and fitted values</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>shorebird.pgls<b>)</b>
<b><span style='color:#dede49;'>abline</span></b><b>(</b><b><span style='color:#54ff54;'>a=</span></b><b><span style='color:#ff54ff;'>0</span></b>, <b><span style='color:#54ff54;'>b=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>anova</span></b><b>(</b>shorebird.pgls<b>)</b> <b><span style='color:#54ffff;'># ANOVA view of the model</span></b>
<b><span style='color:#dede49;'>AIC</span></b><b>(</b>shorebird.pgls<b>)</b> <b><span style='color:#54ffff;'># Akaike's information criterion (smaller = better)</span></b>

<b><span style='color:#b00000;'>## Generalized Estimating Equations</span></b>

<b><span style='color:#54ffff;'># Calculate the model</span></b>
<b><span style='color:#dede49;'>compar.gee</span></b><b>(</b><b><span style='color:#54ff54;'>formula=</span></b>primates.longevity <b>~</b><b><span style='color:#ff54ff;'> </span></b>primates.body, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree<b>)</b>
<b><span style='color:#54ffff;'># or with correlation matrix:</span></b>
<b><span style='color:#dede49;'>compar.gee</span></b><b>(</b><b><span style='color:#54ff54;'>formula=</span></b>primates.longevity <b>~</b><b><span style='color:#ff54ff;'> </span></b>primates.body, <b><span style='color:#54ff54;'>corStruct=</span></b><b><span style='color:#dede49;'>corMartins</span></b><b>(</b><b><span style='color:#54ff54;'>value=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree, <b><span style='color:#54ff54;'>fixed=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>))</b>
<b><span style='color:#54ffff;'># for corStruct there are similar functions corBlomberg, corMartins, corPagel, corBrownian - see manuals for differences</span></b>

<b><span style='color:#b00000;'>## Phylogenetic Signal</span></b>

<b><span style='color:#dede49;'>library</span></b><b>(</b>picante<b>)</b>
<b><span style='color:#54ffff;'># If Blomberg's values of 1 correspond to a Brownian motion process, which implies some degree of phylogenetic signal or conservatism. K values closer to zero correspond to a random or convergent pattern of evolution, while K values greater than 1 indicate strong phylogenetic signal and conservatism of traits.</span></b>
<b><span style='color:#54ffff;'># Test for Bloomberg's K statistics</span></b>
<b><span style='color:#dede49;'>Kcalc</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>apiaceae.eco, <b><span style='color:#54ff54;'>phy=</span></b>apiaceae.tree, <b><span style='color:#54ff54;'>checkdata=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Test with permutations</span></b>
<b><span style='color:#dede49;'>phylosignal</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>apiaceae.eco, <b><span style='color:#54ff54;'>phy=</span></b>apiaceae.tree, <b><span style='color:#54ff54;'>reps=</span></b><b><span style='color:#ff54ff;'>1000</span></b>, <b><span style='color:#54ff54;'>checkdata=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># sapply performs analysis on list of variables (numeric vectors)</span></b>
<b><span style='color:#dede49;'>sapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b><b><span style='color:#dede49;'>list</span></b><b>(</b><b><span style='color:#54ff54;'>body=</span></b>primates.body, <b><span style='color:#54ff54;'>longevity=</span></b>primates.longevity<b>)</b>, <b><span style='color:#54ff54;'>FUN=</span></b>Kcalc, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree, <b><span style='color:#54ff54;'>checkdata=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>
<b><span style='color:#dede49;'>sapply</span></b><b>(</b><b><span style='color:#54ff54;'>X=</span></b><b><span style='color:#dede49;'>list</span></b><b>(</b><b><span style='color:#54ff54;'>body=</span></b>primates.body, <b><span style='color:#54ff54;'>longevity=</span></b>primates.longevity<b>)</b>, <b><span style='color:#54ff54;'>FUN=</span></b>phylosignal, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree, <b><span style='color:#54ff54;'>reps=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>)</b>
<b><span style='color:#54ffff;'># Alternative to use phylosignal with sapply:</span></b>
<b><span style='color:#dede49;'>multiPhylosignal</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>as.data.frame</span></b><b>(</b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, primates.longevity<b>))</b>, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree, <b><span style='color:#54ff54;'>reps=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>)</b>
<b><span style='color:#54ffff;'># Note sapply() and multiPhylosignal() return same data, but the matrices are transposed - use t() to transpose one to look like the other:</span></b>
<b><span style='color:#dede49;'>t</span></b><b>(</b><b><span style='color:#dede49;'>multiPhylosignal</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>as.data.frame</span></b><b>(</b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, primates.longevity<b>))</b>, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree, <b><span style='color:#54ff54;'>reps=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>))</b>

<b><span style='color:#54ffff;'># When there are vectors with standard errors of measurements</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>phytools<b>)</b>
?phylosig <b><span style='color:#54ffff;'># See for details</span></b>
<b><span style='color:#54ffff;'># Test for phylogenetic signal (here without SE)</span></b>
<b><span style='color:#dede49;'>phylosig</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>apiaceae.tree, <b><span style='color:#54ff54;'>x=</span></b>apiaceae.eco, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;K&quot;</span></b>, <b><span style='color:#54ff54;'>test=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>nsim=</span></b><b><span style='color:#ff54ff;'>1000</span></b><b>)</b>
<b><span style='color:#dede49;'>phylosig</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>primates.tree, <b><span style='color:#54ff54;'>x=</span></b>primates.body, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;lambda&quot;</span></b>, <b><span style='color:#54ff54;'>test=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># phylosig() can be used as an alternative to phylosignal() - the functions are similar in basic usage</span></b>

<b><span style='color:#54ffff;'># Examples of usage of GLS for testing of phylogenetic signal</span></b>
<b><span style='color:#dede49;'>summary</span></b><b>(</b><b><span style='color:#dede49;'>gls</span></b><b>(</b><b><span style='color:#54ff54;'>model=</span></b>primates.longevity <b>~</b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>data=</span></b><b><span style='color:#dede49;'>as.data.frame</span></b><b>(</b>primates.longevity<b>)</b>, <b><span style='color:#54ff54;'>correlation=</span></b><b><span style='color:#dede49;'>corBrownian</span></b><b>(</b><b><span style='color:#54ff54;'>value=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree<b>)))</b>
<b><span style='color:#dede49;'>summary</span></b><b>(</b><b><span style='color:#dede49;'>pgls</span></b><b>(</b><b><span style='color:#54ff54;'>formula=</span></b>shorebird.data<b>[[</b><b><span style='color:#ff54ff;'>&quot;M.Mass&quot;</span></b><b>]]</b> <b>~</b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>data=</span></b><b><span style='color:#dede49;'>comparative.data</span></b><b>(</b><b><span style='color:#54ff54;'>phy=</span></b>shorebird.tree, <b><span style='color:#54ff54;'>data=</span></b><b><span style='color:#dede49;'>as.data.frame</span></b><b>(</b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>shorebird.data<b>[[</b><b><span style='color:#ff54ff;'>&quot;M.Mass&quot;</span></b><b>]]</b>, shorebird.data<b>[[</b><b><span style='color:#ff54ff;'>&quot;Species&quot;</span></b><b>]]))</b>, <b><span style='color:#54ff54;'>names.col=</span></b>V2, <b><span style='color:#54ff54;'>vcv=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)))</b>

<b><span style='color:#b00000;'>## Phylogenetic PCA</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>adephylo<b>)</b> <b><span style='color:#54ffff;'># Library needed to create phylo4d object required by ppca</span></b>
<b><span style='color:#54ffff;'># Calculate pPCA</span></b>
primates.ppca <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>ppca</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>phylo4d</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>primates.tree, <b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, primates.longevity<b>))</b>, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;patristic&quot;</span></b>, <b><span style='color:#54ff54;'>center=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>scale=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>scannf=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>nfposi=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>nfnega=</span></b><b><span style='color:#ff54ff;'>0</span></b><b>)</b>
<b><span style='color:#54ffff;'># Print results</span></b>
<b><span style='color:#dede49;'>print</span></b><b>(</b>primates.ppca<b>)</b>
<b><span style='color:#54ffff;'># See summary information</span></b>
<b><span style='color:#dede49;'>summary</span></b><b>(</b>primates.ppca<b>)</b>
<b><span style='color:#54ffff;'># See PCA scores for variables on phylogenetic tree</span></b>
<b><span style='color:#dede49;'>scatter</span></b><b>(</b>primates.ppca<b>)</b>
<b><span style='color:#54ffff;'># See decomposition of pPCA eigenvalues</span></b>
<b><span style='color:#dede49;'>screeplot</span></b><b>(</b>primates.ppca<b>)</b>
<b><span style='color:#54ffff;'># Plot pPCA results - global vs. local structure, decomposition of pPCA eigenvalues, PCA plot of variables and PCA scores for variables on phylogenetic tree</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>primates.ppca<b>)</b>

<b><span style='color:#b00000;'>## Ancestral state reconstruction</span></b>

<b><span style='color:#54ffff;'># By default performs estimation for continuous characters assuming a Brownian motion model fit by maximum likelihood</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>ape<b>)</b>
<b><span style='color:#54ffff;'># See ?ace for possible settings</span></b>
primates.body.ace <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>ace</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>primates.body, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;continuous&quot;</span></b>, <b><span style='color:#54ff54;'>method=</span></b><b><span style='color:#ff54ff;'>&quot;REML&quot;</span></b>, <b><span style='color:#54ff54;'>corStruct=</span></b><b><span style='color:#dede49;'>corBrownian</span></b><b>(</b><b><span style='color:#54ff54;'>value=</span></b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#54ff54;'>phy=</span></b>primates.tree<b>))</b>
<b><span style='color:#54ffff;'># See result - reconstructions are in $ace - to be plotted on nodes - 1st column are node numbers</span></b>
primates.body.ace
<b><span style='color:#54ffff;'># Other implementations are available in packages geiger (functions fitContinuousMCMC and fitDiscrete), phangorn and more in ape (MPR)</span></b>
<b><span style='color:#54ffff;'># Plot it</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>primates.tree, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>tiplabels</span></b><b>(</b><b><span style='color:#dede49;'>round</span></b><b>(</b>primates.body, <b><span style='color:#54ff54;'>digits=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>, <b><span style='color:#54ff54;'>adj=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>0</span></b>, <b>-</b><b><span style='color:#ff54ff;'>1</span></b><b>)</b>, <b><span style='color:#54ff54;'>frame=</span></b><b><span style='color:#ff54ff;'>&quot;none&quot;</span></b>, <b><span style='color:#54ff54;'>col=</span></b><b><span style='color:#ff54ff;'>&quot;blue&quot;</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b><b><span style='color:#dede49;'>round</span></b><b>(</b>primates.body.ace<b>$</b>ace, <b><span style='color:#54ff54;'>digits=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>, <b><span style='color:#54ff54;'>frame=</span></b><b><span style='color:#ff54ff;'>&quot;circle&quot;</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#54ffff;'># ACE returns long numbers - truncate them by e.g. </span></b>
<b><span style='color:#dede49;'>round</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>..., <b><span style='color:#54ff54;'>digits=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b> <b><span style='color:#54ffff;'># &quot;x&quot; is vector with ACE values</span></b>

<b><span style='color:#dede49;'>library</span></b><b>(</b>phytools<b>)</b>
<b><span style='color:#54ffff;'># More possibilities</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>primates.tree, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#54ffff;'># ML estimation of a continuous trait, can compute confidence interval (used by some functions, see further)</span></b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b><b><span style='color:#dede49;'>fastAnc</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>primates.tree, <b><span style='color:#54ff54;'>x=</span></b>primates.body<b>))</b>
<b><span style='color:#54ffff;'># ACE for Brownian evolution with directional trend</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>primates.tree, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b><b><span style='color:#dede49;'>anc.trend</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>primates.tree, <b><span style='color:#54ff54;'>x=</span></b>primates.body, <b><span style='color:#54ff54;'>maxit=</span></b><b><span style='color:#ff54ff;'>100000</span></b><b>)</b><b>$</b>ace<b>)</b>
<b><span style='color:#54ffff;'># ACE for Brownian evolution using likelihood</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>primates.tree, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b><b><span style='color:#dede49;'>round</span></b><b>(</b><b><span style='color:#dede49;'>anc.ML</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>primates.tree, <b><span style='color:#54ff54;'>x=</span></b>primates.body, <b><span style='color:#54ff54;'>maxit=</span></b><b><span style='color:#ff54ff;'>100000</span></b>, <b><span style='color:#54ff54;'>model=</span></b><b><span style='color:#ff54ff;'>&quot;BM&quot;</span></b><b>)</b><b>$</b>ace, <b><span style='color:#54ff54;'>digits=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>

<b><span style='color:#54ffff;'># Bayesian ancestral character estimation</span></b>
primates.body.ace.bayes <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>anc.Bayes</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>primates.tree, <b><span style='color:#54ff54;'>x=</span></b>primates.body, <b><span style='color:#54ff54;'>ngen=</span></b><b><span style='color:#ff54ff;'>100000</span></b><b>)</b> <b><span style='color:#54ffff;'># Use more MCMC generations</span></b>
primates.body.ace.bayes
<b><span style='color:#54ffff;'># Get end of ancestral states from Bayesian posterior distribution (it should converge to certain values)</span></b>
<b><span style='color:#dede49;'>tail</span></b><b>(</b>primates.body.ace.bayes<b>[[</b><b><span style='color:#ff54ff;'>&quot;mcmc&quot;</span></b><b>]])</b>
primates.body.ace.bayes<b>[[</b><b><span style='color:#ff54ff;'>&quot;mcmc&quot;</span></b><b>]][</b><b><span style='color:#ff54ff;'>1001</span></b>,<b><span style='color:#ff54ff;'>3</span></b><b>:</b><b><span style='color:#ff54ff;'>6</span></b><b>]</b>
<b><span style='color:#54ffff;'># Get means of ancestral states from Bayesian posterior distribution</span></b>
<b><span style='color:#dede49;'>colMeans</span></b><b>(</b>primates.body.ace.bayes<b>[[</b><b><span style='color:#ff54ff;'>&quot;mcmc&quot;</span></b><b>]][</b><b><span style='color:#ff54ff;'>201</span></b><b>:</b><b><span style='color:#dede49;'>nrow</span></b><b>(</b>primates.body.ace.bayes<b>[[</b><b><span style='color:#ff54ff;'>&quot;mcmc&quot;</span></b><b>]])</b>,<b><span style='color:#dede49;'>as.character</span></b><b>(</b><b><span style='color:#ff54ff;'>1</span></b><b>:</b>primates.tree<b>$</b>Nnode<b>+</b><b><span style='color:#dede49;'>length</span></b><b>(</b>primates.tree<b>$</b>tip.label<b>))])</b>
<b><span style='color:#54ffff;'># Plot the ancestral states from posterior distribution (it should converge to certain values)</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>primates.body.ace.bayes<b>)</b>
<b><span style='color:#54ffff;'># Plot the tree and reconstructed ancestral states</span></b>
<b><span style='color:#dede49;'>plot.phylo</span></b><b>(</b>primates.tree, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b>
<b><span style='color:#dede49;'>nodelabels</span></b><b>(</b><b><span style='color:#dede49;'>round</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>primates.body.ace.bayes<b>[[</b><b><span style='color:#ff54ff;'>&quot;mcmc&quot;</span></b><b>]][</b><b><span style='color:#ff54ff;'>1001</span></b>,<b><span style='color:#ff54ff;'>3</span></b><b>:</b><b><span style='color:#ff54ff;'>6</span></b><b>]</b>, <b><span style='color:#54ff54;'>digits=</span></b><b><span style='color:#ff54ff;'>3</span></b><b>)</b>, <b><span style='color:#54ff54;'>cex=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#54ffff;'># Another possibility for ancestral character reconstruction</span></b>
?phangorn<b>::</b>ancestral.pml

<b><span style='color:#54ffff;'># Continuous map</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>phytools<b>)</b>
<b><span style='color:#dede49;'>contMap</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>primates.tree, <b><span style='color:#54ff54;'>x=</span></b>primates.body<b>)</b>
<b><span style='color:#54ffff;'># Change colors with setMap()</span></b>
primates.contmap <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>setMap</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>contMap</span></b><b>(</b>primates.tree, primates.body<b>)</b>, <b><span style='color:#54ff54;'>colors=</span></b><b><span style='color:#dede49;'>c</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b>, <b><span style='color:#ff54ff;'>&quot;black&quot;</span></b><b>))</b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>primates.contmap<b>)</b>
<b><span style='color:#54ffff;'># See ?par for more settings</span></b>

<b><span style='color:#b00000;'>## Phenogram</span></b>

<b><span style='color:#dede49;'>library</span></b><b>(</b>adephylo<b>)</b>
<b><span style='color:#dede49;'>table.phylo4d</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>phylo4d</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>primates.tree, <b><span style='color:#54ff54;'>tip.data=</span></b><b><span style='color:#dede49;'>as.data.frame</span></b><b>(</b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, primates.longevity<b>)))</b>, <b><span style='color:#54ff54;'>treetype=</span></b><b><span style='color:#ff54ff;'>&quot;cladogram&quot;</span></b>, <b><span style='color:#54ff54;'>symbol=</span></b><b><span style='color:#ff54ff;'>&quot;circles&quot;</span></b>, <b><span style='color:#54ff54;'>scale=</span></b><b><span style='color:#00ff00;'>FALSE</span></b>, <b><span style='color:#54ff54;'>ratio.tree=</span></b><b><span style='color:#ff54ff;'>0.5</span></b><b>)</b>
<b><span style='color:#dede49;'>table.phylo4d</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#dede49;'>phylo4d</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b>shorebird.tree, <b><span style='color:#54ff54;'>tip.data=</span></b>shorebird.data<b>)</b>, <b><span style='color:#54ff54;'>treetype=</span></b><b><span style='color:#ff54ff;'>&quot;cladogram&quot;</span></b>, <b><span style='color:#54ff54;'>symbol=</span></b><b><span style='color:#ff54ff;'>&quot;circles&quot;</span></b>, <b><span style='color:#54ff54;'>scale=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>ratio.tree=</span></b><b><span style='color:#ff54ff;'>0.5</span></b><b>)</b>
<b><span style='color:#dede49;'>phenogram</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>primates.tree, <b><span style='color:#54ff54;'>x=</span></b>primates.longevity, <b><span style='color:#54ff54;'>fsize=</span></b><b><span style='color:#ff54ff;'>1.2</span></b>, <b><span style='color:#54ff54;'>ftype=</span></b><b><span style='color:#ff54ff;'>&quot;i&quot;</span></b>, <b><span style='color:#54ff54;'>colors=</span></b><b><span style='color:#ff54ff;'>&quot;red&quot;</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;Longevity&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>fancyTree</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>primates.tree, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;phenogram95&quot;</span></b>, <b><span style='color:#54ff54;'>x=</span></b>primates.longevity, <b><span style='color:#54ff54;'>fsize=</span></b><b><span style='color:#ff54ff;'>1.2</span></b>, <b><span style='color:#54ff54;'>ftype=</span></b><b><span style='color:#ff54ff;'>&quot;i&quot;</span></b>, <b><span style='color:#54ff54;'>main=</span></b><b><span style='color:#ff54ff;'>&quot;95-percentile of longevity&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># 2 characters on 2 axis</span></b>
<b><span style='color:#dede49;'>phylomorphospace</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>primates.tree, <b><span style='color:#54ff54;'>X=</span></b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, primates.longevity<b>)</b>, <b><span style='color:#54ff54;'>label=</span></b><b><span style='color:#ff54ff;'>&quot;horizontal&quot;</span></b>, <b><span style='color:#54ff54;'>lwd=</span></b><b><span style='color:#ff54ff;'>2</span></b>, <b><span style='color:#54ff54;'>fsize=</span></b><b><span style='color:#ff54ff;'>1.5</span></b><b>)</b>
<b><span style='color:#54ffff;'># 3D (3rd character is fake here)</span></b>
<b><span style='color:#54ffff;'># 3 characters it a rotating cube</span></b>
<b><span style='color:#dede49;'>phylomorphospace3d</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>primates.tree, <b><span style='color:#54ff54;'>X=</span></b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, primates.longevity, <b><span style='color:#dede49;'>abs</span></b><b>(</b>primates.body<b>-</b>primates.longevity<b>))</b>, <b><span style='color:#54ff54;'>label=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># 2 characters on 2 axis</span></b>
<b><span style='color:#dede49;'>fancyTree</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>primates.tree, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;scattergram&quot;</span></b>, <b><span style='color:#54ff54;'>X=</span></b><b><span style='color:#dede49;'>cbind</span></b><b>(</b>primates.body, primates.longevity<b>)</b>, <b><span style='color:#54ff54;'>res=</span></b><b><span style='color:#ff54ff;'>500</span></b>, <b><span style='color:#54ff54;'>ftype=</span></b><b><span style='color:#ff54ff;'>&quot;i&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># See manuals for more settings</span></b>
?fancyTree
?phenogram
?phylomorphospace
?phylomorphospace3d
?contMap
?setMap
?par

<b><span style='color:#54ffff;'># Ploting traits on trees</span></b>
<b><span style='color:#54ffff;'># Load training data to display</span></b>
<b><span style='color:#dede49;'>data</span></b><b>(</b>shorebird, <b><span style='color:#54ff54;'>package=</span></b><b><span style='color:#ff54ff;'>&quot;caper&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># See information about the data</span></b>
?caper<b>::</b>shorebird
<b><span style='color:#54ffff;'># Load library</span></b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>phytools<b>)</b>
<b><span style='color:#54ffff;'># Prepare named vector of values to plot</span></b>
shorebird.toplot <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>shorebird.data<b>[[</b><b><span style='color:#ff54ff;'>&quot;Egg.Mass&quot;</span></b><b>]]</b>
<b><span style='color:#dede49;'>names</span></b><b>(</b>shorebird.toplot<b>)</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>rownames</span></b><b>(</b>shorebird.data<b>)</b>
shorebird.toplot <b><span style='color:#54ffff;'># See it</span></b>
<b><span style='color:#54ffff;'># See options for plotting functions</span></b>
?plotTree.wBars
?dotTree
?plotSimmap
<b><span style='color:#54ffff;'># Plot the trees</span></b>
<b><span style='color:#dede49;'>plotTree.wBars</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>shorebird.tree, <b><span style='color:#54ff54;'>x=</span></b>shorebird.toplot, <b><span style='color:#54ff54;'>tip.labels=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#dede49;'>dotTree</span></b><b>(</b><b><span style='color:#54ff54;'>tree=</span></b>shorebird.tree, <b><span style='color:#54ff54;'>x=</span></b>shorebird.toplot, <b><span style='color:#54ff54;'>tip.labels=</span></b><b><span style='color:#00ff00;'>TRUE</span></b>, <b><span style='color:#54ff54;'>type=</span></b><b><span style='color:#ff54ff;'>&quot;cladogram&quot;</span></b><b>)</b>

<b><span style='color:#b00000;'>## Graphics</span></b>
<b><span style='color:#54ffff;'># Output figure will be saved to the disk as OutputFile.png</span></b>
<b><span style='color:#dede49;'>png</span></b><b>(</b><b><span style='color:#54ff54;'>filename=</span></b><b><span style='color:#ff54ff;'>&quot;OutputFile.png&quot;</span></b>, <b><span style='color:#54ff54;'>width=</span></b><b><span style='color:#ff54ff;'>720</span></b>, <b><span style='color:#54ff54;'>height=</span></b><b><span style='color:#ff54ff;'>720</span></b>, <b><span style='color:#54ff54;'>bg=</span></b><b><span style='color:#ff54ff;'>&quot;white&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># Here can go any number of functions making plots...</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b>...<b>)</b> <b><span style='color:#54ffff;'># Whatever...</span></b>
<b><span style='color:#54ffff;'># When using plotting commands, nothing is shown on the screen</span></b>
<b><span style='color:#54ffff;'># The final plot(s) will be saved by:</span></b>
<b><span style='color:#dede49;'>dev.off</span></b><b>()</b> <b><span style='color:#54ffff;'># Closes graphical device - needed after use of plotting functions png(), svg(), pdf(), ... followed by any function like plot() to write the file(s) to the disk</span></b>
<b><span style='color:#54ffff;'># filename=&quot;OutFiles_%03d.png&quot; # Returns list of files named OutFiles_001.png, OutFiles_002.png, ... Useful for functions returning more graphs.</span></b>
?png <b><span style='color:#54ffff;'># These functions have various possibilities to set size, whatever.</span></b>
?svg <b><span style='color:#54ffff;'># Exact possibilities of all 3 functions vary from system to system</span></b>
?pdf <b><span style='color:#54ffff;'># according to graphical libraries available in the computer.</span></b>

<b><span style='color:#b00000;'>## Install packages from GitHub</span></b>
<b><span style='color:#54ffff;'># Needed library</span></b>
<b><span style='color:#dede49;'>install.packages</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;devtools&quot;</span></b><b>)</b>
<b><span style='color:#dede49;'>library</span></b><b>(</b>devtools<b>)</b>
<b><span style='color:#dede49;'>dev_mode</span></b><b>(</b><b><span style='color:#54ff54;'>on=</span></b><b><span style='color:#00ff00;'>TRUE</span></b><b>)</b>
<b><span style='color:#54ffff;'># Install selected package from GitHub (user/project)</span></b>
<b><span style='color:#dede49;'>install_github</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;thibautjombart/adegenet&quot;</span></b><b>)</b>
<b><span style='color:#54ffff;'># when finished go back to normal version</span></b>
<b><span style='color:#dede49;'>dev_mode</span></b><b>(</b><b><span style='color:#54ff54;'>on=</span></b><b><span style='color:#00ff00;'>FALSE</span></b><b>)</b>

<b><span style='color:#b00000;'>## Functions</span></b>
MyFunction <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ffff54;'>function</span></b> <b>(</b>x, y<b>)</b> <b>{</b>
	<b><span style='color:#54ffff;'># Any commands can be here...</span></b>
	x <b>+</b><b><span style='color:#ff54ff;'> </span></b>y
	<b>}</b>

<b><span style='color:#54ffff;'># Use as usually:</span></b>
<b><span style='color:#dede49;'>MyFunction</span></b><b>(</b><b><span style='color:#ff54ff;'>5</span></b>, <b><span style='color:#ff54ff;'>8</span></b><b>)</b>
<b><span style='color:#dede49;'>MyFunction</span></b><b>(</b><b><span style='color:#ff54ff;'>1</span></b>, <b><span style='color:#ff54ff;'>4</span></b><b>)</b>
<b><span style='color:#dede49;'>MyFunction</span></b><b>(</b><b><span style='color:#54ff54;'>x=</span></b><b><span style='color:#ff54ff;'>4</span></b>, <b><span style='color:#54ff54;'>y=</span></b><b><span style='color:#ff54ff;'>7</span></b><b>)</b>
MF <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>MyFunction</span></b><b>(</b><b><span style='color:#ff54ff;'>9</span></b>, <b><span style='color:#ff54ff;'>15</span></b><b>)</b>
MF <b><span style='color:#54ffff;'># See it works</span></b>

<b><span style='color:#b00000;'>## Loops</span></b>

<b><span style='color:#54ffff;'># For loop</span></b>
<b><span style='color:#54ffff;'># Simplest loop - print value of &quot;i&quot; in each step</span></b>
<b><span style='color:#54ffff;'># &quot;i&quot; is commonly used for various indexing</span></b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>i <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>5</span></b><b>)</b> <b>{</b> <b><span style='color:#dede49;'>print</span></b><b>(</b>i<b>)</b> <b>}</b>

<b><span style='color:#54ffff;'># In every step modify value of variable &quot;X&quot; (add 1 to previous value)</span></b>
X <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>0</span></b> <b><span style='color:#54ffff;'># Set initial value</span></b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>i <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#ff54ff;'>10</span></b><b>)</b> <b>{</b>
	<b><span style='color:#54ffff;'># Any commands can be here...</span></b>
	<b><span style='color:#dede49;'>print</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Loop turn&quot;</span></b><b>)</b> <b><span style='color:#54ffff;'># Some message for user</span></b>
	<b><span style='color:#dede49;'>print</span></b><b>(</b>i<b>)</b> <b><span style='color:#54ffff;'># Print number of turn - note it is decreasing</span></b>
	X <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>X<b>+</b>i <b><span style='color:#54ffff;'># Rise value of &quot;X&quot; by current value of &quot;i&quot; (see previous line)</span></b>
	<b><span style='color:#dede49;'>print</span></b><b>(</b><b><span style='color:#dede49;'>paste</span></b><b>(</b><b><span style='color:#ff54ff;'>&quot;Variable value:&quot;</span></b>, X<b>))</b> <b><span style='color:#54ffff;'># Print current value of &quot;X&quot;</span></b>
	<b>}</b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>i <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>10</span></b><b>:</b><b><span style='color:#ff54ff;'>5</span></b><b>)</b> <b>{</b> <b><span style='color:#dede49;'>print</span></b><b>(</b>i<b>)</b> <b>}</b> <b><span style='color:#54ffff;'># Can be descending...</span></b>

<b><span style='color:#54ffff;'># Work on each item of a list object</span></b>
<b><span style='color:#54ffff;'># Print length of each sequence in nothofagus.sequences</span></b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>L <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#dede49;'>length</span></b><b>(</b>nothofagus.sequences<b>))</b> <b>{</b>
	<b><span style='color:#dede49;'>print</span></b><b>(</b><b><span style='color:#dede49;'>length</span></b><b>(</b>nothofagus.sequences<b>[[</b>L<b>]]))</b>
	<b>}</b>

<b><span style='color:#54ffff;'># While loop - it is done while the condition is valid</span></b>
<b><span style='color:#54ffff;'># While value of &quot;Q&quot; is &lt; 5 (starting from 0), print it and add 1</span></b>
Q <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>0</span></b>
<b><span style='color:#ffff54;'>while</span></b> <b>(</b>Q <b>&lt;</b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>5</span></b><b>)</b> <b>{</b> <b><span style='color:#dede49;'>print</span></b><b>(</b>Q <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>Q<b>+</b><b><span style='color:#ff54ff;'>1</span></b><b>)</b> <b>}</b>

<b><span style='color:#b00000;'>## If-else branching</span></b>
XX <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>seq</span></b><b>(</b><b><span style='color:#54ff54;'>from=</span></b><b>-</b><b><span style='color:#ff54ff;'>3</span></b>, <b><span style='color:#54ff54;'>to=</span></b><b><span style='color:#ff54ff;'>6.5</span></b>, <b><span style='color:#54ff54;'>by=</span></b><b><span style='color:#ff54ff;'>0.1</span></b><b>)</b>
XX
YY <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#dede49;'>c</span></b><b>()</b>
<b><span style='color:#ffff54;'>for</span></b> <b>(</b>II <b><span style='color:#ffff54;'>in</span></b> <b><span style='color:#ff54ff;'>1</span></b><b>:</b><b><span style='color:#dede49;'>length</span></b><b>(</b>XX<b>))</b> <b>{</b>
	<b><span style='color:#ffff54;'>if</span></b><b>(</b>XX<b>[</b>II<b>]</b> <b>&lt;=</b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b> <b>{</b> <b><span style='color:#54ffff;'># Executed for XX &lt;= 2</span></b>
		YY<b>[</b>II<b>]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>XX<b>[</b>II<b>]</b><b>^</b><b><span style='color:#ff54ff;'>2</span></b>
		<b>}</b> <b><span style='color:#ffff54;'>else</span></b> <b><span style='color:#ffff54;'>if</span></b><b>(</b>XX<b>[</b>II<b>]</b> <b>&gt;</b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b> <b>{</b> <b><span style='color:#54ffff;'># Executed for XX &gt; 2</span></b>
			YY<b>[</b>II<b>]</b> <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>6</span></b><b>-</b>XX<b>[</b>II<b>]</b>
			<b>}</b>
	<b>}</b>
YY
<b><span style='color:#dede49;'>plot</span></b><b>(</b>XX, YY<b>)</b> <b><span style='color:#54ffff;'># See the result</span></b>

<b><span style='color:#54ffff;'># Or (different possibility to get very same result)</span></b>
<b><span style='color:#54ffff;'># Note &quot;XX&quot; is reused from the previous variant</span></b>
CC <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ffff54;'>function</span></b><b>(</b>AA<b>)</b> <b>{</b>
	<b><span style='color:#ffff54;'>if</span></b><b>(</b>AA <b>&lt;=</b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>2</span></b><b>)</b> <b>{</b> <b><span style='color:#54ffff;'># Executed for XX &lt;= 2</span></b>
		BB <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b>AA<b>^</b><b><span style='color:#ff54ff;'>2</span></b>
	<b>}</b> <b><span style='color:#ffff54;'>else</span></b> <b>{</b> <b><span style='color:#54ffff;'># Executed for XX &gt; 2</span></b>
		BB <b><span style='color:#336366;'>&lt;-</span></b><b><span style='color:#ff54ff;'> </span></b><b><span style='color:#ff54ff;'>6</span></b><b>-</b>AA
		<b>}</b>
	<b><span style='color:#dede49;'>return</span></b><b>(</b>BB<b>)</b> <b><span style='color:#54ffff;'># The output value</span></b>
	<b>}</b>
CC <b><span style='color:#54ffff;'># Previously, &quot;YY&quot; contained values to plot made by the for loop, here &quot;CC&quot; contains function to by used by sapply() when plotting</span></b>
<b><span style='color:#dede49;'>plot</span></b><b>(</b><b><span style='color:#dede49;'>sapply</span></b><b>(</b>XX, CC<b>))</b> <b><span style='color:#54ffff;'># See the result</span></b>
</pre>
</body>
</html>
