<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8" />
<meta name="Generator" content="Kate, the KDE Advanced Text Editor" />
<title>course_commands.r</title>
</head>
<body>
<pre style='color:#000000;background-color:#ffffff;'>
<b><span style='color:#b00000;'>## Getting help</span></b>
<b>help</b>(rep) <i><span style='color:#755019;'># Help for particular function (package must be loaded)</span></i>
?rep <i><span style='color:#755019;'># Help for particular function (package must be loaded)</span></i>
??ANOVA <i><span style='color:#755019;'># Search for the term within all installed packages</span></i>
<b>help.search</b>(<span style='color:#bf0303;'>&quot;analysis of variance&quot;</span>) <i><span style='color:#755019;'># Search for the phrase within all installed packages - return list of hits sorted according to type and package (i.e. package::function)</span></i>

<b><span style='color:#b00000;'>## Set working directory</span></b>
<b>setwd</b>(<span style='color:#bf0303;'>&quot;/home/vojta/dokumenty/fakulta/vyuka/r_mol_data/priklady/&quot;</span>) <i><span style='color:#755019;'># Create empty directory and modify the path accordingly</span></i>
<b>getwd</b>() <i><span style='color:#755019;'># Verifies where we are</span></i>
<b>dir</b>() <i><span style='color:#755019;'># Lists files and folders on the disk</span></i>
<b>ls</b>() <i><span style='color:#755019;'># Lists currently available R objects</span></i>

<b><span style='color:#b00000;'>## Basic operations</span></b>
x <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>c</b>(<span style='color:#2b74c7;'>1</span>, <span style='color:#2b74c7;'>2</span>, <span style='color:#2b74c7;'>3</span>, <span style='color:#2b74c7;'>4</span>, <span style='color:#2b74c7;'>5</span>) <i><span style='color:#755019;'># Creates vector (see also ?rep)</span></i>
<b>c</b>() <i><span style='color:#755019;'># Is generic function to concatenate objects into new one</span></i>
<b>length</b>(x) <i><span style='color:#755019;'># Length of the object - for matrices and DF use dim()</span></i>
<b>str</b>(x) <i><span style='color:#755019;'># Information about structure of the object, try also ls.str()</span></i>
<b>mode</b>(x) <i><span style='color:#755019;'># Gets type of storage mode of the object</span></i>
<b>class</b>(x) <i><span style='color:#755019;'># Shows class of the object</span></i>
x[<span style='color:#2b74c7;'>2</span>] <i><span style='color:#755019;'># Shows second element of the object</span></i>
x <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span>x[<span style='color:#803f00;'>-</span><span style='color:#2b74c7;'>5</span>] <i><span style='color:#755019;'># Removes fifth element</span></i>
y <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>matrix</b>(<span style='color:#e85290;'>data=</span><span style='color:#2b74c7;'>5</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>20</span>, <span style='color:#e85290;'>nrow=</span><span style='color:#2b74c7;'>4</span>, <span style='color:#e85290;'>ncol=</span><span style='color:#2b74c7;'>4</span>) <i><span style='color:#755019;'># Creates a matrix</span></i>
<b>is.matrix</b>(y) <i><span style='color:#755019;'># Is it matrix? Try is.&lt;TAB&gt;&lt;TAB&gt; TAB key shows available functions and objects starting by typed text</span></i>
y <i><span style='color:#755019;'># Prints the matrix</span></i>
y[,<span style='color:#2b74c7;'>2</span>] <i><span style='color:#755019;'># Prints second column</span></i>
y[<span style='color:#2b74c7;'>3</span>,] <i><span style='color:#755019;'># Prints third row</span></i>
y[<span style='color:#2b74c7;'>4</span>,<span style='color:#2b74c7;'>3</span>] <i><span style='color:#755019;'># Prints element from fourth row and third column</span></i>
x <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span>y[<span style='color:#2b74c7;'>2</span>,] <i><span style='color:#755019;'># Replaces &quot;x&quot; by second row of &quot;y&quot; (no warning) - R doesn't ask neither notifies when overwriting objects! Be careful!</span></i>
<b>rm</b>(x) <i><span style='color:#755019;'># Deletes x</span></i>
y[,<span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>3</span>] <i><span style='color:#755019;'># Prints first through third column of the matrix</span></i>
y[<span style='color:#2b74c7;'>3</span>,] <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>rep</b>(<span style='color:#e85290;'>x=</span><span style='color:#2b74c7;'>20</span>, <span style='color:#e85290;'>each=</span><span style='color:#2b74c7;'>4</span>) <i><span style='color:#755019;'># Replaces third line by value of 20</span></i>
y[y<span style='color:#803f00;'>==</span><span style='color:#2b74c7;'>20</span>] <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><span style='color:#2b74c7;'>10</span> <i><span style='color:#755019;'># If value of y's element is 20, replace it by 10</span></i>
<b>summary</b>(y) <i><span style='color:#755019;'># Basic statistics - according to columns</span></i>
<b>colnames</b>(y) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>c</b>(<span style='color:#bf0303;'>&quot;A&quot;</span>, <span style='color:#bf0303;'>&quot;B&quot;</span>, <span style='color:#bf0303;'>&quot;C&quot;</span>, <span style='color:#bf0303;'>&quot;D&quot;</span>) <i><span style='color:#755019;'># Set column names - Objects and functions are without quotation marks; files, text with</span></i>
<b>colnames</b>(y) <i><span style='color:#755019;'># Prints column names, use rownames() in same way</span></i>
y[,<span style='color:#bf0303;'>&quot;C&quot;</span>] <i><span style='color:#755019;'># Prints column C (R is case sensitive!)</span></i>
<b>t</b>(y) <i><span style='color:#755019;'># Transposes the matrix</span></i>
y <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>as.data.frame</b>(y) <i><span style='color:#755019;'># Turns into DF (see other functions as.*)</span></i>
y[y<span style='color:#803f00;'>==</span><span style='color:#2b74c7;'>17</span>] <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> &quot;NA&quot;</span> <i><span style='color:#755019;'># Removes values of 17</span></i>
y<span style='color:#803f00;'>$</span>B <i><span style='color:#755019;'># Gets variable B of data frame y ($ works similarly in S3 objects)</span></i>
<b>save</b>(<span style='color:#e85290;'>list=</span><b>ls</b>(), <span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;test.RData&quot;</span>) <i><span style='color:#755019;'># Saves all objects during the work</span></i>
<b>load</b>(<span style='color:#bf0303;'>&quot;test.RData&quot;</span>) <i><span style='color:#755019;'># Loads saved R environment with all objects</span></i>
<i><span style='color:#755019;'># When loading saved project, you have to load again libraries and scripts (see further), data objects are restored</span></i>
<b>rm</b>(y)

<b><span style='color:#b00000;'>## Get information about object classes</span></b>
<b>getClassDef</b>(<span style='color:#bf0303;'>&quot;genind&quot;</span>) <i><span style='color:#755019;'># Or select any other class name</span></i>

<b><span style='color:#b00000;'>## Packages and repositories</span></b>
<i><span style='color:#755019;'># Set repositories</span></i>
<b>options</b>(<span style='color:#e85290;'>repos=</span><b>c</b>(<span style='color:#bf0303;'>&quot;http://cran.at.r-project.org&quot;</span>, <span style='color:#bf0303;'>&quot;http://r-forge.r-project.org/&quot;</span>, <span style='color:#bf0303;'>&quot;http://www.rforge.net/&quot;</span>, <span style='color:#bf0303;'>&quot;http://bioconductor.statistik.tu-dortmund.de/packages/3.1/bioc&quot;</span>, <span style='color:#bf0303;'>&quot;http://bioconductor.statistik.tu-dortmund.de/packages/3.1/data/annotation&quot;</span>, <span style='color:#bf0303;'>&quot;http://bioconductor.statistik.tu-dortmund.de/packages/3.1/data/experiment&quot;</span>, <span style='color:#bf0303;'>&quot;http://bioconductor.statistik.tu-dortmund.de/packages/3.1/extra&quot;</span>))
<b>getOption</b>(<span style='color:#bf0303;'>&quot;repos&quot;</span>) <i><span style='color:#755019;'># Shows actual repositories</span></i>
<i><span style='color:#755019;'># Install packages</span></i>
<i><span style='color:#755019;'># Installation of multiple packages may sometimes fail --- install then packages in smaller groups or one by one</span></i>
<b>install.packages</b>(<span style='color:#e85290;'>pkgs=</span><b>c</b>(<span style='color:#bf0303;'>&quot;Geneland&quot;</span>, <span style='color:#bf0303;'>&quot;PBSmapping&quot;</span>, <span style='color:#bf0303;'>&quot;ParallelStructure&quot;</span>, <span style='color:#bf0303;'>&quot;RandomFields&quot;</span>, <span style='color:#bf0303;'>&quot;RgoogleMaps&quot;</span>, <span style='color:#bf0303;'>&quot;TeachingDemos&quot;</span>, <span style='color:#bf0303;'>&quot;XML&quot;</span>, <span style='color:#bf0303;'>&quot;ade4&quot;</span>, <span style='color:#bf0303;'>&quot;adegenet&quot;</span>, <span style='color:#bf0303;'>&quot;adephylo&quot;</span>, <span style='color:#bf0303;'>&quot;akima&quot;</span>, <span style='color:#bf0303;'>&quot;ape&quot;</span>, <span style='color:#bf0303;'>&quot;colorspace&quot;</span>, <span style='color:#bf0303;'>&quot;combinat&quot;</span>, <span style='color:#bf0303;'>&quot;corrplot&quot;</span>, <span style='color:#bf0303;'>&quot;fields&quot;</span>, <span style='color:#bf0303;'>&quot;gplots&quot;</span>, <span style='color:#bf0303;'>&quot;grid&quot;</span>, <span style='color:#bf0303;'>&quot;ips&quot;</span>, <span style='color:#bf0303;'>&quot;lattice&quot;</span>, <span style='color:#bf0303;'>&quot;mapdata&quot;</span>, <span style='color:#bf0303;'>&quot;mapproj&quot;</span>, <span style='color:#bf0303;'>&quot;maps&quot;</span>, <span style='color:#bf0303;'>&quot;maptools&quot;</span>, <span style='color:#bf0303;'>&quot;muscle&quot;</span>, <span style='color:#bf0303;'>&quot;pegas&quot;</span>, <span style='color:#bf0303;'>&quot;permute&quot;</span>, <span style='color:#bf0303;'>&quot;phangorn&quot;</span>, <span style='color:#bf0303;'>&quot;phyloch&quot;</span>, <span style='color:#bf0303;'>&quot;phytools&quot;</span>, <span style='color:#bf0303;'>&quot;polysat&quot;</span>, <span style='color:#bf0303;'>&quot;poppr&quot;</span>, <span style='color:#bf0303;'>&quot;rworldmap&quot;</span>, <span style='color:#bf0303;'>&quot;seqinr&quot;</span>, <span style='color:#bf0303;'>&quot;sp&quot;</span>, <span style='color:#bf0303;'>&quot;spam&quot;</span>, <span style='color:#bf0303;'>&quot;tcltk&quot;</span>, <span style='color:#bf0303;'>&quot;vegan&quot;</span>), <span style='color:#e85290;'>repos=</span><b>getOption</b>(<span style='color:#bf0303;'>&quot;repos&quot;</span>), <span style='color:#e85290;'>dependencies=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># We will load packages by library(package) one by one when needed</span></i>
<i><span style='color:#755019;'># Update packages</span></i>
<b>update.packages</b>(<span style='color:#e85290;'>repos=</span><b>getOption</b>(<span style='color:#bf0303;'>&quot;repos&quot;</span>))

<b><span style='color:#b00000;'>## Libraries for population - genetic analysis</span></b>

<i><span style='color:#755019;'># Load needed libraries</span></i>
<b>library</b>(ape)
<b>library</b>(ade4)
<b>library</b>(adegenet)
<b>library</b>(pegas)
<b>library</b>(poppr)
<b>library</b>(vegan)

<b><span style='color:#b00000;'>## Data</span></b>
<i><span style='color:#755019;'># Load training data</span></i>
hauss.loci <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.loci</b>(<span style='color:#bf0303;'>&quot;http://trapa.cz/sites/default/rcourse/haussknechtii_ssrs.txt&quot;</span>, <span style='color:#e85290;'>header=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>loci.sep=</span><span style='color:#bf0303;'>&quot;</span><span style='color:#bf5c00;'>\t</span><span style='color:#bf0303;'>&quot;</span>, <span style='color:#e85290;'>allele.sep=</span><span style='color:#bf0303;'>&quot;/&quot;</span>, <span style='color:#e85290;'>col.pop=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>col.loci=</span><span style='color:#2b74c7;'>3</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>14</span>, <span style='color:#e85290;'>row.names=</span><span style='color:#2b74c7;'>1</span>)
<i><span style='color:#755019;'># Data control</span></i>
hauss.loci
<b>print</b>(hauss.loci, <span style='color:#e85290;'>details=</span><span style='color:#00892b;'>TRUE</span>)

<i><span style='color:#755019;'># Conversion of loci to genind - used for many analysis</span></i>
hauss.genind <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>loci2genind</b>(hauss.loci)
<i><span style='color:#755019;'># See population names</span></i>
<b>pop</b>(hauss.genind)
hauss.genind<span style='color:#803f00;'>$</span>pop.names

<i><span style='color:#755019;'># Removes missing data - see ?missingno for types of dealing with them</span></i>
hauss.genind <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>missingno</b>(<span style='color:#e85290;'>pop=</span>hauss.genind, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;mean&quot;</span>, <span style='color:#e85290;'>cutoff=</span><span style='color:#2b74c7;'>0.1</span>, <span style='color:#e85290;'>quiet=</span><span style='color:#00892b;'>FALSE</span>)

<i><span style='color:#755019;'># Convert corrected genind to loci</span></i>
hauss.loci.cor <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>genind2loci</b>(hauss.genind)

<i><span style='color:#755019;'># Writes loci file to the disk</span></i>
<b>write.loci</b>(hauss.loci.cor, <span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;hauss.loci.cor.txt&quot;</span>, <span style='color:#e85290;'>loci.sep=</span><span style='color:#bf0303;'>&quot;</span><span style='color:#bf5c00;'>\t</span><span style='color:#bf0303;'>&quot;</span>, <span style='color:#e85290;'>allele.sep=</span><span style='color:#bf0303;'>&quot;/&quot;</span>)

<i><span style='color:#755019;'># Read coordinates</span></i>
hauss.coord <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.csv</b>(<span style='color:#bf0303;'>&quot;http://trapa.cz/sites/default/rcourse/haussknechtii_coordinates.csv&quot;</span>, <span style='color:#e85290;'>header=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;</span><span style='color:#bf5c00;'>\t</span><span style='color:#bf0303;'>&quot;</span>, <span style='color:#e85290;'>quote=</span><span style='color:#bf0303;'>&quot;&quot;</span>, <span style='color:#e85290;'>dec=</span><span style='color:#bf0303;'>&quot;.&quot;</span>, <span style='color:#e85290;'>row.names=</span><span style='color:#2b74c7;'>1</span>)
hauss.coord

<i><span style='color:#755019;'># Add coordinates - note identification of slots within object</span></i>
hauss.genind<span style='color:#803f00;'>$</span>other<span style='color:#803f00;'>$</span>xy <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span>hauss.coord
<i><span style='color:#755019;'># See result</span></i>
hauss.genind<span style='color:#803f00;'>$</span>other<span style='color:#803f00;'>$</span>xy
hauss.genind

<i><span style='color:#755019;'># Conversion to genpop - for population-level analysis</span></i>
hauss.genpop <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>genind2genpop</b>(hauss.genind, <span style='color:#e85290;'>missing=</span><span style='color:#bf0303;'>&quot;NA&quot;</span>, <span style='color:#e85290;'>process.other=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># See result</span></i>
hauss.genpop

<i><span style='color:#755019;'># Read datasets from various software</span></i>
<b>read.genalex</b>() <i><span style='color:#755019;'># poppr - reads *.csv file</span></i>
<b>read.fstat</b>() <i><span style='color:#755019;'># adegenet - reads *.dat files, only haploid/diploid data</span></i>
<b>read.genetix</b>() <i><span style='color:#755019;'># adegenet - reads *.gtx files, only haploid/diploid data</span></i>
<b>read.genepop</b>() <i><span style='color:#755019;'># adegenet - reads *.gen files, only haploid/diploid data</span></i>
<b>read.structure</b>() <i><span style='color:#755019;'># adegenet - reads *.str files, only haploid/diploid data</span></i>

<i><span style='color:#755019;'># Import of triploid microsattelites</span></i>
tarax3n.table <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.table</b>(<span style='color:#bf0303;'>&quot;http://trapa.cz/sites/default/rcourse/tarax3n.txt&quot;</span>, <span style='color:#e85290;'>header=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;</span><span style='color:#bf5c00;'>\t</span><span style='color:#bf0303;'>&quot;</span>, <span style='color:#e85290;'>quote=</span><span style='color:#bf0303;'>&quot;&quot;</span>, <span style='color:#e85290;'>row.names=</span><span style='color:#2b74c7;'>1</span>)
<i><span style='color:#755019;'># Check the data</span></i>
tarax3n.table
<b>class</b>(tarax3n.table)
<b>dim</b>(tarax3n.table)

tarax3n.genind <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>df2genind</b>(<span style='color:#e85290;'>X=</span>tarax3n.table[,<span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>6</span>], <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;/&quot;</span>, <span style='color:#e85290;'>ncode=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>pop=</span>tarax3n.table[[<span style='color:#bf0303;'>&quot;pop&quot;</span>]], <span style='color:#e85290;'>missing=</span><span style='color:#00892b;'>NA</span>, <span style='color:#e85290;'>ploidy=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;codom&quot;</span>)
<i><span style='color:#755019;'># See resulting genind object</span></i>
tarax3n.genind
<b>summary</b>(tarax3n.genind)

<i><span style='color:#755019;'># Import of AFLP data</span></i>
amara.aflp <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.table</b>(<span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;http://trapa.cz/sites/default/rcourse/amara_aflp.txt&quot;</span>, <span style='color:#e85290;'>header=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;</span><span style='color:#bf5c00;'>\t</span><span style='color:#bf0303;'>&quot;</span>, <span style='color:#e85290;'>quote=</span><span style='color:#bf0303;'>&quot;&quot;</span>)
amara.aflp
<b>dim</b>(amara.aflp)
<b>class</b>(amara.aflp) <i><span style='color:#755019;'># Must be matrix or data frame</span></i>
<i><span style='color:#755019;'># Populations - just one column with population names for all inds</span></i>
amara.pop <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.table</b>(<span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;http://trapa.cz/sites/default/rcourse/amara_pop.txt&quot;</span>, <span style='color:#e85290;'>header=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;</span><span style='color:#bf5c00;'>\t</span><span style='color:#bf0303;'>&quot;</span>, <span style='color:#e85290;'>quote=</span><span style='color:#bf0303;'>&quot;&quot;</span>)
amara.pop
<b>dim</b>(amara.pop)
<i><span style='color:#755019;'># Create genind object - ind.names and loc.names are taken from X</span></i>
aflp.genind <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>df2genind</b>(<span style='color:#e85290;'>X=</span>amara.aflp, <span style='color:#e85290;'>sep=</span><span style='color:#00892b;'>NULL</span>, <span style='color:#e85290;'>ind.names=</span><span style='color:#00892b;'>NULL</span>, <span style='color:#e85290;'>loc.names=</span><span style='color:#00892b;'>NULL</span>, <span style='color:#e85290;'>pop=</span>amara.pop[,<span style='color:#2b74c7;'>1</span>], <span style='color:#e85290;'>missing=</span><span style='color:#00892b;'>NA</span>, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;PA&quot;</span>)
aflp.genind
<b>summary</b>(amara.aflp)

<i><span style='color:#755019;'># Some useful functions for data manipulations</span></i>
<b>genind2df</b>() <i><span style='color:#755019;'># adegenet - export into data frame</span></i>
<b>genind2genalex</b>() <i><span style='color:#755019;'># poppr - export for genalex</span></i>
<b>splitcombine</b>() <i><span style='color:#755019;'># poppr - edits population hierarchy</span></i>
<b>popsub</b>() <i><span style='color:#755019;'># poppr - extracts only selected population(s)</span></i>
<b>clonecorrect</b>() <i><span style='color:#755019;'># poppr - corrects for clones</span></i>
<b>informloci</b>() <i><span style='color:#755019;'># poppr - removes uninformative loci</span></i>
<b>seppop</b>() <i><span style='color:#755019;'># adegenet - separates populations from genind or genlight object</span></i>
<b>seploc</b>() <i><span style='color:#755019;'># adegenet - splits genind, genpop or genlight by markers</span></i>
<i><span style='color:#755019;'># seppop and seploc return lists of objects - for further analysis</span></i>
<i><span style='color:#755019;'># read manual pages (?...) of those functions before usage</span></i>

<i><span style='color:#755019;'># Reading FASTA (reads also another formats, see ?read.dna), sequences of flu viruses from various years</span></i>
usflu.dna <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.dna</b>(<span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;http://adegenet.r-forge.r-project.org/files/usflu.fasta&quot;</span>, <span style='color:#e85290;'>format=</span><span style='color:#bf0303;'>&quot;fasta&quot;</span>)
<i><span style='color:#755019;'># Another possibility (only for FASTA alignments, same result):</span></i>
usflu.dna <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>fasta2DNAbin</b>(<span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;http://adegenet.r-forge.r-project.org/files/usflu.fasta&quot;</span>) <i><span style='color:#755019;'># Normally keeps only SNP -- see ?fasta2DNAbin</span></i>
usflu.dna
<i><span style='color:#755019;'># Check the object</span></i>
<b>class</b>(usflu.dna)
<b>as.character</b>(usflu.dna)[<span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>5</span>,<span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>10</span>]
<b>dim</b>(usflu.dna)
<i><span style='color:#755019;'># Read anotations</span></i>
usflu.annot <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.csv</b>(<span style='color:#bf0303;'>&quot;http://adegenet.r-forge.r-project.org/files/usflu.annot.csv&quot;</span>, <span style='color:#e85290;'>header=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>row.names=</span><span style='color:#2b74c7;'>1</span>)
<b>head</b>(usflu.annot)
<i><span style='color:#755019;'># Convert DNAbin to genind - only polymorphic loci are retained</span></i>
usflu.genind <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>DNAbin2genind</b>(<span style='color:#e85290;'>x=</span>usflu.dna, <span style='color:#e85290;'>pop=</span>usflu.annot[[<span style='color:#bf0303;'>&quot;year&quot;</span>]])
<i><span style='color:#755019;'># read.fasta() from seqinr package reads DNA or AA in FASTA format - returns a list (DNAbin is for us now better choice)</span></i>
usflu.dna <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span>seqinr<span style='color:#803f00;'>::</span><b>read.fasta</b>(<span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;http://adegenet.r-forge.r-project.org/files/usflu.fasta&quot;</span>, <span style='color:#e85290;'>seqtype=</span><span style='color:#bf0303;'>&quot;DNA&quot;</span>)

<i><span style='color:#755019;'># Importing DNA sequences from GeneBank - according to sequence ID, data from http://www.ncbi.nlm.nih.gov/popset/608602125</span></i>
meles.dna <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.GenBank</b>(<b>c</b>(<span style='color:#bf0303;'>&quot;KJ161355.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161354.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161353.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161352.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161351.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161350.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161349.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161348.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161347.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161346.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161345.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161344.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161343.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161342.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161341.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161340.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161339.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161338.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161337.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161336.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161335.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161334.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161333.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161332.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161331.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161330.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161329.1&quot;</span>, <span style='color:#bf0303;'>&quot;KJ161328.1&quot;</span>))
meles.dna
<b>class</b>(meles.dna)
<i><span style='color:#755019;'># Converts DNAbin to genind - extracts SNP - for large datasets can be computationally very intensive</span></i>
meles.genind <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>DNAbin2genind</b>(meles.dna)

<i><span style='color:#755019;'># Query on-line sequence databases</span></i>
<b>library</b>(seqinr)
<b>choosebank</b>() <i><span style='color:#755019;'># Genetic banks available for seqinr</span></i>
<b>choosebank</b>(<span style='color:#bf0303;'>&quot;embl&quot;</span>) <i><span style='color:#755019;'># Choose some bank</span></i>
<i><span style='color:#755019;'># Query selected database - there are much possibilities</span></i>
?query <i><span style='color:#755019;'># See how to construct the query</span></i>
nothofagus <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>query</b>(<span style='color:#e85290;'>listname=</span><span style='color:#bf0303;'>&quot;nothofagus&quot;</span>, <span style='color:#e85290;'>query=</span><span style='color:#bf0303;'>&quot;SP=Nothofagus AND K=rbcl&quot;</span>, <span style='color:#e85290;'>verbose=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># See the sequences information</span></i>
nothofagus<span style='color:#803f00;'>$</span>req
<i><span style='color:#755019;'># Get the sequences as a list</span></i>
nothofagus.sequences <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>getSequence</b>(nothofagus<span style='color:#803f00;'>$</span>req)
<i><span style='color:#755019;'># See sequences</span></i>
nothofagus.sequences
<i><span style='color:#755019;'># Get annotations</span></i>
nothofagus.annot <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>getAnnot</b>(nothofagus<span style='color:#803f00;'>$</span>req)
<i><span style='color:#755019;'># Close the bank when work is over</span></i>
<b>closebank</b>()
<i><span style='color:#755019;'># Convert sequences from a list to DNAbin (functions as.DNAbin*)</span></i>
nothofagus.dna <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>as.DNAbin.list</b>(nothofagus.sequences)

<i><span style='color:#755019;'># Checking SNPs</span></i>
<i><span style='color:#755019;'># Position of polymorphism within alignment - snpposi.plot() requires input data in form of matrix</span></i>
<b>snpposi.plot</b>(<span style='color:#e85290;'>x=</span><b>as.matrix</b>(meles.dna), <span style='color:#e85290;'>codon=</span><span style='color:#00892b;'>FALSE</span>)
<i><span style='color:#755019;'># Position of polymorphism within alignment - differentiating codons</span></i>
<b>snpposi.plot</b>(<b>as.matrix</b>(meles.dna))
<i><span style='color:#755019;'># When converting to genind object, only polymorphic loci are kept - threshold for polymorphism can be arbitrary</span></i>
meles.genind <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>DNAbin2genind</b>(<span style='color:#e85290;'>x=</span>meles.dna, <span style='color:#e85290;'>polyThres=</span><span style='color:#2b74c7;'>0.01</span>)
<i><span style='color:#755019;'># Test of random distribution of SNP</span></i>
<b>snpposi.test</b>(<span style='color:#e85290;'>x=</span><b>as.matrix</b>(meles.dna))

<i><span style='color:#755019;'># Importing SNP</span></i>
<b>read.PLINK</b>(<span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;PLINKfile&quot;</span>, ...)
?readPLINK <i><span style='color:#755019;'># See it for available options</span></i>
<i><span style='color:#755019;'># Extract SNP from FASTA alignment</span></i>
usflu.genlight <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>fasta2genlight</b>(<span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;http://adegenet.r-forge.r-project.org/files/usflu.fasta&quot;</span>, <span style='color:#e85290;'>quiet=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>saveNbAlleles=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># If it crashes (on Windows), try add parameter &quot;parallel=FALSE&quot;</span></i>
<i><span style='color:#755019;'># Function has several options to speed up reading</span></i>
?fasta2genlight
<i><span style='color:#755019;'># Convert genlight to genind</span></i>
usflu.genind <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>as.genind</b>(usflu.genlight)
<i><span style='color:#755019;'># Resulting data matrix might be slightly different than when using DNAbin2genind - depends on used settings</span></i>

<i><span style='color:#755019;'># Exporting data</span></i>
<i><span style='color:#755019;'># Convert genind into DF using genind2genotype() or genind2df()</span></i>
hauss.df <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>genind2genotype</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>pop=</span>hauss.genind<span style='color:#803f00;'>@</span>pop, <span style='color:#e85290;'>res.type=</span><span style='color:#bf0303;'>&quot;matrix&quot;</span>)
hauss.df <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>genind2df</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>pop=</span><span style='color:#00892b;'>NULL</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;/&quot;</span>, <span style='color:#e85290;'>usepop=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>oneColPerAll=</span><span style='color:#00892b;'>FALSE</span>)
<b>write.table</b>(<span style='color:#e85290;'>x=</span>hauss.df, <span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;haussdata.txt&quot;</span>, <span style='color:#e85290;'>quote=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;</span><span style='color:#bf5c00;'>\t</span><span style='color:#bf0303;'>&quot;</span>, <span style='color:#e85290;'>na=</span><span style='color:#bf0303;'>&quot;NA&quot;</span>, <span style='color:#e85290;'>dec=</span><span style='color:#bf0303;'>&quot;.&quot;</span>, <span style='color:#e85290;'>row.names=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>col.names=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Export of DNA sequences into FASTA format</span></i>
<b>write.dna</b>(<span style='color:#e85290;'>x=</span>usflu.dna, <span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;usflu.fasta&quot;</span>, <span style='color:#e85290;'>format=</span><span style='color:#bf0303;'>&quot;fasta&quot;</span>, <span style='color:#e85290;'>append=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>nbcol=</span><span style='color:#2b74c7;'>6</span>)
seqinr<span style='color:#803f00;'>::</span><b>write.fasta</b>(<span style='color:#e85290;'>sequences=</span>meles.dna, <span style='color:#e85290;'>names=</span><b>names</b>(meles.dna), <span style='color:#e85290;'>file.out=</span><span style='color:#bf0303;'>&quot;meles.fasta&quot;</span>, <span style='color:#e85290;'>open=</span><span style='color:#bf0303;'>&quot;w&quot;</span>)
<i><span style='color:#755019;'># Export trees (objects of class phylo)</span></i>
<b>write.tree</b>(<span style='color:#e85290;'>phy=</span>hauss.nj.bruvo, <span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;haussknechtii.nwk&quot;</span>) <i><span style='color:#755019;'># Writes tree(s) in NEWICK format</span></i>
<b>write.nexus</b>(hauss.nj.bruvo, <span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;haussknechtii.nexus&quot;</span>) <i><span style='color:#755019;'># Writes tree(s) in NEXUS format</span></i>

<b><span style='color:#b00000;'>## Descriptive statistics</span></b>

<i><span style='color:#755019;'># Get summary - names and sizes of populations, heterozygosity, some info about loci</span></i>
hauss.summ <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>summary</b>(hauss.genind)
<i><span style='color:#755019;'># Plot expected vs. observed heterozygosity - it looks like big difference</span></i>
<b>plot</b>(<span style='color:#e85290;'>x=</span>hauss.summ<span style='color:#803f00;'>$</span>Hexp, <span style='color:#e85290;'>y=</span>hauss.summ<span style='color:#803f00;'>$</span>Hobs, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Observed vs expected heterozygosity&quot;</span>, <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;Expected heterozygosity&quot;</span>, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;Observed heterozygosity&quot;</span>)
<b>abline</b>(<span style='color:#2b74c7;'>0</span>, <span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;red&quot;</span>)
<i><span style='color:#755019;'># T-test of difference between observed and expected heterozygosity - strongly significant</span></i>
<b>t.test</b>(<span style='color:#e85290;'>x=</span>hauss.summ<span style='color:#803f00;'>$</span>Hexp, <span style='color:#e85290;'>y=</span>hauss.summ<span style='color:#803f00;'>$</span>Hobs, <span style='color:#e85290;'>paired=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>var.equal=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Bartlett's K-squared of difference between observed and expected heterozygosity - not significant</span></i>
<b>bartlett.test</b>(<b>list</b>(hauss.summ<span style='color:#803f00;'>$</span>Hexp,hauss.summ<span style='color:#803f00;'>$</span>Hobs))
<i><span style='color:#755019;'># Create pane with some information</span></i>
<b>par</b>(<span style='color:#e85290;'>mfrow=</span><b>c</b>(<span style='color:#2b74c7;'>2</span>,<span style='color:#2b74c7;'>2</span>)) <i><span style='color:#755019;'># Divide graphical devices into 4 smaller spaces</span></i>
<i><span style='color:#755019;'># Plot alleles number vs. population sizes</span></i>
<b>plot</b>(<span style='color:#e85290;'>x=</span>hauss.summ<span style='color:#803f00;'>$</span>pop.eff, <span style='color:#e85290;'>y=</span>hauss.summ<span style='color:#803f00;'>$</span>pop.nall, <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;Populations sample size&quot;</span>, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;Number of alleles&quot;</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Alleles numbers and sample sizes&quot;</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;red&quot;</span>, <span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>20</span>)
<i><span style='color:#755019;'># Add text description to the point</span></i>
<b>text</b>(<span style='color:#e85290;'>x=</span>hauss.summ<span style='color:#803f00;'>$</span>pop.eff, <span style='color:#e85290;'>y=</span>hauss.summ<span style='color:#803f00;'>$</span>pop.nall, <span style='color:#e85290;'>lab=</span><b>names</b>(hauss.summ<span style='color:#803f00;'>$</span>pop.eff), <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.5</span>)
<i><span style='color:#755019;'># Barplots of various data</span></i>
<b>barplot</b>(<span style='color:#e85290;'>height=</span>hauss.summ<span style='color:#803f00;'>$</span>loc.nall, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;Number of alleles&quot;</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Number of alleles per locus&quot;</span>, <span style='color:#e85290;'>names.arg=</span>hauss.genind<span style='color:#803f00;'>@</span>loc.names, <span style='color:#e85290;'>las=</span><span style='color:#2b74c7;'>3</span>)
<b>barplot</b>(<span style='color:#e85290;'>height=</span>hauss.summ<span style='color:#803f00;'>$</span>Hexp<span style='color:#803f00;'>-</span>hauss.summ<span style='color:#803f00;'>$</span>Hobs, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Heterozygosity: expected-observed&quot;</span>, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;Hexp - Hobs&quot;</span>, <span style='color:#e85290;'>names.arg=</span>hauss.genind<span style='color:#803f00;'>@</span>loc.names, <span style='color:#e85290;'>las=</span><span style='color:#2b74c7;'>3</span>)
<b>barplot</b>(<span style='color:#e85290;'>height=</span>hauss.summ[[<span style='color:#bf0303;'>&quot;pop.eff&quot;</span>]], <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Sample sizes per population&quot;</span>, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;Number of genotypes&quot;</span>, <span style='color:#e85290;'>las=</span><span style='color:#2b74c7;'>3</span>)
<b>dev.off</b>() <i><span style='color:#755019;'># Closes graphical device - otherwise following graphs would still be divided into 4 parts</span></i>
<i><span style='color:#755019;'># poppr returns various population statistics for populations - png() will produce set of figures for each population</span></i>
<b>png</b>(<span style='color:#e85290;'>filename=</span><span style='color:#bf0303;'>&quot;poppr_%03d.png&quot;</span>, <span style='color:#e85290;'>width=</span><span style='color:#2b74c7;'>720</span>, <span style='color:#e85290;'>height=</span><span style='color:#2b74c7;'>720</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;white&quot;</span>) <i><span style='color:#755019;'># Output figures will be saved to the disk</span></i>
<b>poppr</b>(<span style='color:#e85290;'>dat=</span>hauss.genind, <span style='color:#e85290;'>total=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>sample=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>method=</span><span style='color:#2b74c7;'>4</span>, <span style='color:#e85290;'>missing=</span><span style='color:#bf0303;'>&quot;geno&quot;</span>, <span style='color:#e85290;'>cutoff=</span><span style='color:#2b74c7;'>0.15</span>, <span style='color:#e85290;'>quiet=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>clonecorrect=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>hist=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>minsamp=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>legend=</span><span style='color:#00892b;'>TRUE</span>)
<b>dev.off</b>() <i><span style='color:#755019;'># Closes graphical device - needed after use of functions like png(), svg(), pdf(), ...</span></i>
<i><span style='color:#755019;'># Pop - Population name (Note that “Total” also means “Pooled”)</span></i>
<i><span style='color:#755019;'># N - Number of individuals observed</span></i>
<i><span style='color:#755019;'># MLG - Number of multilocus genotypes (MLG) observed</span></i>
<i><span style='color:#755019;'># eMLG - The number of expected MLG at the smallest sample size ≥ 10 based on rarefaction.</span></i>
<i><span style='color:#755019;'># SE - Standard error based on eMLG</span></i>
<i><span style='color:#755019;'># H - Shannon-Wiener Index of MLG diversity</span></i>
<i><span style='color:#755019;'># G - Stoddart and Taylor’s Index of MLG diversity</span></i>
<i><span style='color:#755019;'># Hexp - Nei’s 1978 genotypic diversity (corrected for sample size), or Expected Heterozygosity</span></i>
<i><span style='color:#755019;'># E.5 - Evenness, E5</span></i>
<i><span style='color:#755019;'># Ia - The index of association, IA</span></i>
<i><span style='color:#755019;'># rbarD - The standardized index of association</span></i>

<i><span style='color:#755019;'># Departure from HWE</span></i>
hauss.hwe <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>HWE.test.genind</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>permut=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>nsim=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>res.type=</span><span style='color:#bf0303;'>&quot;matrix&quot;</span>)
<i><span style='color:#755019;'># According to populations</span></i>
<b>barplot</b>(<b>apply</b>(<span style='color:#e85290;'>X=</span>hauss.hwe<span style='color:#803f00;'>&lt;</span><span style='color:#2b74c7;'>0.05</span>, <span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>FUN=</span>mean, <span style='color:#e85290;'>na.rm=</span><span style='color:#00892b;'>TRUE</span>), <span style='color:#e85290;'>las=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;Proportion of departure from HWE&quot;</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;By population&quot;</span>, <span style='color:#e85290;'>names.arg=</span>hauss.genind<span style='color:#803f00;'>@</span>pop.names)
<i><span style='color:#755019;'># According to loci</span></i>
<b>barplot</b>(<b>apply</b>(<span style='color:#e85290;'>X=</span>hauss.hwe<span style='color:#803f00;'>&lt;</span><span style='color:#2b74c7;'>0.05</span>, <span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>FUN=</span>mean, <span style='color:#e85290;'>na.rm=</span><span style='color:#00892b;'>TRUE</span>), <span style='color:#e85290;'>las=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;Proportion of departure from HWE&quot;</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;By marker&quot;</span>)
hauss.hwe.test <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>hw.test</b>(<span style='color:#e85290;'>x=</span>hauss.loci, <span style='color:#e85290;'>B=</span><span style='color:#2b74c7;'>1000</span>)
hauss.hwe.test
<i><span style='color:#755019;'># List of lists for all combinations of loci and populations - very looong list</span></i>
hauss.hwe.full <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>HWE.test.genind</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>permut=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>nsim=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>res.type=</span><span style='color:#bf0303;'>&quot;full&quot;</span>)
hauss.hwe.full

<i><span style='color:#755019;'># FST</span></i>
<i><span style='color:#755019;'># Fit, Fst and Fis for each locus</span></i>
<b>Fst</b>(<span style='color:#e85290;'>x=</span>hauss.loci.cor, <span style='color:#e85290;'>pop=</span><span style='color:#2b74c7;'>1</span>)
<i><span style='color:#755019;'># multilocus estimators of variance components and F-statistics, alternative to Fst</span></i>
<b>fstat</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>pop=</span><span style='color:#00892b;'>NULL</span>, <span style='color:#e85290;'>fstonly=</span><span style='color:#00892b;'>FALSE</span>)
<i><span style='color:#755019;'># Nei's pairwise Fst between all pairs of populations. Difference in res.type=&quot;dist&quot;/&quot;matrix&quot; is only in format of output</span></i>
<b>pairwise.fst</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>pop=</span><span style='color:#00892b;'>NULL</span>, <span style='color:#e85290;'>res.type=</span><span style='color:#bf0303;'>&quot;matrix&quot;</span>, <span style='color:#e85290;'>truenames=</span><span style='color:#00892b;'>TRUE</span>)

<i><span style='color:#755019;'># Estimates of population theta according to Kimmel et al. 1998</span></i>
<b>theta.msat</b>(hauss.loci.cor)

<b><span style='color:#b00000;'>## Multi locus genotypes</span></b>
<i><span style='color:#755019;'># Total number of MLGs (simple value)</span></i>
<b>mlg</b>(<span style='color:#e85290;'>pop=</span>hauss.genind, <span style='color:#e85290;'>quiet=</span><span style='color:#00892b;'>FALSE</span>)
<i><span style='color:#755019;'># MLGs shared among populations (a list)</span></i>
<b>mlg.crosspop</b>(<span style='color:#e85290;'>pop=</span>hauss.genind, <span style='color:#e85290;'>df=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>quiet=</span><span style='color:#00892b;'>FALSE</span>)
<i><span style='color:#755019;'># Detailed view on distribution of MLGs into populations (table and/or plot)</span></i>
<b>mlg.table</b>(<span style='color:#e85290;'>pop=</span>hauss.genind, <span style='color:#e85290;'>bar=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>total=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>quiet=</span><span style='color:#00892b;'>FALSE</span>)

<b><span style='color:#b00000;'>## Genetic distances</span></b>

<i><span style='color:#755019;'># Euclidean distance for individuals</span></i>
hauss.dist <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>dist</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;euclidean&quot;</span>, <span style='color:#e85290;'>diag=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>upper=</span><span style='color:#00892b;'>TRUE</span>)
hauss.dist

<i><span style='color:#755019;'># Nei's distance (not Euclidean) for populations (other methods are available, see ?dist.genpop)</span></i>
hauss.dist.pop <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>dist.genpop</b>(<span style='color:#e85290;'>x=</span>hauss.genpop, <span style='color:#e85290;'>method=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>diag=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>upper=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Test if it is Euclidean</span></i>
<b>is.euclid</b>(hauss.dist.pop, <span style='color:#e85290;'>plot=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>print=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>tol=</span><span style='color:#2b74c7;'>1e-10</span>)
<i><span style='color:#755019;'># Turns to be Euclidean</span></i>
hauss.dist.pop <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>cailliez</b>(<span style='color:#e85290;'>distmat=</span>hauss.dist.pop, <span style='color:#e85290;'>print=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>tol=</span><span style='color:#2b74c7;'>1e-07</span>, <span style='color:#e85290;'>cor.zero=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Test if it is Euclidean</span></i>
<b>is.euclid</b>(hauss.dist.pop, <span style='color:#e85290;'>plot=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>print=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>tol=</span><span style='color:#2b74c7;'>1e-10</span>)
<i><span style='color:#755019;'># Show it</span></i>
hauss.dist.pop

<i><span style='color:#755019;'># Bruvo's distances weighting SSRs repeats - take care about replen parameter - requires repetition length for every SSRs locus</span></i>
hauss.dist.bruvo <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>bruvo.dist</b>(<span style='color:#e85290;'>pop=</span>hauss.genind, <span style='color:#e85290;'>replen=</span><b>rep</b>(<span style='color:#2b74c7;'>2</span>, <span style='color:#2b74c7;'>12</span>), <span style='color:#e85290;'>loss=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Test if it is Euclidean</span></i>
<b>is.euclid</b>(hauss.dist.bruvo, <span style='color:#e85290;'>plot=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>print=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>tol=</span><span style='color:#2b74c7;'>1e-10</span>)
<i><span style='color:#755019;'># Turn to be Euclidean</span></i>
hauss.dist.bruvo <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>cailliez</b>(<span style='color:#e85290;'>distmat=</span>hauss.dist.bruvo, <span style='color:#e85290;'>print=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>tol=</span><span style='color:#2b74c7;'>1e-07</span>, <span style='color:#e85290;'>cor.zero=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Test if it is Euclidean</span></i>
<b>is.euclid</b>(hauss.dist.bruvo, <span style='color:#e85290;'>plot=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>print=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>tol=</span><span style='color:#2b74c7;'>1e-10</span>)
<i><span style='color:#755019;'># Show it</span></i>
hauss.dist.bruvo

<i><span style='color:#755019;'># Nei's distance (not Euclidean) for individuals (other methods are available, see ?nei.dist from poppr package)</span></i>
hauss.dist.nei <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span>poppr<span style='color:#803f00;'>::</span><b>nei.dist</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>warning=</span><span style='color:#00892b;'>TRUE</span>)
<b>is.euclid</b>(hauss.dist.nei, <span style='color:#e85290;'>plot=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>print=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>tol=</span><span style='color:#2b74c7;'>1e-10</span>) <i><span style='color:#755019;'># It has so much NAs, so it it impossible to check it and convert into Euclidean distance</span></i>
<i><span style='color:#755019;'># Show it</span></i>
hauss.dist.nei

<i><span style='color:#755019;'># Dissimilarity matrix returns a distance reflecting the number of allelic differences between two individuals</span></i>
hauss.dist.diss <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>diss.dist</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>percent=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>mat=</span><span style='color:#00892b;'>TRUE</span>)
hauss.dist.diss

<i><span style='color:#755019;'># Import custom distance matrix</span></i>
MyDistance <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.csv</b>(<span style='color:#bf0303;'>&quot;distances.txt&quot;</span>, <span style='color:#e85290;'>header=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;</span><span style='color:#bf5c00;'>\t</span><span style='color:#bf0303;'>&quot;</span>, <span style='color:#e85290;'>dec=</span><span style='color:#bf0303;'>&quot;.&quot;</span>, <span style='color:#e85290;'>row.names=</span><span style='color:#2b74c7;'>1</span>)
MyDistance <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>as.dist</b>(MyDistance)
<b>class</b>(MyDistance)
<b>is.euclid</b>(MyDistance, <span style='color:#e85290;'>plot=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>print=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>tol =</span> <span style='color:#2b74c7;'>1e-10</span>)
<b>dim</b>(MyDistance)
MyDistance
MyDistance <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>cailliez</b>(MyDistance, <span style='color:#e85290;'>print=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>tol =</span> <span style='color:#2b74c7;'>1e-10</span>, <span style='color:#e85290;'>cor.zero=</span><span style='color:#00892b;'>TRUE</span>)
<b>is.euclid</b>(MyDistance, <span style='color:#e85290;'>plot=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>print=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>tol =</span> <span style='color:#2b74c7;'>1e-10</span>)
MyDistance

<i><span style='color:#755019;'># DNA distances - there are various models available</span></i>
?dist.dna
usflu.dist <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>dist.dna</b>(<span style='color:#e85290;'>x=</span>usflu.dna, <span style='color:#e85290;'>model=</span><span style='color:#bf0303;'>&quot;TN93&quot;</span>)
usflu.dist
<b>class</b>(usflu.dist)
<b>dim</b>(<b>as.matrix</b>(usflu.dist))
meles.dist <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>dist.dna</b>(<span style='color:#e85290;'>x=</span>meles.dna, <span style='color:#e85290;'>model=</span><span style='color:#bf0303;'>&quot;F81&quot;</span>)
meles.dist
<b>class</b>(meles.dist)
<b>dim</b>(<b>as.matrix</b>(meles.dist))

<i><span style='color:#755019;'># Pairwise genetic distances for each data block (genlight objects with whole genome data) - sensitive to missing data</span></i>
usflu.dists <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>lapply</b>(<span style='color:#e85290;'>X=</span>usflu.genlight, <span style='color:#e85290;'>FUN=</span><span style='color:#0000bf;'>function</span>(DDD) <b>dist</b>(<b>as.matrix</b>(DDD)))
<b>class</b>(usflu.dists)
<b>names</b>(usflu.dists)
<b>class</b>(usflu.dists[[<span style='color:#2b74c7;'>1</span>]])
usflu.distr <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>Reduce</b>(<span style='color:#e85290;'>f=</span><span style='color:#bf0303;'>&quot;+&quot;</span>, <span style='color:#e85290;'>x=</span>usflu.dists)
<b>class</b>(usflu.distr)
usflu.distr

<i><span style='color:#755019;'># Visualize pairwise genetic similarities</span></i>
<i><span style='color:#755019;'># table.paint() requires data frame, dist can't be directly converted to DF</span></i>
<b>table.paint</b>(<span style='color:#e85290;'>df=</span><b>as.data.frame</b>(<b>as.matrix</b>(usflu.dist)), <span style='color:#e85290;'>cleg=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>clabel.row=</span><span style='color:#2b74c7;'>0.5</span>, <span style='color:#e85290;'>clabel.col=</span><span style='color:#2b74c7;'>0.5</span>)
<i><span style='color:#755019;'># Same visualization, coloured</span></i>
<i><span style='color:#755019;'># heatmap() reorders values</span></i>
<b>heatmap</b>(<span style='color:#e85290;'>x=</span><b>as.matrix</b>(usflu.dist), <span style='color:#e85290;'>Rowv=</span><span style='color:#00892b;'>NA</span>, <span style='color:#e85290;'>Colv=</span><span style='color:#00892b;'>NA</span>, <span style='color:#e85290;'>symm=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Another possibility is to use function corrplot() from package corrplot</span></i>

<b><span style='color:#b00000;'>## Heatmaps</span></b>
<i><span style='color:#755019;'># Based on various distances</span></i>
<b>heatmap</b>(<b>as.matrix</b>(hauss.dist), <span style='color:#e85290;'>symm=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>labRow=</span><b>rownames</b>(<b>as.matrix</b>(hauss.dist.bruvo)), <span style='color:#e85290;'>labCol=</span><b>colnames</b>(<b>as.matrix</b>(hauss.dist.bruvo))) <i><span style='color:#755019;'># hauss.dist doesn't contain names of individuals - we have to add them here</span></i>
<b>heatmap</b>(<b>as.matrix</b>(hauss.dist.pop), <span style='color:#e85290;'>symm=</span><span style='color:#00892b;'>TRUE</span>)
<b>heatmap</b>(<b>as.matrix</b>(hauss.dist.bruvo), <span style='color:#e85290;'>symm=</span><span style='color:#00892b;'>TRUE</span>)
<b>heatmap</b>(<b>as.matrix</b>(hauss.dist.diss), <span style='color:#e85290;'>symm=</span><span style='color:#00892b;'>TRUE</span>)

<b><span style='color:#b00000;'>## Hierarchical clustering</span></b>
<i><span style='color:#755019;'># According to distance used - see ?hclust for available methods</span></i>
<b>plot</b>(<b>hclust</b>(<span style='color:#e85290;'>d=</span>hauss.dist, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;complete&quot;</span>))
<b>plot</b>(<b>hclust</b>(<span style='color:#e85290;'>d=</span>hauss.dist.pop, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;complete&quot;</span>))
<b>plot</b>(<b>hclust</b>(<span style='color:#e85290;'>d=</span>hauss.dist.bruvo, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;complete&quot;</span>))

<b><span style='color:#b00000;'>## UPGMA</span></b>
<i><span style='color:#755019;'># Calculate it</span></i>
<i><span style='color:#755019;'># Saving as phylo object (and not hclust) gives more possibilities for further plotting and manipulations</span></i>
usflu.upgma <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>as.phylo</b>(<b>hclust</b>(<span style='color:#e85290;'>d=</span>usflu.dist, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;average&quot;</span>))
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>usflu.upgma, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>0.75</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;UPGMA tree&quot;</span>)
<i><span style='color:#755019;'># Test quality - tests correlation of original distance in the matrix and reconstructed distance from hclust object</span></i>
<b>plot</b>(<span style='color:#e85290;'>x=</span><b>as.vector</b>(usflu.dist), <span style='color:#e85290;'>y=</span><b>as.vector</b>(<b>as.dist</b>(<b>cophenetic</b>(usflu.upgma))), <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;Original pairwise distances&quot;</span>, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;Pairwise distances on the tree&quot;</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Is UPGMA appropriate?&quot;</span>, <span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>20</span>, <span style='color:#e85290;'>col=</span><b>transp</b>(<span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;black&quot;</span>, <span style='color:#e85290;'>alpha=</span><span style='color:#2b74c7;'>0.1</span>), <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>2</span>)
<b>abline</b>(<b>lm</b>(<b>as.vector</b>(<b>as.dist</b>(<b>cophenetic</b>(usflu.upgma)))<span style='color:#803f00;'>~</span><b>as.vector</b>(usflu.dist)), <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;red&quot;</span>)

<b><span style='color:#b00000;'>## AMOVA</span></b>
<i><span style='color:#755019;'># From package pegas (doesn't directly show percentage of variance)</span></i>
hauss.pop <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span>hauss.genind<span style='color:#803f00;'>$</span>pop
hauss.amova <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>amova</b>(hauss.dist<span style='color:#803f00;'>~</span>hauss.pop, <span style='color:#e85290;'>data=</span><span style='color:#00892b;'>NULL</span>, <span style='color:#e85290;'>nperm=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>is.squared=</span><span style='color:#00892b;'>TRUE</span>)
hauss.amova
<i><span style='color:#755019;'># Another possibility is poppr.amova - for more complicated hierarchy - see ?poppr.amova</span></i>

<b><span style='color:#b00000;'>## MSN based on Bruvo's distance</span></b>
<b>bruvo.msn</b>(<span style='color:#e85290;'>pop=</span>hauss.genind, <span style='color:#e85290;'>replen=</span><b>rep</b>(<span style='color:#2b74c7;'>2</span>, <span style='color:#2b74c7;'>12</span>), <span style='color:#e85290;'>loss=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>palette=</span>rainbow, <span style='color:#e85290;'>vertex.label=</span><span style='color:#bf0303;'>&quot;inds&quot;</span>, <span style='color:#e85290;'>gscale=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>wscale=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>showplot=</span><span style='color:#00892b;'>TRUE</span>)

<b><span style='color:#b00000;'>## NJ</span></b>

<i><span style='color:#755019;'># Calculates the tree (try with various distances)</span></i>
hauss.nj <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>nj</b>(hauss.dist)

<i><span style='color:#755019;'># Test tree quality - plot original vs. reconstructed distance</span></i>
<b>plot</b>(<b>as.vector</b>(hauss.dist), <b>as.vector</b>(<b>as.dist</b>(<b>cophenetic</b>(hauss.nj))), <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;Original distance&quot;</span>, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;Reconstructed distance&quot;</span>)
<b>abline</b>(<b>lm</b>(<b>as.vector</b>(hauss.dist) <span style='color:#803f00;'>~</span><span style='color:#bf0303;'> </span><b>as.vector</b>(<b>as.dist</b>(<b>cophenetic</b>(hauss.nj)))), <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;red&quot;</span>)
<b>summary</b>(<b>lm</b>(<b>as.vector</b>(hauss.dist) <span style='color:#803f00;'>~</span><span style='color:#bf0303;'> </span><b>as.vector</b>(<b>as.dist</b>(<b>cophenetic</b>(hauss.nj))))) <i><span style='color:#755019;'># Prints summary text</span></i>

<i><span style='color:#755019;'># Bootstrap</span></i>
hauss.boot <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>boot.phylo</b>(<span style='color:#e85290;'>phy=</span>hauss.nj, <span style='color:#e85290;'>x=</span>hauss.loci.cor, <span style='color:#e85290;'>FUN=</span><span style='color:#0000bf;'>function</span>(XXX) <b>nj</b>(<b>dist</b>(<b>loci2genind</b>(XXX))), <span style='color:#e85290;'>B=</span><span style='color:#2b74c7;'>1000</span>)

<i><span style='color:#755019;'># Plot a basic tree - see ?plot.phylo for details</span></i>
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>hauss.nj, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;phylogram&quot;</span>)
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>hauss.nj, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;cladogram&quot;</span>, <span style='color:#e85290;'>edge.width=</span><span style='color:#2b74c7;'>2</span>)
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>hauss.nj, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;fan&quot;</span>, <span style='color:#e85290;'>edge.width=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>edge.lty=</span><span style='color:#2b74c7;'>2</span>)
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>hauss.nj, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;radial&quot;</span>, <span style='color:#e85290;'>edge.color=</span><span style='color:#bf0303;'>&quot;red&quot;</span>, <span style='color:#e85290;'>edge.width=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>edge.lty=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>2</span>)

<i><span style='color:#755019;'># Plot a nice tree with colored tips</span></i>
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>hauss.nj, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;unrooted&quot;</span>, <span style='color:#e85290;'>show.tip=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>lwd=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Neighbour-Joining tree&quot;</span>)
<i><span style='color:#755019;'># Labels for nodes - bootstrap - see ?nodelabels for graphical settings</span></i>
<b>nodelabels</b>(<span style='color:#e85290;'>text=</span><b>round</b>(hauss.boot<span style='color:#803f00;'>/</span><span style='color:#2b74c7;'>10</span>))
<i><span style='color:#755019;'># Coloured labels - creates vector of colors according to population information in genind object</span></i>
nj.rainbow <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>colorRampPalette</b>(<b>rainbow</b>(<b>length</b>(<b>levels</b>(<b>pop</b>(hauss.genind)))))
<i><span style='color:#755019;'># Colored tips</span></i>
<b>tiplabels</b>(<span style='color:#e85290;'>text=</span>hauss.genind<span style='color:#803f00;'>$</span>ind.names, <span style='color:#e85290;'>bg=</span><b>fac2col</b>(<span style='color:#e85290;'>x=</span>hauss.genind<span style='color:#803f00;'>$</span>pop, <span style='color:#e85290;'>col.pal=</span>nj.rainbow))

<i><span style='color:#755019;'># Plot BW tree with tip symbols and legend</span></i>
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>hauss.nj, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;cladogram&quot;</span>, <span style='color:#e85290;'>show.tip=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>lwd=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Neighbour-Joining tree&quot;</span>)
<i><span style='color:#755019;'># Add axis with distances</span></i>
<b>axisPhylo</b>()
<i><span style='color:#755019;'># From node labels let's remove unneeded frame</span></i>
<b>nodelabels</b>(<span style='color:#e85290;'>text=</span><b>round</b>(hauss.boot<span style='color:#803f00;'>/</span><span style='color:#2b74c7;'>10</span>), <span style='color:#e85290;'>frame=</span><span style='color:#bf0303;'>&quot;none&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;white&quot;</span>)
<i><span style='color:#755019;'># As tip label we use only symbols - see ?points for graphical details</span></i>
<b>tiplabels</b>(<span style='color:#e85290;'>frame=</span><span style='color:#bf0303;'>&quot;none&quot;</span>, <span style='color:#e85290;'>pch=</span><b>rep</b>(<span style='color:#e85290;'>x=</span><span style='color:#2b74c7;'>0</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>4</span>, <span style='color:#e85290;'>times=</span><b>c</b>(<span style='color:#2b74c7;'>13</span>, <span style='color:#2b74c7;'>17</span>,  <span style='color:#2b74c7;'>2</span>,  <span style='color:#2b74c7;'>6</span>,  <span style='color:#2b74c7;'>9</span>)), <span style='color:#e85290;'>lwd=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>2</span>)
<i><span style='color:#755019;'># Plot a legend explainindg symbols</span></i>
<b>legend</b>(<span style='color:#e85290;'>x=</span><span style='color:#bf0303;'>&quot;topleft&quot;</span>, <span style='color:#e85290;'>legend=</span><b>c</b>(<span style='color:#bf0303;'>&quot;He&quot;</span>, <span style='color:#bf0303;'>&quot;Oh&quot;</span>, <span style='color:#bf0303;'>&quot;Pr&quot;</span>, <span style='color:#bf0303;'>&quot;Ne&quot;</span>, <span style='color:#bf0303;'>&quot;Sk&quot;</span>), <span style='color:#e85290;'>border=</span><span style='color:#bf0303;'>&quot;black&quot;</span>, <span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>0</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>4</span>, <span style='color:#e85290;'>pt.lwd=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>pt.cex=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>bty=</span><span style='color:#bf0303;'>&quot;o&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;lightgrey&quot;</span>, <span style='color:#e85290;'>box.lwd=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.2</span>, <span style='color:#e85290;'>title=</span><span style='color:#bf0303;'>&quot;Populations&quot;</span>)

<i><span style='color:#755019;'># Bruvo's distance - NJ</span></i>
hauss.nj.bruvo <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>bruvo.boot</b>(<span style='color:#e85290;'>pop=</span>hauss.genind, <span style='color:#e85290;'>replen=</span><b>rep</b>(<span style='color:#2b74c7;'>2</span>, <span style='color:#2b74c7;'>12</span>), <span style='color:#e85290;'>sample=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>tree=</span><span style='color:#bf0303;'>&quot;nj&quot;</span>, <span style='color:#e85290;'>showtree=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>cutoff=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>quiet=</span><span style='color:#00892b;'>FALSE</span>)
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>hauss.nj.bruvo, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;unrooted&quot;</span>, <span style='color:#e85290;'>show.tip=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>lwd=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Neighbor-Joining tree.&quot;</span>)
<i><span style='color:#755019;'># bruvo.boot() writes all needed information into resulting object, so there is no need for external bootstrap function</span></i>
<b>nodelabels</b>(hauss.nj.bruvo[[<span style='color:#bf0303;'>&quot;node.labels&quot;</span>]]) <i><span style='color:#755019;'># Note you can call node labels from phylo object as phylo$node.labels or phylo[[&quot;node.labels&quot;]]</span></i>
<b>tiplabels</b>(hauss.nj.bruvo[[<span style='color:#bf0303;'>&quot;tip.label&quot;</span>]], <span style='color:#e85290;'>bg=</span><b>fac2col</b>(<span style='color:#e85290;'>x=</span>hauss.genind<span style='color:#803f00;'>$</span>pop, <span style='color:#e85290;'>col.pal=</span>nj.rainbow))

<i><span style='color:#755019;'># Bruvo's distance - UPGMA</span></i>
hauss.upgma <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>bruvo.boot</b>(<span style='color:#e85290;'>pop=</span>hauss.genind, <span style='color:#e85290;'>replen=</span><b>rep</b>(<span style='color:#2b74c7;'>2</span>, <span style='color:#2b74c7;'>12</span>), <span style='color:#e85290;'>sample=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>tree=</span><span style='color:#bf0303;'>&quot;upgma&quot;</span>, <span style='color:#e85290;'>showtree=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>cutoff=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>quiet=</span><span style='color:#00892b;'>FALSE</span>)
<b>plot.phylo</b>(hauss.upgma, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;unrooted&quot;</span>, <span style='color:#e85290;'>show.tip=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>lwd=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;UPGMA tree&quot;</span>)
<b>nodelabels</b>(hauss.upgma[[<span style='color:#bf0303;'>&quot;node.labels&quot;</span>]])
<b>tiplabels</b>(hauss.upgma[[<span style='color:#bf0303;'>&quot;tip.label&quot;</span>]], <span style='color:#e85290;'>bg=</span><b>fac2col</b>(<span style='color:#e85290;'>x=</span>hauss.genind<span style='color:#803f00;'>@</span>pop, <span style='color:#e85290;'>col.pal=</span>nj.rainbow))

<i><span style='color:#755019;'># Populations</span></i>
hauss.nj.pop <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>nj</b>(hauss.dist.pop)
<i><span style='color:#755019;'># Bootstrap - source() loads external scripts (boot.phylo doesn't work for population trees)</span></i>
<b>source</b>(<span style='color:#bf0303;'>&quot;http://trapa.cz/sites/default/rcourse/boot_phylo_nj_pop.r&quot;</span>)
hauss.boot.pop <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>boot.phylo.nj.pop</b>(hauss.nj.pop, hauss.genind, <span style='color:#2b74c7;'>1000</span>)
<i><span style='color:#755019;'># Plot a tree</span></i>
<b>plot</b>(hauss.nj.pop, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;radial&quot;</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.2</span>, <span style='color:#e85290;'>lwd=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Neighbor-Joining tree of populations&quot;</span>)
<i><span style='color:#755019;'># Labels - bootstrap</span></i>
<b>nodelabels</b>(<b>round</b>(hauss.boot.pop<span style='color:#803f00;'>/</span><span style='color:#2b74c7;'>10</span>), <span style='color:#e85290;'>frame=</span><span style='color:#bf0303;'>&quot;none&quot;</span>)
<b>print.phylo</b>(hauss.nj.pop)

<i><span style='color:#755019;'># Make a tree based on DNA sequences</span></i>
usflu.tree <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>nj</b>(<span style='color:#e85290;'>X=</span>usflu.dist)
<i><span style='color:#755019;'># Plot it</span></i>
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>usflu.tree, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;unrooted&quot;</span>, <span style='color:#e85290;'>show.tip=</span><span style='color:#00892b;'>FALSE</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;Unrooted NJ tree&quot;</span>)
<i><span style='color:#755019;'># Coloured tips</span></i>
usflu.pal <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>colorRampPalette</b>(<b>topo.colors</b>(<b>length</b>(<b>levels</b>(<b>as.factor</b>(usflu.annot<span style='color:#803f00;'>$</span>year)))))
<i><span style='color:#755019;'># Tip labels</span></i>
<b>tiplabels</b>(<span style='color:#e85290;'>text=</span>usflu.annot<span style='color:#803f00;'>$</span>year, <span style='color:#e85290;'>bg=</span><b>num2col</b>(usflu.annot<span style='color:#803f00;'>$</span>year, <span style='color:#e85290;'>col.pal=</span>usflu.pal), <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>0.75</span>)
<i><span style='color:#755019;'># Legend - describing years - pretty() automatically shows best values from given range, num2col() selects colours from colour scale</span></i>
<b>legend</b>(<span style='color:#e85290;'>x=</span><span style='color:#bf0303;'>&quot;bottomright&quot;</span>, <span style='color:#e85290;'>fill=</span><b>num2col</b>(<span style='color:#e85290;'>x=</span><b>pretty</b>(<span style='color:#e85290;'>x=</span><span style='color:#2b74c7;'>1993</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>2008</span>, <span style='color:#e85290;'>n=</span><span style='color:#2b74c7;'>8</span>), <span style='color:#e85290;'>col.pal=</span>usflu.pal), <span style='color:#e85290;'>leg=</span><b>pretty</b>(<span style='color:#e85290;'>x=</span><span style='color:#2b74c7;'>1993</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>2008</span>, <span style='color:#e85290;'>n=</span><span style='color:#2b74c7;'>8</span>), <span style='color:#e85290;'>ncol=</span><span style='color:#2b74c7;'>1</span>)

<i><span style='color:#755019;'># Root the tree - &quot;outgroup&quot; is name of accession (in quotation marks) or number (position within phy object)</span></i>
usflu.tree.rooted <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>root</b>(<span style='color:#e85290;'>phy=</span>usflu.tree, <span style='color:#e85290;'>outgroup=</span><span style='color:#2b74c7;'>1</span>)
<i><span style='color:#755019;'># Plot it</span></i>
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>usflu.tree.rooted, <span style='color:#e85290;'>show.tip=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>edge.width=</span><span style='color:#2b74c7;'>2</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;Rooted NJ tree&quot;</span>)
<i><span style='color:#755019;'># Labeling of tips</span></i>
<b>tiplabels</b>(<span style='color:#e85290;'>text=</span>annot<span style='color:#803f00;'>$</span>year, <span style='color:#e85290;'>bg=</span><b>transp</b>(<b>num2col</b>(<span style='color:#e85290;'>x=</span>annot<span style='color:#803f00;'>$</span>year, <span style='color:#e85290;'>col.pal=</span>usflu.pal), <span style='color:#e85290;'>alpha=</span><span style='color:#2b74c7;'>0.7</span>), <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>0.75</span>, <span style='color:#e85290;'>fg=</span><span style='color:#bf0303;'>&quot;transparent&quot;</span>)
<i><span style='color:#755019;'># Add axis with phylogenetic distance</span></i>
<b>axisPhylo</b>()
<i><span style='color:#755019;'># Legend - describing years - pretty() automatically shows best values from given range, num2col() selects colours from colour scale</span></i>
<b>legend</b>(<span style='color:#e85290;'>x=</span><span style='color:#bf0303;'>&quot;topright&quot;</span>, <span style='color:#e85290;'>fill=</span><b>num2col</b>(<span style='color:#e85290;'>x=</span><b>pretty</b>(<span style='color:#e85290;'>x=</span><span style='color:#2b74c7;'>1993</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>2008</span>, <span style='color:#e85290;'>n=</span><span style='color:#2b74c7;'>8</span>), <span style='color:#e85290;'>col.pal=</span>usflu.pal), <span style='color:#e85290;'>leg=</span><b>pretty</b>(<span style='color:#e85290;'>x=</span><span style='color:#2b74c7;'>1993</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>2008</span>, <span style='color:#e85290;'>n=</span><span style='color:#2b74c7;'>8</span>), <span style='color:#e85290;'>ncol=</span><span style='color:#2b74c7;'>1</span>)

<i><span style='color:#755019;'># Bootstrap</span></i>
<i><span style='color:#755019;'># Calculate it</span></i>
usflu.boot <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>boot.phylo</b>(<span style='color:#e85290;'>phy=</span>usflu.tree.rooted, <span style='color:#e85290;'>x=</span>usflu.dna, <span style='color:#e85290;'>FUN=</span><span style='color:#0000bf;'>function</span>(EEE) <b>root</b>(<b>nj</b>(<b>dist.dna</b>(EEE, <span style='color:#e85290;'>model=</span><span style='color:#bf0303;'>&quot;TN93&quot;</span>)), <span style='color:#e85290;'>outgroup=</span><span style='color:#2b74c7;'>1</span>), <span style='color:#e85290;'>B=</span><span style='color:#2b74c7;'>1000</span>)
<i><span style='color:#755019;'># Plot the tree</span></i>
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>usflu.tree.rooted, <span style='color:#e85290;'>show.tip=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>edge.width=</span><span style='color:#2b74c7;'>2</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;NJ tree + bootstrap values&quot;</span>)
<b>tiplabels</b>(<span style='color:#e85290;'>frame=</span><span style='color:#bf0303;'>&quot;none&quot;</span>, <span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>20</span>, <span style='color:#e85290;'>col=</span><b>transp</b>(<b>num2col</b>(<span style='color:#e85290;'>x=</span>annot<span style='color:#803f00;'>$</span>year, <span style='color:#e85290;'>col.pal=</span>usflu.pal), <span style='color:#e85290;'>alpha=</span><span style='color:#2b74c7;'>0.7</span>), <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>3.5</span>, <span style='color:#e85290;'>fg=</span><span style='color:#bf0303;'>&quot;transparent&quot;</span>)
<b>axisPhylo</b>()
<i><span style='color:#755019;'># Legend - describing years - pretty() automatically shows best values from given range, num2col() selects colours from colour scale</span></i>
<b>legend</b>(<span style='color:#e85290;'>x=</span><span style='color:#bf0303;'>&quot;topright&quot;</span>, <span style='color:#e85290;'>fill=</span><b>num2col</b>(<span style='color:#e85290;'>x=</span><b>pretty</b>(<span style='color:#e85290;'>x=</span><span style='color:#2b74c7;'>1993</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>2008</span>, <span style='color:#e85290;'>n=</span><span style='color:#2b74c7;'>8</span>), <span style='color:#e85290;'>col.pal=</span>usflu.pal), <span style='color:#e85290;'>leg=</span><b>pretty</b>(<span style='color:#e85290;'>x=</span><span style='color:#2b74c7;'>1993</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>2008</span>, <span style='color:#e85290;'>n=</span><span style='color:#2b74c7;'>8</span>), <span style='color:#e85290;'>ncol=</span><span style='color:#2b74c7;'>1</span>)
<i><span style='color:#755019;'># Plots bootstrap support - note usflu.boot contains raw numbers - transform it into percent</span></i>
<b>nodelabels</b>(<span style='color:#e85290;'>text=</span><b>round</b>(usflu.boot<span style='color:#803f00;'>/</span><span style='color:#2b74c7;'>10</span>), <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>0.75</span>)

<i><span style='color:#755019;'># Collapse branches with low bootstrap support</span></i>
usflu.tree.usflu.na.density <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span>usflu.tree.rooted
<i><span style='color:#755019;'># Determine branches with low support - note BS values are in raw numbers - use desired percentage with respect to number of bootstraps</span></i>
usflu.tocollapse <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>match</b>(<span style='color:#e85290;'>x=</span><b>which</b>(usflu.boot <span style='color:#803f00;'>&lt;</span><span style='color:#bf0303;'> </span><span style='color:#2b74c7;'>700</span>)<span style='color:#803f00;'>+</span><b>length</b>(usflu.tree.rooted<span style='color:#803f00;'>$</span>tip.label), <span style='color:#e85290;'>table=</span>usflu.tree.usflu.na.density<span style='color:#803f00;'>$</span>edge[,<span style='color:#2b74c7;'>2</span>])
<i><span style='color:#755019;'># Set length of bad branches to zero</span></i>
usflu.tree.usflu.na.density<span style='color:#803f00;'>$</span>edge.length[usflu.tocollapse] <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><span style='color:#2b74c7;'>0</span>
<i><span style='color:#755019;'># Create new tree</span></i>
usflu.tree.collapsed <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>di2multi</b>(<span style='color:#e85290;'>phy=</span>usflu.tree.usflu.na.density, <span style='color:#e85290;'>tol=</span><span style='color:#2b74c7;'>0.00001</span>)
<i><span style='color:#755019;'># Plot the consensus tree</span></i>
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>usflu.tree.collapsed, <span style='color:#e85290;'>show.tip=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>edge.width=</span><span style='color:#2b74c7;'>2</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;NJ tree after collapsing weak nodes&quot;</span>)
<b>tiplabels</b>(<span style='color:#e85290;'>text=</span>annot<span style='color:#803f00;'>$</span>year, <span style='color:#e85290;'>bg=</span><b>transp</b>(<b>num2col</b>(<span style='color:#e85290;'>x=</span>annot<span style='color:#803f00;'>$</span>year, <span style='color:#e85290;'>col.pal=</span>usflu.pal), <span style='color:#e85290;'>alpha=</span><span style='color:#2b74c7;'>0.7</span>), <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>0.5</span>, <span style='color:#e85290;'>fg=</span><span style='color:#bf0303;'>&quot;transparent&quot;</span>)
<b>axisPhylo</b>()
<b>legend</b>(<span style='color:#e85290;'>x=</span><span style='color:#bf0303;'>&quot;topright&quot;</span>, <span style='color:#e85290;'>fill=</span><b>num2col</b>(<span style='color:#e85290;'>x=</span><b>pretty</b>(<span style='color:#e85290;'>x=</span><span style='color:#2b74c7;'>1993</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>2008</span>, <span style='color:#e85290;'>n=</span><span style='color:#2b74c7;'>8</span>), <span style='color:#e85290;'>col.pal=</span>usflu.pal), <span style='color:#e85290;'>leg=</span><b>pretty</b>(<span style='color:#e85290;'>x=</span><span style='color:#2b74c7;'>1993</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>2008</span>, <span style='color:#e85290;'>n=</span><span style='color:#2b74c7;'>8</span>), <span style='color:#e85290;'>ncol=</span><span style='color:#2b74c7;'>1</span>)

<b><span style='color:#b00000;'>## Replacements of NJ</span></b>
<i><span style='color:#755019;'># Usage is same as nj() Check</span></i>
?njs
?bionj
?bionjs
?NJ
?UNJ
?fastme

<b><span style='color:#b00000;'>## PCoA</span></b>
<i><span style='color:#755019;'># Calculate it - based on various distance matrix</span></i>
hauss.pcoa <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>dudi.pco</b>(<span style='color:#e85290;'>d=</span><b>dist</b>(<span style='color:#e85290;'>x=</span><b>scaleGen</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>center=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>scale=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>truenames=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>missing=</span><span style='color:#bf0303;'>&quot;NA&quot;</span>), <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;euclidean&quot;</span>), <span style='color:#e85290;'>scannf=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>nf=</span><span style='color:#2b74c7;'>3</span>)
<i><span style='color:#755019;'># Basic display</span></i>
<b>s.label</b>(<span style='color:#e85290;'>dfxy=</span>hauss.pcoa<span style='color:#803f00;'>$</span>li, <span style='color:#e85290;'>clabel=</span><span style='color:#2b74c7;'>0.75</span>)
<i><span style='color:#755019;'># To plot different axes use for example &quot;...dfxy=hauss.pcoa$li[c(2, 3)]...&quot;</span></i>
<i><span style='color:#755019;'># Add kernel density</span></i>
<b>s.kde2d</b>(<span style='color:#e85290;'>dfxy=</span>hauss.pcoa<span style='color:#803f00;'>$</span>li, <span style='color:#e85290;'>cpoint=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>add.plot=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Adds histogram of Eigenvalues</span></i>
<b>add.scatter.eig</b>(<span style='color:#e85290;'>w=</span>hauss.pcoa<span style='color:#803f00;'>$</span>eig, <span style='color:#e85290;'>nf=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>xax=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>yax=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>posi=</span><span style='color:#bf0303;'>&quot;bottomleft&quot;</span>, <span style='color:#e85290;'>sub=</span><span style='color:#bf0303;'>&quot;Eigenvalues&quot;</span>)
<i><span style='color:#755019;'># Colored display according to populations</span></i>
hauss.pcoa.col <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>rainbow</b>(<b>length</b>(<b>levels</b>(<b>pop</b>(hauss.genind)))) <i><span style='color:#755019;'># Creates vector of colors according to populations</span></i>
<b>s.class</b>(<span style='color:#e85290;'>dfxy=</span>hauss.pcoa<span style='color:#803f00;'>$</span>li, <span style='color:#e85290;'>fac=</span><b>pop</b>(hauss.genind), <span style='color:#e85290;'>col=</span>hauss.pcoa.col)
<b>add.scatter.eig</b>(<span style='color:#e85290;'>w=</span>hauss.pcoa<span style='color:#803f00;'>$</span>eig, <span style='color:#e85290;'>nf=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>xax=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>yax=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>posi=</span><span style='color:#bf0303;'>&quot;bottomleft&quot;</span>, <span style='color:#e85290;'>sub=</span><span style='color:#bf0303;'>&quot;Eigenvalues&quot;</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;Principal Coordinates Analysis&quot;</span>) <i><span style='color:#755019;'># Adds title to the graph</span></i>
<i><span style='color:#755019;'># Based on Bruvo's distance</span></i>
hauss.pcoa.bruvo <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>dudi.pco</b>(<span style='color:#e85290;'>d=</span><b>bruvo.dist</b>(<span style='color:#e85290;'>pop=</span>hauss.genind, <span style='color:#e85290;'>replen=</span><b>rep</b>(<span style='color:#2b74c7;'>2</span>, <span style='color:#2b74c7;'>12</span>)), <span style='color:#e85290;'>scannf=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>nf=</span><span style='color:#2b74c7;'>3</span>)
<b>s.class</b>(<span style='color:#e85290;'>dfxy=</span>hauss.pcoa.bruvo<span style='color:#803f00;'>$</span>li, <span style='color:#e85290;'>fac=</span><b>pop</b>(hauss.genind), <span style='color:#e85290;'>col=</span>hauss.pcoa.col)
<b>add.scatter.eig</b>(hauss.pcoa.bruvo<span style='color:#803f00;'>$</span>eig, <span style='color:#e85290;'>posi=</span><span style='color:#bf0303;'>&quot;bottomright&quot;</span>, <span style='color:#2b74c7;'>3</span>, <span style='color:#2b74c7;'>1</span>, <span style='color:#2b74c7;'>2</span>)
<i><span style='color:#755019;'># Another possibility for colored plot (see ?colorplot for details)</span></i>
<b>colorplot</b>(<span style='color:#e85290;'>xy=</span>hauss.pcoa<span style='color:#803f00;'>$</span>li[<b>c</b>(<span style='color:#2b74c7;'>1</span>, <span style='color:#2b74c7;'>2</span>)], <span style='color:#e85290;'>X=</span>hauss.pcoa<span style='color:#803f00;'>$</span>li, <span style='color:#e85290;'>transp=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;PC 1&quot;</span>, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;PC 2&quot;</span>)
<b>title</b>(<span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;PCoA, axes 1 and 3&quot;</span>)
<b>abline</b>(<span style='color:#e85290;'>v=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>h=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;grey&quot;</span>, <span style='color:#e85290;'>lty=</span><span style='color:#2b74c7;'>2</span>)

<b><span style='color:#b00000;'>## Clustering analysis</span></b>
<i><span style='color:#755019;'># Graphical web interface for DAPC</span></i>
<b>adegenetServer</b>(<span style='color:#bf0303;'>&quot;DAPC&quot;</span>)
<b><span style='color:#b00000;'>## K-find</span></b>
<i><span style='color:#755019;'># Bayesian K-means clustering</span></i>
<i><span style='color:#755019;'># Retain all informative PC (here about 35). According to second graph select best K (here 2 or 3).</span></i>
hauss.kfind <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>find.clusters</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>stat=</span><span style='color:#bf0303;'>&quot;BIC&quot;</span>, <span style='color:#e85290;'>choose.n.clust=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>max.n.clust=</span><span style='color:#2b74c7;'>10</span>, <span style='color:#e85290;'>n.iter=</span><span style='color:#2b74c7;'>100000</span>, <span style='color:#e85290;'>n.start=</span><span style='color:#2b74c7;'>100</span>, <span style='color:#e85290;'>scale=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>truenames=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># See results as text</span></i>
<b>table</b>(<b>pop</b>(hauss.genind), hauss.kfind<span style='color:#803f00;'>$</span>grp)
hauss.kfind
<i><span style='color:#755019;'># Graph showing table of original and inferred populations and assignment of individuals</span></i>
<b>table.value</b>(<span style='color:#e85290;'>df=</span><b>table</b>(<b>pop</b>(hauss.genind), hauss.kfind<span style='color:#803f00;'>$</span>grp), <span style='color:#e85290;'>col.lab=</span><b>paste</b>(<span style='color:#bf0303;'>&quot;Inferred</span><span style='color:#bf5c00;'>\n</span><span style='color:#bf0303;'>cluster&quot;</span>, <span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span><b>length</b>(hauss.kfind<span style='color:#803f00;'>$</span>size)), <span style='color:#e85290;'>grid=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># For K=3 - note parameters n.pca and n.clust - we just rerun the analysis and when results are stable, no problem here</span></i>
hauss.kfind3 <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>find.clusters</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>n.pca=</span><span style='color:#2b74c7;'>35</span>, <span style='color:#e85290;'>n.clust=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>stat=</span><span style='color:#bf0303;'>&quot;BIC&quot;</span>, <span style='color:#e85290;'>choose.n.clust=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>max.n.clust=</span><span style='color:#2b74c7;'>10</span>, <span style='color:#e85290;'>n.iter=</span><span style='color:#2b74c7;'>100000</span>, <span style='color:#e85290;'>n.start=</span><span style='color:#2b74c7;'>100</span>, <span style='color:#e85290;'>scale=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>truenames=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># See results as text</span></i>
<b>table</b>(<b>pop</b>(hauss.genind), hauss.kfind3<span style='color:#803f00;'>$</span>grp)
hauss.kfind3
<i><span style='color:#755019;'># Graph showing table of original and inferred populations and assignment of individuals</span></i>
<b>table.value</b>(<span style='color:#e85290;'>df=</span><b>table</b>(<b>pop</b>(hauss.genind), hauss.kfind3<span style='color:#803f00;'>$</span>grp), <span style='color:#e85290;'>col.lab=</span><b>paste</b>(<span style='color:#bf0303;'>&quot;Inferred</span><span style='color:#bf5c00;'>\n</span><span style='color:#bf0303;'>cluster&quot;</span>, <span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span><b>length</b>(hauss.kfind3<span style='color:#803f00;'>$</span>size)), <span style='color:#e85290;'>grid=</span><span style='color:#00892b;'>TRUE</span>)

<b><span style='color:#b00000;'>## DAPC</span></b>
<i><span style='color:#755019;'># K=2</span></i>
<i><span style='color:#755019;'># Creates DAPC</span></i>
<i><span style='color:#755019;'># Number of informative PC (Here 15, adegenet recommends &lt; N/3). Select number of informative DA (here 1 - only one is available - it won't produce PCA graph).</span></i>
hauss.dapc <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>dapc</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>pop=</span>hauss.kfind<span style='color:#803f00;'>$</span>grp, <span style='color:#e85290;'>center=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>scale=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>var.contrib=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>pca.info=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>truenames=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Information</span></i>
hauss.dapc
<i><span style='color:#755019;'># Density functions</span></i>
<b>scatter</b>(<span style='color:#e85290;'>x=</span>hauss.dapc, <span style='color:#e85290;'>xax=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>yax=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;DAPC&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;white&quot;</span>, <span style='color:#e85290;'>solid=</span><span style='color:#2b74c7;'>0.5</span>, <span style='color:#e85290;'>leg=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>txt.leg=</span><b>c</b>(<span style='color:#bf0303;'>&quot;Group 1&quot;</span>, <span style='color:#bf0303;'>&quot;Group 2&quot;</span>), <span style='color:#e85290;'>posi.leg=</span><span style='color:#bf0303;'>&quot;topright&quot;</span>)
<i><span style='color:#755019;'># Assignment of individuals to clusters</span></i>
<b>assignplot</b>(<span style='color:#e85290;'>x=</span>hauss.dapc)
<i><span style='color:#755019;'># Structure-like plot</span></i>
<b>compoplot</b>(<span style='color:#e85290;'>x=</span>hauss.dapc, <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;Individuals&quot;</span>, <span style='color:#e85290;'>leg=</span><span style='color:#00892b;'>FALSE</span>)
<i><span style='color:#755019;'># Loadingplot - alleles the most adding to separation of individuals</span></i>
<b>loadingplot</b>(<span style='color:#e85290;'>x=</span>hauss.dapc<span style='color:#803f00;'>$</span>var.contr)
<i><span style='color:#755019;'># alfa-score - according to number of PC axis</span></i>
<b>optim.a.score</b>(<span style='color:#e85290;'>x=</span>hauss.dapc)
<i><span style='color:#755019;'># K=3</span></i>
<i><span style='color:#755019;'># Creates DAPC</span></i>
<i><span style='color:#755019;'># Number of informative PC (Here 15, adegenet recommends &lt; N/3). Select number of informative DA (here 2).</span></i>
hauss.dapc3 <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>dapc</b>(<span style='color:#e85290;'>x=</span>hauss.genind, <span style='color:#e85290;'>pop=</span>hauss.kfind3<span style='color:#803f00;'>$</span>grp, <span style='color:#e85290;'>center=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>scale=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>var.contrib=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>pca.info=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>truenames=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Information</span></i>
hauss.dapc
<i><span style='color:#755019;'># A la PCA graph</span></i>
<b>scatter</b>(<span style='color:#e85290;'>x=</span>hauss.dapc3, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;DAPC, Cardamine&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;white&quot;</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>clab=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>col=</span><b>rainbow</b>(<span style='color:#2b74c7;'>3</span>), <span style='color:#e85290;'>posi.da=</span><span style='color:#bf0303;'>&quot;bottomleft&quot;</span>, <span style='color:#e85290;'>scree.pca=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>posi.pca=</span><span style='color:#bf0303;'>&quot;bottomright&quot;</span>, <span style='color:#e85290;'>leg=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>txt.leg=</span><b>c</b>(<span style='color:#bf0303;'>&quot;Group 1&quot;</span>, <span style='color:#bf0303;'>&quot;Group 2&quot;</span>, <span style='color:#bf0303;'>&quot;Group 3&quot;</span>), <span style='color:#e85290;'>posi.leg=</span><span style='color:#bf0303;'>&quot;topleft&quot;</span>)
<i><span style='color:#755019;'># Same in BW</span></i>
<b>scatter</b>(<span style='color:#e85290;'>x=</span>hauss.dapc3, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;DAPC, Cardamine&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;white&quot;</span>, <span style='color:#e85290;'>pch=</span><b>c</b>(<span style='color:#2b74c7;'>15</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>17</span>), <span style='color:#e85290;'>cell=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>cstar=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>solid=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>2.5</span>, <span style='color:#e85290;'>clab=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>col=</span><b>grey.colors</b>(<span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>start=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>end=</span><span style='color:#2b74c7;'>0.8</span>, <span style='color:#e85290;'>gamma=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>alpha=</span><span style='color:#2b74c7;'>0</span>), <span style='color:#e85290;'>posi.da=</span><span style='color:#bf0303;'>&quot;bottomleft&quot;</span>, <span style='color:#e85290;'>scree.pca=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>posi.pca=</span><span style='color:#bf0303;'>&quot;bottomright&quot;</span>, <span style='color:#e85290;'>leg=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>txt.leg=</span><b>c</b>(<span style='color:#bf0303;'>&quot;Group 1&quot;</span>, <span style='color:#bf0303;'>&quot;Group 2&quot;</span>, <span style='color:#bf0303;'>&quot;Group 3&quot;</span>), <span style='color:#e85290;'>posi.leg=</span><span style='color:#bf0303;'>&quot;topleft&quot;</span>)
<i><span style='color:#755019;'># Density functions</span></i>
<b>scatter</b>(<span style='color:#e85290;'>x=</span>hauss.dapc3, <span style='color:#e85290;'>xax=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>yax=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;DAPC&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;white&quot;</span>, <span style='color:#e85290;'>solid=</span><span style='color:#2b74c7;'>0.5</span>, <span style='color:#e85290;'>leg=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>txt.leg=</span><b>c</b>(<span style='color:#bf0303;'>&quot;Group 1&quot;</span>, <span style='color:#bf0303;'>&quot;Group 2&quot;</span>, <span style='color:#bf0303;'>&quot;Group 3&quot;</span>), <span style='color:#e85290;'>posi.leg=</span><span style='color:#bf0303;'>&quot;topleft&quot;</span>)
<i><span style='color:#755019;'># Assignment of individuals to clusters</span></i>
<b>assignplot</b>(hauss.dapc3)
<i><span style='color:#755019;'># Structure-like plot</span></i>
<b>compoplot</b>(hauss.dapc3, <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;Individuals&quot;</span>, <span style='color:#e85290;'>leg=</span><span style='color:#00892b;'>FALSE</span>)
<i><span style='color:#755019;'># Loadingplot - alleles the most adding to separation of individuals</span></i>
<b>loadingplot</b>(hauss.dapc3<span style='color:#803f00;'>$</span>var.contr)
<i><span style='color:#755019;'># alfa-score - according to number of PC axis</span></i>
<b>optim.a.score</b>(hauss.dapc3)

<b><span style='color:#b00000;'>## SNP</span></b>

<i><span style='color:#755019;'># Plot of missing data (white) and number of 2nd alleles</span></i>
<b>glPlot</b>(<span style='color:#e85290;'>x=</span>usflu.genlight, <span style='color:#e85290;'>legend=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>posi=</span><span style='color:#bf0303;'>&quot;topleft&quot;</span>)
<i><span style='color:#755019;'># Sum of the number of second allele in each SNP</span></i>
usflu.freq <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>glSum</b>(usflu.genlight)
<i><span style='color:#755019;'># Plot distribution of (second) allele frequencies</span></i>
<b>hist</b>(<span style='color:#e85290;'>x=</span>usflu.freq, <span style='color:#e85290;'>proba=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;gold&quot;</span>, <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;Allele frequencies&quot;</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Distribution of (second) allele frequencies&quot;</span>)
<b>lines</b>(<span style='color:#e85290;'>x=</span><b>density</b>(usflu.freq)<span style='color:#803f00;'>$</span>x, <span style='color:#e85290;'>y=</span><b>density</b>(usflu.freq)<span style='color:#803f00;'>$</span>y<span style='color:#803f00;'>*</span><span style='color:#2b74c7;'>1.5</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;red&quot;</span>, <span style='color:#e85290;'>lwd=</span><span style='color:#2b74c7;'>3</span> )
<i><span style='color:#755019;'># Mean number of second allele in each SNP</span></i>
usflu.mean <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>glMean</b>(usflu.genlight)
usflu.mean <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>c</b>(usflu.mean, <span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>-</span>usflu.mean)
<i><span style='color:#755019;'># Plot distribution of allele frequencies - play with parameters to get nice image</span></i>
<b>hist</b>(<span style='color:#e85290;'>x=</span>usflu.mean, <span style='color:#e85290;'>proba=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;darkseagreen3&quot;</span>, <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;Allele frequencies&quot;</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Distribution of allele frequencies&quot;</span>, <span style='color:#e85290;'>nclass=</span><span style='color:#2b74c7;'>20</span>)
<b>lines</b>(<span style='color:#e85290;'>x=</span><b>density</b>(usflu.mean, <span style='color:#e85290;'>bw=</span><span style='color:#2b74c7;'>0.05</span>)<span style='color:#803f00;'>$</span>x, <span style='color:#e85290;'>y=</span><b>density</b>(usflu.mean, <span style='color:#e85290;'>bw=</span><span style='color:#2b74c7;'>0.05</span>)<span style='color:#803f00;'>$</span>y<span style='color:#803f00;'>*</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>lwd=</span><span style='color:#2b74c7;'>3</span>)

<i><span style='color:#755019;'># Number of missing values in each locus</span></i>
usflu.na.density <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>density</b>(<b>glNA</b>(usflu.genlight), <span style='color:#e85290;'>bw=</span><span style='color:#2b74c7;'>10</span>) <i><span style='color:#755019;'># Play with bw parameter to get optimal image</span></i>
<i><span style='color:#755019;'># Set range of xlim parameter from 0 to the length of original alignment</span></i>
<b>plot</b>(<span style='color:#e85290;'>x=</span>usflu.na.density, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;n&quot;</span>, <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;Position in the alignment&quot;</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Location of the missing values (NA)&quot;</span>, <span style='color:#e85290;'>xlim=</span><b>c</b>(<span style='color:#2b74c7;'>0</span>, <span style='color:#2b74c7;'>1701</span>))
<b>polygon</b>(<b>c</b>(usflu.na.density<span style='color:#803f00;'>$</span>x, <b>rev</b>(usflu.na.density<span style='color:#803f00;'>$</span>x)), <b>c</b>(usflu.na.density<span style='color:#803f00;'>$</span>y, <b>rep</b>(<span style='color:#2b74c7;'>0</span>, <b>length</b>(usflu.na.density<span style='color:#803f00;'>$</span>x))), <span style='color:#e85290;'>col=</span><b>transp</b>(<span style='color:#bf0303;'>&quot;blue&quot;</span>, <span style='color:#e85290;'>alpha=</span><span style='color:#2b74c7;'>0.3</span>))
<b>points</b>(<b>glNA</b>(usflu.genlight), <b>rep</b>(<span style='color:#2b74c7;'>0</span>, <b>nLoc</b>(usflu.genlight)), <span style='color:#e85290;'>pch=</span><span style='color:#bf0303;'>&quot;|&quot;</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;blue&quot;</span>)

<i><span style='color:#755019;'># PCA - select number of retained PC axes, about 10 here</span></i>
usflu.pca <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>glPca</b>(<span style='color:#e85290;'>x=</span>usflu.genlight, <span style='color:#e85290;'>center=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>scale=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>loadings=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Plot PCA</span></i>
<b>scatter.glPca</b>(<span style='color:#e85290;'>x=</span>usflu.pca, <span style='color:#e85290;'>posi=</span><span style='color:#bf0303;'>&quot;bottomright&quot;</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;PCA of the US influenza data&quot;</span>)
<i><span style='color:#755019;'># Coloured plot</span></i>
<b>colorplot</b>(usflu.pca<span style='color:#803f00;'>$</span>scores, usflu.pca<span style='color:#803f00;'>$</span>scores, <span style='color:#e85290;'>transp=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>4</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;PCA of the US influenza data&quot;</span>)
<b>abline</b>(<span style='color:#e85290;'>h=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>v=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;grey&quot;</span>)
<b>add.scatter.eig</b>(usflu.pca<span style='color:#803f00;'>$</span>eig[<span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>40</span>], <span style='color:#2b74c7;'>2</span>, <span style='color:#2b74c7;'>1</span>, <span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>posi=</span><span style='color:#bf0303;'>&quot;topright&quot;</span>, <span style='color:#e85290;'>inset=</span><span style='color:#2b74c7;'>0.05</span>, <span style='color:#e85290;'>ratio=</span><span style='color:#2b74c7;'>0.3</span>)
<i><span style='color:#755019;'># Calculate phylogenetic tree</span></i>
usflu.tree.genlight <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>nj</b>(<b>dist</b>(<b>as.matrix</b>(usflu.genlight)))
<i><span style='color:#755019;'># Plot coloured phylogenetic tree</span></i>
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>usflu.tree.genlight, <span style='color:#e85290;'>typ=</span><span style='color:#bf0303;'>&quot;fan&quot;</span>, <span style='color:#e85290;'>show.tip=</span><span style='color:#00892b;'>FALSE</span>)
<b>tiplabels</b>(<span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>20</span>, <span style='color:#e85290;'>col=</span>usflu.col, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>4</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;NJ tree of the US influenza data&quot;</span>)

<b><span style='color:#b00000;'>## Moran's I</span></b>
<i><span style='color:#755019;'># Creates connection network</span></i>
hauss.connectivity <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>chooseCN</b>(<span style='color:#e85290;'>xy=</span>hauss.genind<span style='color:#803f00;'>$</span>other<span style='color:#803f00;'>$</span>xy, <span style='color:#e85290;'>type=</span><span style='color:#2b74c7;'>5</span>, <span style='color:#e85290;'>d1=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>d2=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>plot.nb=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>result.type=</span><span style='color:#bf0303;'>&quot;listw&quot;</span>, <span style='color:#e85290;'>edit.nb=</span><span style='color:#00892b;'>FALSE</span>)
hauss.connectivity
<i><span style='color:#755019;'># Test of Moran's I for 1st PCoA axis</span></i>
<b>moran.test</b>(<span style='color:#e85290;'>x=</span>hauss.pcoa<span style='color:#803f00;'>$</span>li[,<span style='color:#2b74c7;'>1</span>], <span style='color:#e85290;'>listw=</span>hauss.connectivity, <span style='color:#e85290;'>alternative=</span><span style='color:#bf0303;'>&quot;greater&quot;</span>, <span style='color:#e85290;'>randomisation=</span><span style='color:#00892b;'>TRUE</span>)
hauss.pcoa1.mctest <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>moran.mc</b>(<span style='color:#e85290;'>x=</span>hauss.pcoa<span style='color:#803f00;'>$</span>li[,<span style='color:#2b74c7;'>1</span>], <span style='color:#e85290;'>listw=</span>hauss.connectivity, <span style='color:#e85290;'>alternative=</span><span style='color:#bf0303;'>&quot;greater&quot;</span>, <span style='color:#e85290;'>nsim=</span><span style='color:#2b74c7;'>1000</span>)
hauss.pcoa1.mctest
<b>plot</b>(hauss.pcoa1.mctest)
<b>moran.plot</b>(<span style='color:#e85290;'>x=</span>hauss.pcoa<span style='color:#803f00;'>$</span>li[,<span style='color:#2b74c7;'>1</span>], <span style='color:#e85290;'>listw=</span>hauss.connectivity)
<i><span style='color:#755019;'># Test of Moran's I for 2nd PCoA axis</span></i>
<b>moran.test</b>(<span style='color:#e85290;'>x=</span>hauss.pcoa<span style='color:#803f00;'>$</span>li[,<span style='color:#2b74c7;'>2</span>], <span style='color:#e85290;'>listw=</span>hauss.connectivity, <span style='color:#e85290;'>alternative=</span><span style='color:#bf0303;'>&quot;greater&quot;</span>, <span style='color:#e85290;'>randomisation=</span><span style='color:#00892b;'>TRUE</span>)
hauss.pcoa2.mctest <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>moran.mc</b>(<span style='color:#e85290;'>x=</span>hauss.pcoa<span style='color:#803f00;'>$</span>li[,<span style='color:#2b74c7;'>2</span>], <span style='color:#e85290;'>listw=</span>hauss.connectivity, <span style='color:#e85290;'>alternative=</span><span style='color:#bf0303;'>&quot;greater&quot;</span>, <span style='color:#e85290;'>nsim=</span><span style='color:#2b74c7;'>1000</span>)
hauss.pcoa2.mctest
<b>plot</b>(hauss.pcoa2.mctest)
<b>moran.plot</b>(<span style='color:#e85290;'>x=</span>hauss.pcoa[[<span style='color:#bf0303;'>&quot;li&quot;</span>]][,<span style='color:#2b74c7;'>2</span>], <span style='color:#e85290;'>listw=</span>hauss.connectivity)

<b><span style='color:#b00000;'>## sPCA</span></b>

<i><span style='color:#755019;'># Calculates sPCA - here is only 1 positive and 3 negative factors</span></i>
hauss.spca <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>spca</b>(<span style='color:#e85290;'>obj=</span>hauss.genind, <span style='color:#e85290;'>cn=</span>hauss.connectivity, <span style='color:#e85290;'>scale=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>scannf=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Plot eigenvalues of sPCA - gobal vs. local structure</span></i>
<b>barplot</b>(<span style='color:#e85290;'>height=</span>hauss.spca<span style='color:#803f00;'>$</span>eig, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Eigenvalues of sPCA&quot;</span>, <span style='color:#e85290;'>col=</span><b>spectral</b>(<b>length</b>(hauss.spca[[<span style='color:#bf0303;'>&quot;eig&quot;</span>]])))
<b>legend</b>(<span style='color:#bf0303;'>&quot;topright&quot;</span>, <span style='color:#e85290;'>fill=</span><b>spectral</b>(<span style='color:#2b74c7;'>2</span>), <span style='color:#e85290;'>leg=</span><b>c</b>(<span style='color:#bf0303;'>&quot;Global structures&quot;</span>, <span style='color:#bf0303;'>&quot;Local structures&quot;</span>)) <i><span style='color:#755019;'># Add legend</span></i>
<b>abline</b>(<span style='color:#e85290;'>h=</span><span style='color:#2b74c7;'>0</span>,<span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;grey&quot;</span>) <i><span style='color:#755019;'># Add line showing zero</span></i>
<i><span style='color:#755019;'># Information about sPCA</span></i>
<b>print.spca</b>(hauss.spca)
<i><span style='color:#755019;'># Summary of sPCA results</span></i>
<b>summary.spca</b>(hauss.spca)
<i><span style='color:#755019;'># Shows connectivity network, 3 different scores, barplot of eigenvalues and eigenvalues decomposition</span></i>
<b>plot.spca</b>(hauss.spca)
<i><span style='color:#755019;'># Display of scores in color canals (two analogous variants)</span></i>
<b>colorplot.spca</b>(hauss.spca, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>4</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;sPCA - colorplot of PC 1 and 2 (lagged scores)&quot;</span>, <span style='color:#e85290;'>line=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.5</span>)
<b>colorplot</b>(<span style='color:#e85290;'>x=</span>hauss.genind<span style='color:#803f00;'>$</span>other<span style='color:#803f00;'>$</span>xy, hauss.spca<span style='color:#803f00;'>$</span>ls, <span style='color:#e85290;'>axes=</span><span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span><b>ncol</b>(hauss.spca<span style='color:#803f00;'>$</span>li), <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>4</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;sPCA - colorplot of PC 1 and 2 (lagged scores)&quot;</span>, <span style='color:#e85290;'>line=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.5</span>)
<i><span style='color:#755019;'># Spatial and variance components of the eigenvalues</span></i>
<b>screeplot.spca</b>(<span style='color:#e85290;'>x=</span>hauss.spca, <span style='color:#e85290;'>main=</span><span style='color:#00892b;'>NULL</span>)

<i><span style='color:#755019;'># Map of genetic clines</span></i>
<b>library</b>(akima) <i><span style='color:#755019;'># This library is needed for some manipulation with coordinates</span></i>
hauss.spca.usflu.na.density <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>interp</b>(<b>other</b>(hauss.genind)<span style='color:#803f00;'>$</span>xy[,<span style='color:#2b74c7;'>1</span>], <b>other</b>(hauss.genind)<span style='color:#803f00;'>$</span>xy[,<span style='color:#2b74c7;'>2</span>], hauss.spca<span style='color:#803f00;'>$</span>ls[,<span style='color:#2b74c7;'>1</span>], <span style='color:#e85290;'>xo=</span><b>seq</b>(<b>min</b>(<b>other</b>(hauss.genind)<span style='color:#803f00;'>$</span>xy[,<span style='color:#2b74c7;'>1</span>]), <b>max</b>(<b>other</b>(hauss.genind)<span style='color:#803f00;'>$</span>xy[,<span style='color:#2b74c7;'>1</span>]), <span style='color:#e85290;'>le=</span><span style='color:#2b74c7;'>200</span>), <span style='color:#e85290;'>yo=</span><b>seq</b>(<b>min</b>(<b>other</b>(hauss.genind)<span style='color:#803f00;'>$</span>xy[,<span style='color:#2b74c7;'>2</span>]), <b>max</b>(<b>other</b>(hauss.genind)<span style='color:#803f00;'>$</span>xy[,<span style='color:#2b74c7;'>2</span>]), <span style='color:#e85290;'>le=</span><span style='color:#2b74c7;'>200</span>), <span style='color:#e85290;'>duplicate=</span><span style='color:#bf0303;'>&quot;median&quot;</span>)
<i><span style='color:#755019;'># For 1st axis</span></i>
<b>image</b>(<span style='color:#e85290;'>x=</span>hauss.spca.usflu.na.density, <span style='color:#e85290;'>col=</span><b>spectral</b>(<span style='color:#2b74c7;'>100</span>))
<b>s.value</b>(<span style='color:#e85290;'>dfxy=</span>hauss.genind<span style='color:#803f00;'>$</span>other<span style='color:#803f00;'>$</span>xy, <span style='color:#e85290;'>z=</span>hauss.pcoa<span style='color:#803f00;'>$</span>li[,<span style='color:#2b74c7;'>1</span>], <span style='color:#e85290;'>add.p=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>csize=</span><span style='color:#2b74c7;'>0.5</span>, <span style='color:#e85290;'>sub=</span><span style='color:#bf0303;'>&quot;PCoA - first PC&quot;</span>, <span style='color:#e85290;'>csub=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>possub=</span><span style='color:#bf0303;'>&quot;topleft&quot;</span>)
<i><span style='color:#755019;'># For 2nd axis</span></i>
<b>image</b>(<span style='color:#e85290;'>x=</span>hauss.spca.usflu.na.density, <span style='color:#e85290;'>col=</span><b>spectral</b>(<span style='color:#2b74c7;'>100</span>))
<b>s.value</b>(<span style='color:#e85290;'>dfxy=</span>hauss.genind<span style='color:#803f00;'>$</span>other<span style='color:#803f00;'>$</span>xy, <span style='color:#e85290;'>z=</span>hauss.pcoa[[<span style='color:#bf0303;'>&quot;li&quot;</span>]][,<span style='color:#2b74c7;'>2</span>], <span style='color:#e85290;'>add.p=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>csize=</span><span style='color:#2b74c7;'>0.5</span>, <span style='color:#e85290;'>sub=</span><span style='color:#bf0303;'>&quot;PCoA - second PC&quot;</span>, <span style='color:#e85290;'>csub=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>possub=</span><span style='color:#bf0303;'>&quot;topleft&quot;</span>)
<i><span style='color:#755019;'># Interpolated lagged score on a map</span></i>
hauss.spca.annot <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><span style='color:#0000bf;'>function</span>() {
    <b>title</b>(<span style='color:#bf0303;'>&quot;sPCA - interpolated map of individual scores&quot;</span>)
    <b>points</b>(<b>other</b>(hauss.genind)<span style='color:#803f00;'>$</span>xy[,<span style='color:#2b74c7;'>1</span>], <b>other</b>(hauss.genind)<span style='color:#803f00;'>$</span>xy[,<span style='color:#2b74c7;'>2</span>])
  }
<b>filled.contour</b>(hauss.spca.usflu.na.density, <span style='color:#e85290;'>color.pal=</span><b>colorRampPalette</b>(<b>lightseasun</b>(<span style='color:#2b74c7;'>100</span>)), <span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>20</span>, <span style='color:#e85290;'>nlevels=</span><span style='color:#2b74c7;'>100</span>, <span style='color:#e85290;'>key.title=</span><b>title</b>(<span style='color:#bf0303;'>&quot;Lagged</span><span style='color:#bf5c00;'>\n</span><span style='color:#bf0303;'>score 1&quot;</span>), <span style='color:#e85290;'>plot.title=</span><b>hauss.spca.annot</b>())

<i><span style='color:#755019;'># Loading plots - which alleles contribute the most?</span></i>
hauss.spca.loadings <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span>hauss.spca<span style='color:#803f00;'>$</span>c1[,<span style='color:#2b74c7;'>1</span>]<span style='color:#803f00;'>^</span><span style='color:#2b74c7;'>2</span>
<b>names</b>(hauss.spca.loadings) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>rownames</b>(hauss.spca<span style='color:#803f00;'>$</span>c1)
<b>loadingplot</b>(<span style='color:#e85290;'>x=</span>hauss.spca.loadings, <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;Alleles&quot;</span>, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;Weight of the alleles&quot;</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Contribution of alleles to the first sPCA axis&quot;</span>)
<b>boxplot</b>(<span style='color:#e85290;'>formula=</span>hauss.spca.loadings<span style='color:#803f00;'>~</span>hauss.genind<span style='color:#803f00;'>$</span>loc.fac, <span style='color:#e85290;'>las=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;Contribution&quot;</span>, <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;Marker&quot;</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Contribution by markers into the first global score&quot;</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;grey&quot;</span>)

<b><span style='color:#b00000;'>## Monmonier - genetic boundaries</span></b>
<i><span style='color:#755019;'># It requires every point to have unique coordinates (one can use jitter() or difference in scale of meters). Example here is on population level, which is not ideal.</span></i>
<i><span style='color:#755019;'># Calculates Monmonier's function (for threshold use 'd')</span></i>
hauss.monmonier <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>monmonier</b>(<span style='color:#e85290;'>xy=</span>hauss.genpop<span style='color:#803f00;'>$</span>other<span style='color:#803f00;'>$</span>xy, <span style='color:#e85290;'>dist=</span><b>dist</b>(hauss.genpop<span style='color:#803f00;'>$</span>tab), <span style='color:#e85290;'>cn=</span><b>chooseCN</b>(hauss.genpop<span style='color:#803f00;'>$</span>other<span style='color:#803f00;'>$</span>xy, <span style='color:#e85290;'>ask=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>type=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>plot.nb=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>edit.nb=</span><span style='color:#00892b;'>FALSE</span>), <span style='color:#e85290;'>nrun=</span><span style='color:#2b74c7;'>1</span>)
<b>coords.monmonier</b>(hauss.monmonier) <i><span style='color:#755019;'># See result as text</span></i>
<i><span style='color:#755019;'># Plot genetic boundaries</span></i>
<b>plot.monmonier</b>(hauss.monmonier, <span style='color:#e85290;'>add.arrows=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>bwd=</span><span style='color:#2b74c7;'>10</span>, <span style='color:#e85290;'>sub=</span><span style='color:#bf0303;'>&quot;Monmonier plot&quot;</span>, <span style='color:#e85290;'>csub=</span><span style='color:#2b74c7;'>2</span>)
<b>points</b>(hauss.genpop<span style='color:#803f00;'>$</span>other<span style='color:#803f00;'>$</span>xy, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>2.5</span>, <span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>20</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;red&quot;</span>)
<b>text</b>(<span style='color:#e85290;'>x=</span>hauss.genpop<span style='color:#803f00;'>$</span>other<span style='color:#803f00;'>$</span>xy<span style='color:#803f00;'>$</span>lon, <span style='color:#e85290;'>y=</span>hauss.genpop<span style='color:#803f00;'>$</span>other<span style='color:#803f00;'>$</span>xy<span style='color:#803f00;'>$</span>lat, <span style='color:#e85290;'>labels=</span>hauss.genpop<span style='color:#803f00;'>$</span>pop.names, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>3</span>)
<b>legend</b>(<span style='color:#bf0303;'>&quot;bottomright&quot;</span>, <span style='color:#e85290;'>leg=</span><span style='color:#bf0303;'>&quot;Genetic boundaries</span><span style='color:#bf5c00;'>\n</span><span style='color:#bf0303;'>among populations&quot;</span>)

<b><span style='color:#b00000;'>## Mantel test - isolation by distance</span></b>
<i><span style='color:#755019;'># Geographical distance</span></i>
hauss.gdist <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>dist</b>(<span style='color:#e85290;'>x=</span>hauss.genind<span style='color:#803f00;'>$</span>other<span style='color:#803f00;'>$</span>xy, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;euclidean&quot;</span>, <span style='color:#e85290;'>diag=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>upper=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Mantel test</span></i>
hauss.mantel <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>mantel.randtest</b>(<span style='color:#e85290;'>m1=</span>hauss.dist, <span style='color:#e85290;'>m2=</span>hauss.gdist, <span style='color:#e85290;'>nrepet=</span><span style='color:#2b74c7;'>1000</span>)
hauss.mantel <i><span style='color:#755019;'># See text output</span></i>
<b>plot</b>(hauss.mantel, <span style='color:#e85290;'>nclass=</span><span style='color:#2b74c7;'>30</span>)
<i><span style='color:#755019;'># Libraries required by mantel.correlog:</span></i>
<b>library</b>(permute)
<b>library</b>(lattice)
<b>library</b>(vegan)
<i><span style='color:#755019;'># Different implementation of Mantel test testing distance classes</span></i>
hauss.mantel.cor <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>mantel.correlog</b>(<span style='color:#e85290;'>D.eco=</span>hauss.dist, <span style='color:#e85290;'>D.geo=</span>hauss.gdist, <span style='color:#e85290;'>XY=</span><span style='color:#00892b;'>NULL</span>, <span style='color:#e85290;'>n.class=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>break.pts=</span><span style='color:#00892b;'>NULL</span>, <span style='color:#e85290;'>cutoff=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>r.type=</span><span style='color:#bf0303;'>&quot;pearson&quot;</span>, <span style='color:#e85290;'>nperm=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>mult=</span><span style='color:#bf0303;'>&quot;holm&quot;</span>, <span style='color:#e85290;'>progressive=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># See results for respective classes</span></i>
hauss.mantel.cor
<b>summary</b>(hauss.mantel.cor)
<i><span style='color:#755019;'># Plot it</span></i>
<b>plot</b>(hauss.mantel.cor)

<b><span style='color:#b00000;'>## Geneland</span></b>
<i><span style='color:#755019;'># Haploid and diploid codominant markers (microsattelites or SNPs)</span></i>
<i><span style='color:#755019;'># Load needed libraries</span></i>
<b>library</b>(PBSmapping)
<b>library</b>(RandomFields)
<b>library</b>(fields)
<b>library</b>(spam)
<b>library</b>(grid)
<b>library</b>(maps)
<b>library</b>(tcltk)
<b>library</b>(Geneland)
<i><span style='color:#755019;'># Graphical interface available for Geneland - we will use only command line</span></i>
<b>Geneland.GUI</b>()
<i><span style='color:#755019;'># Loading and conversions of coordinates - Geneland requires specific coordinate space</span></i>
hauss.geneland.coord <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>as.matrix</b>(hauss.coord) <i><span style='color:#755019;'># hauss.cord is DF, we need just plain matrix</span></i>
<b>colnames</b>(hauss.geneland.coord) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>c</b>(<span style='color:#bf0303;'>&quot;X&quot;</span>, <span style='color:#bf0303;'>&quot;Y&quot;</span>)
<b>attr</b>(hauss.geneland.coord, <span style='color:#bf0303;'>&quot;projection&quot;</span>) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> &quot;LL&quot;</span>
<b>attr</b>(hauss.geneland.coord, <span style='color:#bf0303;'>&quot;zone&quot;</span>) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><span style='color:#00892b;'>NA</span>
hauss.geneland.coord.utm <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>convUL</b>(hauss.geneland.coord)
<b>dim</b>(hauss.geneland.coord)
hauss.geneland.coord
<b>dim</b>(hauss.geneland.coord.utm)
hauss.geneland.coord.utm <i><span style='color:#755019;'># Final coordinates</span></i>
<i><span style='color:#755019;'># Load data (only haploid or diploid data are supported) - only plain table with alleles</span></i>
hauss.geneland.data <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.table</b>(<span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;http://trapa.cz/sites/default/rcourse/haussknechtii_geneland.txt&quot;</span>, <span style='color:#e85290;'>na.string=</span><span style='color:#bf0303;'>&quot;-999&quot;</span>, <span style='color:#e85290;'>header=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;</span><span style='color:#bf5c00;'>\t</span><span style='color:#bf0303;'>&quot;</span>)
<b>dim</b>(hauss.geneland.data)
hauss.geneland.data
<i><span style='color:#755019;'># Set number of independent runs</span></i>
hauss.geneland.nrun <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><span style='color:#2b74c7;'>5</span>
<i><span style='color:#755019;'># Set length of burnin chain</span></i>
hauss.geneland.burnin <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><span style='color:#2b74c7;'>100</span>
<i><span style='color:#755019;'># Set maximal K (number of populations)</span></i>
hauss.geneland.maxpop <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><span style='color:#2b74c7;'>10</span>
<i><span style='color:#755019;'># FOR loop will run several independent runs and produce output maps of genetic clusters - outputs are written into subdirectory within geneland directory</span></i>
<span style='color:#0000bf;'>for</span> (hauss.geneland.irun <span style='color:#0000bf;'>in</span> <span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span>hauss.geneland.nrun) {
  hauss.geneland.path.mcmc <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>paste</b>(<span style='color:#bf0303;'>&quot;geneland/&quot;</span>, hauss.geneland.irun, <span style='color:#bf0303;'>&quot;/&quot;</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;&quot;</span>) <i><span style='color:#755019;'># paste is good especially for joining several text chains into one</span></i>
  <i><span style='color:#755019;'># On Windows, remove following line and create subdirectories from 1 to max K manually (creating subdirs in Windows is complicated)</span></i>
  <b>system</b>(<b>paste</b>(<span style='color:#bf0303;'>&quot;mkdir &quot;</span>, hauss.geneland.path.mcmc)) <i><span style='color:#755019;'># Creates subdirectory</span></i>
  <i><span style='color:#755019;'># Inferrence - MCMC chain - see ?MCMC for details</span></i>
  <b>MCMC</b>(<span style='color:#e85290;'>coordinates=</span>hauss.geneland.coord.utm, <span style='color:#e85290;'>geno.dip.codom=</span>hauss.geneland.data, <span style='color:#e85290;'>path.mcmc=</span>hauss.geneland.path.mcmc, <span style='color:#e85290;'>delta.coord=</span><span style='color:#2b74c7;'>0.001</span>, <span style='color:#e85290;'>varnpop=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>npopmin=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>npopmax=</span>hauss.geneland.maxpop, <span style='color:#e85290;'>nit=</span><span style='color:#2b74c7;'>10000</span>, <span style='color:#e85290;'>thinning=</span><span style='color:#2b74c7;'>10</span>, <span style='color:#e85290;'>freq.model=</span><span style='color:#bf0303;'>&quot;Uncorrelated&quot;</span>, <span style='color:#e85290;'>spatial=</span><span style='color:#00892b;'>TRUE</span>)
  <i><span style='color:#755019;'># Post-process chains</span></i>
  <b>PostProcessChain</b>(<span style='color:#e85290;'>coordinates=</span>hauss.geneland.coord.utm, <span style='color:#e85290;'>path.mcmc=</span>hauss.geneland.path.mcmc, <span style='color:#e85290;'>nxdom=</span><span style='color:#2b74c7;'>500</span>, <span style='color:#e85290;'>nydom=</span><span style='color:#2b74c7;'>500</span>, <span style='color:#e85290;'>burnin=</span>hauss.geneland.burnin)
  <i><span style='color:#755019;'># Output</span></i>
  <i><span style='color:#755019;'># Simulated number of populations</span></i>
  <b>Plotnpop</b>(<span style='color:#e85290;'>path.mcmc=</span>hauss.geneland.path.mcmc, <span style='color:#e85290;'>printit=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>file=</span><b>paste</b>(hauss.geneland.path.mcmc, <span style='color:#bf0303;'>&quot;/geneland-number_of_clusters.pdf&quot;</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;&quot;</span>), <span style='color:#e85290;'>format=</span><span style='color:#bf0303;'>&quot;pdf&quot;</span>, <span style='color:#e85290;'>burnin=</span>hauss.geneland.burnin)
  <b>dev.off</b>() <i><span style='color:#755019;'># We must close graphical device manually</span></i>
  <i><span style='color:#755019;'># Map of estimated population membership</span></i>
  <b>PosteriorMode</b>(<span style='color:#e85290;'>coordinates=</span>hauss.geneland.coord.utm, <span style='color:#e85290;'>path.mcmc=</span>hauss.geneland.path.mcmc, <span style='color:#e85290;'>printit=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>format=</span><span style='color:#bf0303;'>&quot;pdf&quot;</span>, <span style='color:#e85290;'>file=</span><b>paste</b>(hauss.geneland.path.mcmc,<span style='color:#bf0303;'>&quot;/geneland-map.pdf&quot;</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;&quot;</span>))
  <b>dev.off</b>() <i><span style='color:#755019;'># We must close graphical device manually</span></i>
  }
<i><span style='color:#755019;'># Prepare list to record values of Fst for all runs</span></i>
hauss.geneland.fstat <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>list</b>()
<i><span style='color:#755019;'># Estimate Fst</span></i>
<span style='color:#0000bf;'>for</span> (hauss.geneland.irun <span style='color:#0000bf;'>in</span> <span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span>hauss.geneland.nrun) {
  hauss.geneland.path.mcmc <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>paste</b>(<span style='color:#bf0303;'>&quot;geneland/&quot;</span>, hauss.geneland.irun, <span style='color:#bf0303;'>&quot;/&quot;</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;&quot;</span>)
  <i><span style='color:#755019;'># F-statistics - Fis and Fst</span></i>
  hauss.geneland.fstat[[hauss.geneland.irun]] <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>Fstat.output</b>(<span style='color:#e85290;'>coordinates=</span>hauss.geneland.coord.utm, <span style='color:#e85290;'>genotypes=</span>hauss.geneland.data, <span style='color:#e85290;'>burnin=</span>hauss.geneland.burnin, <span style='color:#e85290;'>ploidy=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>path.mcmc=</span>hauss.geneland.path.mcmc)
  }
  <i><span style='color:#755019;'># Print Fst output</span></i>
  hauss.geneland.fstat
<i><span style='color:#755019;'># MCMC inference under the admixture model</span></i>
<span style='color:#0000bf;'>for</span> (hauss.geneland.irun <span style='color:#0000bf;'>in</span> <span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span>hauss.geneland.nrun) {
  hauss.geneland.path.mcmc <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>paste</b>(<span style='color:#bf0303;'>&quot;geneland/&quot;</span>, hauss.geneland.irun, <span style='color:#bf0303;'>&quot;/&quot;</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;&quot;</span>)
  hauss.geneland.path.mcmc.adm <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>paste</b>(hauss.geneland.path.mcmc, <span style='color:#bf0303;'>&quot;admixture&quot;</span>, <span style='color:#bf0303;'>&quot;/&quot;</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;&quot;</span>)
  <i><span style='color:#755019;'># On Windows, remove following line of code and create in each result directory (from 1 to max K) new subdirectory &quot;admixture&quot; (creating subdirs in Windows is complicated)  system(paste(&quot;mkdir &quot;, hauss.geneland.path.mcmc.adm))</span></i>
  <b>HZ</b>(<span style='color:#e85290;'>coordinates=</span>hauss.geneland.coord.utm, <span style='color:#e85290;'>geno.dip.codom=</span>hauss.geneland.data, <span style='color:#e85290;'>path.mcmc.noadm=</span>hauss.geneland.path.mcmc, <span style='color:#e85290;'>nit=</span><span style='color:#2b74c7;'>10000</span>, <span style='color:#e85290;'>thinning=</span><span style='color:#2b74c7;'>10</span>, <span style='color:#e85290;'>path.mcmc.adm=</span>hauss.geneland.path.mcmc.adm)
  }
<i><span style='color:#755019;'># Produce maps of respective inferred clusters  </span></i>
<span style='color:#0000bf;'>for</span> (hauss.geneland.irun <span style='color:#0000bf;'>in</span> <span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span>hauss.geneland.nrun) {
  hauss.geneland.path.mcmc <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>paste</b>(<span style='color:#bf0303;'>&quot;geneland/&quot;</span>, hauss.geneland.irun, <span style='color:#bf0303;'>&quot;/&quot;</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;&quot;</span>)
  <i><span style='color:#755019;'># Maps - tesselations</span></i>
  <b>PlotTessellation</b>(<span style='color:#e85290;'>coordinates=</span>hauss.geneland.coord.utm, <span style='color:#e85290;'>path.mcmc=</span>hauss.geneland.path.mcmc, <span style='color:#e85290;'>printit=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>path=</span>hauss.geneland.path.mcmc)
  <span style='color:#0000bf;'>for</span> (hauss.geneland.irun.img <span style='color:#0000bf;'>in</span> <span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span>hauss.geneland.maxpop) { <b>dev.off</b>() } <i><span style='color:#755019;'># We must close graphical device manually</span></i>
  }
<i><span style='color:#755019;'># Calculate average posterior probability</span></i>
hauss.geneland.lpd <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>rep</b>(<span style='color:#00892b;'>NA</span>, hauss.geneland.nrun)
<span style='color:#0000bf;'>for</span> (hauss.geneland.irun <span style='color:#0000bf;'>in</span> <span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span>hauss.geneland.nrun) {
  hauss.geneland.path.mcmc <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>paste</b>(<span style='color:#bf0303;'>&quot;geneland/&quot;</span>, hauss.geneland.irun, <span style='color:#bf0303;'>&quot;/&quot;</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;&quot;</span>)
  hauss.geneland.path.lpd <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>paste</b>(hauss.geneland.path.mcmc,<span style='color:#bf0303;'>&quot;log.posterior.density.txt&quot;</span>,<span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;&quot;</span>)
  hauss.geneland.lpd[hauss.geneland.irun] <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>mean</b>(<b>scan</b>(hauss.geneland.path.lpd)[<span style='color:#803f00;'>-</span>(<span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span>hauss.geneland.burnin)])
  }
<i><span style='color:#755019;'># Sorts runs according to decreasing posterior probability - the first one is the best</span></i>
<b>order</b>(hauss.geneland.lpd, <span style='color:#e85290;'>decreasing=</span><span style='color:#00892b;'>TRUE</span>)
hauss.geneland.lpd

<b><span style='color:#b00000;'>## Maps</span></b>

<i><span style='color:#755019;'># Load libraries</span></i>
<b>library</b>(sp)
<b>library</b>(rworldmap) <i><span style='color:#755019;'># Basic world maps</span></i>
<b>library</b>(TeachingDemos) <i><span style='color:#755019;'># To be able to move text little bit</span></i>
<b>library</b>(RgoogleMaps) <i><span style='color:#755019;'># Google and OpenStreetMaps</span></i>
<i><span style='color:#755019;'># Plot basic map with state boundaries within selected range</span></i>
<b>plot</b>(<span style='color:#e85290;'>x=</span><b>getMap</b>(<span style='color:#e85290;'>resolution=</span><span style='color:#bf0303;'>&quot;high&quot;</span>), <span style='color:#e85290;'>xlim=</span><b>c</b>(<span style='color:#2b74c7;'>19</span>, <span style='color:#2b74c7;'>24</span>), <span style='color:#e85290;'>ylim=</span><b>c</b>(<span style='color:#2b74c7;'>39</span>, <span style='color:#2b74c7;'>44</span>), <span style='color:#e85290;'>asp=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>lwd=</span><span style='color:#2b74c7;'>1.5</span>)
<b>box</b>() <i><span style='color:#755019;'># Add frame around the map</span></i>
<i><span style='color:#755019;'># Plot location points</span></i>
<b>points</b>(<span style='color:#e85290;'>x=</span>hauss.genpop<span style='color:#803f00;'>@</span>other<span style='color:#803f00;'>$</span>xy[[<span style='color:#bf0303;'>&quot;lon&quot;</span>]], <span style='color:#e85290;'>y=</span>hauss.genpop<span style='color:#803f00;'>@</span>other<span style='color:#803f00;'>$</span>xy[[<span style='color:#bf0303;'>&quot;lat&quot;</span>]], <span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>15</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>19</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;red&quot;</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>4</span>)
<i><span style='color:#755019;'># Add text descriptions for points. Text is aside and with background</span></i>
<b>shadowtext</b>(<span style='color:#e85290;'>x=</span>hauss.genpop<span style='color:#803f00;'>@</span>other<span style='color:#803f00;'>$</span>xy[[<span style='color:#bf0303;'>&quot;lon&quot;</span>]], <span style='color:#e85290;'>y=</span>hauss.genpop<span style='color:#803f00;'>@</span>other<span style='color:#803f00;'>$</span>xy[[<span style='color:#bf0303;'>&quot;lat&quot;</span>]], <span style='color:#e85290;'>labels=</span><b>as.vector</b>(hauss.genind<span style='color:#803f00;'>@</span>pop.names), <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;black&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;white&quot;</span>, <span style='color:#e85290;'>theta=</span><b>seq</b>(pi<span style='color:#803f00;'>/</span><span style='color:#2b74c7;'>4</span>, <span style='color:#2b74c7;'>2</span><span style='color:#803f00;'>*</span>pi, <span style='color:#e85290;'>length.out=</span><span style='color:#2b74c7;'>8</span>), <span style='color:#e85290;'>r=</span><span style='color:#2b74c7;'>0.15</span>, <span style='color:#e85290;'>pos=</span><b>c</b>(<span style='color:#2b74c7;'>1</span>, <span style='color:#2b74c7;'>3</span>, <span style='color:#2b74c7;'>2</span>, <span style='color:#2b74c7;'>4</span>, <span style='color:#2b74c7;'>4</span>), <span style='color:#e85290;'>offset=</span><span style='color:#2b74c7;'>0.75</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.5</span>)
<i><span style='color:#755019;'># Insert legend</span></i>
<b>legend</b>(<span style='color:#e85290;'>x=</span><span style='color:#bf0303;'>&quot;topright&quot;</span>, <span style='color:#e85290;'>inset=</span><span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>/</span><span style='color:#2b74c7;'>50</span>, <span style='color:#e85290;'>legend=</span><b>c</b>(<span style='color:#bf0303;'>&quot;He&quot;</span>, <span style='color:#bf0303;'>&quot;Oh&quot;</span>, <span style='color:#bf0303;'>&quot;Pr&quot;</span>, <span style='color:#bf0303;'>&quot;Ne&quot;</span>, <span style='color:#bf0303;'>&quot;Sk&quot;</span>), <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;red&quot;</span>, <span style='color:#e85290;'>border=</span><span style='color:#bf0303;'>&quot;black&quot;</span>, <span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>15</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>19</span>, <span style='color:#e85290;'>pt.cex=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>bty=</span><span style='color:#bf0303;'>&quot;o&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;lightgrey&quot;</span>, <span style='color:#e85290;'>box.lwd=</span><span style='color:#2b74c7;'>1.5</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.5</span>, <span style='color:#e85290;'>title=</span><span style='color:#bf0303;'>&quot;Populations&quot;</span>)

<i><span style='color:#755019;'># Google map is produced into a file. Parametr markers contain data frame with coordinates and possibly with another information</span></i>
hauss.gmap <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>GetMap</b>(<span style='color:#e85290;'>center=</span><b>c</b>(<span style='color:#e85290;'>lat=</span><span style='color:#2b74c7;'>41</span>, <span style='color:#e85290;'>lon=</span><span style='color:#2b74c7;'>21</span>), <span style='color:#e85290;'>size=</span><b>c</b>(<span style='color:#2b74c7;'>640</span>, <span style='color:#2b74c7;'>640</span>), <span style='color:#e85290;'>destfile=</span><span style='color:#bf0303;'>&quot;gmap.png&quot;</span>, <span style='color:#e85290;'>zoom=</span><span style='color:#2b74c7;'>8</span>, <span style='color:#e85290;'>markers=</span>hauss.coord, <span style='color:#e85290;'>maptype=</span><span style='color:#bf0303;'>&quot;terrain&quot;</span>)

<i><span style='color:#755019;'># Plot on OpenStreeMap - server is commonly overloaded and doesn't respond correctly</span></i>
<b>GetMap.OSM</b>(<span style='color:#e85290;'>lonR=</span><b>c</b>(<span style='color:#2b74c7;'>18</span>, <span style='color:#2b74c7;'>24</span>), <span style='color:#e85290;'>latR=</span><b>c</b>(<span style='color:#2b74c7;'>39</span>, <span style='color:#2b74c7;'>44</span>), <span style='color:#e85290;'>scale=</span><span style='color:#2b74c7;'>200000</span>, <span style='color:#e85290;'>destfile=</span><span style='color:#bf0303;'>&quot;osmmap.png&quot;</span>, <span style='color:#e85290;'>format=</span><span style='color:#bf0303;'>&quot;png&quot;</span>, <span style='color:#e85290;'>RETURNIMAGE=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>GRAYSCALE=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>NEWMAP=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>verbose=</span><span style='color:#2b74c7;'>1</span>)

<b>library</b>(maps) <i><span style='color:#755019;'># Various maping tools (plotting, ...)</span></i>
<b>library</b>(mapdata) <i><span style='color:#755019;'># More detailed maps, but political boundaries often outdatet, see http://cran.r-project.org/web/packages/mapdata/</span></i>
<b>library</b>(mapproj) <i><span style='color:#755019;'># Converts latitude/longitude into projected coordinates</span></i>
<i><span style='color:#755019;'># Plot a map, check parameters, among others projection and ?mapproject for its details</span></i>
<b>map</b>(<span style='color:#e85290;'>database=</span><span style='color:#bf0303;'>&quot;worldHires&quot;</span>, <span style='color:#e85290;'>boundary=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>interior=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>fill=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;lightgrey&quot;</span>, <span style='color:#e85290;'>plot=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>xlim=</span><b>c</b>(<span style='color:#2b74c7;'>16</span>, <span style='color:#2b74c7;'>27</span>), <span style='color:#e85290;'>ylim=</span><b>c</b>(<span style='color:#2b74c7;'>37</span>, <span style='color:#2b74c7;'>46</span>))
<i><span style='color:#755019;'># If you'd use projection, use - mapproject() to convert also coordinates! See ?mapproject for details</span></i>
<b>points</b>(<span style='color:#e85290;'>x=</span>hauss.genpop<span style='color:#803f00;'>@</span>other<span style='color:#803f00;'>$</span>xy[[<span style='color:#bf0303;'>&quot;lon&quot;</span>]], <span style='color:#e85290;'>y=</span>hauss.genpop<span style='color:#803f00;'>@</span>other<span style='color:#803f00;'>$</span>xy[[<span style='color:#bf0303;'>&quot;lat&quot;</span>]], <span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>15</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>19</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;red&quot;</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>3</span>)

<i><span style='color:#755019;'># Plotting on SHP files</span></i>
<b>library</b>(maptools)
<i><span style='color:#755019;'># Load SHP file</span></i>
<i><span style='color:#755019;'># Data from http://download.geofabrik.de/europe/macedonia.html</span></i>
<i><span style='color:#755019;'># Directory has to contain also respective DBF and SHX files (same name, only different extension)</span></i>
<i><span style='color:#755019;'># Get from http://trapa.cz/sites/default/rcourse/macedonia.zip</span></i>
<i><span style='color:#755019;'># There are several functions readShape* - select appropriate according to data stored in respective SHP file</span></i>
macedonia_building <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>readShapeLines</b>(<span style='color:#e85290;'>fn=</span><span style='color:#bf0303;'>&quot;macedonia_buildings.shp&quot;</span>)
<b>plot</b>(macedonia_building)
macedonia_landuse <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>readShapeLines</b>(<span style='color:#e85290;'>fn=</span><span style='color:#bf0303;'>&quot;macedonia_landuse.shp&quot;</span>)
<b>plot</b>(macedonia_landuse)
macedonia_natural <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>readShapeLines</b>(<span style='color:#e85290;'>fn=</span><span style='color:#bf0303;'>&quot;macedonia_natural.shp&quot;</span>)
<b>plot</b>(macedonia_natural)
macedonia_railways <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>readShapeLines</b>(<span style='color:#e85290;'>fn=</span><span style='color:#bf0303;'>&quot;macedonia_railways.shp&quot;</span>)
<b>plot</b>(macedonia_railways)
macedonia_roads <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>readShapeLines</b>(<span style='color:#e85290;'>fn=</span><span style='color:#bf0303;'>&quot;macedonia_roads.shp&quot;</span>)
<b>plot</b>(macedonia_roads)
macedonia_waterways <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>readShapeLines</b>(<span style='color:#e85290;'>fn=</span><span style='color:#bf0303;'>&quot;macedonia_waterways.shp&quot;</span>)
<b>plot</b>(macedonia_waterways)
<i><span style='color:#755019;'># Plot all layers into single image</span></i>
<b>plot</b>(macedonia_building)
<b>plot</b>(macedonia_landuse, <span style='color:#e85290;'>add=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;darkgreen&quot;</span>, <span style='color:#e85290;'>fill=</span><span style='color:#00892b;'>TRUE</span>)
<b>plot</b>(macedonia_natural, <span style='color:#e85290;'>add=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;green&quot;</span>, <span style='color:#e85290;'>fill=</span><span style='color:#00892b;'>TRUE</span>)
<b>plot</b>(macedonia_railways, <span style='color:#e85290;'>add=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;brown&quot;</span>, <span style='color:#e85290;'>lty=</span><span style='color:#bf0303;'>&quot;dotted&quot;</span>)
<b>plot</b>(macedonia_roads, <span style='color:#e85290;'>add=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;orange&quot;</span>)
<b>plot</b>(macedonia_waterways, <span style='color:#e85290;'>add=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;blue&quot;</span>, <span style='color:#e85290;'>lwd=</span><span style='color:#2b74c7;'>2</span>)
<i><span style='color:#755019;'># Add sampling points</span></i>
<b>points</b>(<span style='color:#e85290;'>x=</span>hauss.genpop<span style='color:#803f00;'>@</span>other<span style='color:#803f00;'>$</span>xy[[<span style='color:#bf0303;'>&quot;lon&quot;</span>]], <span style='color:#e85290;'>y=</span>hauss.genpop<span style='color:#803f00;'>@</span>other<span style='color:#803f00;'>$</span>xy[[<span style='color:#bf0303;'>&quot;lat&quot;</span>]], <span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>15</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>19</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;red&quot;</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>4</span>)
<i><span style='color:#755019;'># Add description of psampling points</span></i>
<b>shadowtext</b>(<span style='color:#e85290;'>x=</span>hauss.genpop<span style='color:#803f00;'>@</span>other<span style='color:#803f00;'>$</span>xy[[<span style='color:#bf0303;'>&quot;lon&quot;</span>]], <span style='color:#e85290;'>y=</span>hauss.genpop<span style='color:#803f00;'>@</span>other<span style='color:#803f00;'>$</span>xy[[<span style='color:#bf0303;'>&quot;lat&quot;</span>]], <span style='color:#e85290;'>labels=</span><b>as.vector</b>(hauss.genind<span style='color:#803f00;'>@</span>pop.names), <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;black&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;white&quot;</span>, <span style='color:#e85290;'>theta=</span><b>seq</b>(pi<span style='color:#803f00;'>/</span><span style='color:#2b74c7;'>4</span>, <span style='color:#2b74c7;'>2</span><span style='color:#803f00;'>*</span>pi, <span style='color:#e85290;'>length.out=</span><span style='color:#2b74c7;'>8</span>), <span style='color:#e85290;'>r=</span><span style='color:#2b74c7;'>0.15</span>, <span style='color:#e85290;'>pos=</span><b>c</b>(<span style='color:#2b74c7;'>1</span>, <span style='color:#2b74c7;'>3</span>, <span style='color:#2b74c7;'>2</span>, <span style='color:#2b74c7;'>4</span>, <span style='color:#2b74c7;'>4</span>), <span style='color:#e85290;'>offset=</span><span style='color:#2b74c7;'>0.75</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.5</span>)
<i><span style='color:#755019;'># Add legend</span></i>
<b>legend</b>(<span style='color:#e85290;'>x=</span><span style='color:#bf0303;'>&quot;topright&quot;</span>, <span style='color:#e85290;'>inset=</span><span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>/</span><span style='color:#2b74c7;'>50</span>, <span style='color:#e85290;'>legend=</span><b>c</b>(<span style='color:#bf0303;'>&quot;He&quot;</span>, <span style='color:#bf0303;'>&quot;Oh&quot;</span>, <span style='color:#bf0303;'>&quot;Pr&quot;</span>, <span style='color:#bf0303;'>&quot;Ne&quot;</span>, <span style='color:#bf0303;'>&quot;Sk&quot;</span>), <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;red&quot;</span>, <span style='color:#e85290;'>border=</span><span style='color:#bf0303;'>&quot;black&quot;</span>, <span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>15</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>19</span>, <span style='color:#e85290;'>pt.cex=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>bty=</span><span style='color:#bf0303;'>&quot;o&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;lightgrey&quot;</span>, <span style='color:#e85290;'>box.lwd=</span><span style='color:#2b74c7;'>1.5</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.5</span>, <span style='color:#e85290;'>title=</span><span style='color:#bf0303;'>&quot;Populations&quot;</span>)

<b><span style='color:#b00000;'>## Structure</span></b>

<i><span style='color:#755019;'># Inspiration: http://www.molecularecologist.com/913/09/using-r-to-run-parallel-analyses-of-population-genetic-data-in-structure-parallelstructure/</span></i>
<i><span style='color:#755019;'># Install ParallelStructure, see https://r-forge.r-project.org/R/?group_id=1636 and http://www.plosone.org/article/info%3Adoi%2F10.1371%2Fjournal.pone.0070651</span></i>
<i><span style='color:#755019;'># get input data from https://trapa.cz/sites/default/rcourse/hauss_stru.in and joblist https://trapa.cz/sites/default/rcourse/joblist.txt</span></i>
<i><span style='color:#755019;'># Set working directory</span></i>
<b>setwd</b>(<span style='color:#bf0303;'>&quot;~/dokumenty/fakulta/vyuka/r_mol_data/priklady/structure/&quot;</span>)
<b>getwd</b>()
<i><span style='color:#755019;'># Load ParallelStructure package</span></i>
<b>library</b>(ParallelStructure)
<i><span style='color:#755019;'># See manual for the R package and Structure http://pritchardlab.stanford.edu/structure.html</span></i>
?parallel_structure
<i><span style='color:#755019;'># Run Structure</span></i>
<b>parallel_structure</b>(<span style='color:#e85290;'>joblist=</span><span style='color:#bf0303;'>&quot;joblist.txt&quot;</span>, <span style='color:#e85290;'>n_cpu=</span><span style='color:#2b74c7;'>7</span>, <span style='color:#e85290;'>structure_path=</span><span style='color:#bf0303;'>&quot;~/bin/&quot;</span>, <span style='color:#e85290;'>infile=</span><span style='color:#bf0303;'>&quot;hauss_stru.in&quot;</span>, <span style='color:#e85290;'>outpath=</span><span style='color:#bf0303;'>&quot;results/&quot;</span>, <span style='color:#e85290;'>numinds=</span><span style='color:#2b74c7;'>47</span>, <span style='color:#e85290;'>numloci=</span><span style='color:#2b74c7;'>12</span>, <span style='color:#e85290;'>plot_output=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>label=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>popdata=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>popflag=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>phenotypes=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>markernames=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>mapdist=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>onerowperind=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>phaseinfo=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>extracol=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>missing=</span><span style='color:#803f00;'>-</span><span style='color:#2b74c7;'>9</span>, <span style='color:#e85290;'>ploidy=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>usepopinfo=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>revert_convert=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>printqhat=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>locdata=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>recessivealleles=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>phased=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>noadmix=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>linkage=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>locprior=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>inferalpha=</span><span style='color:#2b74c7;'>1</span>)

<i><span style='color:#755019;'># When running on Windows, parallel support may be missing - install Rmpi library required by ParallelStructure for parallelisation on Windows</span></i>
<b>install.packages</b>(<span style='color:#bf0303;'>&quot;Rmpi&quot;</span>)
<b>library</b>(Rmpi)
<i><span style='color:#755019;'># Instead of parallel_structure() use MPI_structure() with same arguments</span></i>
<b>MPI_structure</b>(...) <i><span style='color:#755019;'># Same arguments as on previous slide</span></i>
<i><span style='color:#755019;'># If this fails, look for some UNIX machine...</span></i>

<i><span style='color:#755019;'># Postprocess results with Structure sum R script by Dorothee Ehrich</span></i>
<b>source</b>(<span style='color:#bf0303;'>&quot;http://trapa.cz/sites/default/rcourse/structure-sum-2009.r&quot;</span>)
<i><span style='color:#755019;'># Create new directory with result files results_job_*_f and set working directory accordingly</span></i>
<b>setwd</b>(<span style='color:#bf0303;'>&quot;/home/vojta/dokumenty/fakulta/vyuka/r_mol_data/priklady/structure/structure_sum/&quot;</span>)
<b>dir</b>()
<i><span style='color:#755019;'># Prepare file list_k.txt containing on each line K and name of output &quot;_f&quot; file - get it from https://trapa.cz/sites/default/rcourse/list_k.txt</span></i>
<i><span style='color:#755019;'># See documentation for details. Functions take as an argument list_k file and number of populations</span></i>
<b>Structure.table</b>(<span style='color:#bf0303;'>&quot;list_k.txt&quot;</span>, <span style='color:#2b74c7;'>5</span>)
<b>Structure.simil</b>(<span style='color:#bf0303;'>&quot;list_k.txt&quot;</span>, <span style='color:#2b74c7;'>5</span>)
<b>Structure.deltaK</b>(<span style='color:#bf0303;'>&quot;list_k.txt&quot;</span>, <span style='color:#2b74c7;'>5</span>)
<b>graphics.off</b>() <i><span style='color:#755019;'># Close graphics</span></i>
<b>Structure.cluster</b>(<span style='color:#bf0303;'>&quot;list_k.txt&quot;</span>, <span style='color:#2b74c7;'>5</span>)
<i><span style='color:#755019;'># Reordering (&quot;alignment&quot;) of runs to get same clusters in same columns (prepare respective list_k files - one for each K)</span></i>
<b>Structure.order</b>(<span style='color:#bf0303;'>&quot;list_k_02.txt&quot;</span>, <span style='color:#2b74c7;'>5</span>)
<b>Structure.order</b>(<span style='color:#bf0303;'>&quot;list_k_03.txt&quot;</span>, <span style='color:#2b74c7;'>5</span>)
<b>Structure.order</b>(<span style='color:#bf0303;'>&quot;list_k_04.txt&quot;</span>, <span style='color:#2b74c7;'>5</span>)
<b>Structure.order</b>(<span style='color:#bf0303;'>&quot;list_k_05.txt&quot;</span>, <span style='color:#2b74c7;'>5</span>)
<b>Structure.order</b>(<span style='color:#bf0303;'>&quot;list_k_06.txt&quot;</span>, <span style='color:#2b74c7;'>5</span>)
<b>Structure.order</b>(<span style='color:#bf0303;'>&quot;list_k_07.txt&quot;</span>, <span style='color:#2b74c7;'>5</span>)
<i><span style='color:#755019;'># Continue with CLUMPP and distruct</span></i>
<i><span style='color:#755019;'># Details: https://trapa.cz/en/structure-r-linux</span></i>

<b><span style='color:#b00000;'>## Multiple sequence alignment</span></b>

<b>library</b>(muscle) <i><span style='color:#755019;'># Muscle has its own functions read.fasta and write.fasta</span></i>
<i><span style='color:#755019;'># Reads local file and writes R object or external file with alignment, you can use any command line argument available for muscle, see ?muscle and its manual</span></i>
<b>muscle</b>(<span style='color:#e85290;'>seqs=</span><span style='color:#bf0303;'>&quot;usflu.fasta&quot;</span>, <span style='color:#e85290;'>out=</span><span style='color:#bf0303;'>&quot;usflu.aln&quot;</span>, <span style='color:#e85290;'>quiet=</span><span style='color:#00892b;'>FALSE</span>) <i><span style='color:#755019;'># Write result to external file</span></i>
usflu.muscle <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>muscle</b>(<span style='color:#e85290;'>seqs=</span><span style='color:#bf0303;'>&quot;usflu.fasta&quot;</span>, <span style='color:#e85290;'>out=</span><span style='color:#00892b;'>NULL</span>, <span style='color:#e85290;'>quiet=</span><span style='color:#00892b;'>FALSE</span>) <i><span style='color:#755019;'># Write result to R object</span></i>
<i><span style='color:#755019;'># Prints the alignment</span></i>
<b>print</b>(usflu.muscle, <span style='color:#e85290;'>from=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>to=</span>usflu.muscle<span style='color:#803f00;'>$</span>length)
<b>class</b>(usflu.muscle)

<i><span style='color:#755019;'># Libraries </span></i>
<b>library</b>(colorspace)
<b>library</b>(XML)
<b>library</b>(phyloch) <i><span style='color:#755019;'># Alignment with mafft, you can also try package ips</span></i>
<i><span style='color:#755019;'># Requires mafft binary on specific location - might be needed to copy it or make symlink</span></i>
<i><span style='color:#755019;'># read ?mafft and mafft's documentation</span></i>
meles.mafft <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>mafft</b>(<span style='color:#e85290;'>x=</span>meles.dna, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;localpair&quot;</span>, <span style='color:#e85290;'>maxiterate=</span><span style='color:#2b74c7;'>100</span>)
meles.mafft
<b>class</b>(meles.mafft)
<i><span style='color:#755019;'># Multiple sequence alignments using clustal, muscle and t-coffee are available in package ape</span></i>
<i><span style='color:#755019;'># read ?clustal and documentation of Clustal and Muscle to set correct parameters</span></i>
<b>clustal</b>(<span style='color:#e85290;'>x=</span>meles.dna, <span style='color:#e85290;'>pw.gapopen=</span><span style='color:#2b74c7;'>10</span>, <span style='color:#e85290;'>pw.gapext=</span><span style='color:#2b74c7;'>0.1</span>, <span style='color:#e85290;'>gapopen=</span><span style='color:#2b74c7;'>10</span>, <span style='color:#e85290;'>gapext=</span><span style='color:#2b74c7;'>0.2</span>, <span style='color:#e85290;'>quiet=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>original.ordering=</span><span style='color:#00892b;'>TRUE</span>)
meles.muscle <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span>ape<span style='color:#803f00;'>::</span><b>muscle</b>(<span style='color:#e85290;'>x=</span>meles.dna, <span style='color:#e85290;'>exec=</span><span style='color:#bf0303;'>&quot;muscle&quot;</span>, <span style='color:#e85290;'>quiet=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>original.ordering=</span><span style='color:#00892b;'>TRUE</span>)
meles.muscle
<b>class</b>(meles.muscle)
<i><span style='color:#755019;'># Plot the alignment - you can select which bases to plot and/or modify colours</span></i>
<b>image</b>(<span style='color:#e85290;'>x=</span>meles.muscle, <b>c</b>(<span style='color:#bf0303;'>&quot;a&quot;</span>, <span style='color:#bf0303;'>&quot;t&quot;</span>, <span style='color:#bf0303;'>&quot;c&quot;</span> ,<span style='color:#bf0303;'>&quot;g&quot;</span>, <span style='color:#bf0303;'>&quot;n&quot;</span>), <span style='color:#e85290;'>col=</span><b>rainbow</b>(<span style='color:#2b74c7;'>5</span>))
<i><span style='color:#755019;'># Add grey dotted grid</span></i>
<b>grid</b>(<span style='color:#e85290;'>nx=</span><b>ncol</b>(meles.muscle), <span style='color:#e85290;'>ny=</span><b>nrow</b>(meles.muscle), <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;lightgrey&quot;</span>)
<i><span style='color:#755019;'># Remove gaps from alignment - destroy it</span></i>
meles.nogaps <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>del.gaps</b>(meles.muscle)

<b><span style='color:#b00000;'>## Tree manipulations</span></b>

<i><span style='color:#755019;'># Read trees in NEWICK format - single or multiple tree(s)</span></i>
oxalis.trees <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.tree</b>(<span style='color:#bf0303;'>&quot;http://trapa.cz/sites/default/rcourse/oxalis.nwk&quot;</span>)
<b>summary</b>(oxalis.trees)
<b>length</b>(oxalis.trees)
<b>names</b>(oxalis.trees)
<i><span style='color:#755019;'># Export trees in NEWICK format</span></i>
<b>write.tree</b>(<span style='color:#e85290;'>phy=</span>oxalis.trees, <span style='color:#e85290;'>file=</span><span style='color:#bf0303;'>&quot;trees.nwk&quot;</span>)

<i><span style='color:#755019;'># Drop a tip</span></i>
<b>plot.phylo</b>(hauss.nj)
hauss.nj.drop <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>drop.tip</b>(<span style='color:#e85290;'>phy=</span>hauss.nj, <span style='color:#e85290;'>tip=</span><span style='color:#2b74c7;'>47</span>)
<b>plot.phylo</b>(hauss.nj.drop)

<i><span style='color:#755019;'># Drop a tip from multiPhylo</span></i>
<b>plot.multiPhylo</b>(<span style='color:#e85290;'>x=</span>oxalis.trees)
oxalis.trees.drop <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>lapply</b>(<span style='color:#e85290;'>X=</span>oxalis.trees, <span style='color:#e85290;'>FUN=</span>drop.tip, <span style='color:#bf0303;'>&quot;1&quot;</span>)
<b>class</b>(oxalis.trees.drop) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> &quot;multiPhylo&quot;</span>
<b>plot.multiPhylo</b>(<span style='color:#e85290;'>x=</span>oxalis.trees.drop)

<i><span style='color:#755019;'># Interactively extract tree</span></i>
<i><span style='color:#755019;'># Plot source tree</span></i>
<b>plot.phylo</b>(hauss.nj)
<i><span style='color:#755019;'># Select clade to extract by clicking on it</span></i>
hauss.nj.extracted <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>extract.clade</b>(<span style='color:#e85290;'>phy=</span>hauss.nj, <span style='color:#e85290;'>interactive=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># See new extracted tree</span></i>
<b>plot.phylo</b>(hauss.nj.extracted)
<i><span style='color:#755019;'># Non-interactively extract tree</span></i>
hauss.nj.extracted <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>extract.clade</b>(<span style='color:#e85290;'>phy=</span>hauss.nj, <span style='color:#e85290;'>node=</span><span style='color:#2b74c7;'>60</span>, <span style='color:#e85290;'>interactive=</span><span style='color:#00892b;'>FALSE</span>)
<i><span style='color:#755019;'># See new extracted tree</span></i>
<b>plot.phylo</b>(hauss.nj.extracted)

<i><span style='color:#755019;'># Drop &quot;extinct&quot; tips - those who don't reach end the tree</span></i>
<i><span style='color:#755019;'># tolerance is respective to the used metrics</span></i>
<b>plot.phylo</b>(hauss.nj)
<b>axisPhylo</b>()
hauss.nj.fossil <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>drop.fossil</b>(<span style='color:#e85290;'>phy=</span>hauss.nj, <span style='color:#e85290;'>tol=</span><span style='color:#2b74c7;'>0.25</span>)
<b>plot.phylo</b>(hauss.nj.fossil)

<i><span style='color:#755019;'># Bind two trees into one</span></i>
hauss.nj.bind <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>bind.tree</b>(<span style='color:#e85290;'>x=</span>hauss.nj.fossil, <span style='color:#e85290;'>y=</span>hauss.nj.extracted, <span style='color:#e85290;'>where=</span><span style='color:#bf0303;'>&quot;root&quot;</span>, <span style='color:#e85290;'>position=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>interactive=</span><span style='color:#00892b;'>FALSE</span>)
<b>plot.phylo</b>(hauss.nj.bind)
<i><span style='color:#755019;'># Bind two trees interactively</span></i>
<i><span style='color:#755019;'># Plot tree receiving the new one</span></i>
<b>plot.phylo</b>(hauss.nj.fossil)
<i><span style='color:#755019;'># Select where to bind new tree to</span></i>
hauss.nj.bind <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>bind.tree</b>(<span style='color:#e85290;'>x=</span>hauss.nj.fossil, <span style='color:#e85290;'>y=</span>hauss.nj.extracted, <span style='color:#e85290;'>interactive=</span><span style='color:#00892b;'>TRUE</span>)
<b>plot.phylo</b>(hauss.nj.bind)

<i><span style='color:#755019;'># Rotate tree</span></i>
<b>plot.phylo</b>(hauss.nj)
hauss.nj.rotated <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>rotate</b>(<span style='color:#e85290;'>phy=</span>hauss.nj, <span style='color:#e85290;'>node=</span><span style='color:#bf0303;'>&quot;70&quot;</span>)
<b>plot.phylo</b>(hauss.nj.rotated)

<i><span style='color:#755019;'># Ladderize the tree</span></i>
<b>plot.phylo</b>(hauss.nj)
hauss.nj.ladderized <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>ladderize</b>(hauss.nj)
<b>plot.phylo</b>(hauss.nj.ladderized)

<i><span style='color:#755019;'># Root the tree</span></i>
<b>plot.phylo</b>(hauss.nj)
<b>print.phylo</b>(hauss.nj)
hauss.nj.rooted <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>root</b>(<span style='color:#e85290;'>phy=</span>hauss.nj, <span style='color:#e85290;'>outgroup=</span><span style='color:#bf0303;'>&quot;10&quot;</span>)
<b>print.phylo</b>(hauss.nj.rooted)
<b>plot.phylo</b>(hauss.nj.rooted)
<i><span style='color:#755019;'># Root the tree interactive</span></i>
<b>plot.phylo</b>(hauss.nj)
hauss.nj.rooted <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>root</b>(<span style='color:#e85290;'>phy=</span>hauss.nj, <span style='color:#e85290;'>interactive=</span><span style='color:#00892b;'>TRUE</span>)
<b>plot.phylo</b>(hauss.nj.rooted)
<i><span style='color:#755019;'># unroot the tree</span></i>
<b>unroot</b>()
<i><span style='color:#755019;'># Check if it is rooted</span></i>
<b>is.rooted</b>()

<i><span style='color:#755019;'># Check if the tree is ultrametric - is variance of distances of all tips to node 0? It is required for some analysis</span></i>
<b>is.ultrametric</b>()
<i><span style='color:#755019;'># Make tree ultrametric</span></i>
<b>chronos</b>()
?chronos <i><span style='color:#755019;'># Check it for mode how to calculate the lengths</span></i>

<i><span style='color:#755019;'># Compute branch lengths for trees withou branch lengths</span></i>
<b>compute.brlen</b>()
?compute.brlen <i><span style='color:#755019;'># Check it for mode how to calculate the lengths</span></i>

<i><span style='color:#755019;'># Computes the branch lengths of a tree giving its branching times (aka node ages or heights)</span></i>
<b>compute.brtime</b>()
?compute.brtima <i><span style='color:#755019;'># Check it for mode how to calculate the lengths</span></i>

<b><span style='color:#b00000;'>## Topographical distances among trees</span></b>

<b>library</b>(gplots)
<b>library</b>(corrplot)
<b>library</b>(phytools)

<i><span style='color:#755019;'># Prepare matrix for distances</span></i>
oxalis.trees.d <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>matrix</b>(<span style='color:#e85290;'>nrow=</span><b>length</b>(oxalis.trees), <span style='color:#e85290;'>ncol=</span><b>length</b>(oxalis.trees))

<i><span style='color:#755019;'># Calculate pairwise topographic distances</span></i>
<span style='color:#0000bf;'>for</span> (i <span style='color:#0000bf;'>in</span> <span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span><b>length</b>(oxalis.trees)) {
  <span style='color:#0000bf;'>for</span> (j <span style='color:#0000bf;'>in</span> i<span style='color:#803f00;'>:</span><b>length</b>(oxalis.trees)) {
    <b>print</b>(<b>c</b>(i,j))
    oxalis.trees.d[i,j] <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>dist.topo</b>(oxalis.trees[[i]], oxalis.trees[[j]])
   }
 }

<i><span style='color:#755019;'># Basic information about the distance matrix</span></i>
<b>dim</b>(oxalis.trees.d)
<b>head.matrix</b>(oxalis.trees.d)

<i><span style='color:#755019;'># Add names of columns and rows</span></i>
<b>colnames</b>(oxalis.trees.d) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>names</b>(oxalis.trees)
<b>rownames</b>(oxalis.trees.d) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>names</b>(oxalis.trees)

<i><span style='color:#755019;'># Make matrix symetric</span></i>
oxalis.trees.d[<b>lower.tri</b>(oxalis.trees.d)]<b><span style='color:#336366;'>=</span></b><b>t</b>(oxalis.trees.d)[<b>lower.tri</b>(oxalis.trees.d)]

<i><span style='color:#755019;'># Create heatmaps using heatmap.2 function from gplots package</span></i>
<b>heatmap.2</b>(<span style='color:#e85290;'>x=</span>oxalis.trees.d, <span style='color:#e85290;'>Rowv=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>Colv=</span><span style='color:#bf0303;'>&quot;Rowv&quot;</span>, <span style='color:#e85290;'>dendrogram=</span><span style='color:#bf0303;'>&quot;none&quot;</span>, <span style='color:#e85290;'>symm=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>scale=</span><span style='color:#bf0303;'>&quot;none&quot;</span>, <span style='color:#e85290;'>na.rm=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>revC=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>col=</span><b>rainbow</b>(<span style='color:#2b74c7;'>15</span>), <span style='color:#e85290;'>cellnote=</span>oxalis.trees.d, <span style='color:#e85290;'>notecex=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>notecol=</span><span style='color:#bf0303;'>&quot;white&quot;</span>, <span style='color:#e85290;'>trace=</span><span style='color:#bf0303;'>&quot;row&quot;</span>, <span style='color:#e85290;'>linecol=</span><span style='color:#bf0303;'>&quot;black&quot;</span>, <span style='color:#e85290;'>labRow=</span><b>names</b>(oxalis.trees), <span style='color:#e85290;'>labCol=</span><b>names</b>(oxalis.trees), <span style='color:#e85290;'>key=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>keysize=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>density.info=</span><span style='color:#bf0303;'>&quot;density&quot;</span>, <span style='color:#e85290;'>symkey=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>main=</span><span style='color:#bf0303;'>&quot;Correlation matrix of topographical distances&quot;</span>, <span style='color:#e85290;'>xlab=</span><b>names</b>(oxalis.trees), <span style='color:#e85290;'>ylab=</span><b>names</b>(oxalis.trees))

<i><span style='color:#755019;'># Robinsons-Foulds distance</span></i>
oxalis.trees.d.rf <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>multiRF</b>(oxalis.trees)

<i><span style='color:#755019;'># Add names of columns and rows</span></i>
<b>colnames</b>(oxalis.trees.d.rf) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>names</b>(oxalis.trees)
<b>rownames</b>(oxalis.trees.d.rf) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>names</b>(oxalis.trees)

<i><span style='color:#755019;'># Create heatmap using corrplot function from corrplot package</span></i>
<b>corrplot</b>(<span style='color:#e85290;'>corr=</span>oxalis.trees.d.rf, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;circle&quot;</span>, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;upper&quot;</span>, <span style='color:#e85290;'>col=</span><b>rainbow</b>(<span style='color:#2b74c7;'>15</span>), <span style='color:#e85290;'>title=</span><span style='color:#bf0303;'>&quot;Correlation matrix of topographical distances&quot;</span>, <span style='color:#e85290;'>is.corr=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>diag=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>outline=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>order=</span><span style='color:#bf0303;'>&quot;alphabet&quot;</span>, <span style='color:#e85290;'>tl.pos=</span><span style='color:#bf0303;'>&quot;lt&quot;</span>, <span style='color:#e85290;'>tl.col=</span><span style='color:#bf0303;'>&quot;black&quot;</span>)
<b>corrplot</b>(<span style='color:#e85290;'>corr=</span>oxalis.trees.d.rf, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;number&quot;</span>, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;lower&quot;</span>, <span style='color:#e85290;'>add=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>col=</span><b>rainbow</b>(<span style='color:#2b74c7;'>15</span>), <span style='color:#e85290;'>title=</span><span style='color:#bf0303;'>&quot;Correlation matrix of topographical distances&quot;</span>, <span style='color:#e85290;'>is.corr=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>diag=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>outline=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>order=</span><span style='color:#bf0303;'>&quot;alphabet&quot;</span>, <span style='color:#e85290;'>tl.pos=</span><span style='color:#bf0303;'>&quot;ld&quot;</span>, <span style='color:#e85290;'>cl.pos=</span><span style='color:#bf0303;'>&quot;n&quot;</span>)

<b><span style='color:#b00000;'>## Seeing trees in the forest</span></b>

<i><span style='color:#755019;'># Root all trees</span></i>
oxalis.trees.rooted <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>lapply</b>(<span style='color:#e85290;'>X=</span>oxalis.trees, <span style='color:#e85290;'>FUN=</span>root, <span style='color:#bf0303;'>&quot;TaxaOut&quot;</span>)
<b>class</b>(oxalis.trees.rooted) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> &quot;multiPhylo&quot;</span>
<i><span style='color:#755019;'># Consenus tree (50 % rule)</span></i>
oxalis.tree.con <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span>ape<span style='color:#803f00;'>::</span><b>consensus</b>(oxalis.trees.rooted, <span style='color:#e85290;'>p=</span><span style='color:#2b74c7;'>0.5</span>, <span style='color:#e85290;'>check.labels=</span><span style='color:#00892b;'>TRUE</span>)
<b>print.phylo</b>(oxalis.tree.con)
<i><span style='color:#755019;'># Plot the tree</span></i>
<b>plot.phylo</b>(oxalis.tree.con, <span style='color:#e85290;'>edge.width=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>label.offset=</span><span style='color:#2b74c7;'>0.3</span>)
<b>axisPhylo</b>(<span style='color:#e85290;'>side=</span><span style='color:#2b74c7;'>1</span>)

<i><span style='color:#755019;'># Species tree</span></i>
<i><span style='color:#755019;'># All trees must be ultrametric - chronos scale them</span></i>
oxalis.trees.ultra <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>lapply</b>(<span style='color:#e85290;'>X=</span>oxalis.trees.rooted, <span style='color:#e85290;'>FUN=</span>chronos, <span style='color:#e85290;'>model=</span><span style='color:#bf0303;'>&quot;correlated&quot;</span>)
<b>class</b>(oxalis.trees.ultra) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> &quot;multiPhylo&quot;</span>
<i><span style='color:#755019;'># Mean distances</span></i>
oxalis.tree.sp.mean <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>speciesTree</b>(oxalis.trees.ultra, mean)
<b>print.phylo</b>(oxalis.tree.sp.mean)
<i><span style='color:#755019;'># Plot the tree</span></i>
<b>plot.phylo</b>(oxalis.tree.sp.mean, <span style='color:#e85290;'>edge.width=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>label.offset=</span><span style='color:#2b74c7;'>0.01</span>)
<b>edgelabels</b>(<span style='color:#e85290;'>text=</span><b>round</b>(oxalis.tree.sp.mean[[<span style='color:#bf0303;'>&quot;edge.length&quot;</span>]], <span style='color:#e85290;'>digits=</span><span style='color:#2b74c7;'>2</span>), <span style='color:#e85290;'>frame=</span><span style='color:#bf0303;'>&quot;none&quot;</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;red&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;none&quot;</span>)
<b>axisPhylo</b>(<span style='color:#e85290;'>side=</span><span style='color:#2b74c7;'>1</span>)

<i><span style='color:#755019;'># Parsimony super tree</span></i>
<b>library</b>(phangorn)
oxalis.tree.sp <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>superTree</b>(<span style='color:#e85290;'>tree=</span>oxalis.trees.rooted, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;optim.parsimony&quot;</span>, <span style='color:#e85290;'>rooted=</span><span style='color:#00892b;'>TRUE</span>)
<b>print.phylo</b>(oxalis.tree.sp)
<b>plot.phylo</b>(oxalis.tree.sp, <span style='color:#e85290;'>edge.width=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>label.offset=</span><span style='color:#2b74c7;'>0.01</span>)
<b>axisPhylo</b>(<span style='color:#e85290;'>side=</span><span style='color:#2b74c7;'>1</span>)

<i><span style='color:#755019;'># Density tree</span></i>
<b>densiTree</b>(<span style='color:#e85290;'>x=</span>oxalis.trees.ultra, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;cladogram&quot;</span>, <span style='color:#e85290;'>alpha=</span><span style='color:#2b74c7;'>0.5</span>, <span style='color:#e85290;'>consensus=</span>oxalis.tree.sp.mean, <span style='color:#e85290;'>scaleX=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>col=</span><b>c</b>(<span style='color:#bf0303;'>&quot;black&quot;</span>, <span style='color:#bf0303;'>&quot;green&quot;</span>, <span style='color:#bf0303;'>&quot;blue&quot;</span>, <span style='color:#bf0303;'>&quot;red&quot;</span>), <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.5</span>)

<i><span style='color:#755019;'># Networks</span></i>
oxalis.tree.net <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>consensusNet</b>(oxalis.trees.rooted, <span style='color:#e85290;'>prob=</span><span style='color:#2b74c7;'>0.25</span>)
<b>plot.networx</b>(<span style='color:#e85290;'>x=</span>oxalis.tree.net, <span style='color:#e85290;'>planar=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;2D&quot;</span>, <span style='color:#e85290;'>use.edge.length=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>show.tip.label=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>show.edge.label=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>show.node.label=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>show.nodes=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>edge.color=</span><span style='color:#bf0303;'>&quot;black&quot;</span>, <span style='color:#e85290;'>tip.color=</span><span style='color:#bf0303;'>&quot;blue&quot;</span>)
<b>plot.networx</b>(<span style='color:#e85290;'>x=</span>oxalis.tree.net, <span style='color:#e85290;'>planar=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;3D&quot;</span>, <span style='color:#e85290;'>use.edge.length=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>show.tip.label=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>show.edge.label=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>show.node.label=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>show.nodes=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>edge.color=</span><span style='color:#bf0303;'>&quot;black&quot;</span>, <span style='color:#e85290;'>tip.color=</span><span style='color:#bf0303;'>&quot;blue&quot;</span>)

<i><span style='color:#755019;'># Plot all trees on same scale</span></i>
<b>kronoviz</b>(<span style='color:#e85290;'>x=</span>oxalis.trees.rooted, <span style='color:#e85290;'>layout=</span><b>length</b>(oxalis.trees.rooted), <span style='color:#e85290;'>horiz=</span><span style='color:#00892b;'>TRUE</span>)
<i><span style='color:#755019;'># Close graphical device to cancel division of plotting device</span></i>
<b>dev.off</b>()

<b><span style='color:#b00000;'>## Maximum parsimony</span></b>
<i><span style='color:#755019;'># Conversion to phyDat for phangorn</span></i>
meles.phydat <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>as.phyDat</b>(meles.dna)
<i><span style='color:#755019;'># Prepare starting tree</span></i>
meles.tre.ini <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>nj</b>(<b>dist.dna</b>(<span style='color:#e85290;'>x=</span>meles.dna, <span style='color:#e85290;'>model=</span><span style='color:#bf0303;'>&quot;raw&quot;</span>))
<i><span style='color:#755019;'># Parsimony</span></i>
?parsimony
<i><span style='color:#755019;'># Returns maximum parsimony score</span></i>
<b>parsimony</b>(<span style='color:#e85290;'>tree=</span>meles.tre.ini, <span style='color:#e85290;'>data=</span>meles.phydat)
<i><span style='color:#755019;'># Optimisation - returns maximum parsimony tree</span></i>
meles.tre.pars <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>optim.parsimony</b>(<span style='color:#e85290;'>tree=</span>meles.tre.ini, <span style='color:#e85290;'>data=</span>meles.phydat)
<i><span style='color:#755019;'># Draw a tree</span></i>
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>meles.tre.pars, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;cladogram&quot;</span>, <span style='color:#e85290;'>edge.width=</span><span style='color:#2b74c7;'>2</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;Maximum-parsimony tree of Meles&quot;</span>)

<b><span style='color:#b00000;'>## More about plotting the trees</span></b>
?plot.phylo <i><span style='color:#755019;'># check it for various possibilities what to influence</span></i>
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>hauss.nj, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;cladogram&quot;</span>, <span style='color:#e85290;'>use.edge.length=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>direction=</span><span style='color:#bf0303;'>&quot;leftwards&quot;</span>)
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>hauss.nj, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;cladogram&quot;</span>, <span style='color:#e85290;'>use.edge.length=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>direction=</span><span style='color:#bf0303;'>&quot;rightwards&quot;</span>)

<i><span style='color:#755019;'># Nice tiplabels and higlighted tiplabel</span></i>
trape <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.tree</b>(<span style='color:#e85290;'>text =</span> <span style='color:#bf0303;'>&quot;((Homo, Pan), Gorilla);&quot;</span>)
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>trape, <span style='color:#e85290;'>show.tip.label=</span><span style='color:#00892b;'>FALSE</span>)
<b>tiplabels</b>(trape<span style='color:#803f00;'>$</span>tip.label, <span style='color:#e85290;'>bg=</span><b>c</b>(<span style='color:#bf0303;'>&quot;white&quot;</span>, <span style='color:#bf0303;'>&quot;black&quot;</span>, <span style='color:#bf0303;'>&quot;white&quot;</span>), <span style='color:#e85290;'>col=</span><b>c</b>(<span style='color:#bf0303;'>&quot;black&quot;</span>, <span style='color:#bf0303;'>&quot;white&quot;</span>, <span style='color:#bf0303;'>&quot;black&quot;</span>), <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>2</span>)
<b>nodelabels</b>(<span style='color:#e85290;'>text=</span><b>c</b>(<span style='color:#bf0303;'>&quot;6.4 Ma&quot;</span>, <span style='color:#bf0303;'>&quot;5.4 Ma&quot;</span>), <span style='color:#e85290;'>frame=</span><span style='color:#bf0303;'>&quot;circle&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;yellow&quot;</span>)
<b>add.scale.bar</b>() 

<b><span style='color:#b00000;'>## PIC</span></b>

<i><span style='color:#755019;'># Prepare the data</span></i>
<i><span style='color:#755019;'># Body mass of primates</span></i>
primates.body <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>c</b>(<span style='color:#2b74c7;'>4.09434</span>, <span style='color:#2b74c7;'>3.61092</span>, <span style='color:#2b74c7;'>2.37024</span>, <span style='color:#2b74c7;'>2.02815</span>, <span style='color:#2b74c7;'>1.46968</span>)
<i><span style='color:#755019;'># Longevity of primates</span></i>
primates.longevity <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>c</b>(<span style='color:#2b74c7;'>4.74493</span>, <span style='color:#2b74c7;'>3.3322</span>, <span style='color:#2b74c7;'>3.3673</span>, <span style='color:#2b74c7;'>2.89037</span>, <span style='color:#2b74c7;'>2.30259</span>)
<i><span style='color:#755019;'># Add names to the values</span></i>
<b>names</b>(primates.body) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>names</b>(primates.longevity) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>c</b>(<span style='color:#bf0303;'>&quot;Homo&quot;</span>, <span style='color:#bf0303;'>&quot;Pongo&quot;</span>, <span style='color:#bf0303;'>&quot;Macaca&quot;</span>, <span style='color:#bf0303;'>&quot;Ateles&quot;</span>, <span style='color:#bf0303;'>&quot;Galago&quot;</span>)
<i><span style='color:#755019;'># Create a tree in Newick format</span></i>
primates.tree <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.tree</b>(<span style='color:#e85290;'>text=</span><span style='color:#bf0303;'>&quot;((((Homo:0.21, Pongo:0.21):0.28, Macaca:0.49):0.13, Ateles:0.62):0.38, Galago:1.00);&quot;</span>)
<b>plot.phylo</b>(primates.tree)

<i><span style='color:#755019;'># PIC - phylogenetically independent contrasts</span></i>
primates.pic.body <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>pic</b>(<span style='color:#e85290;'>x=</span>primates.body, <span style='color:#e85290;'>phy=</span>primates.tree, <span style='color:#e85290;'>scaled=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>var.contrasts=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>rescaled.tree=</span><span style='color:#00892b;'>FALSE</span>)
primates.pic.longevity <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>pic</b>(<span style='color:#e85290;'>x=</span>primates.longevity, <span style='color:#e85290;'>phy=</span>primates.tree, <span style='color:#e85290;'>scaled=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>var.contrasts=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>rescaled.tree=</span><span style='color:#00892b;'>FALSE</span>)

<i><span style='color:#755019;'># Plot a tree with PIC values</span></i>
<b>plot.phylo</b>(<span style='color:#e85290;'>x=</span>primates.tree, <span style='color:#e85290;'>lwd=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.5</span>, <span style='color:#e85290;'>offset=</span><span style='color:#2b74c7;'>0.2</span>)
<b>nodelabels</b>(<b>round</b>(primates.pic.body, <span style='color:#e85290;'>digits=</span><span style='color:#2b74c7;'>3</span>), <span style='color:#e85290;'>adj=</span><b>c</b>(<span style='color:#2b74c7;'>0</span>, <span style='color:#803f00;'>-</span><span style='color:#2b74c7;'>0.5</span>), <span style='color:#e85290;'>frame=</span><span style='color:#bf0303;'>&quot;none&quot;</span>)
<b>nodelabels</b>(<b>round</b>(primates.pic.longevity, <span style='color:#e85290;'>digits=</span><span style='color:#2b74c7;'>3</span>), <span style='color:#e85290;'>adj=</span><b>c</b>(<span style='color:#2b74c7;'>0</span>, <span style='color:#2b74c7;'>1</span>), <span style='color:#e85290;'>frame=</span><span style='color:#bf0303;'>&quot;none&quot;</span>)
<b>add.scale.bar</b>()

<i><span style='color:#755019;'># Plot PIC</span></i>
<b>plot</b>(<span style='color:#e85290;'>x=</span>primates.pic.body, <span style='color:#e85290;'>y=</span>primates.pic.longevity, <span style='color:#e85290;'>pch=</span><span style='color:#2b74c7;'>16</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.5</span>)
<b>abline</b>(<span style='color:#e85290;'>a=</span><span style='color:#2b74c7;'>0</span>, <span style='color:#e85290;'>b=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>lty=</span><span style='color:#2b74c7;'>2</span>) <i><span style='color:#755019;'># x=y line</span></i>

<i><span style='color:#755019;'># Correlation of PIC of body mass and longevity</span></i>
<b>cor</b>(<span style='color:#e85290;'>x=</span>primates.pic.body, <span style='color:#e85290;'>y=</span>primates.pic.longevity, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;pearson&quot;</span>)
<b>lm</b>(<span style='color:#e85290;'>formula=</span>primates.pic.longevity<span style='color:#803f00;'>~</span>primates.pic.body)
<i><span style='color:#755019;'># Because PICs have expected mean zero - such linear regressions should be done through the origin (i.e. the intercept is set to zero)</span></i>
<b>lm</b>(<span style='color:#e85290;'>formula=</span>primates.pic.longevity<span style='color:#803f00;'>~</span>primates.pic.body<span style='color:#803f00;'>-</span><span style='color:#2b74c7;'>1</span>)

<i><span style='color:#755019;'># Permutation procedure to test PIC</span></i>
<b>lmorigin</b>(<span style='color:#e85290;'>formula=</span>primates.pic.longevity<span style='color:#803f00;'>~</span>primates.pic.body, <span style='color:#e85290;'>nperm=</span><span style='color:#2b74c7;'>1000</span>)

<b><span style='color:#b00000;'>## Phylogenetic autocorrelation</span></b>

<i><span style='color:#755019;'># Autocorrelation coefficient to quantify whether the distribution of a trait among a set of species is affected or not by their phylogenetic relationships</span></i>
<i><span style='color:#755019;'># In the absence of phylogenetic autocorrelation, the mean expected value of I and its variance are known - it is thus possible to test the null hypothesis of the absence of dependence among observations</span></i>
<i><span style='color:#755019;'># Let's choose weights as wij = 1/dij, where the d’s is the distances measured on the tree - cophenetic() calculates cophenetic distances</span></i>
primates.weights <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>/</span><b>cophenetic</b>(primates.tree) <i><span style='color:#755019;'># can be just cophenetic(primates.tree) or some other transformation</span></i>
<i><span style='color:#755019;'># See it</span></i>
primates.weights
<b>class</b>(primates.weights)
<i><span style='color:#755019;'># Set diagonal to 0</span></i>
<b>diag</b>(primates.weights) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><span style='color:#2b74c7;'>0</span>
<i><span style='color:#755019;'># Calculate Moran's I</span></i>
<b>Moran.I</b>(<span style='color:#e85290;'>x=</span>primates.body, <span style='color:#e85290;'>weight=</span>primates.weights, <span style='color:#e85290;'>alternative=</span><span style='color:#bf0303;'>&quot;greater&quot;</span>) <i><span style='color:#755019;'># Slighly significant positive phylogenetic correlation among body mass</span></i>
<b>Moran.I</b>(<span style='color:#e85290;'>x=</span>primates.longevity, <span style='color:#e85290;'>weight=</span>primates.weights, <span style='color:#e85290;'>alternative=</span><span style='color:#bf0303;'>&quot;greater&quot;</span>) <i><span style='color:#755019;'># Positive, but non-significant</span></i>

<i><span style='color:#755019;'># Test of Moran's with randomisation procedure</span></i>
<b>gearymoran</b>(<span style='color:#e85290;'>bilis=</span>primates.weights, <span style='color:#e85290;'>X=</span><b>data.frame</b>(primates.body, primates.longevity), <span style='color:#e85290;'>nrepet=</span><span style='color:#2b74c7;'>1000</span>) <i><span style='color:#755019;'># Body is significant - nonrandom, longevity not (random)</span></i>

<i><span style='color:#755019;'># Test of Abouheif designed to detect phylogenetic autocorrelation in a quantitative trait - in fact Moran's I test using a particular phylogenetic proximity between tips</span></i>
<b>library</b>(adephylo)
<b>abouheif.moran</b>(<span style='color:#e85290;'>x=</span><b>cbind</b>(primates.body, primates.longevity), <span style='color:#e85290;'>W=</span>primates.weights, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;oriAbouheif&quot;</span>, <span style='color:#e85290;'>a=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>nrepet=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>alter=</span><span style='color:#bf0303;'>&quot;greater&quot;</span>)

<i><span style='color:#755019;'># correlogram can be used to visualize the results of phylogenetic autocorrelative analysis</span></i>
<i><span style='color:#755019;'># Loads training data set</span></i>
<b>data</b>(carnivora)
<i><span style='color:#755019;'># Look at the data</span></i>
<b>head</b>(carnivora)
<i><span style='color:#755019;'># Calculate the correlogram</span></i>
carnivora.correlogram <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>correlogram.formula</b>(<span style='color:#e85290;'>formula=</span>SW<span style='color:#803f00;'>~</span>Order<span style='color:#803f00;'>/</span>SuperFamily<span style='color:#803f00;'>/</span>Family<span style='color:#803f00;'>/</span>Genus, <span style='color:#e85290;'>data=</span>carnivora)
<i><span style='color:#755019;'># See results</span></i>
carnivora.correlogram
<i><span style='color:#755019;'># Calculate the correlogram - test for both body masses</span></i>
carnivora.correlogram2 <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>correlogram.formula</b>(<span style='color:#e85290;'>formula=</span>SW<span style='color:#803f00;'>+</span>FW<span style='color:#803f00;'>~</span>Order<span style='color:#803f00;'>/</span>SuperFamily<span style='color:#803f00;'>/</span>Family<span style='color:#803f00;'>/</span>Genus, <span style='color:#e85290;'>data=</span>carnivora)
<i><span style='color:#755019;'># See results</span></i>
carnivora.correlogram2
<i><span style='color:#755019;'># Plot it</span></i>
<b>plot.correlogram</b>(<span style='color:#e85290;'>x=</span>carnivora.correlogram, <span style='color:#e85290;'>legend=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>test.level=</span><span style='color:#2b74c7;'>0.05</span>, <span style='color:#e85290;'>col=</span><b>c</b>(<span style='color:#bf0303;'>&quot;white&quot;</span>, <span style='color:#bf0303;'>&quot;black&quot;</span>))
<i><span style='color:#755019;'># Plot it - test for both body masses</span></i>
<i><span style='color:#755019;'># Two graphs</span></i>
<b>plot.correlogramList</b>(<span style='color:#e85290;'>x=</span>carnivora.correlogram2, <span style='color:#e85290;'>lattice=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>legend=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>test.level=</span><span style='color:#2b74c7;'>0.05</span>)
<i><span style='color:#755019;'># Only one graph</span></i>
<b>plot.correlogramList</b>(<span style='color:#e85290;'>x=</span>carnivora.correlogram2, <span style='color:#e85290;'>lattice=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>legend=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>test.level=</span><span style='color:#2b74c7;'>0.05</span>)

<b><span style='color:#b00000;'>## phylogenetic PCA</span></b>
<b>library</b>(phylobase) <i><span style='color:#755019;'># Library needed to create phylo4d object required by ppca</span></i>
<i><span style='color:#755019;'># Calculate pPCA</span></i>
primates.ppca <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>ppca</b>(<span style='color:#e85290;'>x=</span><b>phylo4d</b>(<span style='color:#e85290;'>x=</span>primates.tree, <b>cbind</b>(primates.body, primates.longevity)), <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;patristic&quot;</span>, <span style='color:#e85290;'>a=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>center=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>scale=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>scannf=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>nfposi=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>nfnega=</span><span style='color:#2b74c7;'>0</span>)
<i><span style='color:#755019;'># Print results</span></i>
<b>print.ppca</b>(primates.ppca)
<i><span style='color:#755019;'># See summary information</span></i>
<b>summary.ppca</b>(primates.ppca)
<i><span style='color:#755019;'># See PCA scores for variables on phylogenetic tree</span></i>
<b>scatter.ppca</b>(primates.ppca)
<i><span style='color:#755019;'># See decomposition of pPCA eigenvalues</span></i>
<b>screeplot.ppca</b>(primates.ppca)
<i><span style='color:#755019;'># Plot pPCA results - global vs. local structure, decomposition of pPCA eigenvalues, PCA plot of variables and PCA scores for variables on phylogenetic tree</span></i>
<b>plot.ppca</b>(primates.ppca)

<b><span style='color:#b00000;'>## Ancestral state reconstruction</span></b>
<i><span style='color:#755019;'># By default performs estimation for continuous characters assuming a Brownian motion model fit by maximum likelihood</span></i>
<i><span style='color:#755019;'># See ?ace for possible settings</span></i>
primates.body.ace <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>ace</b>(<span style='color:#e85290;'>x=</span>primates.body, <span style='color:#e85290;'>phy=</span>primates.tree, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;continuous&quot;</span>, <span style='color:#e85290;'>method=</span><span style='color:#bf0303;'>&quot;REML&quot;</span>, <span style='color:#e85290;'>corStruct=</span><b>corBrownian</b>(<span style='color:#e85290;'>value=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>phy=</span>primates.tree))
<i><span style='color:#755019;'># See result - reconstructions are in $ace - to be plotted on nodes - 1st column are node numbers</span></i>
primates.body.ace
<i><span style='color:#755019;'># Other implementations are available in packages geiger (functions fitContinuousMCMC and fitDiscrete), phangorn and more in ape (MPR)</span></i>
<i><span style='color:#755019;'># Plot it</span></i>
<b>plot</b>(primates.tree, <span style='color:#e85290;'>lwd=</span><span style='color:#2b74c7;'>2</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>2</span>)
<b>tiplabels</b>(<b>round</b>(primates.body, <span style='color:#e85290;'>digits=</span><span style='color:#2b74c7;'>3</span>), <span style='color:#e85290;'>adj=</span><b>c</b>(<span style='color:#2b74c7;'>0</span>, <span style='color:#803f00;'>-</span><span style='color:#2b74c7;'>1</span>), <span style='color:#e85290;'>frame=</span><span style='color:#bf0303;'>&quot;none&quot;</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;blue&quot;</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>2</span>)
<b>nodelabels</b>(<b>round</b>(primates.body.ace<span style='color:#803f00;'>$</span>ace, <span style='color:#e85290;'>digits=</span><span style='color:#2b74c7;'>3</span>), <span style='color:#e85290;'>frame=</span><span style='color:#bf0303;'>&quot;circle&quot;</span>, <span style='color:#e85290;'>bg=</span><span style='color:#bf0303;'>&quot;red&quot;</span>, <span style='color:#e85290;'>cex=</span><span style='color:#2b74c7;'>1.5</span>)



<b><span style='color:#b00000;'>################################################################################</span></b>

<i><span style='color:#755019;'># - populační indexy po populacích - seppop a lapply</span></i>
<i><span style='color:#755019;'># - bootstrap Fst</span></i>
<i><span style='color:#755019;'># - konverze muscle do DNAbin/něčeho jiného</span></i>

<b><span style='color:#b00000;'>## RAxML</span></b>
meles.raxml <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>raxml</b>(<span style='color:#e85290;'>DNAbin=</span>usflu.dna, <span style='color:#e85290;'>N=</span><span style='color:#bf0303;'>&quot;autoFC&quot;</span>, <span style='color:#e85290;'>exec=</span><span style='color:#bf0303;'>&quot;~/bin/&quot;</span>)

<b><span style='color:#b00000;'>## Extra</span></b>
<b>axisGeo</b>() <i><span style='color:#755019;'># package phyloch - adds scale in geological time, has many options</span></i>

<b><span style='color:#b00000;'>## Polysat - microsattelites and mixed ploidies</span></b>
<b>library</b>(combinat)
<b>library</b>(polysat)
polysattest <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>read.GeneMapper</b>(<span style='color:#bf0303;'>&quot;http://trapa.cz/sites/default/rcourse/GeneMapperExample.txt&quot;</span>)
<i><span style='color:#755019;'># Basic information</span></i>
<b>summary</b>(polysattest)
<b>Loci</b>(polysattest) <i><span style='color:#755019;'># Information about loci</span></i>
<b>class</b>(polysattest)
<i><span style='color:#755019;'># Add information missing in input data file</span></i>
<b>Description</b>(polysattest) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> &quot;Dataset for the tutorial&quot;</span> <i><span style='color:#755019;'># Description of dataset</span></i>
<b>PopNames</b>(polysattest) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>c</b>(<span style='color:#bf0303;'>&quot;PopA&quot;</span>, <span style='color:#bf0303;'>&quot;PopB&quot;</span>, <span style='color:#bf0303;'>&quot;PopC&quot;</span>) <i><span style='color:#755019;'># Population names</span></i>
<b>PopInfo</b>(polysattest) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>rep</b>(<span style='color:#e85290;'>x=</span><span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>each=</span><span style='color:#2b74c7;'>100</span>) <i><span style='color:#755019;'># Which individuals belong to which population</span></i>
<b>PopInfo</b>(polysattest) <i><span style='color:#755019;'># Which individuals belong to which population</span></i>
<b>Usatnts</b>(polysattest) <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>c</b>(<span style='color:#2b74c7;'>2</span>, <span style='color:#2b74c7;'>3</span>, <span style='color:#2b74c7;'>2</span>) <i><span style='color:#755019;'># Ploidies</span></i>
<b>Usatnts</b>(polysattest) <i><span style='color:#755019;'># Get information about ploidy levels for populations</span></i>
<i><span style='color:#755019;'># Show samples</span></i>
<b>Samples</b>(polysattest)
<i><span style='color:#755019;'># Subsetting of populations:</span></i>
polysattest2 <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>Samples</b>(<span style='color:#e85290;'>object=</span>polysattest, <span style='color:#e85290;'>populations=</span><span style='color:#bf0303;'>&quot;PopA&quot;</span>)
<b>summary</b>(polysattest2)
<i><span style='color:#755019;'># View genotypes (for locus loc1) of samples A1-A20</span></i>
<b>viewGenotypes</b>(polysattest, <span style='color:#e85290;'>samples=</span><b>paste</b>(<span style='color:#bf0303;'>&quot;A&quot;</span>, <span style='color:#2b74c7;'>1</span><span style='color:#803f00;'>:</span><span style='color:#2b74c7;'>20</span>, <span style='color:#e85290;'>sep=</span><span style='color:#bf0303;'>&quot;&quot;</span>), <span style='color:#e85290;'>loci=</span><span style='color:#bf0303;'>&quot;loc1&quot;</span>)
<i><span style='color:#755019;'># Show missing data</span></i>
<b>find.missing.gen</b>(polysattest)
<b>summary</b>(polysattest)
testmat <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>meandistance.matrix</b>(polysattest)
pca <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>cmdscale</b>(testmat)
mycol <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>c</b>(<span style='color:#bf0303;'>&quot;red&quot;</span>, <span style='color:#bf0303;'>&quot;green&quot;</span>, <span style='color:#bf0303;'>&quot;blue&quot;</span>)
<b>plot</b>(pca[,<span style='color:#2b74c7;'>1</span>], pca[,<span style='color:#2b74c7;'>2</span>], <span style='color:#e85290;'>col=</span>mycol[<b>PopInfo</b>(polysattest)], <span style='color:#803f00;'>+</span><span style='color:#bf0303;'> </span><span style='color:#e85290;'>main =</span> <span style='color:#bf0303;'>&quot;PCA with Bruvo distance&quot;</span>)
testmat2 <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>meandistance.matrix</b>(polysattest, <span style='color:#e85290;'>distmetric=</span>Lynch.distance)
pca2 <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>cmdscale</b>(testmat2)
<b>plot</b>(pca2[,<span style='color:#2b74c7;'>1</span>], pca2[,<span style='color:#2b74c7;'>2</span>], <span style='color:#e85290;'>col=</span><b>rep</b>(<b>c</b>(<span style='color:#bf0303;'>&quot;red&quot;</span>, <span style='color:#bf0303;'>&quot;green&quot;</span>, <span style='color:#bf0303;'>&quot;blue&quot;</span>), <span style='color:#e85290;'>each=</span><span style='color:#2b74c7;'>100</span>), <span style='color:#803f00;'>+</span><span style='color:#bf0303;'> </span><span style='color:#e85290;'>main =</span> <span style='color:#bf0303;'>&quot;PCA with Lynch distance&quot;</span>)
polysattest2 <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>deleteSamples</b>(polysattest, <b>c</b>(<span style='color:#bf0303;'>&quot;B59&quot;</span>, <span style='color:#bf0303;'>&quot;C30&quot;</span>))
polysattest2 <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>deleteLoci</b>(polysattest2, <span style='color:#bf0303;'>&quot;loc2&quot;</span>)

<b><span style='color:#b00000;'>## ML</span></b>
<i><span style='color:#755019;'># Konverze dat do formátu pro phangorn</span></i>
meles.dna2 <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>as.phyDat</b>(meles.dna)
<b>class</b>(meles.dna2)
meles.dna2
meles.tre.ini <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>nj</b>(<b>dist.dna</b>(meles.dna,<span style='color:#e85290;'>model=</span><span style='color:#bf0303;'>&quot;TN93&quot;</span>))
meles.tre.ini
<b>pml</b>(meles.tre.ini, meles.dna2, <span style='color:#e85290;'>k=</span><span style='color:#2b74c7;'>4</span>)
<i><span style='color:#755019;'># Inicializace optimalizační procedury</span></i>
?pml
<b>table</b>(<b>as.character</b>(meles.dna2))
<i><span style='color:#755019;'># Nalezení chybějících dat</span></i>
meles.na.posi <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>which</b>(<b>apply</b>(<b>as.character</b>(meles.dna),<span style='color:#2b74c7;'>2</span>, <span style='color:#0000bf;'>function</span>(e) <b>any</b>(<span style='color:#803f00;'>!</span>e <span style='color:#803f00;'>%in%</span><span style='color:#bf0303;'> </span><b>c</b>(<span style='color:#bf0303;'>&quot;a&quot;</span>,<span style='color:#bf0303;'>&quot;t&quot;</span>,<span style='color:#bf0303;'>&quot;g&quot;</span>,<span style='color:#bf0303;'>&quot;c&quot;</span>))))
<i><span style='color:#755019;'># Graf chybějících dat</span></i>
meles.na.density <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>apply</b>(<b>as.character</b>(meles.dna),<span style='color:#2b74c7;'>2</span>, <span style='color:#0000bf;'>function</span>(e) <b>sum</b>(<span style='color:#803f00;'>!</span>e <span style='color:#803f00;'>%in%</span><span style='color:#bf0303;'> </span><b>c</b>(<span style='color:#bf0303;'>&quot;a&quot;</span>,<span style='color:#bf0303;'>&quot;t&quot;</span>,<span style='color:#bf0303;'>&quot;g&quot;</span>,<span style='color:#bf0303;'>&quot;c&quot;</span>)))
<b>plot</b>(meles.na.density, <span style='color:#e85290;'>type=</span><span style='color:#bf0303;'>&quot;l&quot;</span>, <span style='color:#e85290;'>col=</span><span style='color:#bf0303;'>&quot;blue&quot;</span>, <span style='color:#e85290;'>xlab=</span><span style='color:#bf0303;'>&quot;Position in HA segment&quot;</span>, <span style='color:#e85290;'>ylab=</span><span style='color:#bf0303;'>&quot;Number of NAs&quot;</span>)
<i><span style='color:#755019;'># Odstranění chybějících dat</span></i>
meles.dna3 <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span>meles.dna[,<span style='color:#803f00;'>-</span>na.posi]
<i><span style='color:#755019;'># Alignment dat bez chybějících hodnot</span></i>
meles.dna3
<b>table</b>(<b>as.character</b>(meles.dna3))
<i><span style='color:#755019;'># Konverze alignmentu do formátu pro výpočet ML</span></i>
<i><span style='color:#755019;'># Znovu spočítat likelihood</span></i>
meles.dna4 <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>as.phyDat</b>(meles.dna3)
meles.tre.ini <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>nj</b>(<b>dist.dna</b>(meles.dna3,<span style='color:#e85290;'>model=</span><span style='color:#bf0303;'>&quot;TN93&quot;</span>))
meles.fit.ini <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>pml</b>(meles.tre.ini, meles.dna4, <span style='color:#e85290;'>k=</span><span style='color:#2b74c7;'>4</span>)
meles.fit.ini
meles.fit <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>optim.pml</b>(meles.fit.ini, <span style='color:#e85290;'>optNni=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>optBf=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>optQ=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>optGamma=</span><span style='color:#00892b;'>TRUE</span>)
meles.fit
<b>class</b>(meles.fit)
<b>names</b>(meles.fit)
meles.fit<span style='color:#803f00;'>$</span>tree
<i><span style='color:#755019;'># Porovnání starého a nového fitu</span></i>
<b>anova</b>(meles.fit.ini, meles.fit)
<i><span style='color:#755019;'># AIC - nižší = lepší</span></i>
<b>AIC</b>(meles.fit.ini)
<b>AIC</b>(meles.fit)
<i><span style='color:#755019;'># Extrakce a nakreslení stromu</span></i>
meles.tre4 <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>root</b>(meles.fit<span style='color:#803f00;'>$</span>tree,<span style='color:#2b74c7;'>1</span>)
<b>plot</b>(meles.tre4, <span style='color:#e85290;'>show.tip=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>edge.width=</span><span style='color:#2b74c7;'>2</span>)
<b>title</b>(<span style='color:#bf0303;'>&quot;Maximum-likelihood tree&quot;</span>)
<b>axisPhylo</b>()

<b><span style='color:#b00000;'>## Diversity</span></b>
nuphar.divpart <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>fastDivPart</b>(<span style='color:#e85290;'>infile=</span><span style='color:#bf0303;'>&quot;diversity/nuphar.genepop&quot;</span>, <span style='color:#e85290;'>outfile=</span><span style='color:#bf0303;'>&quot;diversity/&quot;</span>, <span style='color:#e85290;'>gp=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>pairwise=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>WC_Fst=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>bs_locus=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>bs_pairwise=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>bootstraps=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>plot=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>parallel=</span><span style='color:#00892b;'>TRUE</span>)
<b>inCalc</b>(<span style='color:#e85290;'>infile=</span><span style='color:#bf0303;'>&quot;diversity/nuphar.genepop&quot;</span>, <span style='color:#e85290;'>outfile=</span><span style='color:#bf0303;'>&quot;diversity/&quot;</span>, <span style='color:#e85290;'>pairwise=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>xlsx=</span><span style='color:#00892b;'>FALSE</span>, <span style='color:#e85290;'>bootstraps=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>parallel=</span><span style='color:#00892b;'>TRUE</span>)
nuphar.genepop <b><span style='color:#336366;'>&lt;-</span></b><span style='color:#bf0303;'> </span><b>readGenepop</b>(<span style='color:#e85290;'>infile=</span><span style='color:#bf0303;'>&quot;diversity/nuphar.genepop&quot;</span>, <span style='color:#e85290;'>gp=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>bootstrap=</span><span style='color:#00892b;'>FALSE</span>)
<b>corPlot</b>(nuphar.genepop, nuphar.divpart)
<b>difPlot</b>(nuphar.divpart, <span style='color:#e85290;'>outfile=</span><span style='color:#bf0303;'>&quot;diversity/&quot;</span>, <span style='color:#e85290;'>interactive=</span><span style='color:#00892b;'>TRUE</span>)
<b>chiCalc</b>(<span style='color:#e85290;'>infile=</span><span style='color:#bf0303;'>&quot;diversity/nuphar.genepop&quot;</span>, <span style='color:#e85290;'>outfile=</span><span style='color:#bf0303;'>&quot;diversity/&quot;</span>, <span style='color:#e85290;'>gp=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>minFreq=</span><span style='color:#2b74c7;'>0.01</span>)
<b>divRatio</b>(<span style='color:#e85290;'>infile=</span><span style='color:#bf0303;'>&quot;diversity/nuphar.genepop&quot;</span>, <span style='color:#e85290;'>outfile=</span><span style='color:#bf0303;'>&quot;diversity/&quot;</span>, <span style='color:#e85290;'>gp=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>pop_stats=</span>nuphar.genepop, <span style='color:#e85290;'>refPos=</span><span style='color:#2b74c7;'>10</span>, <span style='color:#e85290;'>bootstraps=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>parallel=</span><span style='color:#00892b;'>TRUE</span>)
<b>divMigrate</b>(<span style='color:#e85290;'>infile=</span><span style='color:#bf0303;'>&quot;diversity/nuphar.genepop&quot;</span>, <span style='color:#e85290;'>nbs=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>plot=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>para=</span><span style='color:#00892b;'>TRUE</span>)

<b><span style='color:#b00000;'>## Private alleles apod.</span></b>
<b>convert</b>(<span style='color:#bf0303;'>&quot;popgenkit/nuphar.genepop&quot;</span>, <span style='color:#e85290;'>ndigit=</span><span style='color:#2b74c7;'>3</span>)
<b>popgen</b>(<span style='color:#bf0303;'>&quot;popgenkit/nuphar.arp&quot;</span>, <span style='color:#e85290;'>ndigit=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>freq.overall=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>freq.by.pop=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>genetic.stats=</span><span style='color:#00892b;'>TRUE</span>, <span style='color:#e85290;'>pairwise.fst=</span><span style='color:#00892b;'>TRUE</span>)
<b>bootstrapHet</b>(<span style='color:#bf0303;'>&quot;popgenkit/nuphar.arp&quot;</span>, <span style='color:#e85290;'>ndigit=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>nrepet=</span><span style='color:#2b74c7;'>100</span>) <i><span style='color:#755019;'># Má problémy s množstvím chybějících dat (hlavně při vyšším BS)</span></i>
<b>jackmsatpop</b>(<span style='color:#bf0303;'>&quot;popgenkit/nuphar.arp&quot;</span>, <span style='color:#e85290;'>ndigit=</span><span style='color:#2b74c7;'>3</span>, <span style='color:#e85290;'>interval=</span><span style='color:#2b74c7;'>1</span>, <span style='color:#e85290;'>nrepet=</span><span style='color:#2b74c7;'>1000</span>, <span style='color:#e85290;'>richness=</span><span style='color:#00892b;'>FALSE</span>) <i><span style='color:#755019;'># Nefunguje (nejspíš kvůli množství chybějících dat)!</span></i>
</pre>
</body>
</html>
